diff -BurpN dnsmasq-2.61/src/config.h dnsmasq/src/config.h
--- dnsmasq-2.61/src/config.h	2012-05-17 11:50:30.640000002 +0000
+++ dnsmasq/src/config.h	2012-05-17 12:00:00.450000002 +0000
@@ -118,9 +118,9 @@ RESOLVFILE
 #define HAVE_DHCP
 #define HAVE_DHCP6 
 #define HAVE_TFTP
-#define HAVE_SCRIPT
+/* #define HAVE_SCRIPT */
 /* #define HAVE_LUASCRIPT */
-/* #define HAVE_BROKEN_RTC */
+#define HAVE_BROKEN_RTC
 /* #define HAVE_DBUS */
 /* #define HAVE_IDN */
 /* #define HAVE_CONNTRACK */
@@ -206,10 +206,12 @@ HAVE_SOCKADDR_SA_LEN
 #if !defined(__ARCH_HAS_MMU__) && !defined(__UCLIBC_HAS_MMU__)
 #  define NO_FORK
 #endif
-#if defined(__UCLIBC_HAS_IPV6__)
+#if defined(__UCLIBC_HAS_IPV6__) && defined(USE_IPV6)
 #  ifndef IPV6_V6ONLY
 #    define IPV6_V6ONLY 26
 #  endif
+#elif !defined(NO_IPV6)
+#  define NO_IPV6
 #endif
 
 /* This is for glibc 2.x */
