diff -BurpN dnsmasq.orig/Makefile dnsmasq/Makefile
--- dnsmasq.orig/Makefile	2013-03-22 15:11:53.000000000 +0000
+++ dnsmasq/Makefile	2013-03-24 11:20:26.899101907 +0000
@@ -24,8 +24,8 @@ MANDIR        = $(PREFIX)/share/man
 LOCALEDIR     = $(PREFIX)/share/locale
 BUILDDIR      = $(SRC)
 DESTDIR       = 
-CFLAGS        = -Wall -W -O2
-LDFLAGS       = 
+CFLAGS        = -Wall -W -Os -ffunction-sections -fdata-sections $(EXTRACFLAGS)
+LDFLAGS       = -ffunction-sections -fdata-sections -Wl,--gc-sections
 COPTS         = 
 RPM_OPT_FLAGS = 
 LIBS          = 
