diff -BurpN dnsmasq-2.63.orig/src/netlink.c dnsmasq-2.63/src/netlink.c
--- dnsmasq-2.63.orig/src/netlink.c	2012-08-16 19:04:05.000000000 +0600
+++ dnsmasq-2.63/src/netlink.c	2012-08-19 19:21:57.159852805 +0600
@@ -336,7 +336,9 @@ static int nl_async(struct nlmsghdr *h)
   if (h->nlmsg_type == NLMSG_ERROR)
     {
       struct nlmsgerr *err = NLMSG_DATA(h);
-      my_syslog(LOG_ERR, _("netlink returns error: %s"), strerror(-(err->error)));
+
+      if (err->error != 0)
+        my_syslog(LOG_ERR, _("netlink returns error: %s"), strerror(-(err->error)));
       return 0;
     }
   else if (h->nlmsg_pid == 0 && h->nlmsg_type == RTM_NEWROUTE) 
