--- a/src/radv.c
+++ b/src/radv.c
@@ -184,7 +184,7 @@ void icmp6_packet(time_t now)
 	  mac = daemon->namebuff;
 	}
          
-      my_syslog(MS_DHCP | LOG_INFO, "RTR-SOLICIT(%s) %s", interface, mac);
+      my_syslog(MS_DHCP | LOG_DEBUG, "RTR-SOLICIT(%s) %s", interface, mac);
       /* source address may not be valid in solicit request. */
       send_ra(now, if_index, interface, !IN6_IS_ADDR_UNSPECIFIED(&from.sin6_addr) ? &from.sin6_addr : NULL);
     }
@@ -281,7 +281,7 @@ static void send_ra(time_t now, int iface, char *iface_name, struct in6_addr *de
 		  opt->prefix = local;
 		  
 		  inet_ntop(AF_INET6, &local, daemon->addrbuff, ADDRSTRLEN);
-		  my_syslog(MS_DHCP | LOG_INFO, "RTR-ADVERT(%s) %s old prefix", iface_name, daemon->addrbuff); 		    
+		  my_syslog(MS_DHCP | LOG_DEBUG, "RTR-ADVERT(%s) %s old prefix", iface_name, daemon->addrbuff); 		    
 		}
 	   
 	      up = &context->next;
@@ -529,7 +529,7 @@ static int add_prefixes(struct in6_addr *local,  int prefix,
 		  opt->prefix = *local;
 		  
 		  inet_ntop(AF_INET6, local, daemon->addrbuff, ADDRSTRLEN);
-		  my_syslog(MS_DHCP | LOG_INFO, "RTR-ADVERT(%s) %s", param->if_name, daemon->addrbuff); 		    
+		  my_syslog(MS_DHCP | LOG_DEBUG, "RTR-ADVERT(%s) %s", param->if_name, daemon->addrbuff); 		    
 		}
 
 	    }
