From 03431d637301861ed13a4f5d7fc9be682bf28c71 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Thu, 20 Mar 2014 16:25:43 +0000
Subject: [PATCH 10/10] Initialise uid when creating CNAME cache record.

---
 src/rfc1035.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/src/rfc1035.c b/src/rfc1035.c
index ee11ff0..236d726 100644
--- a/src/rfc1035.c
+++ b/src/rfc1035.c
@@ -1065,6 +1065,8 @@ int extract_addresses(struct dns_header *header, size_t qlen, char *name, time_t
 		      if (newc)
 			{
 			  newc->addr.cname.target.cache = NULL;
+			  /* anything other than zero, to avoid being mistaken for CNAME to interface-name */ 
+			  newc->addr.cname.uid = 1; 
 			  if (cpp)
 			    {
 			      cpp->addr.cname.target.cache = newc;
-- 
1.7.2.5

