From 8767ceecd49f8f935e8b2f269d20be5c5bef0da8 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Mon, 21 May 2012 20:54:19 +0100
Subject: [PATCH 3/3] Make libvirt-style access control work when only doing RA.

---
 src/dhcp-common.c |    2 +-
 src/radv.c        |    2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/dhcp-common.c b/src/dhcp-common.c
index 9d1290c..402f644 100644
--- a/src/dhcp-common.c
+++ b/src/dhcp-common.c
@@ -407,7 +407,7 @@ static int join_multicast_worker(struct in6_addr *local, int prefix,
       return 1;
 
   /* weird libvirt-inspired access control */
-  for (context = daemon->dhcp6; context; context = context->next)
+  for (context = daemon->ra_contexts; context; context = context->next)
     if (!context->interface || strcmp(context->interface, ifrn_name) == 0)
       break;
 
diff --git a/src/radv.c b/src/radv.c
index edbef05..a9cfccc 100644
--- a/src/radv.c
+++ b/src/radv.c
@@ -158,7 +158,7 @@ void icmp6_packet(void)
       return;
  
   /* weird libvirt-inspired access control */
-  for (context = daemon->dhcp6; context; context = context->next)
+  for (context = daemon->ra_contexts; context; context = context->next)
     if (!context->interface || strcmp(context->interface, interface) == 0)
       break;
   
-- 
1.7.2.5

