--- a/src/rfc3315.c
+++ b/src/rfc3315.c
@@ -1596,6 +1596,8 @@ static void calculate_times(struct dhcp_context *context, unsigned int *min_time
 {
   unsigned int preferred_time, valid_time;
 
+  if (requested_time == 0)
+    requested_time = lease_time;
   if (requested_time < 120u )
     requested_time = 120u; /* sanity */
   if (lease_time == 0xffffffff || (requested_time != 0xffffffff && requested_time < lease_time))
