From 2b127a1eab194b39ea65e4b69570f233b0abbafc Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Tue, 18 Sep 2012 21:51:22 +0100
Subject: [PATCH 17/20] Flag DHCP or DHCPv6 in starup logging.

---
 CHANGELOG     |    3 +++
 src/dnsmasq.c |    9 +++++----
 2 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/CHANGELOG b/CHANGELOG
index 8dbcee9..a4e1d9e 100644
--- a/CHANGELOG
+++ b/CHANGELOG
@@ -16,6 +16,9 @@ version 2.64
 	    Don't report spurious netlink errors, regression in
 	    2.63. Thanks to Vladislav Grishenko for the patch.
 
+	    Flag DHCP or DHCPv6 in starup logging. Thanks to 
+	    Vladislav Grishenko for the patch.
+
 
 version 2.63
             Do duplicate dhcp-host address check in --test mode.
diff --git a/src/dnsmasq.c b/src/dnsmasq.c
index 6abd201..0023a4f 100644
--- a/src/dnsmasq.c
+++ b/src/dnsmasq.c
@@ -658,12 +658,13 @@ int main (int argc, char **argv)
 	  if ((dhcp_tmp->flags & CONTEXT_DHCP) || family == AF_INET) 
 	    my_syslog(MS_DHCP | LOG_INFO, 
 		      (dhcp_tmp->flags & CONTEXT_RA_STATELESS) ? 
-		      _("stateless DHCPv6 on %s%.0s%.0s") :
+		      _("%s stateless on %s%.0s%.0s") :
 		      (dhcp_tmp->flags & CONTEXT_STATIC) ? 
-		      _("DHCP, static leases only on %.0s%s, %s") :
+		      _("%s, static leases only on %.0s%s, %s") :
 		      (dhcp_tmp->flags & CONTEXT_PROXY) ?
-		      _("DHCP, proxy on subnet %.0s%s%.0s") :
-		      _("DHCP, IP range %s -- %s, %s"),
+		      _("%s, proxy on subnet %.0s%s%.0s") :
+		      _("%s, IP range %s -- %s, %s"),
+		      (family != AF_INET) ? "DHCPv6" : "DHCP",
 		      daemon->dhcp_buff, daemon->dhcp_buff3, daemon->namebuff);
 
 	  if (dhcp_tmp->flags & CONTEXT_RA_NAME)
-- 
1.7.2.5

