From 23780dd577aa2003370f0d6b31bd5f5b9c51c494 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Tue, 23 Oct 2012 17:04:37 +0100
Subject: [PATCH 10/10] Set tag "dhcpv6" rather than "DHCPv6", hardwired tags in lower-case is consistent.

---
 src/rfc3315.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/rfc3315.c b/src/rfc3315.c
index 77e0900..8c3cae0 100644
--- a/src/rfc3315.c
+++ b/src/rfc3315.c
@@ -194,8 +194,8 @@ static int dhcp6_no_relay(int msg_type, struct in6_addr *link_address, struct dh
   iface_id.next = tags;
   tags = &iface_id; 
 
-  /* set tag "DHCPv6" */
-  v6_id.net = "DHCPv6";
+  /* set tag "dhcpv6" */
+  v6_id.net = "dhcpv6";
   v6_id.next = tags;
   tags = &v6_id;
 
-- 
1.7.2.5

