diff -urB ppp.orig/chat/Makefile.linux ppp/chat/Makefile.linux
--- ppp.orig/chat/Makefile.linux	2008-11-26 13:02:29.000000000 +0300
+++ ppp/chat/Makefile.linux	2009-02-09 18:08:34.000000000 +0300
@@ -11,7 +11,7 @@
 CDEFS=	$(CDEF1) $(CDEF2) $(CDEF3) $(CDEF4)
 
 COPTS=	-O2 -g -pipe
-CFLAGS=	$(COPTS) $(CDEFS)
+CFLAGS=	$(COPTS) $(CDEFS) $(EXTRACFLAGS)
 
 INSTALL= install
 
diff -urB ppp.orig/pppd/Makefile.linux ppp/pppd/Makefile.linux
--- ppp.orig/pppd/Makefile.linux	2008-11-26 13:02:29.000000000 +0300
+++ ppp/pppd/Makefile.linux	2009-02-09 18:10:19.000000000 +0300
@@ -48,21 +48,21 @@
 # Uncomment the next line to include support for PPP packet filtering.
 # This requires that the libpcap library and headers be installed
 # and that the kernel driver support PPP packet filtering.
-FILTER=y
+#FILTER=y
 
 # Uncomment the next line to enable multilink PPP (enabled by default)
 # Linux distributions: Please leave multilink ENABLED in your builds
 # of pppd!
-HAVE_MULTILINK=y
+#HAVE_MULTILINK=y
 
 # Uncomment the next line to enable the TDB database (enabled by default.)
 # If you enable multilink, then TDB is automatically enabled also.
 # Linux distributions: Please leave TDB ENABLED in your builds.
-USE_TDB=y
+#USE_TDB=y
 
-HAS_SHADOW=y
+#HAS_SHADOW=y
 #USE_PAM=y
-HAVE_INET6=y
+#HAVE_INET6=y
 
 # Enable plugins
 PLUGIN=y
@@ -74,15 +74,16 @@
 #USE_SRP=y
 
 # Use libutil
-USE_LIBUTIL=y
+#USE_LIBUTIL=y
 
-MAXOCTETS=y
+#MAXOCTETS=y
 
 INCLUDE_DIRS= -I../include
 
-COMPILE_FLAGS= -DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MMAP
+COMPILE_FLAGS= #-DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MMAP
 
-CFLAGS= $(COPTS) $(COMPILE_FLAGS) $(INCLUDE_DIRS) '-DDESTDIR="@DESTDIR@"'
+CFLAGS= $(COPTS) $(COMPILE_FLAGS) $(EXTRACFLAGS) $(INCLUDE_DIRS) '-DDESTDIR="@DESTDIR@"'
+LDFLAGS += $(EXTRA_LDFLAGS)
 
 ifdef CHAPMS
 CFLAGS   += -DCHAPMS=1
@@ -117,10 +117,10 @@
 #LIBS     += -lshadow $(LIBS)
 endif
 
-ifneq ($(wildcard /usr/include/crypt.h),)
+#ifneq ($(wildcard /usr/include/crypt.h),)
 CFLAGS  += -DHAVE_CRYPT_H=1
 LIBS	+= -lcrypt
-endif
+#endif
 
 ifdef NEEDDES
 ifndef USE_CRYPT
diff -urB ppp.orig/pppd/plugins/Makefile.linux ppp/pppd/plugins/Makefile.linux
--- ppp.orig/pppd/plugins/Makefile.linux	2008-11-26 13:02:29.000000000 +0300
+++ ppp/pppd/plugins/Makefile.linux	2009-02-09 18:10:50.000000000 +0300
@@ -1,6 +1,6 @@
 #CC	= gcc
 COPTS	= -O2 -g
-CFLAGS	= $(COPTS) -I.. -I../../include -fPIC
+CFLAGS	= $(COPTS) -I.. -I../../include -fPIC $(EXTRACFLAGS)
 LDFLAGS	= -shared
 INSTALL	= install
 
diff -uBp a/pppd/plugins/pppol2tp/Makefile.linux b/pppd/plugins/pppol2tp/Makefile.linux
--- a/pppd/plugins/pppol2tp/Makefile.linux	2009-10-08 10:00:57.000000000 +0400
+++ b/pppd/plugins/pppol2tp/Makefile.linux	2010-11-22 21:56:19.000000000 +0300
@@ -1,6 +1,6 @@
 #CC	= gcc
 COPTS	= -O2 -g
-CFLAGS	= $(COPTS) -I. -I../.. -I../../../include -fPIC
+CFLAGS	= $(COPTS) -I. -I../.. -I$(LINUXDIR)/include -fPIC $(EXTRACFLAGS)
 LDFLAGS	= -shared
 INSTALL	= install
 
diff -uBp a/pppd/plugins/radius/Makefile.linux b/pppd/plugins/radius/Makefile.linux
--- a/pppd/plugins/radius/Makefile.linux	2009-11-17 01:26:07.000000000 +0300
+++ b/pppd/plugins/radius/Makefile.linux	2010-11-22 21:54:24.000000000 +0300
@@ -12,7 +12,7 @@ VERSION = $(shell awk -F '"' '/VERSION/
 INSTALL	= install
 
 PLUGIN=radius.so radattr.so radrealms.so
-CFLAGS=-I. -I../.. -I../../../include -O2 -fPIC -DRC_LOG_FACILITY=LOG_DAEMON
+CFLAGS=-I. -I../.. -I../../../include -O2 -fPIC -DRC_LOG_FACILITY=LOG_DAEMON $(EXTRACFLAGS)
 
 # Uncomment the next line to include support for Microsoft's
 # MS-CHAP authentication protocol.
diff -urBp a/pppd/plugins/rp-pppoe/Makefile.linux b/pppd/plugins/rp-pppoe/Makefile.linux
--- a/pppd/plugins/rp-pppoe/Makefile.linux	2009-02-09 17:22:47.000000000 +0300
+++ b/pppd/plugins/rp-pppoe/Makefile.linux	2010-11-22 21:53:54.000000000 +0300
@@ -26,7 +26,7 @@ INSTALL	= install
 RP_VERSION=3.10p
 
 COPTS=-O2 -g
-CFLAGS=$(COPTS) -I../../../include '-DRP_VERSION="$(RP_VERSION)"'
+CFLAGS=$(COPTS) -I../../../include '-DRP_VERSION="$(RP_VERSION)"' $(EXTRACFLAGS)
 all: rp-pppoe.so pppoe-discovery
 
 pppoe-discovery: pppoe-discovery.o debug.o
