diff -ur ppp-2.4.2/pppd/auth.c pppd/pppd/auth.c
--- ppp-2.4.2/pppd/auth.c	2003-06-12 03:56:26.000000000 +0400
+++ pppd/pppd/auth.c	2005-01-12 22:18:24.637707136 +0300
@@ -241,6 +241,7 @@
 bool allow_any_ip = 0;		/* Allow peer to use any IP address */
 bool explicit_remote = 0;	/* User specified explicit remote name */
 char remote_name[MAXNAMELEN];	/* Peer's name for authentication */
+bool tx_only;			/* JYWeng 20031216: idle time counting on tx traffic */
 
 static char *uafname;		/* name of most recent +ua file */
 
@@ -1050,6 +1051,10 @@
     if (idle_time_hook != 0) {
 	tlim = idle_time_hook(&idle);
     } else {
+/* JYWeng 20031216: replace itime with idle.xmit_idle for only outgoing traffic is counted*/
+	if(tx_only) 
+		itime = idle.xmit_idle;
+	else
 	itime = MIN(idle.xmit_idle, idle.recv_idle);
 	tlim = idle_time_limit - itime;
     }
diff -ur ppp-2.4.2/pppd/fsm.c pppd/pppd/fsm.c
--- ppp-2.4.2/pppd/fsm.c	2003-06-29 14:06:14.000000000 +0400
+++ pppd/pppd/fsm.c	2005-01-12 22:18:24.639706832 +0300
@@ -70,6 +70,8 @@
 
 int peer_mru[NUM_PPP];
 
+/* JYWeng 20031216: add to wanstatus.log */
+void saveWANStatus(char *currentstatus, int statusindex);
 
 /*
  * fsm_init - Initialize fsm.
@@ -280,6 +282,10 @@
     case ACKRCVD:
     case ACKSENT:
 	if (f->retransmits <= 0) {
+/* JYWeng 20031216: add to wanstatus.log */
+	    int statusindex=0; 
+	    saveWANStatus("No response from ISP.", statusindex);
+/* JYWeng 20031216: add to wanstatus.log */
 	    warn("%s: timeout sending Config-Requests\n", PROTO_NAME(f));
 	    f->state = STOPPED;
 	    if( (f->flags & OPT_PASSIVE) == 0 && f->callbacks->finished )
diff -ur ppp-2.4.2/pppd/lcp.c pppd/pppd/lcp.c
--- ppp-2.4.2/pppd/lcp.c	2003-11-18 13:42:56.000000000 +0300
+++ pppd/pppd/lcp.c	2005-01-12 22:18:24.687699536 +0300
@@ -68,6 +68,9 @@
 
 static void lcp_delayed_up __P((void *));
 
+/* JYWeng 20031216: add to wanstatus.log */
+void saveWANStatus(char *currentstatus, int statusindex);
+
 /*
  * LCP-related command-line options.
  */
@@ -400,6 +403,7 @@
     char *reason;
 {
     fsm *f = &lcp_fsm[unit];
+    int statusindex = 0;/* JYWeng 20031216: add to wanstatus.log */
 
     if (phase != PHASE_DEAD)
 	new_phase(PHASE_TERMINATE);
@@ -415,6 +419,16 @@
 
     } else
 	fsm_close(f, reason);
+/* JYWeng 20031216: add to wanstatus.log */
+    if(strstr(reason, "Link inactive")) {
+	    statusindex = 1;
+	    saveWANStatus("Terminating connection due to lack of activity.", statusindex);
+    }
+    else {
+	    statusindex = 2;
+	    saveWANStatus(reason, statusindex);
+    }
+/* JYWeng 20031216: add to wanstatus.log */
 }
 
 
diff -ur ppp-2.4.2/pppd/options.c pppd/pppd/options.c
--- ppp-2.4.2/pppd/options.c	2004-01-13 07:02:07.000000000 +0300
+++ pppd/pppd/options.c	2005-01-12 22:18:24.690699080 +0300
@@ -65,6 +65,7 @@
 #if defined(ultrix) || defined(NeXT)
 char *strdup __P((char *));
 #endif
+bool tx_only;			/* JYWeng 20031216: idle time counting on tx traffic */
 
 static const char rcsid[] = RCSID;
 
@@ -309,6 +310,10 @@
       "Check for traffic limit every N seconds", OPT_PRIO | OPT_LLIMIT | 1 },
 #endif
 
+/* JYWeng 20031216: add for tx_only option*/
+    { "tx_only", o_bool, &tx_only,
+      "set idle time counting on tx_only or not", 1 },
+
     { NULL }
 };
 
diff -ur ppp-2.4.2/pppd/pathnames.h pppd/pppd/pathnames.h
--- ppp-2.4.2/pppd/pathnames.h	2002-11-09 14:24:42.000000000 +0300
+++ pppd/pppd/pathnames.h	2005-01-12 22:18:02.074137320 +0300
@@ -9,39 +9,39 @@
 
 #else /* HAVE_PATHS_H */
 #ifndef _PATH_VARRUN
-#define _PATH_VARRUN 	"/etc/ppp/"
+#define _PATH_VARRUN 	"/var/run/"
 #endif
 #define _PATH_DEVNULL	"/dev/null"
 #endif /* HAVE_PATHS_H */
 
 #ifndef _ROOT_PATH
-#define _ROOT_PATH
+#define _ROOT_PATH	"/tmp"
 #endif
 
-#define _PATH_UPAPFILE 	 _ROOT_PATH "/etc/ppp/pap-secrets"
-#define _PATH_CHAPFILE 	 _ROOT_PATH "/etc/ppp/chap-secrets"
-#define _PATH_SRPFILE 	 _ROOT_PATH "/etc/ppp/srp-secrets"
-#define _PATH_SYSOPTIONS _ROOT_PATH "/etc/ppp/options"
-#define _PATH_IPUP	 _ROOT_PATH "/etc/ppp/ip-up"
-#define _PATH_IPDOWN	 _ROOT_PATH "/etc/ppp/ip-down"
-#define _PATH_AUTHUP	 _ROOT_PATH "/etc/ppp/auth-up"
-#define _PATH_AUTHDOWN	 _ROOT_PATH "/etc/ppp/auth-down"
-#define _PATH_TTYOPT	 _ROOT_PATH "/etc/ppp/options."
-#define _PATH_CONNERRS	 _ROOT_PATH "/etc/ppp/connect-errors"
-#define _PATH_PEERFILES	 _ROOT_PATH "/etc/ppp/peers/"
-#define _PATH_RESOLV	 _ROOT_PATH "/etc/ppp/resolv.conf"
+#define _PATH_UPAPFILE 	 _ROOT_PATH "/ppp/pap-secrets"
+#define _PATH_CHAPFILE 	 _ROOT_PATH "/ppp/chap-secrets"
+#define _PATH_SRPFILE 	 _ROOT_PATH "/ppp/srp-secrets"
+#define _PATH_SYSOPTIONS _ROOT_PATH "/ppp/options"
+#define _PATH_IPUP	 _ROOT_PATH "/ppp/ip-up"
+#define _PATH_IPDOWN	 _ROOT_PATH "/ppp/ip-down"
+#define _PATH_AUTHUP	 _ROOT_PATH "/ppp/auth-up"
+#define _PATH_AUTHDOWN	 _ROOT_PATH "/ppp/auth-down"
+#define _PATH_TTYOPT	 _ROOT_PATH "/ppp/options."
+#define _PATH_CONNERRS	 "/var/log/ppp-connect-errors"
+#define _PATH_PEERFILES	 _ROOT_PATH "/ppp/peers/"
+#define _PATH_RESOLV	 _ROOT_PATH "/ppp/resolv.conf"
 
 #define _PATH_USEROPT	 ".ppprc"
 #define	_PATH_PSEUDONYM	 ".ppp_pseudonym"
 
 #ifdef INET6
-#define _PATH_IPV6UP     _ROOT_PATH "/etc/ppp/ipv6-up"
-#define _PATH_IPV6DOWN   _ROOT_PATH "/etc/ppp/ipv6-down"
+#define _PATH_IPV6UP     _ROOT_PATH "/ppp/ipv6-up"
+#define _PATH_IPV6DOWN   _ROOT_PATH "/ppp/ipv6-down"
 #endif
 
 #ifdef IPX_CHANGE
-#define _PATH_IPXUP	 _ROOT_PATH "/etc/ppp/ipx-up"
-#define _PATH_IPXDOWN	 _ROOT_PATH "/etc/ppp/ipx-down"
+#define _PATH_IPXUP	 _ROOT_PATH "/ppp/ipx-up"
+#define _PATH_IPXDOWN	 _ROOT_PATH "/ppp/ipx-down"
 #endif /* IPX_CHANGE */
 
 #ifdef __STDC__
@@ -50,10 +50,10 @@
 #ifdef HAVE_PATHS_H
 #define _PATH_PPPDB	"/var/run/pppd.tdb"
 #else
-#define _PATH_PPPDB	"/etc/ppp/pppd.tdb"
+#define _PATH_PPPDB	"/var/run/pppd.tdb"
 #endif
 #endif /* __STDC__ */
 
 #ifdef PLUGIN
-#define _PATH_PLUGIN	"/usr/lib/pppd/" VERSION
+#define _PATH_PLUGIN	"/usr/lib/pppd"
 #endif /* PLUGIN */
diff -urbB ppp-2.4.2/chat/Makefile.linux pppd/chat/Makefile.linux
--- ppp-2.4.2/chat/Makefile.linux	2004-01-13 06:57:55.000000000 +0300
+++ pppd/chat/Makefile.linux	2008-12-23 12:34:42.000000000 +0300
@@ -7,7 +7,7 @@
 CDEFS=	$(CDEF1) $(CDEF2) $(CDEF3) $(CDEF4)
 
 COPTS=	-O2 -g -pipe $(CDEFS)
-CFLAGS=	$(COPTS) $(CDEFS)
+CFLAGS=	$(COPTS) $(CDEFS) $(EXTRACFLAGS)
 
 INSTALL= install
 
diff -ur ppp-2.4.2/pppd/plugins/Makefile.linux pppd/pppd/plugins/Makefile.linux
--- ppp-2.4.2/pppd/plugins/Makefile.linux	2004-01-13 06:56:24.000000000 +0300
+++ pppd/pppd/plugins/Makefile.linux	2008-12-23 12:29:44.000000000 +0300
@@ -1,6 +1,6 @@
-CC	= gcc
+#CC	= gcc
 COPTS	= -O2 -g
-CFLAGS	= $(COPTS) -I.. -I../../include -fPIC
+CFLAGS	= $(COPTS) -I.. -I../../include -fPIC $(EXTRACFLAGS)
 LDFLAGS	= -shared
 INSTALL	= install -o root
 
diff -ur ppp-2.4.2/pppd/plugins/rp-pppoe/Makefile.linux pppd/pppd/plugins/rp-pppoe/Makefile.linux
--- ppp-2.4.2/pppd/plugins/rp-pppoe/Makefile.linux	2004-01-13 06:57:55.000000000 +0300
+++ pppd/pppd/plugins/rp-pppoe/Makefile.linux	2005-01-12 22:15:04.761093000 +0300
@@ -21,7 +21,7 @@
 all: rp-pppoe.so
 
 rp-pppoe.so: libplugin.a plugin.o
-	gcc -o rp-pppoe.so -shared plugin.o libplugin.a
+	$(LD) -o rp-pppoe.so -shared plugin.o libplugin.a
 
 install: all
 	$(INSTALL) -d -m 755 $(LIBDIR)
@@ -31,20 +31,20 @@
 	rm -f *.o *.so
 
 plugin.o: plugin.c
-	gcc '-DRP_VERSION="$(VERSION)"' $(CFLAGS) -I../../..  -c -o plugin.o -fPIC plugin.c
+	$(CC) '-DRP_VERSION="$(VERSION)"' $(CFLAGS) -I../../..  -c -o plugin.o -fPIC plugin.c
 
 libplugin.a: discovery.o if.o common.o debug.o
-	ar -rc $@ $^
+	$(AR) -rc $@ $^
 
 discovery.o: discovery.c
-	gcc $(CFLAGS) '-DVERSION="$(VERSION)"' -c -o discovery.o -fPIC discovery.c
+	$(CC) $(CFLAGS) '-DVERSION="$(VERSION)"' -c -o discovery.o -fPIC discovery.c
 
 if.o: if.c
-	gcc $(CFLAGS) '-DVERSION="$(VERSION)"' -c -o if.o -fPIC if.c
+	$(CC) $(CFLAGS) '-DVERSION="$(VERSION)"' -c -o if.o -fPIC if.c
 
 debug.o: debug.c
-	gcc $(CFLAGS) '-DVERSION="$(VERSION)"' -c -o debug.o -fPIC debug.c
+	$(CC) $(CFLAGS) '-DVERSION="$(VERSION)"' -c -o debug.o -fPIC debug.c
 
 common.o: common.c
-	gcc $(CFLAGS) '-DVERSION="$(VERSION)"' -c -o common.o -fPIC common.c
+	$(CC) $(CFLAGS) '-DVERSION="$(VERSION)"' -c -o common.o -fPIC common.c
 
diff -ur ppp-2.4.2/pppd/plugins/rp-pppoe/plugin.c pppd/pppd/plugins/rp-pppoe/plugin.c
--- ppp-2.4.2/pppd/plugins/rp-pppoe/plugin.c	2004-01-13 07:03:58.000000000 +0300
+++ pppd/pppd/plugins/rp-pppoe/plugin.c	2005-01-12 22:18:24.711695888 +0300
@@ -52,7 +52,7 @@
 #include "if_ppp.h"
 #include "if_pppox.h"
 
-#define _PATH_ETHOPT         _ROOT_PATH "/etc/ppp/options."
+#define _PATH_ETHOPT         _ROOT_PATH "/ppp/options."
 
 char pppd_version[] = VERSION;
 
@@ -191,7 +191,8 @@
     struct ifreq ifr;
 
     if (mtu > MAX_PPPOE_MTU) {
-	warn("Couldn't increase MTU to %d", mtu);
+    	if (mtu != PPP_MTU)
+	    warn("Couldn't increase MTU to %d", mtu);
 	mtu = MAX_PPPOE_MTU;
     }
     sock = socket(AF_INET, SOCK_DGRAM, 0);
@@ -215,7 +216,7 @@
 		int pcomp,
 		int accomp)
 {
-    if (mru > MAX_PPPOE_MTU)
+    if (mru > MAX_PPPOE_MTU && mru != PPP_MRU)
 	warn("Couldn't increase MRU to %d", mru);
 }
 
diff -ur ppp-2.4.2/pppd/pppd.h pppd/pppd/pppd.h
--- ppp-2.4.2/pppd/pppd.h	2003-04-07 04:01:46.000000000 +0400
+++ pppd/pppd/pppd.h	2005-01-12 22:18:24.732692696 +0300
@@ -209,6 +209,7 @@
 /*
  * Global variables.
  */
+extern bool tx_only;			/* JYWeng 20031216: idle time counting on tx traffic */
 
 extern int	hungup;		/* Physical layer has disconnected */
 extern int	ifunit;		/* Interface unit number */
diff -ur ppp-2.4.2/pppd/utils.c pppd/pppd/utils.c
--- ppp-2.4.2/pppd/utils.c	2004-01-13 07:13:08.000000000 +0300
+++ pppd/pppd/utils.c	2005-01-12 22:18:24.734692392 +0300
@@ -1057,3 +1057,23 @@
     }
 }
 
+/* JYWeng 20031216: add to wanstatus.log */
+
+void saveWANStatus(char *currentstatus, int statusindex)
+{
+	FILE *STATUSFILE;
+#ifdef ONWL500G_SHELL
+	if ((req_unit == 0) && (STATUSFILE = fopen("/etc/linuxigd/wanstatus.log", "w"))!=NULL)
+	{
+		fprintf(STATUSFILE, "StatusCode=\"%d\"\n", statusindex);
+		fprintf(STATUSFILE, "StatusReason=\"%s\"\n", currentstatus);
+		fclose(STATUSFILE);
+	}
+#else
+	if ((req_unit == 0) && (STATUSFILE = fopen("/tmp/wanstatus.log", "w"))!=NULL)
+	{
+		fprintf(STATUSFILE, "%d,%s\n", statusindex, currentstatus);
+		fclose(STATUSFILE);
+	}
+#endif
+}
--- ppp-2.4.2/pppd/Makefile.linux	2003-11-28 00:55:19.000000000 +0300
+++ pppd/pppd/Makefile.linux	2008-12-23 12:14:09.588515312 +0300
@@ -47,19 +47,19 @@
 # Uncomment the next line to include support for PPP packet filtering.
 # This requires that the libpcap library and headers be installed
 # and that the kernel driver support PPP packet filtering.
-FILTER=y
+#FILTER=y
 
 # Uncomment the next line to enable multilink PPP (enabled by default)
 # Linux distributions: Please leave multilink ENABLED in your builds
 # of pppd!
-HAVE_MULTILINK=y
+#HAVE_MULTILINK=y
 
 # Uncomment the next line to enable the TDB database (enabled by default.)
 # If you enable multilink, then TDB is automatically enabled also.
 # Linux distributions: Please leave TDB ENABLED in your builds.
-USE_TDB=y
+#USE_TDB=y
 
-HAS_SHADOW=y
+#HAS_SHADOW=y
 #USE_PAM=y
 #HAVE_INET6=y
 
@@ -72,13 +72,13 @@
 # Enable EAP SRP-SHA1 authentication (requires libsrp)
 #USE_SRP=y
 
-MAXOCTETS=y
+#MAXOCTETS=y
 
 INCLUDE_DIRS= -I../include
 
-COMPILE_FLAGS= -DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MMAP
+COMPILE_FLAGS= #-DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MMAP
 
-CFLAGS= $(COPTS) $(COMPILE_FLAGS) $(INCLUDE_DIRS)
+CFLAGS= $(COPTS) $(COMPILE_FLAGS) $(INCLUDE_DIRS) $(EXTRACFLAGS)
 
 ifdef CHAPMS
 CFLAGS   += -DCHAPMS=1
@@ -116,12 +116,12 @@
 #LIBS     += -lshadow $(LIBS)
 endif
 
-ifneq ($(wildcard /usr/include/crypt.h),)
+#ifneq ($(wildcard /usr/include/crypt.h),)
 CFLAGS   += -DHAVE_CRYPT_H=1
-endif
-ifneq ($(wildcard /usr/lib/libcrypt.*),)
+#endif
+#ifneq ($(wildcard /usr/lib/libcrypt.*),)
 LIBS	+= -lcrypt
-endif
+#endif
 
 ifdef NEEDDES
 ifndef USE_CRYPT
