From f5d4bddfbee4d9a5af479e20f4e17b5e2fb4a3ed Mon Sep 17 00:00:00 2001
From: Denys Vlasenko <vda.linux@googlemail.com>
Date: Sat, 27 Jun 2009 00:24:35 +0200
Subject: [PATCH] trylink: don't use ld --gc-sections if ld doesn't support it

Signed-off-by: Ian Abbott <abbotti@mev.co.uk>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
Signed-off-by: Mike Frysinger <vapier@gentoo.org>
---
 scripts/trylink |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/scripts/trylink b/scripts/trylink
index 7ea1d5c..8c88546 100755
--- a/scripts/trylink
+++ b/scripts/trylink
@@ -99,6 +99,11 @@ else
 fi
 )`
 
+# The --gc-sections option is not supported by older versions of ld
+if test -n "$GC_SECTIONS"; then
+    GC_SECTIONS=`check_cc "$GC_SECTIONS" ""`
+fi
+
 # Sanitize lib list (dups, extra spaces etc)
 LDLIBS=`echo "$LDLIBS" | xargs -n1 | sort | uniq | xargs`
 
-- 
1.6.3.3

