From 5b3430017895001e134b6c37a0ea4d2c082ba209
From: Vladimir Dronnikov
Date: Mon, 04 Oct 2010 23:21:32 +0000
Subject: sendmail: fix assumption that headers should have a space after the colon

Signed-off-by: Vladimir Dronnikov <dronnikov@gmail.com>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/mailutils/sendmail.c b/mailutils/sendmail.c
index 59591ee..ec97cf8 100644
--- a/mailutils/sendmail.c
+++ b/mailutils/sendmail.c
@@ -221,12 +221,12 @@ int sendmail_main(int argc UNUSED_PARAM, char **argv)
 
 		// analyze headers
 		// To: or Cc: headers add recipients
-		if (0 == strncasecmp("To: ", s, 4) || 0 == strncasecmp("Bcc: " + 1, s, 4)) {
-			rcptto(sane_address(s+4));
+		if (0 == strncasecmp("To:", s, 3) || 0 == strncasecmp("Bcc:" + 1, s, 3)) {
+			rcptto(sane_address(s+3));
 			goto addheader;
 		// Bcc: header adds blind copy (hidden) recipient
-		} else if (0 == strncasecmp("Bcc: ", s, 5)) {
-			rcptto(sane_address(s+5));
+		} else if (0 == strncasecmp("Bcc:", s, 4)) {
+			rcptto(sane_address(s+4));
 			free(s);
 			// N.B. Bcc: vanishes from headers!
 
--
cgit v0.8.2.1
