From 786635e62e4c8a8646d01ede1c7bb98ce604c7fa
From: Denys Vlasenko
Date: Sun, 17 Oct 2010 10:44:39 +0000
Subject: seq: exit on write errors

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/coreutils/seq.c b/coreutils/seq.c
index 8be2536..22bf3ec 100644
--- a/coreutils/seq.c
+++ b/coreutils/seq.c
@@ -86,7 +86,8 @@ int seq_main(int argc, char **argv)
 	v = first;
 	n = 0;
 	while (increment >= 0 ? v <= last : v >= last) {
-		printf("%s%0*.*f", sep, width, frac_part, v);
+		if (printf("%s%0*.*f", sep, width, frac_part, v) < 0)
+			break; /* I/O error, bail out (yes, this really happens) */
 		sep = opt_s;
 		/* v += increment; - would accumulate floating point errors */
 		n++;
--
cgit v0.8.2.1
