From 3205c24a822c61d74b0f58090606433c52bc4796
From: Denys Vlasenko
Date: Sun, 07 Feb 2010 22:28:23 +0000
Subject: shaNNNsum: accept and ignore -b and -t

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/coreutils/md5_sha1_sum.c b/coreutils/md5_sha1_sum.c
index a64026d..3d50bb0 100644
--- a/coreutils/md5_sha1_sum.c
+++ b/coreutils/md5_sha1_sum.c
@@ -101,8 +101,10 @@ int md5_sha1_sum_main(int argc UNUSED_PARAM, char **argv)
 	unsigned flags;
 	/*hash_algo_t hash_algo = applet_name[3];*/
 
-	if (ENABLE_FEATURE_MD5_SHA1_SUM_CHECK)
-		flags = getopt32(argv, "scw");
+	if (ENABLE_FEATURE_MD5_SHA1_SUM_CHECK) {
+		/* -b "binary", -t "text" are ignored (shaNNNsum compat) */
+		flags = getopt32(argv, "scwbt");
+	}
 	else optind = 1;
 	argv += optind;
 	//argc -= optind;
--
cgit v0.8.2.1
