From 5a163b26451c591187482f99659e5fe639a0616a
From: Denys Vlasenko
Date: Tue, 11 Jan 2011 12:07:22 +0000
Subject: lsmod: fox unsafe usage of strlen() - 1

Signed-off-by: Denys Vlasenko <dvlasenk@redhat.com>
---
diff --git a/modutils/lsmod.c b/modutils/lsmod.c
index ab7c11f..d7e1668 100644
--- a/modutils/lsmod.c
+++ b/modutils/lsmod.c
@@ -87,7 +87,8 @@ int lsmod_main(int argc UNUSED_PARAM, char **argv UNUSED_PARAM)
 			// N.B. token[3] is either '-' (module is not used by others)
 			// or comma-separated list ended by comma
 			// so trimming the trailing char is just what we need!
-			token[3][strlen(token[3])-1] = '\0';
+			if (token[3][0])
+				token[3][strlen(token[3]) - 1] = '\0';
 # if ENABLE_UNICODE_SUPPORT
 			{
 				uni_stat_t uni_stat;
--
cgit v0.8.2.1
