diff -BurpN busybox-1.17.1/Makefile busybox/Makefile
--- busybox-1.17.1/Makefile	2010-07-24 22:13:44.000000000 +0000
+++ busybox/Makefile	2010-07-25 18:29:20.000000000 +0000
@@ -315,7 +315,7 @@ AFLAGS_KERNEL	=
 
 # Use LINUXINCLUDE when you must reference the include/ directory.
 # Needed to be compatible with the O= option
-CFLAGS		:= $(CFLAGS)
+CFLAGS		:= $(CFLAGS) $(EXTRA_CFLAGS)
 # Added only to final link stage of busybox binary
 CFLAGS_busybox	:= $(CFLAGS_busybox)
 CPPFLAGS	:= $(CPPFLAGS)
@@ -970,6 +970,11 @@ MRPROPER_FILES += .config .config.old include/asm 
 		  include/usage_compressed.h \
 		  include/applet_tables.h \
 		  include/applets.h \
+		  docs/busybox.1 \
+		  docs/BusyBox.html \
+		  docs/busybox.net/BusyBox.html \
+		  docs/busybox.pod \
+		  docs/BusyBox.txt \
 		  include/usage.h \
 		  applets/usage \
 		  .kernelrelease Module.symvers tags TAGS cscope* \
diff -BurpN busybox-1.17.1/include/applets.src.h busybox/include/applets.src.h
--- busybox-1.17.1/include/applets.src.h	2010-07-25 18:32:06.000000000 +0000
+++ busybox/include/applets.src.h	2010-07-25 18:32:16.000000000 +0000
@@ -83,6 +83,7 @@ IF_BEEP(APPLET(beep, BB_DIR_USR_BIN, BB_
 IF_BEEP(APPLET(beep, BB_DIR_USR_BIN, BB_SUID_DROP))
 IF_BLKID(APPLET(blkid, BB_DIR_SBIN, BB_SUID_DROP))
 IF_BRCTL(APPLET(brctl, BB_DIR_USR_SBIN, BB_SUID_DROP))
+IF_HTTPD(APPLET_ODDNAME(busybox_httpd, httpd, BB_DIR_USR_SBIN, BB_SUID_DROP, httpd))
 IF_BZIP2(APPLET(bzip2, BB_DIR_USR_BIN, BB_SUID_DROP))
 IF_CAL(APPLET(cal, BB_DIR_USR_BIN, BB_SUID_DROP))
 IF_CAT(APPLET_NOFORK(cat, cat, BB_DIR_BIN, BB_SUID_DROP, cat))
@@ -182,7 +183,6 @@ IF_HEAD(APPLET(head, BB_DIR_USR_BIN, BB
 IF_HEXDUMP(APPLET_NOEXEC(hexdump, hexdump, BB_DIR_USR_BIN, BB_SUID_DROP, hexdump))
 IF_HOSTID(APPLET_NOFORK(hostid, hostid, BB_DIR_USR_BIN, BB_SUID_DROP, hostid))
 IF_HOSTNAME(APPLET(hostname, BB_DIR_BIN, BB_SUID_DROP))
-IF_HTTPD(APPLET(httpd, BB_DIR_USR_SBIN, BB_SUID_DROP))
 IF_HWCLOCK(APPLET(hwclock, BB_DIR_SBIN, BB_SUID_DROP))
 IF_ID(APPLET_NOEXEC(id, id, BB_DIR_USR_BIN, BB_SUID_DROP, id))
 IF_IFCONFIG(APPLET(ifconfig, BB_DIR_SBIN, BB_SUID_DROP))
diff -BurpN busybox-1.17.1/include/libbb.h busybox/include/libbb.h
--- busybox-1.17.1/include/libbb.h	2010-07-24 22:12:43.000000000 +0000
+++ busybox/include/libbb.h	2010-07-25 18:29:36.000000000 +0000
@@ -1577,9 +1577,9 @@ extern const char bb_dev_null[];
 extern const char bb_busybox_exec_path[];
 /* util-linux manpage says /sbin:/bin:/usr/sbin:/usr/bin,
  * but I want to save a few bytes here */
-extern const char bb_PATH_root_path[] ALIGN1; /* "PATH=/sbin:/usr/sbin:/bin:/usr/bin" */
+extern const char bb_PATH_root_path[] ALIGN1; /* "PATH=/opt/sbin:/sbin:/usr/sbin:/opt/bin:/usr/local/bin:/bin:/usr/bin" */
 #define bb_default_root_path (bb_PATH_root_path + sizeof("PATH"))
-#define bb_default_path      (bb_PATH_root_path + sizeof("PATH=/sbin:/usr/sbin"))
+#define bb_default_path      (bb_PATH_root_path + sizeof("PATH=/opt/sbin:/sbin:/usr/sbin"))
 
 extern const int const_int_0;
 extern const int const_int_1;
diff -BurpN busybox-1.17.1/libbb/messages.c busybox/libbb/messages.c
--- busybox-1.17.1/libbb/messages.c	2010-07-06 02:25:54.000000000 +0000
+++ busybox/libbb/messages.c	2010-07-25 18:29:36.000000000 +0000
@@ -49,7 +49,7 @@ const char bb_default_login_shell[] ALIG
 /* util-linux manpage says /sbin:/bin:/usr/sbin:/usr/bin,
  * but I want to save a few bytes here. Check libbb.h before changing! */
 const char bb_PATH_root_path[] ALIGN1 =
-	"PATH=/sbin:/usr/sbin:/bin:/usr/bin" BB_ADDITIONAL_PATH;
+	"PATH=/opt/sbin:/sbin:/usr/sbin:/opt/bin:/usr/local/bin:/bin:/usr/bin" BB_ADDITIONAL_PATH;
 
 
 const int const_int_1 = 1;
diff -BurpN busybox-1.17.1/scripts/Makefile.lib busybox/scripts/Makefile.lib
--- busybox-1.17.1/scripts/Makefile.lib	2010-07-06 02:25:54.000000000 +0000
+++ busybox/scripts/Makefile.lib	2010-07-25 18:29:36.000000000 +0000
@@ -119,9 +119,9 @@ cpp_flags      = -Wp,-MD,$(depfile) $(NO
 
 # Seems to be a wrong thing to do. LDFLAGS contains gcc's flags,
 # yet ld_flags is fed to ld.
-#ld_flags       = $(LDFLAGS) $(EXTRA_LDFLAGS)
+ld_flags       = $(LDFLAGS) $(EXTRA_LDFLAGS)
 # Remove the -Wl, prefix from linker options normally passed through gcc
-ld_flags       = $(filter-out -Wl$(comma)%,$(LDFLAGS) $(EXTRA_LDFLAGS))
+#ld_flags       = $(filter-out -Wl$(comma)%,$(LDFLAGS) $(EXTRA_LDFLAGS))
 
 
 # Finds the multi-part object the current object will be linked into
