--- busybox-1.15.2/libbb/getopt32.c	2009-10-08 04:59:09.000000000 +0400
+++ busybox/libbb/getopt32.c	2009-11-20 18:16:50.000000000 +0300
@@ -536,6 +536,12 @@ getopt32(char **argv, const char *applet
 	/* optreset = 1; */
 #endif
 	/* optarg = NULL; opterr = 0; optopt = 0; - do we need this?? */
+#ifdef __UCLIBC__
+	/* Ugly hack for applets with broken argv[0] - uClibc segfault
+	 * on invalid pointer in getopt() in case of wrong commandline
+	 */
+	argv[0] = applet_name;
+#endif
 
 	pargv = NULL;
 
