--- busybox/networking/httpd.c.orig	2008-11-09 20:27:59.000000000 +0300
+++ busybox/networking/httpd.c	2009-02-07 12:02:05.000000000 +0300
@@ -253,6 +253,7 @@ struct globals {
 	USE_FEATURE_HTTPD_BASIC_AUTH(const char *g_realm;)
 	USE_FEATURE_HTTPD_BASIC_AUTH(char *remoteuser;)
 	USE_FEATURE_HTTPD_CGI(char *referer;)
+	USE_FEATURE_HTTPD_CGI(char *host;)
 	USE_FEATURE_HTTPD_CGI(char *user_agent;)
 	USE_FEATURE_HTTPD_CGI(char *http_accept;)
 	USE_FEATURE_HTTPD_CGI(char *http_accept_language;)
@@ -300,6 +301,7 @@ struct globals {
 #define g_realm           (G.g_realm          )
 #define remoteuser        (G.remoteuser       )
 #define referer           (G.referer          )
+#define g_host            (G.host             )
 #define user_agent        (G.user_agent       )
 #define http_accept       (G.http_accept      )
 #define http_accept_language (G.http_accept_language)
@@ -1406,6 +1408,8 @@ static void send_cgi_and_exit(
 #endif
 	if (referer)
 		setenv1("HTTP_REFERER", referer);
+	if (g_host)
+		setenv1("HTTP_HOST", g_host);
 
 	xpiped_pair(fromCgi);
 	xpiped_pair(toCgi);
@@ -2011,6 +2015,8 @@ static void handle_incoming_and_exit(con
 				content_type = xstrdup(skip_whitespace(iobuf + sizeof("Content-Type:")-1));
 			} else if (STRNCASECMP(iobuf, "Referer:") == 0) {
 				referer = xstrdup(skip_whitespace(iobuf + sizeof("Referer:")-1));
+			} else if (STRNCASECMP(iobuf, "Host:") == 0) {
+				g_host = xstrdup(skip_whitespace(iobuf + sizeof("Host:")-1));
 			} else if (STRNCASECMP(iobuf, "User-Agent:") == 0) {
 				user_agent = xstrdup(skip_whitespace(iobuf + sizeof("User-Agent:")-1));
 			} else if (STRNCASECMP(iobuf, "Accept:") == 0) {
