--- busybox/networking/udhcp/clientpacket.c.orig	2008-11-09 20:27:58.000000000 +0300
+++ busybox/networking/udhcp/clientpacket.c	2009-02-07 11:38:33.000000000 +0300
@@ -48,7 +48,7 @@ static void init_packet(struct dhcpMessa
 		add_option_string(packet->options, client_config.hostname);
 	if (client_config.fqdn)
 		add_option_string(packet->options, client_config.fqdn);
-	if ((type != DHCPDECLINE) && (type != DHCPRELEASE))
+	if ((type != DHCPDECLINE) && (type != DHCPRELEASE) && (client_config.vendorclass))
 		add_option_string(packet->options, client_config.vendorclass);
 }
 
