diff -BurpN busybox-1.14.2/e2fsprogs/e2fsck.c busybox/e2fsprogs/e2fsck.c
--- busybox-1.14.2/e2fsprogs/e2fsck.c	2009-07-22 15:37:26.000000000 +0000
+++ busybox/e2fsprogs/e2fsck.c	2009-07-22 15:38:10.000000000 +0000
@@ -12732,7 +12732,8 @@ static void check_if_skip(e2fsck_t ctx)
 		if (batt && (fs->super->s_mnt_count <
 			     (unsigned) fs->super->s_max_mnt_count*2))
 			reason = 0;
-	} else if (fs->super->s_checkinterval &&
+	} else if (!(ctx->options & E2F_OPT_PREEN) &&
+		   fs->super->s_checkinterval &&
 		   ((now - fs->super->s_lastcheck) >=
 		    fs->super->s_checkinterval)) {
 		reason = _(" has gone %u days without being checked");
