diff -urBp igmpproxy-0.1_beta5/src/config.c igmpproxy/src/config.c
--- igmpproxy-0.1_beta5/src/config.c	2009-08-16 19:46:04.000000000 +0400
+++ igmpproxy/src/config.c	2009-09-15 11:17:41.000000000 +0400
@@ -336,15 +336,18 @@ struct SubnetList *parseSubnetAddress(ch
     tmpStr = strtok(NULL, "/");
     if(tmpStr != NULL) {
         int bitcnt = atoi(tmpStr);
-        if(bitcnt <= 0 || bitcnt > 32) {
+        if(bitcnt < 0 || bitcnt > 32) {
             my_log(LOG_WARNING, 0, "The bits part of the address is invalid : %d.",tmpStr);
             return NULL;
         }
 
-        mask <<= (32 - bitcnt);
+	if (bitcnt == 0)
+	    mask = 0;
+	else
+    	    mask <<= (32 - bitcnt);
     }
 
-    if(addr == -1 || addr == 0) {
+    if(addr == -1) {
         my_log(LOG_WARNING, 0, "Unable to parse address token '%s'.", addrstr);
         return NULL;
     }
