diff -BurpN accel-pptp.orig/pppd_plugin/src/util.c accel-pptp/pppd_plugin/src/util.c
--- accel-pptp.orig/pppd_plugin/src/util.c	2009-08-16 08:36:25.000000000 +0000
+++ accel-pptp/pppd_plugin/src/util.c	2009-08-21 13:12:22.000000000 +0000
@@ -64,8 +64,10 @@ void _warn(const char *func, const char 
 void _fatal(const char *func, const char *file, int line, const char *format, ...)
 {
     MAKE_STRING("fatal");
+    open_log();
     fprintf(stderr, "%s\n", string);
     syslog(LOG_CRIT, "%s", string);
+    close_log();
     exit(1);
 }
 
