diff -BurpN accel-pptp.orig/pppd_plugin/src/pptp.c accel-pptp/pppd_plugin/src/pptp.c
--- accel-pptp.orig/pppd_plugin/src/pptp.c	2009-08-21 14:35:08.000000000 +0000
+++ accel-pptp/pppd_plugin/src/pptp.c	2009-08-21 14:34:03.000000000 +0000
@@ -178,8 +178,10 @@ static int pptp_start_client(void)
 		fatal("PPTP: failed to create PPTP socket (%s)\n",strerror(errno));
 		return -1;
 	}
+	/* obsolete from 0.7
 	if (setsockopt(pptp_fd,0,PPTP_SO_TIMEOUT,&pptp_timeout,sizeof(pptp_timeout)))
 		warn("PPTP: failed to setsockopt PPTP_SO_TIMEOUT (%s)\n",strerror(errno));
+	*/
 	if (bind(pptp_fd,(struct sockaddr*)&src_addr,sizeof(src_addr)))
 	{
 		fatal("PPTP: failed to bind PPTP socket (%s)\n",strerror(errno));
