diff -BurpN accel-pptp.orig/pppd_plugin/src/pptp.c accel-pptp/pppd_plugin/src/pptp.c
--- accel-pptp.orig/pppd_plugin/src/pptp.c	2009-10-03 09:09:55.000000000 +0000
+++ accel-pptp/pppd_plugin/src/pptp.c	2010-12-27 09:52:22.705982002 +0000
@@ -54,6 +54,8 @@
 #include <stdlib.h>
 
 #include <net/route.h>
+#include <features.h>
+#include <resolv.h>
 
 extern char** environ;
 
@@ -131,6 +133,12 @@ static int pptp_start_client(void)
 	struct sockaddr_pppox src_addr,dst_addr;
 	struct hostent *hostinfo;
 
+#if !defined(__UCLIBC__) \
+ || (__UCLIBC_MAJOR__ == 0 \
+ && (__UCLIBC_MINOR__ < 9 || (__UCLIBC_MINOR__ == 9 && __UCLIBC_SUBLEVEL__ < 31)))
+	/* force ns refresh from resolv.conf with uClibc pre-0.9.31 */
+	res_init();
+#endif
 	hostinfo=gethostbyname(pptp_server);
   if (!hostinfo)
 	{
