diff -urB linux-2.4.37.orig/include/linux/icmpv6.h linux/include/linux/icmpv6.h
--- linux-2.4.37.orig/include/linux/icmpv6.h	2010-02-02 00:04:00.000000000 +0300
+++ linux/include/linux/icmpv6.h	2010-02-14 14:52:10.000000000 +0300
@@ -104,6 +104,14 @@
 #define ICMPV6_MLD2_REPORT		143
 #define MLD2_ALL_MCR_INIT { { { 0xff,0x02,0,0,0,0,0,0,0,0,0,0,0,0,0,0x16 } } }
 
+#define ICMPV6_NI_QUERY			139
+#define ICMPV6_NI_REPLY			140
+
+#define ICMPV6_DHAAD_REQUEST		144
+#define ICMPV6_DHAAD_REPLY		145
+#define ICMPV6_MOBILE_PREFIX_SOL	146
+#define ICMPV6_MOBILE_PREFIX_ADV	147
+
 /*
  *	Codes for Destination Unreachable
  */
diff -urB linux-2.4.37.orig/include/net/ipv6.h linux/include/net/ipv6.h
--- linux-2.4.37.orig/include/net/ipv6.h	2010-02-02 00:04:46.000000000 +0300
+++ linux/include/net/ipv6.h	2010-02-14 15:08:14.000000000 +0300
@@ -22,6 +22,8 @@
 
 #define SIN6_LEN_RFC2133	24
 
+#define IPV6_MAXPLEN           65535
+
 /*
  *	NextHeader field of IPv6 header
  */
@@ -37,6 +39,7 @@
 #define NEXTHDR_ICMP		58	/* ICMP for IPv6. */
 #define NEXTHDR_NONE		59	/* No next header */
 #define NEXTHDR_DEST		60	/* Destination options header. */
+#define NEXTHDR_MOBILITY	135	/* Mobility header. */
 
 #define NEXTHDR_MAX		255
 
@@ -92,12 +95,14 @@
  */
 
 struct frag_hdr {
-	unsigned char	nexthdr;
-	unsigned char	reserved;	
-	unsigned short	frag_off;
-	__u32		identification;
+	__u8	nexthdr;
+	__u8	reserved;
+	__u16	frag_off;
+	__u32	identification;
 };
 
+#define        IP6_MF  0x0001
+
 #ifdef __KERNEL__
 
 #include <net/sock.h>
