--- samba/source/rpc_parse/parse_rpc.c.orig	2008-02-02 21:45:36.000000000 +0300
+++ samba/source/rpc_parse/parse_rpc.c	2008-02-02 21:47:10.000000000 +0300
@@ -327,6 +327,9 @@
 
 BOOL smb_io_rpc_hdr_rb(char *desc, RPC_HDR_RB *rpc, prs_struct *ps, int depth)
 {
+	RPC_HDR_RB rpc2;
+	int i;
+	
 	if (rpc == NULL)
 		return False;
 
@@ -347,6 +350,20 @@
 		return False;
 	if(!smb_io_rpc_iface("", &rpc->transfer, ps, depth))
 		return False;
+		
+	/* just chew through extra context id's for now */
+	
+	for ( i=1; i<rpc->num_elements; i++ ) {
+		if(!prs_uint16("context_id  ", ps, depth, &rpc2.context_id ))
+			return False;
+		if(!prs_uint8 ("num_syntaxes", ps, depth, &rpc2.num_syntaxes))
+			return False;
+
+		if(!smb_io_rpc_iface("", &rpc2.abstract, ps, depth))
+			return False;
+		if(!smb_io_rpc_iface("", &rpc2.transfer, ps, depth))
+			return False;
+	}	
 
 	return True;
 }
