Subject: [PATCH] usb-storage: always set the allow_restart flag
From: Mauro Carvalho Chehab <mchehab@infradead.org>

from 2.6.25, kernel.org commit f09e495df27d80ae77005ddb2e93df18ec24d04a
---
 drivers/usb/storage/scsiglue.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/drivers/usb/storage/scsiglue.c b/drivers/usb/storage/scsiglue.c
--- a/drivers/usb/storage/scsiglue.c
+++ b/drivers/usb/storage/scsiglue.c
@@ -183,6 +183,10 @@ static int slave_configure(struct scsi_device *sdev)
 		 * is an occasional series of retries that will all fail. */
 		sdev->retry_hwerror = 1;
 
+		/* USB disks should allow restart.  Some drives spin down
+		 * automatically, requiring a START-STOP UNIT command. */
+		sdev->allow_restart = 1;
+
 	} else {
 
 		/* Non-disk-type devices don't need to blacklist any pages
-- 
1.6.5.GIT

