diff --git a/net/netfilter/xt_conntrack.c b/net/netfilter/xt_conntrack.c
--- a/net/netfilter/xt_conntrack.c
+++ b/net/netfilter/xt_conntrack.c
@@ -192,17 +192,6 @@ conntrack_mt(const struct sk_buff *skb,
 	return 1;
 }
 
-static bool
-conntrack_mt_v1(const struct sk_buff *skb, const struct net_device *in,
-             const struct net_device *out, const struct xt_match *match,
-             const void *matchinfo, int offset, unsigned int protoff,
-             bool *hotdrop)
-{
-	const struct xt_conntrack_mtinfo1 *info = matchinfo;
-
-	return conntrack_mt(skb, match, matchinfo, info->state_mask, info->status_mask);
-}
-
 static bool
 conntrack_mt_v2(const struct sk_buff *skb, const struct net_device *in,
              const struct net_device *out, const struct xt_match *match,
@@ -235,26 +224,6 @@ conntrack_mt_destroy(const struct xt_mat
 static struct xt_match conntrack_mt_reg[] __read_mostly = {
 	{
 		.name       = "conntrack",
-		.revision   = 1,
-		.family     = AF_INET,
-		.matchsize  = sizeof(struct xt_conntrack_mtinfo1),
-		.match      = conntrack_mt_v1,
-		.checkentry = conntrack_mt_check,
-		.destroy    = conntrack_mt_destroy,
-		.me         = THIS_MODULE,
-	},
-	{
-		.name       = "conntrack",
-		.revision   = 1,
-		.family     = AF_INET6,
-		.matchsize  = sizeof(struct xt_conntrack_mtinfo1),
-		.match      = conntrack_mt_v1,
-		.checkentry = conntrack_mt_check,
-		.destroy    = conntrack_mt_destroy,
-		.me         = THIS_MODULE,
-	},
-	{
-		.name       = "conntrack",
 		.revision   = 2,
 		.family     = AF_INET,
 		.matchsize  = sizeof(struct xt_conntrack_mtinfo2),
