diff -BurpN rp-l2tp-0.4/Makefile.in rp-l2tp/Makefile.in
--- rp-l2tp-0.4/Makefile.in	2004-07-01 15:17:11.000000000 +0000
+++ rp-l2tp/Makefile.in	2009-08-31 05:46:52.000000000 +0000
@@ -25,12 +25,13 @@ docdir=@prefix@/doc/l2tp-$(VERSION)
 install=@INSTALL@
 install_dir=@INSTALL@ -d
 sbindir=@sbindir@
+sysconfdir=@sysconfdir@
 
 OBJS=auth.o debug.o dgram.o main.o md5.o network.o options.o peer.o session.o tunnel.o utils.o
 EXTRA_LIBS=@LIBEVENT@/*.o -ldl
 
 SRCS=$(OBJS:.o=.c)
-CFLAGS=-g -I.. -I@LIBEVENT@ -Wall -Wstrict-prototypes -ansi -pedantic -D_GNU_SOURCE -DVERSION=\"$(VERSION)\" -DPREFIX=\"$(prefix)\"
+CFLAGS=@CFLAGS@ -I.. -I@LIBEVENT@ -Wall -Wstrict-prototypes -ansi -pedantic -D_GNU_SOURCE -DVERSION=\"$(VERSION)\" -DPREFIX=\"$(prefix)\" -DSYSCONFDIR=\"$(sysconfdir)\"
 
 TARGETS=l2tpd libl2tp.a handlers
 
@@ -38,8 +39,8 @@ all: $(TARGETS)
 
 libl2tp.a: $(OBJS)
 	rm -f $@
-	ar -rc $@ $^
-	ranlib $@ || true
+	@AR@ -rc $@ $^
+	@RANLIB@ $@ || true
 
 l2tpd: libl2tp.a libevent/libevent.a
 	$(MAKE) -C handlers
@@ -69,21 +70,24 @@ release: FORCE
 	gzip -v -9 rp-l2tp-$(VERSION).tar
 
 clean:
+	rm -f .depend
 	rm -f *.o *.a *~
 	rm -f l2tpd
 	$(MAKE) -C handlers clean
+	test -d libevent && $(MAKE) -C libevent clean
 
 distclean: clean
 	rm -f Makefile config.cache
 	$(MAKE) -C handlers distclean
+	test -d libevent && $(MAKE) -C libevent distclean
 
 -include .depend
 
 install: all
 	-mkdir -p $(RPM_INSTALL_ROOT)$(DESTDIR)$(sbindir)
 	$(install) -m 755 -s l2tpd $(RPM_INSTALL_ROOT)$(DESTDIR)$(sbindir)
-	-mkdir -p $(RPM_INSTALL_ROOT)$(DESTDIR)/etc/l2tp
-	$(install) -m 644 l2tp.conf $(RPM_INSTALL_ROOT)$(DESTDIR)/etc/l2tp/l2tp.conf.example
+	-mkdir -p $(RPM_INSTALL_ROOT)$(DESTDIR)$(sysconfdir)/l2tp
+	$(install) -m 644 l2tp.conf $(RPM_INSTALL_ROOT)$(DESTDIR)$(sysconfdir)/l2tp/l2tp.conf.example
 	-mkdir -p $(RPM_INSTALL_ROOT)$(DESTDIR)$(mandir)/man8
 	-mkdir -p $(RPM_INSTALL_ROOT)$(DESTDIR)$(mandir)/man5
 	$(install) -m 644 man/l2tpd.8 $(RPM_INSTALL_ROOT)$(DESTDIR)$(mandir)/man8
diff -BurpN rp-l2tp-0.4/configure rp-l2tp/configure
--- rp-l2tp-0.4/configure	2002-09-30 18:46:52.000000000 +0000
+++ rp-l2tp/configure	2009-08-31 05:46:52.000000000 +0000
@@ -1766,6 +1766,7 @@ s%@oldincludedir@%$oldincludedir%g
 s%@infodir@%$infodir%g
 s%@mandir@%$mandir%g
 s%@CC@%$CC%g
+s%@AR@%$AR%g
 s%@INSTALL_PROGRAM@%$INSTALL_PROGRAM%g
 s%@INSTALL_SCRIPT@%$INSTALL_SCRIPT%g
 s%@INSTALL_DATA@%$INSTALL_DATA%g
diff -BurpN rp-l2tp-0.4/handlers/Makefile.in rp-l2tp/handlers/Makefile.in
--- rp-l2tp-0.4/handlers/Makefile.in	2002-09-30 19:45:00.000000000 +0000
+++ rp-l2tp/handlers/Makefile.in	2009-08-31 05:46:52.000000000 +0000
@@ -5,7 +5,7 @@
 prefix=@prefix@
 exec_prefix=@exec_prefix@
 mandir=@mandir@
-docdir=@prefix@/doc/rp-pppoe-$(VERSION)
+docdir=@prefix@/doc/l2tp-$(VERSION)
 install=@INSTALL@
 install_dir=@INSTALL@ -d
 sbindir=@sbindir@
@@ -14,24 +14,24 @@ HANDLERS=sync-pppd.so cmd.so
 
 OBJS=pty.o sync-pppd.o dstring.o
 SRCS=$(OBJS:.o=.c)
-CFLAGS=-g -O2 -I.. -I../libevent -I../../libevent -Wall
+CFLAGS=@CFLAGS@ -I.. -I../@LIBEVENT@ -Wall
 
 all: $(HANDLERS) l2tp-control
 
 l2tp-control: cmd-control.o
-	gcc -o l2tp-control cmd-control.o
+	@CC@ -o l2tp-control cmd-control.o
 
 cmd-control.o: cmd-control.c
-	gcc -c -o $@ $^
+	@CC@ -c -o $@ $^
 
 %.o: %.c
-	gcc $(CFLAGS) -fPIC -c -o $@ $<
+	@CC@ $(CFLAGS) -fPIC -c -o $@ $<
 
 sync-pppd.so: pty.o sync-pppd.o
-	gcc -shared -o $@ $^
+	@CC@ -shared -o $@ $^
 
 cmd.so: cmd.o dstring.o
-	gcc -shared -o $@ $^
+	@CC@ -shared -o $@ $^
 
 clean:
 	rm -f *.so *.o *~
@@ -42,4 +42,6 @@ install: all
 	-mkdir -p $(RPM_INSTALL_ROOT)$(prefix)/lib/l2tp/plugins
 	$(install) -m 755 -s l2tp-control $(RPM_INSTALL_ROOT)$(sbindir)
 	$(install) -m 755 $(HANDLERS) $(RPM_INSTALL_ROOT)$(prefix)/lib/l2tp/plugins
+
 distclean: clean
+	rm -f Makefile
diff -BurpN rp-l2tp-0.4/libevent/Makefile.in rp-l2tp/libevent/Makefile.in
--- rp-l2tp-0.4/libevent/Makefile.in	2002-09-30 19:45:00.000000000 +0000
+++ rp-l2tp/libevent/Makefile.in	2009-08-31 05:46:52.000000000 +0000
@@ -18,24 +18,27 @@ all: libevent.a
 
 libevent.a: $(OBJS)
 	rm -f libevent.a
-	ar -cq libevent.a $(OBJS)
+	@AR@ -cq libevent.a $(OBJS)
 	@RANLIB@ libevent.a
 
 event.o: event.c $(HDRS)
-	gcc $(CFLAGS) -c -o event.o event.c
+	@CC@ $(CFLAGS) -c -o event.o event.c
 
 hash.o: hash.c $(HDRS)
-	gcc $(CFLAGS) -c -o hash.o hash.c
+	@CC@ $(CFLAGS) -c -o hash.o hash.c
 
 event_sig.o: event_sig.c $(HDRS)
-	gcc $(CFLAGS) -c -o event_sig.o event_sig.c
+	@CC@ $(CFLAGS) -c -o event_sig.o event_sig.c
 
 event_tcp.o: event_tcp.c $(HDRS)
-	gcc $(CFLAGS) -c -o event_tcp.o event_tcp.c
+	@CC@ $(CFLAGS) -c -o event_tcp.o event_tcp.c
 
 clean: FORCE
 	rm -f *.a *.o *~
 
+distclean: clean
+	rm -f Makefile
+
 FORCE:
 
 .phony: FORCE
diff -BurpN rp-l2tp-0.4/main.c rp-l2tp/main.c
--- rp-l2tp-0.4/main.c	2002-09-30 19:45:00.000000000 +0000
+++ rp-l2tp/main.c	2011-04-09 23:00:01.675157000 +0000
@@ -68,7 +68,7 @@ main(int argc, char *argv[])
     l2tp_peer_init();
     l2tp_debug_set_bitmask(debugmask);
 
-    if (l2tp_parse_config_file(es, "/etc/l2tp/l2tp.conf") < 0) {
+    if (l2tp_parse_config_file(es, SYSCONFDIR"/l2tp/l2tp.conf") < 0) {
 	l2tp_die();
     }
 
diff -BurpN rp-l2tp-0.4/utils.c rp-l2tp/utils.c
--- rp-l2tp-0.4/utils.c	2002-09-30 19:45:00.000000000 +0000
+++ rp-l2tp/utils.c	2011-04-09 22:58:20.235157000 +0000
@@ -196,7 +196,7 @@ l2tp_load_handler(EventSelector *es,
     if (*fname == '/') {
 	handle = dlopen(fname, RTLD_NOW);
     } else {
-	snprintf(buf, sizeof(buf), "%s/lib/l2tp/plugins/%s", PREFIX, fname);
+	snprintf(buf, sizeof(buf), PREFIX"/lib/l2tp/plugins/%s", fname);
 	buf[sizeof(buf)-1] = 0;
 	handle = dlopen(buf, RTLD_NOW);
     }
