diff -BurpN rp-l2tp-0.4/handlers/sync-pppd.c rp-l2tp/handlers/sync-pppd.c
--- rp-l2tp-0.4/handlers/sync-pppd.c	2004-07-01 14:58:56.000000000 +0000
+++ rp-l2tp/handlers/sync-pppd.c	2009-08-31 05:51:04.000000000 +0000
@@ -183,7 +183,8 @@ close_session(l2tp_session *ses, char co
     sl->event = NULL;
 
     /* Re-establish session if desired */
-    if (may_reestablish && tunnel->peer->persist && tunnel->peer->fail < tunnel->peer->maxfail) {
+    if (may_reestablish && tunnel->peer->persist && 
+        (tunnel->peer->maxfail == 0 || tunnel->peer->fail++ < tunnel->peer->maxfail)) {
         struct timeval t;
 
         t.tv_sec = tunnel->peer->holdoff;
@@ -214,8 +215,21 @@ slave_exited(pid_t pid, int status, void
 
     if (sl->fd >= 0) close(sl->fd);
     if (sl->event) Event_DelHandler(sl->es, sl->event);
+    sl->fd = -1;
+    sl->event = NULL;
 
     if (ses) {
+        l2tp_tunnel *tunnel = ses->tunnel;
+        
+        /* Re-establish session if desired */
+        if (tunnel->peer->persist) {
+            struct timeval t;
+
+            t.tv_sec = tunnel->peer->holdoff;
+            t.tv_usec = 0;
+            Event_AddTimerHandler(tunnel->es, t, l2tp_tunnel_reestablish, tunnel->peer);
+        }
+        
 	ses->private = NULL;
 	l2tp_session_send_CDN(ses, RESULT_GENERAL_REQUEST, 0,
 			      "pppd process exited");
diff -BurpN rp-l2tp-0.4/session.c rp-l2tp/session.c
--- rp-l2tp-0.4/session.c	2004-07-01 14:58:55.000000000 +0000
+++ rp-l2tp/session.c	2009-08-31 05:51:04.000000000 +0000
@@ -473,7 +473,7 @@ l2tp_session_handle_CDN(l2tp_session *se
     val = l2tp_dgram_search_avp(dgram, ses->tunnel, NULL, NULL, &len,
 				VENDOR_IETF, AVP_RESULT_CODE);
     if (!val || len < 4) {
-	l2tp_tunnel_delete_session(ses, "Received CDN", 0);
+	l2tp_tunnel_delete_session(ses, "Received CDN", 1);
     } else {
 	uint16_t result_code, error_code;
 	char *msg;
@@ -486,7 +486,7 @@ l2tp_session_handle_CDN(l2tp_session *se
 	}
 	snprintf(buf, sizeof(buf), "Received CDN: result-code = %d, error-code = %d, message = '%.*s'", result_code, error_code, (int) len-4, msg);
 	buf[1023] = 0;
-	l2tp_tunnel_delete_session(ses, buf, 0);
+	l2tp_tunnel_delete_session(ses, buf, 1);
     }
 }
 
diff -BurpN rp-l2tp-0.4/tunnel.c rp-l2tp/tunnel.c
--- rp-l2tp-0.4/tunnel.c	2004-07-01 14:58:56.000000000 +0000
+++ rp-l2tp/tunnel.c	2009-08-31 05:51:05.000000000 +0000
@@ -966,6 +966,17 @@ tunnel_handle_timeout(EventSelector *es,
     if (tunnel->retransmissions >= MAX_RETRANSMISSIONS) {
 	l2tp_set_errmsg("Too many retransmissions on tunnel (%s); closing down",
 		   l2tp_debug_tunnel_to_str(tunnel));
+		   
+	if (tunnel->state < TUNNEL_ESTABLISHED && tunnel->peer && tunnel->peer->persist && 
+	    (tunnel->peer->maxfail == 0 || tunnel->peer->fail++ < tunnel->peer->maxfail)) 
+	{
+	    struct timeval t;
+
+	    t.tv_sec = tunnel->peer->holdoff;
+	    t.tv_usec = 0;
+	    Event_AddTimerHandler(tunnel->es, t, l2tp_tunnel_reestablish, tunnel->peer);
+	}
+	
 	/* Close tunnel... */
 	tunnel_free(tunnel);
 	return;
@@ -1061,6 +1072,16 @@ tunnel_handle_StopCCN(l2tp_tunnel *tunne
     l2tp_session *ses;
     void *cursor;
 
+    if (tunnel->state < TUNNEL_ESTABLISHED && tunnel->peer && tunnel->peer->persist && 
+        (tunnel->peer->maxfail == 0 || tunnel->peer->fail++ < tunnel->peer->maxfail)) 
+    {
+	struct timeval t;
+
+	t.tv_sec = tunnel->peer->holdoff;
+	t.tv_usec = 0;
+	Event_AddTimerHandler(tunnel->es, t, l2tp_tunnel_reestablish, tunnel->peer);
+    }
+
     /* Shut down all the sessions */
     for (ses = hash_start(&tunnel->sessions_by_my_id, &cursor);
 	 ses ;
