diff -BurpN rp-l2tp-0.4/dgram.c rp-l2tp/dgram.c
--- rp-l2tp-0.4/dgram.c	2003-01-11 05:29:03.000000000 +0000
+++ rp-l2tp/dgram.c	2009-08-31 05:54:03.000000000 +0000
@@ -535,7 +535,8 @@ l2tp_dgram_send_to_wire(l2tp_dgram const
     unsigned char *len_ptr = NULL;
 
     DBG(l2tp_db(DBG_XMIT_RCV,
-	   "dgram_send_to_wire() -> %s\n",
+	   "dgram_send_to_wire(%s:%d) -> %s\n", 
+	   inet_ntoa(to->sin_addr), ntohs(to->sin_port),
 	   l2tp_debug_describe_dgram(dgram)));
     buf[0] = dgram->bits;
     buf[1] = dgram->version;
diff -BurpN rp-l2tp-0.4/main.c rp-l2tp/main.c
--- rp-l2tp-0.4/main.c	2002-09-30 19:45:00.000000000 +0000
+++ rp-l2tp/main.c	2009-08-31 05:54:03.000000000 +0000
@@ -23,6 +23,7 @@ static char const RCSID[] =
 #include <signal.h>
 #include <fcntl.h>
 #include <stdlib.h>
+#include <syslog.h>
 
 static void
 usage(int argc, char *argv[], int exitcode)
@@ -63,6 +64,7 @@ main(int argc, char *argv[])
 	}
     }
 
+    openlog(argv[0], LOG_PID, LOG_DAEMON);
     l2tp_random_init();
     l2tp_tunnel_init(es);
     l2tp_peer_init();
diff -BurpN rp-l2tp-0.4/utils.c rp-l2tp/utils.c
--- rp-l2tp-0.4/utils.c	2002-09-30 19:45:00.000000000 +0000
+++ rp-l2tp/utils.c	2009-08-31 05:54:03.000000000 +0000
@@ -27,6 +27,7 @@ static char const RCSID[] =
 #include <dlfcn.h>
 #include <string.h>
 #include <errno.h>
+#include <syslog.h>
 
 #include "l2tp.h"
 
@@ -88,6 +89,8 @@ l2tp_set_errmsg(char const *fmt, ...)
     va_end(ap);
     errmsg[MAX_ERRMSG_LEN-1] = 0;
     fprintf(stderr, "Error: %s\n", errmsg);
+    
+    vsyslog(LOG_ERR, fmt, ap);
 }
 
 /**********************************************************************
