diff -BurpN inadyn.orig/include/dyndns.h inadyn/include/dyndns.h
--- inadyn.orig/include/dyndns.h	2010-07-14 18:15:30.000000000 +0000
+++ inadyn/include/dyndns.h	2011-05-01 23:37:24.892944000 +0000
@@ -47,6 +47,7 @@ typedef enum
     DYNDNS_3322_DYNAMIC,
     DNSOMATIC_DEFAULT,
     HE_IPV6TB,
+    HE_DYNDNS,
     LAST_DNS_SYSTEM = -1
 } DYNDNS_SYSTEM_ID;
 
diff -BurpN inadyn.orig/src/dyndns.c inadyn/src/dyndns.c
--- inadyn.orig/src/dyndns.c	2010-07-14 18:15:52.000000000 +0000
+++ inadyn/src/dyndns.c	2011-05-01 23:38:06.572944000 +0000
@@ -151,6 +151,12 @@ DYNDNS_SYSTEM_INFO dns_system_table[] =
 	    (DNS_SYSTEM_REQUEST_FUNC) get_req_for_he_ipv6tb_server,
 	    DYNDNS_MY_IP_SERVER, DYNDNS_MY_IP_SERVER_URL,
 			"ipv4.tunnelbroker.net", "/ipv4_end.php?", NULL}},
+    {HE_DYNDNS,
+	{"dyndns@he.net", NULL,
+	    (DNS_SYSTEM_SRV_RESPONSE_OK_FUNC) is_dyndns_server_rsp_ok,
+	    (DNS_SYSTEM_REQUEST_FUNC) get_req_for_noip_http_dns_server,
+	    DYNDNS_MY_IP_SERVER, DYNDNS_MY_IP_SERVER_URL,
+			"dyn.dns.he.net", "/nic/update?hostname=", NULL}},
 
     {CUSTOM_HTTP_BASIC_AUTH, 
         {"custom@http_svr_basic_auth", NULL,  
