diff -BurpN inadyn.orig/include/dyndns.h inadyn/include/dyndns.h
--- inadyn.orig/include/dyndns.h	2010-07-11 12:07:17.000000000 +0000
+++ inadyn/include/dyndns.h	2010-07-14 18:15:30.000000000 +0000
@@ -283,6 +283,7 @@ typedef struct 
 	DYNDNS_SERVER_NAME proxy_server_name;
 	DYNDNS_ALIAS_INFO alias_info[DYNDNS_MAX_ALIAS_NUMBER];
 	int alias_count;
+	BOOL wildcard;
 } DYNDNS_INFO_TYPE;
 
 typedef struct
@@ -329,7 +330,6 @@ typedef struct DYN_DNS_CLIENT
 	BOOL force_addr_update;
     BOOL use_proxy;
 	BOOL abort;
-	BOOL wildcard;
 
 	char *p_pidfilename;
 	/*dbg*/
diff -BurpN inadyn.orig/src/dyndns.c inadyn/src/dyndns.c
--- inadyn.orig/src/dyndns.c	2010-07-11 12:52:12.000000000 +0000
+++ inadyn/src/dyndns.c	2010-07-14 18:15:52.000000000 +0000
@@ -213,7 +213,7 @@ static int get_req_for_dyndns_server(DYN
 		p_dyndns_specific->p_system,
 		p_self->info[infcnt].alias_info[alcnt].names.name,
 		p_self->info[infcnt].my_ip_address.name,
-		p_self->wildcard ? "ON" : "OFF",
+		p_self->info[infcnt].wildcard ? "ON" : "OFF",
 		p_self->info[infcnt].alias_info[alcnt].names.name,
 		p_self->info[infcnt].dyndns_server_name.name,
 		p_self->info[infcnt].credentials.p_enc_usr_passwd_buffer
@@ -270,7 +270,7 @@ static int get_req_for_easydns_http_dns_
 		p_self->info[infcnt].dyndns_server_url,
 		p_self->info[infcnt].alias_info[alcnt].names.name,
 		p_self->info[infcnt].my_ip_address.name,
-		p_self->wildcard ? "ON" : "OFF",
+		p_self->info[infcnt].wildcard ? "ON" : "OFF",
 		p_self->info[infcnt].credentials.p_enc_usr_passwd_buffer,
 		p_self->info[infcnt].dyndns_server_name.name
 		);
diff -BurpN inadyn.orig/src/inadyn_cmd.c inadyn/src/inadyn_cmd.c
--- inadyn.orig/src/inadyn_cmd.c	2010-07-11 12:51:24.000000000 +0000
+++ inadyn/src/inadyn_cmd.c	2010-07-14 18:17:51.000000000 +0000
@@ -48,7 +48,7 @@ static int curr_info;
 #define DYNDNS_INPUT_FILE_OPT_STRING "--input_file"
 
 static RC_TYPE help_handler(CMD_DATA *p_cmd, int current_nr, void *p_context);
-static RC_TYPE wildcard_handler(CMD_DATA *p_cmd, int current_nr, void *p_context);
+static RC_TYPE get_wildcard_handler(CMD_DATA *p_cmd, int current_nr, void *p_context);
 static RC_TYPE get_username_handler(CMD_DATA *p_cmd, int current_nr, void *p_context);
 static RC_TYPE get_password_handler(CMD_DATA *p_cmd, int current_nr, void *p_context);
 static RC_TYPE get_alias_handler(CMD_DATA *p_cmd, int current_nr, void *p_context);
@@ -146,7 +146,7 @@ static CMD_DESCRIPTION_TYPE cmd_options_
 #else
 	{"--cache_file", 1, {get_cache_file, NULL}, "cache file name. (e.g. /tmp/ddns.cache). Defaults to /tmp/inadyn.cache."},
 #endif
-	{"--wildcard", 0, {wildcard_handler, NULL}, "enable domain wildcarding for dyndns.org, 3322.org, or easydns.com."},
+	{"--wildcard", 0, {get_wildcard_handler, NULL}, "enable domain wildcarding for dyndns.org, 3322.org, or easydns.com."},
 	{NULL,		0,	{0, NULL},	NULL }
 };
 
@@ -193,14 +193,14 @@ static RC_TYPE help_handler(CMD_DATA *p_
 	return RC_OK;
 }
 
-static RC_TYPE wildcard_handler(CMD_DATA *p_cmd, int current_nr, void *p_context)
+static RC_TYPE get_wildcard_handler(CMD_DATA *p_cmd, int current_nr, void *p_context)
 {
 	DYN_DNS_CLIENT *p_self = (DYN_DNS_CLIENT *) p_context;
 	if (p_self == NULL)
 	{
 		return RC_INVALID_POINTER;
 	}
-	p_self->wildcard = TRUE;
+	p_self->info[curr_info].wildcard = TRUE;
 	return RC_OK;
 }
 
