diff -BurpN rp-pppoe-3.11.orig/src/configure rp-pppoe-3.11/src/configure
--- rp-pppoe-3.11.orig/src/configure	2012-08-18 00:31:25.000000000 +0600
+++ rp-pppoe-3.11/src/configure	2012-11-03 20:18:48.335709944 +0600
@@ -4609,9 +4609,13 @@ esac
 
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking packing order of bit fields" >&5
 $as_echo_n "checking packing order of bit fields... " >&6; }
+if test "${rpppoe_cv_pack_bitfields+set}" = set; then :
+  $as_echo_n "(cached) " >&6
+else
+
 if test "${rpppoe_cv_pack_bitfields+set}" != set ; then
 if test "$cross_compiling" = yes; then :
-  $ECHO "no defaults for cross-compiling"; exit 0
+  $ECHO "no defaults for cross-compiling"; exit 1
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
@@ -4650,6 +4654,9 @@ fi
 
 fi
 
+fi
+
+
 if test "$rpppoe_cv_pack_bitfields" = "rev" ; then
 	{ $as_echo "$as_me:${as_lineno-$LINENO}: result: reversed" >&5
 $as_echo "reversed" >&6; }
diff -BurpN rp-pppoe-3.11.orig/src/configure.in rp-pppoe-3.11/src/configure.in
--- rp-pppoe-3.11.orig/src/configure.in	2012-08-18 00:31:25.000000000 +0600
+++ rp-pppoe-3.11/src/configure.in	2012-11-03 20:18:08.000000000 +0600
@@ -231,6 +231,7 @@ esac
 
 dnl Figure out packing order of structures
 AC_MSG_CHECKING([packing order of bit fields])
+AC_CACHE_VAL(rpppoe_cv_pack_bitfields,[
 if test "${rpppoe_cv_pack_bitfields+set}" != set ; then
 AC_TRY_RUN([
 union foo {
@@ -255,8 +256,9 @@ main(void)
 	return 2;
     }
 }], rpppoe_cv_pack_bitfields=normal, rpppoe_cv_pack_bitfields=rev,
-$ECHO "no defaults for cross-compiling"; exit 0)
+$ECHO "no defaults for cross-compiling"; exit 1)
 fi
+])
 
 if test "$rpppoe_cv_pack_bitfields" = "rev" ; then
 	AC_MSG_RESULT(reversed)
