diff -BurpN rp-pppoe.orig/src/plugin.c rp-pppoe/src/plugin.c
--- rp-pppoe.orig/src/plugin.c	2008-06-30 14:00:43.000000000 +0000
+++ rp-pppoe/src/plugin.c	2009-05-05 19:32:19.000000000 +0000
@@ -58,7 +58,7 @@ static char const RCSID[] =
 #define _ROOT_PATH ""
 #endif
 
-#define _PATH_ETHOPT         _ROOT_PATH "/etc/ppp/options."
+#define _PATH_ETHOPT         _ROOT_PATH "/ppp/options."
 
 char pppd_version[] = VERSION;
 
