From 42698cb7ab366f8b92cee61f78e11d5a3b80539e Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Thu, 20 Sep 2012 21:19:35 +0100
Subject: [PATCH 21/25] Log ignored DHCPv6 information-requests.

---
 src/rfc3315.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/rfc3315.c b/src/rfc3315.c
index 716aa4a..f3929f2 100644
--- a/src/rfc3315.c
+++ b/src/rfc3315.c
@@ -1005,10 +1005,10 @@ static int dhcp6_no_relay(int msg_type, struct in6_addr *link_address, struct dh
       
     case DHCP6IREQ:
       {
+	log6_packet("DHCPINFORMATION-REQUEST", clid, clid_len, NULL, xid, iface_name, ignore ? "ignored" : hostname);
 	if (ignore)
 	  return 0;
 	*outmsgtypep = DHCP6REPLY;
-	log6_packet("DHCPINFORMATION-REQUEST", clid, clid_len, NULL, xid, iface_name, hostname);
 	break;
       }
       
-- 
1.7.2.5

