--- wpa_supplicant.orig/src/eapol_supp/eapol_supp_sm.c	2010-01-12 14:55:02.000000000 +0000
+++ wpa_supplicant/src/eapol_supp/eapol_supp_sm.c	2011-09-06 06:35:14.353897002 +0000
@@ -1822,9 +1822,9 @@ struct eapol_sm *eapol_sm_init(struct ea
 	sm->portControl = Auto;
 
 	/* Supplicant PAE state machine */
-	sm->heldPeriod = 60;
-	sm->startPeriod = 30;
-	sm->maxStart = 3;
+	sm->heldPeriod = 30;
+	sm->startPeriod = 5;
+	sm->maxStart = 8;
 
 	/* Supplicant Backend state machine */
 	sm->authPeriod = 30;
