--- mc-4.7.0-pre3/misc/mc.keymap.orig	2009-10-01 13:02:59.000000000 +0000
+++ mc-4.7.0-pre3/misc/mc.keymap	2009-10-19 19:39:57.000000000 +0000
@@ -163,12 +163,12 @@
 [viewer]
 ViewSearch = question; slash; f7
 ViewContinueSearch = ctrl-r; ctrl-s; f17
-ViewMoveToBol = home
-ViewMoveToEol = end
+ViewMoveToBol = ctrl-a
+ViewMoveToEol = ctrl-e
 ViewMoveLeft = h; left
 ViewMoveRight = l; right
 ViewMoveUp = k; y; insert; up
-ViewMoveDown = j; e; delete; down
+ViewMoveDown = j; e; delete; down; enter
 ViewMovePgDn = f; space; pgdn
 ViewMovePgUp = b; pgup
 ViewMoveHalfPgDn = d
