diff -BurpN src.orig/include/emf/emf/emfc.h src/include/emf/emf/emfc.h
--- src.orig/include/emf/emf/emfc.h	2010-01-29 14:41:26.000000000 +0000
+++ src/include/emf/emf/emfc.h	2010-09-21 08:54:36.200000002 +0000
@@ -20,6 +20,7 @@
 
 #define IP_ISMULTI(a)           (((a) & 0xf0000000) == 0xe0000000)
 #define MCAST_ADDR_LINKLOCAL(a) (((a) & 0xffffff00) == 0xe0000000)
+#define MCAST_ADDR_UPNP_SSDP(a) ((a) == 0xeffffffa)
 
 #define EMFC_STATS_INCR(emfc, member) (((emfc)->stats.member)++)
 
diff -BurpN src.orig/emf/emf/emfc.c src/emf/emf/emfc.c
--- src.orig/emf/emf/emfc.c	2010-01-29 14:43:59.000000000 +0000
+++ src/emf/emf/emfc.c	2010-09-21 08:55:25.710000002 +0000
@@ -352,11 +352,13 @@ emfc_input(emfc_info_t *emfc, void *sdu,
 		EMFC_STATS_INCR(emfc, mcast_data_frames);
 
 		/* Packets with destination IP address in the range 224.0.0.x
-		 * must be forwarded on all ports.
+		 * must be forwarded on all ports. Similarly UPnP specific
+		 * protocol traffic such as SSDP must be forwarded on to all
+		 * the ports.
 		 */
-		if (MCAST_ADDR_LINKLOCAL(dest_ip))
+		if (MCAST_ADDR_LINKLOCAL(dest_ip) || MCAST_ADDR_UPNP_SSDP(dest_ip))
 		{
-			EMF_DEBUG("Flooding the frames with link-local address\n");
+			EMF_DEBUG("Flooding the frames with link-local/ssdp address\n");
 			return (EMF_NOP);
 		}
 
