diff -N -u -r -b pptp-linux-orig/pptp_gre.c pptp-linux/pptp_gre.c
--- pptp-linux-orig/pptp_gre.c	2009-02-20 05:30:22.140122481 +0600
+++ pptp-linux/pptp_gre.c	2009-02-20 05:26:13.000000000 +0600
@@ -355,7 +355,16 @@
         return 0;
     }
     /* silently discard packets not for this call */
-    if (ntoh16(header->call_id) != pptp_gre_call_id) return 0;
+    if (ntoh16(header->call_id) != pptp_gre_call_id) {
+        if (log_level >= 2)
+            log("Discarding for other call : %d %d",
+                ntoh16(header->call_id), pptp_gre_call_id);
+        if (pptp_gre_call_id == 0) {
+            pptp_gre_call_id = ntoh16(header->call_id);
+        } else {
+            return 0;
+        }
+    }
     /* test if acknowledgement present */
     if (PPTP_GRE_IS_A(ntoh8(header->ver))) { 
         u_int32_t ack = (PPTP_GRE_IS_S(ntoh8(header->flags)))?
