diff -BurpN -x'*.o' -x'*.orig' router/bridge/Makefile gateway/bridge/Makefile
--- router/bridge/Makefile	2004-10-14 19:11:18.000000000 +0000
+++ gateway/bridge/Makefile	2009-03-24 11:29:01.000000000 +0000
@@ -7,7 +7,6 @@ clean:
 				$(MAKE) -C doc clean
 				$(MAKE) -C libbridge clean
 				$(MAKE) -C misc clean
-				rcsclean *
 				rm -f *~
 
 brctl/brctl:			brctl/brctl.c brctl/brctl.h brctl/brctl_cmd.c brctl/brctl_disp.c libbridge/libbridge.a libbridge/libbridge.h
diff -BurpN -x'*.o' -x'*.orig' router/bridge/brctl/Makefile gateway/bridge/brctl/Makefile
--- router/bridge/brctl/Makefile	2004-10-14 19:11:18.000000000 +0000
+++ gateway/bridge/brctl/Makefile	2009-03-24 11:29:28.000000000 +0000
@@ -1,7 +1,7 @@
 #
 # $Id$
 
-CFLAGS = -Wall -g -O2 -I../libbridge
+CFLAGS = -Wall -g -O2 -I../libbridge $(EXTRACFLAGS)
 LDFLAGS = -g
 ifeq ($(strip $(DOSTATIC)),true)
     LDFLAGS += --static
@@ -16,7 +16,6 @@ brctld:			brctld.o brctl_cmd.o brctl_dis
 			$(CC) $(LDFLAGS) -o brctld brctld.o brctl_cmd.o brctl_disp.o ../libbridge/libbridge.a
 
 clean:
-			rcsclean *
 			rm -f *~
 			rm -f *.o
 			rm -f brctl
diff -BurpN -x'*.o' -x'*.orig' router/bridge/brctl/brctld.c gateway/bridge/brctl/brctld.c
--- router/bridge/brctl/brctld.c	2004-10-14 19:11:18.000000000 +0000
+++ gateway/bridge/brctl/brctld.c	2009-12-04 10:54:10.000000000 +0000
@@ -177,7 +177,7 @@ int main(int argc, char *argv[])
 	addr.sin_family = AF_INET;
 	addr.sin_addr.s_addr = htonl(INADDR_ANY);
 	addr.sin_port = htons(31338);
-	if (bind(sock, &addr, sizeof(addr)) < 0) {
+	if (bind(sock, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
 		perror("bind");
 		return 1;
 	}
@@ -198,7 +198,7 @@ int main(int argc, char *argv[])
 		int newsock;
 
 		len = sizeof(addr);
-		if ((newsock = accept(sock, &addr, &len)) < 0) {
+		if ((newsock = accept(sock, (struct sockaddr *)&addr, &len)) < 0) {
 			perror("accept");
 			return 1;
 		}
diff -BurpN -x'*.o' -x'*.orig' router/bridge/doc/Makefile gateway/bridge/doc/Makefile
--- router/bridge/doc/Makefile	2004-10-14 19:11:18.000000000 +0000
+++ gateway/bridge/doc/Makefile	2009-03-24 11:29:35.000000000 +0000
@@ -1,5 +1,4 @@
 all:			FAQ HOWTO PROJECTS RPM-GPG-KEY TODO
 
 clean:
-			rcsclean *
 			rm -f *~
diff -BurpN -x'*.o' -x'*.orig' router/bridge/libbridge/Makefile gateway/bridge/libbridge/Makefile
--- router/bridge/libbridge/Makefile	2004-10-14 19:11:18.000000000 +0000
+++ gateway/bridge/libbridge/Makefile	2009-03-31 19:02:05.000000000 +0000
@@ -1,9 +1,8 @@
-CC := $(CC) -Wall -c -g
+CC := $(CC) -Wall -c -g $(EXTRACFLAGS)
 
 all:			libbridge.a
 
 clean:
-			rcsclean *
 			rm -f *~
 			rm -f *.o
 			rm -f libbridge.a
diff -BurpN -x'*.o' -x'*.orig' router/bridge/libbridge/libbridge_if.c gateway/bridge/libbridge/libbridge_if.c
--- router/bridge/libbridge/libbridge_if.c	2004-10-14 19:11:18.000000000 +0000
+++ gateway/bridge/libbridge/libbridge_if.c	2009-12-04 10:46:55.000000000 +0000
@@ -20,6 +20,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <errno.h>
 #include <sys/fcntl.h>
 #include <sys/ioctl.h>
diff -BurpN -x'*.o' -x'*.orig' router/bridge/libbridge/libbridge_init.c gateway/bridge/libbridge/libbridge_init.c
--- router/bridge/libbridge/libbridge_init.c	2004-10-14 19:11:18.000000000 +0000
+++ gateway/bridge/libbridge/libbridge_init.c	2009-12-04 10:47:07.000000000 +0000
@@ -20,6 +20,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <errno.h>
 #include <sys/fcntl.h>
 #include <sys/ioctl.h>
diff -BurpN -x'*.o' -x'*.orig' router/bridge/libbridge/libbridge_misc.c gateway/bridge/libbridge/libbridge_misc.c
--- router/bridge/libbridge/libbridge_misc.c	2004-10-14 19:11:18.000000000 +0000
+++ gateway/bridge/libbridge/libbridge_misc.c	2009-12-04 10:49:40.000000000 +0000
@@ -20,6 +20,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <sys/time.h>
 #include <asm/param.h>
 #include "libbridge.h"
diff -BurpN -x'*.o' -x'*.orig' router/bridge/misc/Makefile gateway/bridge/misc/Makefile
--- router/bridge/misc/Makefile	2004-10-14 19:11:18.000000000 +0000
+++ gateway/bridge/misc/Makefile	2009-03-24 11:29:32.000000000 +0000
@@ -9,7 +9,6 @@ endif
 all:		bidi tap
 
 clean:
-		rcsclean *
 		rm -f *~
 		rm -f bidi
 		rm -f tap
