diff -BurN -x'*.o' -x'*.orig' router/config/Config gateway/config/Config
--- router/config/Config	2004-10-14 19:11:24.000000000 +0000
+++ gateway/config/Config	2009-09-28 16:47:39.000000000 +0000
@@ -13,12 +13,6 @@
 
 mainmenu "Router Configuration"
 
-config CONFIG_NAT
-       bool "Network Address Translation (NAT) Support"
-       default y
-       help
-       Answer Y to configure this box as a NAT box.
-
 menu "Base Features"
 
 config CONFIG_RC
@@ -59,9 +53,9 @@
 config CONFIG_BUSYBOX_CONFIG
        string "BusyBox default configuration"
        depends on CONFIG_BUSYBOX
-       default "router"
+       default "defconfig"
        help
-       Override the default BusyBox configuration. File must exist in busybox/configs.
+       Override the default BusyBox configuration. File must exist in busybox/sysdeps/linux.
 
 config CONFIG_WLCONF
        bool "Wireless configuration"
@@ -78,7 +72,6 @@
 
 config CONFIG_VLAN
        bool "VLAN configuration"
-       depends on CONFIG_NAT
        default y
        help
        Answer Y for the user mode VLAN configuration utility.
@@ -96,6 +89,12 @@
        help
        Answer Y for additional web pages.
 
+config CONFIG_NAT
+       bool "Network Address Translation (NAT) Support"
+       default y
+       help
+       Answer Y to configure this box as a NAT box.
+
 config CONFIG_NETCONF
        bool "Firewall configuration"
        depends on (CONFIG_IPTABLES || CONFIG_LIBIPT) && CONFIG_NAT
@@ -108,7 +107,7 @@
 
 choice
 	prompt "Iptables"
-    depends on CONFIG_NAT
+	depends on CONFIG_NAT
 	default CONFIG_LIBIPT
 	optional
 	help
@@ -152,10 +151,6 @@
        help
        Answer Y for vendor-specific customization.
 
-config CONFIG_UDHCPD
-       bool "DHCP client and server"
-       default y
-
 config CONFIG_PPP
        bool "PPPoE client"
        depends on CONFIG_NAT
@@ -223,6 +218,167 @@
 
 endmenu
 
+menu "Asus stuff"
+
+config CONFIG_PPTP
+	bool "PPTP client"
+	default n
+
+config CONFIG_ACCEL_PPTP
+	bool "Accelerated PPTP driver"
+	default y
+
+config CONFIG_PPPOERELAY
+	bool "PPPOE-relay"
+	default y
+
+config CONFIG_INFOSVR
+	bool "CONFIG_INFOSVR"
+	default y
+
+config CONFIG_EZIPUPDATE
+	bool "DynDNS support"
+	default y
+
+config CONFIG_VSFTPD
+	bool "vsftpd server"
+	default y
+
+config CONFIG_JPEG6B
+	bool "jpeg6b library"
+	default y
+
+config CONFIG_RCAMD
+	bool "rcamd daemon (web-camera support)"
+	default y
+
+config CONFIG_SENDMAIL
+	bool "Mini sendmail"
+	default y
+
+config CONFIG_LPRNG
+	bool "LPRng daemon"
+	default y
+
+config CONFIG_OTHERS
+	bool "CONFIG_OTHERS"
+	default y
+
+config CONFIG_TEST
+	bool "CONFIG_TEST"
+	default n
+
+config CONFIG_BPALOGIN
+	bool "Big Pond Broadband support"
+	default y
+
+config CONFIG_WAVESERVER
+	bool "CONFIG_WAVESERVER"
+	default y
+endmenu
+
+menu "Custom firmware features"
+
+config CONFIG_LOADER
+       bool "LZMA compressed kernel loader"
+       default y
+       help
+       Save bunch of flash space
+
+config CONFIG_SSHD
+       bool "Dropbear SSH daemon"
+       default y
+       help
+       Dropbear is the small ssh daemon.
+
+config CONFIG_JETDIRECT
+       bool "p910nd printing daemon"
+       default y
+       help
+       p910nd is the small JetDirect capable printing daemon.
+
+config CONFIG_SAMBA
+       bool "Samba filesharing"
+       default y
+       help
+       Samba filesharing.
+
+config CONFIG_NFSD
+       bool "NFS server"
+       default y
+       help
+       NFS filesharing.
+
+config CONFIG_NFS
+       bool "NFS client"
+       default y
+       help
+       NFS filesharing.
+
+config CONFIG_IPROUTE2
+       bool "iproute2 package"
+       default y
+       help
+       iproute2 package used by wonder shaper.
+
+config CONFIG_IPV6
+       bool "IPv6 support"
+       default y
+       help
+       IPv6 is the next-generation IP protocol. Please ensure IPv6 is enabled in the kernel and busybox config
+
+config CONFIG_SNMP
+       bool "SNMP agent support"
+       default y
+       help
+       SNMP is the network management protocol, usefull for monitoring network devices
+
+choice
+	prompt "L2TP protocol support"
+	default CONFIG_L2TP
+	optional
+	help
+	L2TP is the protocol used by some ISPs
+
+config CONFIG_L2TP
+       bool "rp-l2tp"
+
+config CONFIG_XL2TPD
+       bool "xl2tpd"
+
+endchoice
+
+config CONFIG_IGMPPROXY
+       bool "IGMPproxy"
+       default y
+       help
+       IGMPproxy is a simple dynamic Multicast Routing Daemon
+       using only IGMP signalling (Internet Group Management Protocol). 
+       It's intended for simple forwarding of Multicast traffic 
+       between networks.
+
+config CONFIG_SCSIIDLE
+       bool "SCSI start & stop"
+       default y
+       help
+       Small programs to start and stop SCSI disks.
+
+config CONFIG_USBMODESWITCH
+       bool "USB_Modeswitch"
+       depends on CONFIG_LIBUSB
+       default y
+       help
+       Mode switching tool for controlling "flip flop" (multiple device) USB gear.
+
+config CONFIG_MADWIMAX
+       bool "WiMAX modem support"
+       depends on CONFIG_LIBUSB
+       default y
+       help
+       Include driver for WiMAX modems (Yota).
+
+endmenu
+
 menu "Additional C libraries"
 
 config CONFIG_LIBCRYPT
@@ -269,6 +425,12 @@
        Contains code for "standard" functions used in many different
        Unix utilities.
 
+config CONFIG_LIBUSB
+       bool "libusb"
+       default n
+       help
+       Library for interfacing to the USB subsystem.
+
 endmenu
 
 menu "Environment"
@@ -297,10 +459,26 @@
        help
        Specify the location of the target /usr/lib directory.
 
+config EXTRACFLAGS
+       string "Extra C compiler flags"
+       default ""
+       help
+       Specify the extra C compiler flags (if necessary).
+
 endmenu
 
 menu "Internal Options"
 
+config CONFIG_SENTRY5
+       bool "Sentry5 Features Enable"
+       default y
+       help
+       Answer Y for the Sentry5 Features compilation enable.
 
+config CONFIG_ROBO
+       bool "Robo switch configuration"
+       default y
+       help
+       Answer Y for the user mode Robo switch configuration utility.
 
 endmenu
diff -BurN -x'*.o' -x'*.orig' router/config/defconfig gateway/config/defconfig
--- router/config/defconfig	2004-10-14 19:11:24.000000000 +0000
+++ gateway/config/defconfig	2009-09-28 16:49:52.000000000 +0000
@@ -1,7 +1,6 @@
 #
 # Automatically generated make config: don't edit
 #
-CONFIG_NAT=y
 
 #
 # Base Features
@@ -9,31 +8,30 @@
 CONFIG_RC=y
 CONFIG_NVRAM=y
 CONFIG_SHARED=y
-CONFIG_LIBBCM=y
+# CONFIG_LIBBCM is not set
 CONFIG_BUSYBOX=y
-CONFIG_BUSYBOX_CONFIG=router
+CONFIG_BUSYBOX_CONFIG=defconfig
 CONFIG_WLCONF=y
 CONFIG_BRIDGE=y
+# CONFIG_VLAN is not set
 CONFIG_HTTPD=y
 CONFIG_WWW=y
+CONFIG_NAT=y
 CONFIG_NETCONF=y
-# CONFIG_IPTABLES is not set
-CONFIG_LIBIPT=y
+CONFIG_IPTABLES=y
+# CONFIG_LIBIPT is not set
 # CONFIG_GLIBC is not set
 CONFIG_UCLIBC=y
-CONFIG_LIBOPT=y
-CONFIG_VLAN=y
+# CONFIG_LIBOPT is not set
 
 #
 # Options
 #
-CONFIG_VENDOR=broadcom
-CONFIG_UDHCPD=y
+CONFIG_VENDOR=asus
 CONFIG_PPP=y
 CONFIG_UPNP=y
 CONFIG_NAS=y
-CONFIG_SES=y
-CONFIG_EZC=y
+# CONFIG_EZC is not set
 CONFIG_NTP=y
 CONFIG_DNSMASQ=y
 CONFIG_UTILS=y
@@ -42,15 +40,53 @@
 # CONFIG_WCN is not set
 
 #
+# Asus stuff
+#
+# CONFIG_PPTP is not set
+CONFIG_ACCEL_PPTP=y
+CONFIG_PPPOERELAY=y
+CONFIG_INFOSVR=y
+CONFIG_EZIPUPDATE=y
+CONFIG_VSFTPD=y
+CONFIG_JPEG6B=y
+CONFIG_RCAMD=y
+CONFIG_SENDMAIL=y
+CONFIG_LPRNG=y
+CONFIG_OTHERS=y
+# CONFIG_TEST is not set
+CONFIG_BPALOGIN=y
+CONFIG_WAVESERVER=y
+
+#
+# Custom firmware features
+#
+CONFIG_LOADER=y
+CONFIG_SSHD=y
+CONFIG_JETDIRECT=y
+CONFIG_SAMBA=y
+CONFIG_NFSD=y
+CONFIG_NFS=y
+CONFIG_IPROUTE2=y
+CONFIG_IPV6=y
+CONFIG_SNMP=y
+CONFIG_L2TP=y
+# CONFIG_XL2TPD is not set
+CONFIG_IGMPPROXY=y
+CONFIG_SCSIIDLE=y
+CONFIG_USBMODESWITCH=y
+CONFIG_MADWIMAX=y
+
+#
 # Additional C libraries
 #
-# CONFIG_LIBCRYPT is not set
-# CONFIG_LIBDL is not set
-# CONFIG_LIBM is not set
-# CONFIG_LIBNSL is not set
-# CONFIG_LIBPTHREAD is not set
-# CONFIG_LIBRESOLV is not set
-# CONFIG_LIBUTIL is not set
+CONFIG_LIBCRYPT=y
+CONFIG_LIBDL=y
+CONFIG_LIBM=y
+CONFIG_LIBNSL=y
+CONFIG_LIBPTHREAD=y
+CONFIG_LIBRESOLV=y
+CONFIG_LIBUTIL=y
+CONFIG_LIBUSB=y
 
 #
 # Environment
@@ -59,7 +95,10 @@
 LINUXDIR=$(SRCBASE)/linux/linux
 LIBDIR=$(TOOLCHAIN)/lib
 USRLIBDIR=$(TOOLCHAIN)/usr/lib
+EXTRACFLAGS=-mips32 -mtune=mips32 -Wno-pointer-sign 
 
 #
 # Internal Options
 #
+# CONFIG_SENTRY5 is not set
+# CONFIG_ROBO is not set
diff -BurN -x'*.o' -x'*.orig' router/config/lkc.h gateway/config/lkc.h
--- router/config/lkc.h	2004-10-14 19:11:24.000000000 +0000
+++ gateway/config/lkc.h	2006-12-16 16:30:02.000000000 +0000
@@ -55,9 +55,6 @@
 struct file *file_lookup(const char *name);
 int file_write_dep(const char *name);
 
-extern struct menu *current_entry;
-extern struct menu *current_menu;
-
 /* symbol.c */
 void sym_init(void);
 void sym_clear_all_valid(void);
diff -BurN -x'*.o' -x'*.orig' router/config/menu.c gateway/config/menu.c
--- router/config/menu.c	2004-10-14 19:11:25.000000000 +0000
+++ gateway/config/menu.c	2006-12-16 16:30:02.000000000 +0000
@@ -10,7 +10,6 @@
 #include "lkc.h"
 
 struct menu rootmenu;
-struct menu *current_menu, *current_entry;
 static struct menu **last_entry_ptr;
 
 struct file *file_list;
diff -BurN -x'*.o' -x'*.orig' router/config/zconf.tab.c_shipped gateway/config/zconf.tab.c_shipped
--- router/config/zconf.tab.c_shipped	2004-10-14 19:11:25.000000000 +0000
+++ gateway/config/zconf.tab.c_shipped	2006-12-16 16:30:02.000000000 +0000
@@ -160,6 +160,8 @@
 
 struct symbol *symbol_hash[257];
 
+static struct menu *current_menu, *current_entry;
+
 #define YYERROR_VERBOSE
 
 
diff -BurN -x'*.o' -x'*.orig' router/config/zconf.y gateway/config/zconf.y
--- router/config/zconf.y	2004-10-14 19:11:25.000000000 +0000
+++ gateway/config/zconf.y	2006-12-16 16:30:02.000000000 +0000
@@ -25,6 +25,8 @@
 
 struct symbol *symbol_hash[257];
 
+static struct menu *current_menu, *current_entry;
+
 #define YYERROR_VERBOSE
 %}
 %expect 36
