diff -BurN router/httpd/broadcom.c gateway/httpd/broadcom.c
--- router/httpd/broadcom.c	2004-11-11 12:27:09.000000000 +0500
+++ gateway/httpd/broadcom.c	2008-01-06 16:26:57.000000000 +0500
@@ -592,6 +592,7 @@
 #endif
 
 
+#ifdef REMOVE
 static int
 ej_wl_parse_str(int eid, webs_t wp, int argc, char_t **argv) 
 {
@@ -624,25 +625,30 @@
 
 	return websWrite(wp, "%u", val);
 }
-
+#endif
 
 int
 ej_wl_sta_status(int eid, webs_t wp, char *ifname)
 {
-	int ret, i;
-	char bssid[32];
-	char bssinfobuf[2000];
-	wl_bss_info_t *info;
-	int val;
+	int ret;
+	unsigned char bssid[32];
 
 	// Get bssid
-	ret=wl_ioctl(ifname, WLC_GET_BSSID, bssid, sizeof(bssid));
-	//wl_scan_results();
+	ret = wl_ioctl(ifname, WLC_GET_BSSID, bssid, sizeof(bssid));
 
-	if (ret==0 && !(bssid[0]==0&&bssid[1]==0&&bssid[2]==0
-		&&bssid[3]==0&&bssid[4]==0&&bssid[5]==0)) 	
+	if (ret == 0 && memcmp(bssid, "\x00\x00\x00\x00\x00", 6))
 	{
-		return(websWrite(wp, "Status	: Connect to %s\n", nvram_safe_get("wl0_ssid")));
+		uint32 rssi;
+		
+		if (wl_ioctl(ifname, WLC_GET_RSSI, &rssi, sizeof(rssi)))
+			return(websWrite(wp, "Status	: Connected to %s\n"
+				     "AP	: %02x:%02x:%02x:%02x:%02x:%02x\n", 
+				nvram_safe_get("wl0_ssid"), 
+				bssid[0], bssid[1], bssid[2], bssid[3], bssid[4], bssid[5]));
+		else	return(websWrite(wp, "Status	: Connected to %s\n"
+				     "AP	: %02x:%02x:%02x:%02x:%02x:%02x\n"
+				     "Signal	: %d dBm\n",  nvram_safe_get("wl0_ssid"), 
+				bssid[0], bssid[1], bssid[2], bssid[3], bssid[4], bssid[5], rssi));
 	}
 	return(websWrite(wp, "Status	: Connecting to %s\n", nvram_safe_get("wl0_ssid")));
 }
@@ -693,7 +699,7 @@
 	}
 	else if (nvram_match(strcat_r(prefix, "mode", tmp), "sta"))
 	{
-		ret+=websWrite(wp, "Mode	: Stations\n");
+		ret+=websWrite(wp, "Mode	: Station\n");
 		ret+=websWrite(wp, "Channel	: %d\n", ci.target_channel);
 		ret+=ej_wl_sta_status(eid, wp, name);
 		return ret;
@@ -891,17 +897,17 @@
 
 			if (strstr(buff, "br0"))
 			{
-				ret += websWrite(wp, "%-16s%-16s%-16s%-6s%-6d %-2d %7d LAN\n",
+				ret += websWrite(wp, "%-16s%-16s%-16s%-6s%-6d %-2d %7d LAN %s\n",
 				sdest, sgw,
 				inet_ntoa(mask),
-				flags, metric, ref, use);
+				flags, metric, ref, use, buff);
 			}
 			else if(!strstr(buff, "lo"))
 			{
-				ret += websWrite(wp, "%-16s%-16s%-16s%-6s%-6d %-2d %7d WAN\n",
+				ret += websWrite(wp, "%-16s%-16s%-16s%-6s%-6d %-2d %7d WAN %s\n",
 				sdest, sgw,
 				inet_ntoa(mask),
-				flags, metric, ref, use);
+				flags, metric, ref, use, buff);
 			}
 		}
 		nl++;
diff -BurN router/httpd/httpd.c gateway/httpd/httpd.c
--- router/httpd/httpd.c	2004-11-11 12:27:08.000000000 +0500
+++ gateway/httpd/httpd.c	2008-01-06 16:26:57.000000000 +0500
@@ -97,6 +97,7 @@
 int redirect = 1;
 char wan_if[16];
 int http_port=SERVER_PORT;
+static int server_port = SERVER_PORT; /* Port for SERVER USER interface */
 
 /* Added by Joey for handle one people at the same time */
 unsigned int login_ip=0;
@@ -413,10 +414,10 @@
 
     method = path = line;
     strsep(&path, " ");
-    while (*path == ' ') path++;
+    while (path && *path == ' ') path++;
     protocol = path;
     strsep(&protocol, " ");
-    while (*protocol == ' ') protocol++;
+    while (protocol && *protocol == ' ') protocol++;
     cp = protocol;
     strsep(&cp, " ");
     if ( !method || !path || !protocol ) {
@@ -474,7 +475,7 @@
 
     //printf("File: %s\n", file);
 
-    if (http_port==SERVER_PORT && !http_login_check())
+    if (http_port==server_port && !http_login_check())
     {
 	sprintf(line, "Please log out user, %s, first or wait for session timeout(60 seconds).", inet_ntoa(login_ip));
 
@@ -485,7 +486,7 @@
 	
     if ( (file[0] == '\0' || file[len-1] == '/'))
     {
-	if (strlen(wan_if)>0 && !is_connected() && !is_firsttime() && http_port==SERVER_PORT)
+	if (strlen(wan_if)>0 && !is_connected() && !is_firsttime() && http_port==server_port)
 	{
 		redirect=1;
 		file="WizardDetect.asp";
@@ -560,7 +561,7 @@
 http_login(unsigned int ip)
 {
 
-	if (http_port!=SERVER_PORT || ip == 0x100007f) return;
+	if (http_port!=server_port || ip == 0x100007f) return;
 
 	login_ip = ip;
 	login_try = 0;
@@ -571,7 +572,7 @@
 {
 	time_t now;
 	
-	if (http_port!=SERVER_PORT || login_ip_tmp == 0x100007f) return 1;
+	if (http_port!=server_port || login_ip_tmp == 0x100007f) return 1;
 
 	http_login_timeout(login_ip_tmp);
 
@@ -593,7 +594,7 @@
 
 	//printf("login : %x %x\n", now, login_timestamp);
 
-	if ((unsigned long)(now - login_timestamp)>60) //one minitues
+	if (ip != login_ip && (unsigned long)(now - login_timestamp)>60) //one minitues
 	{
 		http_logout(login_ip);
 	}
@@ -613,7 +614,7 @@
 
 int is_auth(void)
 {
-	if (http_port==SERVER_PORT ||
+	if (http_port==server_port ||
 		strcmp(nvram_get_x("PrinterStatus", "usb_webhttpcheck_x"), "1")==0) return 1;
 	else return 0;
 }
@@ -623,6 +624,9 @@
 	int fd, err;
 	struct ifreq ifr;
 	struct ethtool_cmd ecmd;
+	
+	if (strstr(wan_if, "vlan"))
+		return 1;
 
 	memset(&ifr, 0, sizeof(ifr));
 
@@ -671,7 +675,7 @@
 	char line[128], *reason;
 
 	/* check if physical connection exist */
-	if (!is_phyconnected()) return 0;
+	/* if (!is_phyconnected()) return 0; */
 	
 	/* check if connection static is CONNECTED */
 	if (strcmp(nvram_get_x("WANIPAddress", "wan_status_t"), "Disconnected")==0)
@@ -705,6 +709,14 @@
 	socklen_t sz = sizeof(usa);
 	char pidfile[32];
 
+	server_port = atoi(nvram_get_x("", "http_lanport"));
+    if (server_port)
+      http_port = server_port;
+    else
+      server_port = http_port;
+      
+    
+    
 	// Added by Joey for handling WAN Interface 
 	// usage: httpd [wan interface] [port]
 	if (argc>2) http_port=atoi(argv[2]);
@@ -727,7 +739,7 @@
 	{
 	FILE *pid_fp;
 	/* Daemonize and log PID */
-	//if (http_port==SERVER_PORT)
+	//if (http_port==server_port)
 	//{
 		if (daemon(1, 1) == -1) 
 		{
@@ -735,7 +747,7 @@
 			exit(errno);
 		}
 	//}
-	if (http_port==SERVER_PORT)
+	if (http_port==server_port)
 		strcpy(pidfile, "/var/run/httpd.pid");
 	else sprintf(pidfile, "/var/run/httpd-%d.pid", http_port);
 
diff -BurN router/httpd/Makefile gateway/httpd/Makefile
--- router/httpd/Makefile	2004-11-11 12:27:08.000000000 +0500
+++ gateway/httpd/Makefile	2008-01-06 16:26:57.000000000 +0500
@@ -8,7 +8,7 @@
 
 CFLAGS	+= -I. -I$(TOP)/shared -I$(SRCBASE)/include -Wall -DFLASH_EMULATOR -DLinux -DMIPS
 #CFLAGS	+= -g -DDEBUG
-CFLAGS	+= -s -O2
+CFLAGS	+= -s -O2 $(EXTRACFLAGS)
 LDFLAGS	+= -L$(TOP)/nvram -L$(TOP)/shared -L$(TOP)/netconf
 
 vpath %.c $(TOP)/shared
@@ -24,5 +24,7 @@
 	install -D httpd $(INSTALLDIR)/usr/sbin/httpd
 	$(STRIP) $(INSTALLDIR)/usr/sbin/httpd
 
+common.o: common.c variables.c
+
 httpd: httpd.o ej.o cgi.o web_ex.o shutils.o common.o nvram_f.o nvmap.o broadcom.o ralink.o
 	$(CC) -o $@ $^ $(LDFLAGS) -lnvram -lshared -lnetconf
diff -BurN router/httpd/nvmap.c gateway/httpd/nvmap.c
--- router/httpd/nvmap.c	2004-11-11 12:27:09.000000000 +0500
+++ gateway/httpd/nvmap.c	2008-06-14 21:30:09.000000000 +0600
@@ -1,5 +1,6 @@
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <bcmnvram.h>
 
 
@@ -99,6 +100,8 @@
 { "Storage", "x_AccUser6_Write", "sh_accuser_write_x6"},
 { NULL, NULL, NULL }};
 
+char *mac_conv(char *mac_name, int idx, char *buf);
+
 /* This function is used to map nvram value from asus to Broadcom */
 void readFromNvram()
 {	
diff -BurN router/httpd/shutils.c gateway/httpd/shutils.c
--- router/httpd/shutils.c	2004-11-11 12:27:08.000000000 +0500
+++ gateway/httpd/shutils.c	2008-01-06 16:26:57.000000000 +0500
@@ -125,6 +125,8 @@
 		close(STDIN_FILENO);
 		setsid();
 
+		fd = open("/dev/null", O_RDWR); /* stdin */
+		
 		/* Redirect stdout to <path> */
 		if (path) {
 			flags = O_WRONLY | O_CREAT;
diff -BurN router/httpd/variables.c gateway/httpd/variables.c
--- router/httpd/variables.c	2004-11-15 13:06:01.000000000 +0500
+++ gateway/httpd/variables.c	2009-02-22 18:18:44.000000000 +0500
@@ -455,6 +455,8 @@
                 {"dhcp_staticmac_x", "14", validate_hwaddr, NULL, FALSE, 0x0},
             
                 {"dhcp_staticip_x", "15", validate_ipaddr, NULL, FALSE, 0x0},
+            
+                {"dhcp_staticname_x", "32", validate_string, ARGV("32"), FALSE, 0x0},
                             
       {0,0,0,0,0}
       };      
@@ -477,7 +479,9 @@
         
       struct variable variables_DeviceSecurity11a_ACLList[] = {          
       
-                {"wl_maclist_x", "32", validate_hwaddr, NULL, FALSE, 0x0},
+                {"wl_maclist_x", "20", validate_hwaddr, NULL, FALSE, 0x0},
+      
+                {"wl_macdesc_x", "40", validate_string,  ARGV("80"), FALSE, 0x0},
                             
       {0,0,0,0,0}
       };      
@@ -534,7 +538,9 @@
         
       struct variable variables_DeviceSecurity11b_ACLList[] = {          
       
-                {"wl_maclist_x", "32", validate_hwaddr, NULL, FALSE, 0x0},
+                {"wl_maclist_x", "20", validate_hwaddr, NULL, FALSE, 0x0},
+      
+                {"wl_macdesc_x", "40", validate_string,  ARGV("80"), FALSE, 0x0},
                             
       {0,0,0,0,0}
       };      
@@ -588,8 +594,6 @@
                                                          
                     {"usb_ftppasswd_x", "18", validate_string, ARGV("16"), FALSE, 0x0},                                                                                                                     
                               
-             {"usb_ftpmaxuser_x", "7", validate_range, ARGV("0", "12", ""), FALSE , 0x0},                                          	                                                                      
-        	              
               {"usb_ftprights_x", "16", validate_choice, ARGV(                
               
                    "Read/Write/Erase",
@@ -616,6 +620,26 @@
       {0,0,0,0,0}
       };      
         
+      struct variable variables_PrinterStatus_x_ExportsList[] = {          
+      
+                {"usb_nfslist_x", "80", validate_string, ARGV("80"), FALSE, 0x0},
+                            
+      {0,0,0,0,0}
+      };      
+        
+      struct variable variables_PrinterStatus_x_SharesList[] = {          
+      
+                {"usb_smbdir_x", "20", validate_string, ARGV("80"), FALSE, 0x0},
+      
+                {"usb_smbshare_x", "20", validate_string, ARGV("80"), FALSE, 0x0},
+      
+                {"usb_smblevel_x", "14", validate_choice, ARGV("Read Only", "Read/Write", 0), FALSE, 0x0},
+      
+                {"usb_smbdesc_x", "20", validate_string, ARGV("80"), FALSE, 0x0},
+                            
+      {0,0,0,0,0}
+      };      
+        
       struct variable variables_PrinterStatus_x_QRuleList[] = {          
       
                 {"qos_ipaddr_x", "16", validate_ipaddr, NULL, FALSE, 0x0},
@@ -957,6 +981,10 @@
                     
                 {"wan_pppoe_txonly_x", "", validate_string, ARGV(""), FALSE, FALSE}, 
                          
+                {"wan_pppoe_options_x", "", validate_string, ARGV("255"), FALSE, FALSE}, 
+                         
+                {"wan_pptp_options_x", "", validate_string, ARGV("255"), FALSE, FALSE}, 
+                         
              {"wan_pppoe_mtu", "", validate_range, ARGV("576", "1492", ""), FALSE, FALSE},
                      
              {"wan_pppoe_mru", "", validate_range, ARGV("576", "1492", ""), FALSE, FALSE},
@@ -1074,6 +1102,8 @@
                        {"", "", validate_range, ARGV("0","65535"), FALSE, FALSE},                                                                                            
                                      
                  {"fw_enable_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
+                                     
+                 {"fw_dos_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
               	
               {"fw_log_x", "", validate_choice, ARGV(              
               
@@ -1097,6 +1127,13 @@
                                
                  {"misc_ping_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
                                
+             {"misc_conntrack_x", "", validate_range, ARGV("1024", "16384", ""), FALSE, FALSE},
+
+	{"recent_ssh_enable", "", validate_range, ARGV("0", "1"), FALSE, FALSE},
+	{"recent_ftp_enable", "", validate_range, ARGV("0", "1"), FALSE, FALSE},
+	{"recent_seconds", "", validate_range, ARGV("1", "65535"), FALSE, FALSE},
+	{"recent_hitcount", "", validate_range, ARGV("1", "65535"), FALSE, FALSE},
+
                  {"fw_wl_enable_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
               
                 {"filter_wl_date_x", "", validate_portrange, NULL, FALSE, FALSE},
@@ -1110,6 +1147,14 @@
                    "ACCEPT",
               
               0), FALSE, FALSE},
+            	
+              {"filter_vs_default_x", "", validate_choice, ARGV(              
+              
+                   "DROP",
+              
+                   "ACCEPT",
+              
+              0), FALSE, FALSE},
            
                 {"filter_wl_icmp_x", "", validate_portrange, NULL, FALSE, FALSE},
             
@@ -1318,6 +1363,10 @@
    
       struct variable variables_RouterConfig[] = {
                        
+                 {"dr_enable_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
+                       
+                 {"mr_enable_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
+                       
                  {"sr_enable_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
               	
               {"sr_rip_x", "", validate_choice, ARGV(              
@@ -1470,7 +1519,7 @@
                                              
                        {"", "", validate_range, ARGV("0","65535"), FALSE, FALSE},                                                                                            
                      
-      {"GWStatic", "Group", validate_group, ARGV(variables_RouterConfig_GWStatic, "6", "59", "sr_num_x"), FALSE, FALSE},     
+      {"GWStatic", "Group", validate_group, ARGV(variables_RouterConfig_GWStatic, "16", "59", "sr_num_x"), FALSE, FALSE},     
        
       {"StaticRoute", "Group", validate_group, ARGV(variables_RouterConfig_StaticRoute, "16", "46", "dr_staticnum_x"), FALSE, FALSE},     
        
@@ -1502,6 +1551,8 @@
                     {"lan_hostname", "", validate_string, ARGV("32"), FALSE, FALSE},                                                                                                                   
                  
                 {"lan_gateway", "", validate_ipaddr, NULL, FALSE, FALSE},
+                 
+                {"lan_dns", "", validate_ipaddr, NULL, FALSE, FALSE},
                              
                  {"dhcp_enable_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
                                
@@ -1555,7 +1606,8 @@
                     
       {"DmzDHCPLog", "Status", NULL, ARGV("dleases.log",""), FALSE, FALSE},
                  
-                 {"upnp_enable", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
+                 {"upnp_enable", "", validate_range, ARGV("0","2"), FALSE, FALSE},                                                       
+		{"udpxy_enable_x", "", validate_range, ARGV("0","65535"), FALSE, FALSE},
               
                 {"log_ipaddr", "", validate_ipaddr, NULL, FALSE, FALSE},
             	
@@ -1595,7 +1647,7 @@
               
                    "GMT0BST:(GMT) England",
               
-                   "MET-1METDST:(GMT+01:00) Netherland, France, Italy",
+                   "MET-1METDST,M3.5.0/2,M10.5.0/3:(GMT+01:00) Czech Republic, Netherland, France, Italy",
               
                    "MEZ-1MESZ:(GMT+01:00) Germany",
               
@@ -1623,15 +1675,23 @@
               
                    "SST-8:(GMT+08:00) Singapore",
               
-                   "WAS-8WAD:(GMT+08:00) Australia West",
+                   "AWST-8:(GMT+08:00) Australia (WA)",
               
                    "JST-9:(GMT+09:00) Japan, Korea",
               
                    "KST-9KDT:(GMT+09:00) Korean",
               
+                   "ACST-9:30:(GMT+09:30) Australia (NT)",
+              
+                   "ACST-9:30ACDT-10:30,M10.5.0,M3.5.0:(GMT+09:30) Australia (SA)",
+              
                    "UCT-10:(GMT+10:00) Guam, Russia",
               
-                   "EST-10EDT:(GMT+10:00) Australia",
+                   "AEST-10:(GMT+10:00) Australia (QLD)",
+              
+                   "AEST-10AEDT-11,M10.1.0,M3.5.0:(GMT+10:00) Australia (TAS)",
+              
+                   "AEST-10AEDT-11,M10.5.0,M3.5.0:(GMT+10:00) Australia (NSW,ACT,VIC)",
               
                    "UCT-11:(GMT+11:00) Solomon Islands",
               
@@ -1646,6 +1706,8 @@
                     {"ntp_server0", "", validate_string, ARGV(""), FALSE, FALSE},                                                                                                                   
                                                          
                     {"ntp_server1", "", validate_string, ARGV(""), FALSE, FALSE},                                                                                                                   
+
+             {"ntp_interval_x", "", validate_range, ARGV("1", "144"), FALSE, FALSE},
                                   
                  {"", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
               
@@ -1677,7 +1739,12 @@
               
       {"x_DDNSStatus", "Status", NULL, ARGV("ddns.log","DDNSStatus"), FALSE, FALSE},
  
-      {"ManualDHCPList", "Group", validate_group, ARGV(variables_LANHostConfig_ManualDHCPList, "8", "29", "dhcp_staticnum_x"), FALSE, FALSE},     
+	{"snmp_enable", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
+	{"snmp_community", "", validate_string, ARGV("32"), FALSE,FALSE},
+	{"snmp_contact", "", validate_string, ARGV("80"), FALSE,FALSE},
+	{"snmp_location", "", validate_string, ARGV("80"), FALSE,FALSE},
+	
+      {"ManualDHCPList", "Group", validate_group, ARGV(variables_LANHostConfig_ManualDHCPList, "32", "61", "dhcp_staticnum_x"), FALSE, FALSE},     
        
       {"DmzManualDHCPList", "Group", validate_group, ARGV(variables_LANHostConfig_DmzManualDHCPList, "8", "29", "DmzManualDHCPCount"), FALSE, FALSE},     
        
@@ -1793,7 +1860,7 @@
                     
                 {"", "", validate_string, ARGV(""), FALSE, FALSE}, 
              
-      {"ACLList", "Group", validate_group, ARGV(variables_DeviceSecurity11a_ACLList, "64", "32", "wl_macnum_x"), FALSE, FALSE},     
+      {"ACLList", "Group", validate_group, ARGV(variables_DeviceSecurity11a_ACLList, "64", "60", "wl_macnum_x"), FALSE, FALSE},     
        
       {"AESList", "Group", validate_group, ARGV(variables_DeviceSecurity11a_AESList, "64", "58", "AESListCount"), FALSE, FALSE},     
                         
@@ -1869,9 +1936,11 @@
               
                    "1:Auto",
               
-                   "4:54G Only",
+                   "4:Performance",
               
                    "0:802.11B Only",
+
+                   "5:54G LRS",
               
               0), FALSE, FALSE},
            
@@ -1890,6 +1955,16 @@
                    "radius:Radius with 802.1x",
               
               0), FALSE, FALSE},
+
+              {"wl_wpa_mode", "", validate_choice, ARGV(
+
+                   "0:WPA-Auto-Personal",
+
+                   "1:WPA-Personal",
+
+                   "2:WPA2-Personal",
+
+              0), FALSE, FALSE},
            	
               {"wl_crypto", "", validate_choice, ARGV(              
               
@@ -1980,6 +2059,36 @@
                    "54000000:54",
               
               0), FALSE, FALSE},
+              	
+              {"wl_mrate", "", validate_choice, ARGV(              
+              
+                   "0:Auto",
+              
+                   "1000000:1",
+              
+                   "2000000:2",
+              
+                   "5500000:5.5",
+              
+                   "6000000:6",
+              
+                   "9000000:9",
+              
+                   "11000000:11",
+              
+                   "12000000:12",
+              
+                   "18000000:18",
+              
+                   "24000000:24",
+              
+                   "36000000:36",
+              
+                   "48000000:48",
+              
+                   "54000000:54",
+              
+              0), FALSE, FALSE},
            	
               {"wl_rateset", "", validate_choice, ARGV(              
               
@@ -1990,6 +2099,16 @@
                    "12:1, 2 Mbps",
               
               0), FALSE, FALSE},
+           	
+              {"wl_reg_mode", "", validate_choice, ARGV(              
+              
+                   "off:Off",
+              
+                   "d:802.11d",
+              
+                   "h:802.11h",
+              
+              0), FALSE, FALSE},
                         
              {"wl_frag", "", validate_range, ARGV("256", "2346", ""), FALSE, FALSE},
                      
@@ -2064,6 +2183,16 @@
                  {"wl_preauth", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
                                                       
                     {"wl_net_reauth", "", validate_string, ARGV(""), FALSE, FALSE},                                                                                                                   
+
+		{"wl_nbw", "20", validate_string, ARGV(""), FALSE, FALSE},
+		{"wl_nctrlsb", "none", validate_string, ARGV(""), FALSE, FALSE},
+		{"wl_nband", "2", validate_string, ARGV(""), FALSE, FALSE},
+		{"wl_nmcsidx", "-1", validate_string, ARGV(""), FALSE, FALSE},
+		{"wl_nmode", "-1", validate_string, ARGV(""), FALSE, FALSE},
+		{"wl_leddc", "0x640000", validate_string, ARGV(""), FALSE, FALSE},
+		{"wl_wme_apsd", "on", validate_string, ARGV(""), FALSE, FALSE},
+		{"wl_sta_retry_time", "5", validate_string, ARGV(""), FALSE, FALSE},
+
                   
       {"RBRList", "Group", validate_group, ARGV(variables_WLANConfig11b_RBRList, "16", "32", "wl_wdsnum_x"), FALSE, FALSE},     
                         
@@ -2100,7 +2229,7 @@
                     
                 {"", "", validate_string, ARGV(""), FALSE, FALSE}, 
              
-      {"ACLList", "Group", validate_group, ARGV(variables_DeviceSecurity11b_ACLList, "64", "32", "wl_macnum_x"), FALSE, FALSE},     
+      {"ACLList", "Group", validate_group, ARGV(variables_DeviceSecurity11b_ACLList, "64", "60", "wl_macnum_x"), FALSE, FALSE},     
        
       {"AESList", "Group", validate_group, ARGV(variables_DeviceSecurity11b_AESList, "64", "58", "AESListCount"), FALSE, FALSE},     
                         
@@ -2256,21 +2385,44 @@
             
       {"x_FEject", "Status", NULL, ARGV("ddns.log","DDNSStatus"), FALSE, FALSE},
                  
-                 {"usb_ftpenable_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
+                 {"usb_storage_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
+                 {"usb_smbenable_x", "", validate_range, ARGV("0","2"), FALSE, FALSE},                                                       
+                 {"usb_smbhidden_x", "", validate_range, ARGV("0","2"), FALSE, FALSE},                                                       
+                 {"usb_smbwrkgrp_x", "", validate_string, ARGV(""), FALSE, FALSE},                                                       
+                 {"usb_smbcpage_x", "", validate_string, ARGV(""), FALSE, FALSE},                                                       
+                 {"x_SharesList", "Group", validate_group, ARGV(variables_PrinterStatus_x_SharesList, "32", "80", "usb_smbnum_x"), FALSE, FALSE},                                    
+                 
+                 {"usb_ftpenable_x", "", validate_range, ARGV("0","2"), FALSE, FALSE},                                                       
                                
-                 {"usb_ftpanonymous_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
+                 {"usb_ftpanonymous_x", "", validate_range, ARGV("0","3"), FALSE, FALSE},                                                       
                                
                  {"usb_ftpsuper_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
                            
              {"usb_ftpport_x", "", validate_range, ARGV("1", "65535", ""), FALSE, FALSE},
                      
-             {"usb_ftpmax_x", "", validate_range, ARGV("1", "12", ""), FALSE, FALSE},
+             {"usb_ftppubroot_x", "", validate_string, ARGV(""), FALSE, FALSE},                                                    
+                     
+             {"usb_ftppvtroot_x", "", validate_string, ARGV(""), FALSE, FALSE},                                                    
+                     
+             {"usb_ftpanonroot_x", "", validate_string, ARGV(""), FALSE, FALSE},                                                    
+                     
+             {"usb_ftpdirlist_x", "", validate_range, ARGV("0","2"), FALSE, FALSE},
+
+             {"usb_ftpmax_x", "", validate_range, ARGV("0", "12", ""), FALSE, FALSE},
+                     
+             {"usb_ftpipmax_x", "", validate_range, ARGV("0", "12", ""), FALSE, FALSE},
                                  
                        {"usb_ftptimeout_x", "", validate_range, ARGV("0","65535"), FALSE, FALSE},                                                                                            
                                              
                        {"usb_ftpstaytimeout_x", "", validate_range, ARGV("0","65535"), FALSE, FALSE},                                                                                            
                                                             
                     {"usb_ftpscript_x", "", validate_string, ARGV(""), FALSE, FALSE},                                                                                                                   
+                                                            
+                    {"usb_ftpscript_x", "", validate_string, ARGV(""), FALSE, FALSE},                                                                                                                   
+                     
+             {"usb_ftprate_x", "", validate_range, ARGV("0", "99999", ""), FALSE, FALSE},
+                     
+             {"usb_ftpanonrate_x", "", validate_range, ARGV("0", "99999", ""), FALSE, FALSE},
                  	
               {"", "", validate_choice, ARGV(              
               
@@ -2296,14 +2448,27 @@
                                              
                        {"qos_urulenum_x", "", validate_range, ARGV("0","65535"), FALSE, FALSE},                                                                                            
                      
-      {"x_FUserList", "Group", validate_group, ARGV(variables_PrinterStatus_x_FUserList, "32", "59", "usb_ftpnum_x"), FALSE, FALSE},     
-       
-      {"x_FBanIPList", "Group", validate_group, ARGV(variables_PrinterStatus_x_FBanIPList, "16", "24", "usb_bannum_x"), FALSE, FALSE},     
+      {"x_FUserList", "Group", validate_group, ARGV(variables_PrinterStatus_x_FUserList, "32", "52", "usb_ftpnum_x"), FALSE, FALSE},     
        
-      {"x_QRuleList", "Group", validate_group, ARGV(variables_PrinterStatus_x_QRuleList, "8", "44", "qos_rulenum_x"), FALSE, FALSE},     
+      {"x_FBanIPList", "Group", validate_group, ARGV(variables_PrinterStatus_x_FBanIPList, "16", "24", "usb_bannum_x"), FALSE, FALSE},     
+      
+      {"usb_nfsenable_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},                                                       
+
+      {"x_ExportsList", "Group", validate_group, ARGV(variables_PrinterStatus_x_ExportsList, "32", "80", "usb_nfsnum_x"), FALSE, FALSE},     
+              
+      {"x_QRuleList", "Group", validate_group, ARGV(variables_PrinterStatus_x_QRuleList, "32", "44", "qos_rulenum_x"), FALSE, FALSE},     
        
-      {"x_UQRuleList", "Group", validate_group, ARGV(variables_PrinterStatus_x_UQRuleList, "8", "28", "qos_urulenum_x"), FALSE, FALSE},     
+      {"x_UQRuleList", "Group", validate_group, ARGV(variables_PrinterStatus_x_UQRuleList, "32", "28", "qos_urulenum_x"), FALSE, FALSE},     
                         
+	{"telnet_enable", "", validate_range, ARGV("0","1"), FALSE, FALSE},
+	{"ssh_enable", "", validate_range, ARGV("0","1"), FALSE, FALSE},
+	{"ssh_port", "", validate_range, ARGV("1", "65535", ""), FALSE, FALSE},
+	{"infosvr_enable", "", validate_range, ARGV("0","1"), FALSE, FALSE},
+	{"lpr_enable", "", validate_range, ARGV("0","1"), FALSE, FALSE},
+	{"raw_enable", "", validate_range, ARGV("0","1"), FALSE, FALSE},
+	{"audio_enable", "", validate_range, ARGV("0","1"), FALSE, FALSE},
+	{"usb20_disable_x", "", validate_range, ARGV("0","1"), FALSE, FALSE},
+	
       { 0, 0, 0, 0, 0, 0}
       };
       
@@ -4221,6 +4386,25 @@
 struct action actions_PrinterStatus[];
 struct variable variables_WLANAuthentication11b[];
 
+struct variable variables_IPv6Config[] = {
+
+	{"ipv6_lan_addr", "", validate_string, ARGV("40"), FALSE, FALSE},
+	{"ipv6_lan_netsize", "", validate_range, ARGV("1", "127"), FALSE, FALSE},
+	{"ipv6_wan_addr", "", validate_string, ARGV("40"), FALSE, FALSE},
+	{"ipv6_wan_netsize", "", validate_range, ARGV("1", "127"), FALSE, FALSE},
+	{"ipv6_wan_router", "", validate_string, ARGV("40"), FALSE, FALSE},
+	{"ipv6_sit_enable", "", validate_range, ARGV("0","1"), FALSE, FALSE},
+	{"ipv6_sit_remote", "", validate_ipaddr, NULL, FALSE, FALSE},
+	{"ipv6_sit_localaddr", "", validate_string, ARGV("40"), FALSE, FALSE},
+	{"ipv6_sit_netsize", "", validate_range, ARGV("1", "127"), FALSE, FALSE},
+	{"ipv6_sit_remoteaddr", "", validate_string, ARGV("40"), FALSE, FALSE},
+	{"ipv6_sit_mtu", "", validate_range, ARGV("0", "1480"), FALSE, FALSE},
+	{"ipv6_sit_ttl", "", validate_range, ARGV("0", "255"), FALSE, FALSE},
+	{"ipv6_radvd_enable", "", validate_range, ARGV("0", "1"), FALSE, FALSE},
+	
+	{ 0, 0, 0, 0}
+	};
+struct action actions_IPv6Config[];
 
 struct svcLink svcLinks[] = {            
            {"General", 	"urn:schemas-upnp-org:service:General:1", variables_General, actions_General},
@@ -4239,6 +4423,7 @@
            {"DeviceSecurity11b", "urn:schemas-upnp-org:service:DeviceSecurity:1", variables_DeviceSecurity11b, actions_DeviceSecurity},
            {"WLANAuthentication11b", "urn:schemas-upnp-org:service:WLANAuthentication:1", variables_WLANAuthentication11b, actions_WLANAuthentication},         
            {"PrinterStatus", "urn:schemas-upnp-org:service:PrinterStatus:1", variables_PrinterStatus, actions_PrinterStatus},
+	   {"IPv6Config", "urn::IPv6Config:1", variables_IPv6Config, actions_IPv6Config},
            {0, 0, 0}
       };
 	
diff -BurN router/httpd/web_ex.c gateway/httpd/web_ex.c
--- router/httpd/web_ex.c	2004-12-01 18:58:17.000000000 +0500
+++ gateway/httpd/web_ex.c	2008-12-12 22:49:45.000000000 +0500
@@ -78,6 +78,8 @@
 static int nvram_generate_table(webs_t wp, char *serviceId, char *groupName);
 
 static int ej_select_country(int eid, webs_t wp, int argc, char_t **argv);
+static int wl_channels_in_country(char *abbrev, int channels[]);
+static int wl_channels_in_country_asus(char *abbrev, int channels[]);
 
 char ibuf[8192];
 char ibuf2[8192];
@@ -124,7 +126,7 @@
 	static char s[] = "XXXXX days, XX hours, XX minutes, XX seconds";
 	char *c = s;
 
-#ifdef SHOWALL
+#if 1 //def SHOWALL
 	if (seconds > 60*60*24) {
 		c += sprintf(c, "%d days, ", seconds / (60*60*24));
 		seconds %= 60*60*24;
@@ -170,7 +172,10 @@
 {	
 	//printf("Redirect to : %s\n", url);	
 	websWrite(wp, T("<html><head>\r\n"));
-        websWrite(wp, T("<meta http-equiv=\"refresh\" content=\"0; url=http://%s/%s\">\r\n"), getip((FILE *)wp), url);
+	if (next_host && *next_host)
+        	websWrite(wp, T("<meta http-equiv=\"refresh\" content=\"0; url=http://%s/%s\">\r\n"), next_host, url);
+        else
+        	websWrite(wp, T("<meta http-equiv=\"refresh\" content=\"0; url=%s\">\r\n"), url);
         websWrite(wp, T("<meta http-equiv=\"Content-Type\" content=\"text/html\">\r\n"));
         websWrite(wp, T("</head></html>\r\n"));      
 	
@@ -205,7 +210,7 @@
      {
 	   if (strcmp(SystemCmd, "")!=0)
 	   {
-	   	sprintf(SystemCmd, "%s > /tmp/syscmd.log\n", SystemCmd);
+	   	sprintf(SystemCmd, "%s > /tmp/syscmd.log 2>&1\n", SystemCmd);
 	   	system(SystemCmd);
 	   }	
 	   else
@@ -229,6 +234,7 @@
      else if (strcmp(name, "lpr_remove")==0)
      {
 	   kill_pidfile_s("/var/run/lpdparent.pid", SIGUSR2);
+	   eval("killall","p910nd");
      }
      else if (strcmp(name, "wlan11a.sh")==0 || strcmp(name,"wlan11b.sh")==0)
      {
@@ -238,6 +244,11 @@
      {
 		sys_refresh_lease();	
      }
+     else if (strcmp(name,"dnsmasq.sh")==0)
+     {
+	   kill_pidfile_s("/var/run/dnsmasq.pid", SIGALRM);
+	   sleep(1);
+     }
      else if (strcmp(name,"iptable.sh")==0) 
      {
 		// TODO	
@@ -474,6 +485,25 @@
 	return 0;
 }	
 
+static int
+ej_nvram_double_match_x(int eid, webs_t wp, int argc, char_t **argv)
+{
+	char *sid, *name, *match, *output;
+	char *sid2, *name2, *match2;
+
+	if (ejArgs(argc, argv, "%s %s %s %s %s %s %s", &sid, &name, &match, &sid2, &name2, &match2, &output) < 7) {
+		websError(wp, 400, "Insufficient args\n");
+		return -1;
+	}
+	
+	if (nvram_match_x(sid, name, match) && nvram_match_x(sid2, name2, match2))
+	{
+		return websWrite(wp, output);
+	}	
+
+	return 0;
+}
+
 /*
  * Example: 
  * wan_proto=dhcp
@@ -1035,7 +1065,7 @@
 	while (fgets(buf, MAX_LINE_SIZE, fp)!=NULL)
 	{	 	
 	    //printf("Read time: %s\n", buf);	    	    	    	   
-	    ret += websWrite(wp, buf);
+	    ret += websWrite(wp, "%s", buf);
 	}		    	                             		
 	 
 	fclose(fp);		
@@ -1444,14 +1474,15 @@
 	   	sys_script(script);
 	}   
 	
-
-
-    	    	               
-        if (!strcmp(value, "  Save  ") || !strcmp(value, " Apply "))
+        if (!strcmp(value, "  Save  "))
         {
-
            strcpy(urlcache, next_url);
            websRedirect(wp, next_url);
+        } else 
+        if (!strcmp(value, " Apply "))
+        {
+           strcpy(urlcache, current_url);
+           websRedirect(wp, current_url);
         } 
         
     	else if(!strcmp(value, " Finish ")) websRedirect(wp, "SaveRestart.asp");
@@ -2294,11 +2325,112 @@
 {
 	sys_commit(NULL);
 	sys_download("/tmp/settings");
-	do_file(url, stream);
+	do_file("/tmp/settings", stream);
 	//unlink("/tmp/settings");
 }
 
 
+static void
+do_upload_file(char *upload_file, char *url, FILE *stream, int len, char *boundary)
+{
+	FILE *fifo = NULL;
+	char buf[1024];
+	int count, ret = EINVAL;
+
+	cprintf("Start upload\n");
+	/* Look for our part */
+	while (len > 0) 
+	{
+	     if (!fgets(buf, MIN(len + 1, sizeof(buf)), stream))
+	     {			
+			goto err;
+	     }			
+		
+	     len -= strlen(buf);
+		
+	     if (!strncasecmp(buf, "Content-Disposition:", 20) &&
+		    strstr(buf, "name=\"file\""))
+			break;			
+	}
+
+	/* Skip boundary and headers */
+	while (len > 0) {
+		if (!fgets(buf, MIN(len + 1, sizeof(buf)), stream))
+		{
+			goto err;
+		}		
+		len -= strlen(buf);
+		if (!strcmp(buf, "\n") || !strcmp(buf, "\r\n"))
+		{
+			break;
+		}
+	}
+
+	if (!(fifo = fopen(upload_file, "w"))) goto err;
+				
+	len -= strlen(boundary)+8;
+	cprintf("Upgrading %d\n", len);
+	while (len > 0)
+	{
+		count = fread(buf, 1, MIN(len, sizeof(buf)), stream);
+		fwrite(buf, 1, count, fifo);
+		len -= count;
+	}
+	fread(buf, 1, strlen(boundary)+8, stream);
+	fflush(fifo);
+	ret = 0;
+ err:
+	if (fifo)
+	   fclose(fifo);
+		
+	//printf("Error : %d\n", ret);
+	fcntl(fileno(stream), F_SETOWN, -ret);
+}
+
+static void
+do_uploadflashfs_post(char *url, FILE *stream, int len, char *boundary)
+{
+  char upload_file[]="/tmp/flash.tar.gz";
+  
+  do_upload_file(upload_file, url, stream, len, boundary);
+  
+  if(eval("tar","ztf",upload_file) != 0)
+   fcntl(fileno(stream), F_SETOWN, -EINVAL);
+}
+
+static void
+do_uploadflashfs_cgi(char *url, FILE *stream)
+{
+	int ret;
+	
+	//printf("Upgrade CGI\n");
+	ret = fcntl(fileno(stream), F_GETOWN, 0);
+	/* Reboot if successful */
+	if (ret == 0)
+	{
+		websApply(stream, "Uploading.asp"); 
+		eval("flashfs","commit");
+		eval("flashfs","enable");
+		sys_reboot();
+	}	
+	else    
+	{
+	   	websApply(stream, "UploadError.asp");
+	   	//unlink("/tmp/settings_u.prf");
+	}   	
+	  
+}
+
+static void
+do_flashfs_file(char *url, FILE *stream)
+{
+
+	eval("flashfs", "load");
+	eval("flashfs", "save");
+	do_file("/tmp/flash.tar.gz", stream);
+	//unlink("/tmp/settings");
+}
+
 #elif defined(vxworks)
 
 static void
@@ -2345,6 +2477,7 @@
 	websAspDefine("nvram_get_buf_x", ej_nvram_get_buf_x);	
 	websAspDefine("nvram_get_table_x", ej_nvram_get_table_x);	
 	websAspDefine("nvram_match_x", ej_nvram_match_x);
+	websAspDefine("nvram_double_match_x", ej_nvram_double_match_x);
     	websAspDefine("nvram_match_both_x", ej_nvram_match_both_x);	
 	websAspDefine("nvram_match_list_x", ej_nvram_match_list_x);
 	websAspDefine("select_channel", ej_select_channel);
@@ -2370,9 +2503,11 @@
 	{ "**.js",  "text/javascript", NULL, NULL, do_file, do_auth },
 	{ "**.cab", "text/txt", NULL, NULL, do_file, do_auth },
 	{ "**.CFG", "text/txt", NULL, NULL, do_prf_file, do_auth },
+	{ "**.tar.gz", "text/txt", NULL, NULL, do_flashfs_file, do_auth },
 	{ "apply.cgi*", "text/html", no_cache, do_apply_cgi_post, do_apply_cgi, do_auth },	
 	{ "upgrade.cgi*", "text/html", no_cache, do_upgrade_post, do_upgrade_cgi, do_auth},
 	{ "upload.cgi*", "text/html", no_cache, do_upload_post, do_upload_cgi, do_auth },
+	{ "uploadflashfs.cgi*", "text/html", no_cache, do_uploadflashfs_post, do_uploadflashfs_cgi, do_auth },
  	{ "syslog.cgi*", "text/txt", no_cache, NULL, do_log_cgi, do_auth },
 	{ "webcam.cgi*", "text/html", no_cache, NULL, do_webcam_cgi, do_auth },
 	{ NULL, NULL, NULL, NULL, NULL, NULL }
@@ -2510,6 +2645,7 @@
 	{ "nvram_get_buf_x", ej_nvram_get_buf_x},
 	{ "nvram_get_table_x", ej_nvram_get_table_x},
 	{ "nvram_match_x", ej_nvram_match_x},
+	{ "nvram_double_match_x", ej_nvram_double_match_x},
         { "nvram_match_both_x", ej_nvram_match_both_x},
 	{ "nvram_match_list_x", ej_nvram_match_list_x},
 	{ "select_channel", ej_select_channel},
@@ -2895,7 +3031,7 @@
 
 	cic->buflen = sizeof(ibuf2);
 	strcpy(cic->country_abbrev, abbrev);
-	cic->band = WLC_BAND_B;
+	cic->band = WLC_BAND_2G;
 	cic->count = 0;
 
 
