diff -BurpN '-x*.o' router/jpeg-6b/Makefile gateway/jpeg-6b/Makefile
--- router/jpeg-6b/Makefile	2004-11-11 10:28:33.000000000 +0300
+++ gateway/jpeg-6b/Makefile	2009-03-31 22:26:46.000000000 +0400
@@ -24,7 +24,7 @@
 #CC=gcc
 
 # You may need to adjust these cc options:
-CFLAGS= -O2  -I$(srcdir)
+CFLAGS= -O2  -I$(srcdir) $(EXTRACFLAGS)
 # Generally, we recommend defining any configuration symbols in jconfig.h,
 # NOT via -D switches here.
 # However, any special defines for ansi2knr.c may be included here:
