diff -BurN router/LPRng/Makefile gateway/LPRng/Makefile
--- router/LPRng/Makefile	2004-11-11 10:28:51.000000000 +0300
+++ gateway/LPRng/Makefile	2007-04-04 22:38:06.000000000 +0400
@@ -42,7 +42,7 @@
 LPD_CONF_PATH=\"${sysconfdir}/lpd/lpd.conf\"
 PRINTCAP_PATH=\"${sysconfdir}/printcap\"
 LPD_PATH=\"${sbindir}/lpd\"
-SHELL="/bin/sh"
+#SHELL="/bin/sh"
 LOCKFILE=\"/var/run/lpd\"
 PSHOWALL="-ax"
 FILTER_DIR=${libexecdir}/filters
diff -urbB router/LPRng/src/include/lp.h gateway/LPRng/src/include/lp.h
--- router/LPRng/src/include/lp.h	2004-11-11 10:29:18.000000000 +0300
+++ gateway/LPRng/src/include/lp.h	2008-06-14 20:08:41.000000000 +0400
@@ -479,6 +479,7 @@
     char    busy;       //TRUE if the Printer is busy otherwise FALSE
 } LPT_STATUS;
 LPT_STATUS          lptstatus;  //Golbal variable to store the status of server
+void check_prn_status(char *status_prn, char *cliadd_prn);
 #define FALSE   0
 #define TRUE    1
 #define ONLINE	"On-line"
diff -BurN router/LPRng/src/common/lpd.c gateway/LPRng/src/common/lpd.c
--- router/LPRng/src/common/lpd.c	2004-11-11 10:29:16.000000000 +0300
+++ gateway/LPRng/src/common/lpd.c	2008-12-19 13:14:58.000000000 +0300
@@ -31,7 +31,7 @@
 #else
 #define		PRINT	printf
 #endif
-#define 	Raw_Printing_with_ASUS
+/*#define 	Raw_Printing_with_ASUS*/
 #define 	LOGOPTS		(LOG_PERROR|LOG_PID|LOG_LPR|LOG_ERR)
 
 /* 2004/09/10, added by Joey 
@@ -45,7 +45,9 @@
 
 //JY
 #include "lp.h"
+#ifndef MAX
 #define MAX(a, b) ((a) > (b) ? (a) : (b))//JY1112
+#endif
 
 
 extern int          errno;
@@ -100,7 +100,7 @@
 
 
 
-void main()
+int main(int argc, char **argv)
 {
     int                 sockfd , clisockfd;
     unsigned int        clilen;
@@ -317,7 +317,9 @@
 #endif
 		close(sockfd);
 		close(sockfd_ASUS);
+#ifdef Raw_Printing_with_ASUS //Lisa
 		close(netfd);
+#endif
         	exit(0);
         }
               
@@ -468,7 +470,7 @@
 
 	    if (iCount!=0)
 	    {
-             	(void *)para_buf = malloc(pHdrCmd->ParaLength);
+             	para_buf = malloc(pHdrCmd->ParaLength);
             }
 
 	    //PRINT("HdrCmd Length %d\n", iCount);
@@ -702,7 +704,7 @@
                     /************************************/
                     /************************************/
                     
-                    (void *)pReadbuf = malloc(len + 1);
+                    pReadbuf = malloc(len + 1);
                     
                     if(pReadbuf == NULL)
                     {
@@ -770,7 +772,7 @@
 
             if (iCount!=0)
 	    {
-            	(void *)para_buf = malloc(pHdrData->DataLength);
+            	para_buf = malloc(pHdrData->DataLength);
             }
             //remark 
 	    //PRINT("pHdrData->DataLength -- %d\n",pHdrData->DataLength);
@@ -1085,10 +1087,6 @@
     char    *token;
     FILE    *fp;
 
-#ifdef USBONLY
-    g_useUsb = TRUE;
-    return(g_useUsb);
-#else
     fp=fopen("/proc/sys/dev/parport/parport0/devices/lp/deviceid","r");
 
     if( fp != NULL)
@@ -1126,8 +1124,9 @@
         
         fclose(fp);
     }
-#endif
     
+    g_useUsb = TRUE;
+    return(g_useUsb);
 }
 
 /*1110test status
@@ -1165,7 +1165,7 @@
 	char QueueName_got[32];
 	char *index1;
 	int rps_i=0, rps_j=0;
-	while(index1 = strrchr(input, ' '))
+	while((index1 = strrchr(input, ' ')))
 		index1[0] = 0;
 	rps_i = 0;
 	strcpy(QueueName_got, input);
