diff -BurN -x'*.o' -x'*.orig' router/misc/rootprep.files gateway/misc/rootprep.files
--- router/misc/rootprep.files	1970-01-01 00:00:00.000000000 +0000
+++ gateway/misc/rootprep.files	2010-01-30 17:58:56.000000000 +0000
@@ -0,0 +1,5 @@
+/dev/null c 666 root root 1 3
+/dev/zero c 666 root root 1 5
+/dev/console c 600 root root 5 1
+/dev/nvram c 644 root root 253 0
+/dev/gpio c 660 root root 254 0
diff -BurN router/misc/rootprep.sh gateway/misc/rootprep.sh
--- router/misc/rootprep.sh	2004-10-14 23:11:28.000000000 +0400
+++ gateway/misc/rootprep.sh	2005-07-16 11:14:16.342974160 +0400
@@ -17,18 +17,33 @@
 
 # tmp
 mkdir -p tmp
+mkdir -p tmp/var
 ln -sf tmp/var var
 (cd $ROOTDIR/usr && ln -sf ../tmp)
 
 # dev
 mkdir -p dev
 
+# miscellaneous
+mkdir -p sys
+mkdir -p mnt
+mkdir -p opt
+mkdir -p home
+mkdir -p proc
+
+# usr/local -> tmp/local
+ln -sf ../tmp/local usr/local
+
 # etc
-mkdir -p etc
+mkdir -p usr/etc
+ln -sf ../usr/etc tmp/etc
+ln -sf tmp/etc etc
 echo "/lib" > etc/ld.so.conf
 echo "/usr/lib" >> etc/ld.so.conf
+echo "/opt/lib" >> etc/ld.so.conf
 /sbin/ldconfig -r $ROOTDIR
 
-# miscellaneous
-mkdir -p mnt
-mkdir -p proc
+# the hack to make /opt/lib work with uclibc 0.9.19 ld.so
+# which is ignoring ld.so.conf completely
+mkdir -p usr/X11R6
+ln -sf ../../opt/lib usr/X11R6/lib
