--- nfs-utils/support/lib/Makefile	2004-10-19 04:22:58.000000000 +0400
+++ nfs-utils-libnfs/support/lib/Makefile	2007-11-10 11:55:18.989479793 +0300
@@ -1,9 +1,17 @@
 
 include	$(TOP)rules.mk
 
-LIBS	= libnfs.a libexport.a libmisc.a librpc.a libgssapi.a
+LIBS	= librpc.a libgssapi.a 
 
-all install::	$(LIBS)
+NFSOBJS = ../nfs/exports.o ../nfs/rmtab.o ../nfs/xio.o \
+	  ../nfs/rpcmisc.o ../nfs/rpcdispatch.o ../nfs/xlog.o ../nfs/xmalloc.o ../nfs/wildmat.o \
+	  ../nfs/nfssvc.o ../nfs/nfsclient.o ../nfs/nfsexport.o ../nfs/getfh.o ../nfs/nfsctl.o \
+	  ../nfs/lockdsvc.o ../nfs/svc_socket.o ../nfs/cacheio.o
+EXPOBJS = ../export/mount_clnt.o ../export/mount_xdr.o ../export/client.o ../export/export.o ../export/hostname.o \
+	  ../export/nfsctl.o ../export/rmtab.o ../export/xtab.o
+MSCOBJS = ../misc/from_local.o ../misc/mountpoint.o
+
+all install::	$(LIBS) libnfs.so
 	@:
 
 clean distclean::
@@ -11,3 +19,6 @@
 
 lib%.a:
 	ln -sf ../$*/$@ .
+
+libnfs.so: $(NFSOBJS) $(EXPOBJS) $(MSCOBJS)
+	$(CC) -shared -o $@ $^
--- nfs-utils/utils/exportfs/Makefile	2003-07-03 10:14:43.000000000 +0400
+++ nfs-utils-libnfs/utils/exportfs/Makefile	2007-11-10 11:43:50.596623777 +0300
@@ -4,8 +4,8 @@
 
 PROGRAM	= exportfs
 OBJS	= exportfs.o
-LIBDEPS	= $(TOP)support/lib/libexport.a $(TOP)/support/lib/libnfs.a
-LIBS	= -lexport -lnfs -lmisc
+LIBDEPS	= $(TOP)/support/lib/libnfs.so
+LIBS	= -lnfs
 MAN8	= exportfs
 MAN5	= exports
 MAN7	= nfsd
--- nfs-utils/utils/mountd/Makefile	2003-05-21 10:28:20.000000000 +0400
+++ nfs-utils-libnfs/utils/mountd/Makefile	2007-11-10 11:56:10.486216255 +0300
@@ -5,8 +5,8 @@
 PROGRAM	= mountd
 PREFIX	= rpc.
 OBJS	= mountd.o mount_dispatch.o auth.o rmtab.o cache.o svc_run.o
-LIBDEPS	= $(TOP)support/lib/libexport.a $(TOP)/support/lib/libnfs.a
-LIBS	= -lexport -lnfs -lmisc $(LIBBSD) $(LIBWRAP) $(LIBNSL)
+LIBDEPS	= $(TOP)/support/lib/libnfs.so
+LIBS	= -lnfs $(LIBBSD) $(LIBWRAP) $(LIBNSL)
 MAN8	= mountd
 
 include $(TOP)rules.mk
--- nfs-utils/utils/showmount/Makefile	1999-10-19 03:21:12.000000000 +0400
+++ nfs-utils-libnfs/utils/showmount/Makefile	2007-11-10 11:57:11.311813213 +0300
@@ -4,8 +4,8 @@
 
 PROGRAM	= showmount
 OBJS	= showmount.o
-LIBDEPS	= $(TOP)support/lib/libexport.a
-LIBS	= -lexport
+LIBDEPS	= $(TOP)support/lib/libnfs.so
+LIBS	= -lnfs
 MAN8	= showmount
 
 include $(TOP)rules.mk
--- nfs-utils/utils/statd/Makefile	2003-09-15 03:45:55.000000000 +0400
+++ nfs-utils-libnfs/utils/statd/Makefile	2007-11-10 11:56:46.547827783 +0300
@@ -16,7 +16,7 @@
 PREFIX	= rpc.
 OBJS	= $(SRCS:.c=.o)
 CCOPTS	= $(DEBUG) $(SIMUL)
-LIBS	= -lexport -lnfs -lmisc $(LIBWRAP) $(LIBNSL)
+LIBS	= -lnfs $(LIBWRAP) $(LIBNSL)
 
 SRCS	= $(RPCSRCS) $(SIMSRCS) \
 	  callback.c notlist.c log.c misc.c monitor.c notify.c simu.c \
