diff -BurpN -x'*.o' -x'*.orig' router/nvram/Makefile gateway/nvram/Makefile
--- router/nvram/Makefile	2004-11-11 07:30:36.000000000 +0000
+++ gateway/nvram/Makefile	2009-03-31 19:23:16.000000000 +0000
@@ -14,7 +14,7 @@
 
 CFLAGS	+= -I. -I$(SRCBASE)/include -Wall
 #CFLAGS	+= -g -DDEBUG
-CFLAGS	+= -s -O2 -DASUS
+CFLAGS	+= -s -O2 -DASUS $(EXTRACFLAGS)
 LDFLAGS += -L.
 LDFLAGS	+= -L$(TOP)/shared -L$(INSTALLDIR)/shared/usr/lib -lshared
 
diff -BurpN -x'*.o' -x'*.orig' router/nvram/main.c gateway/nvram/main.c
--- router/nvram/main.c	2004-11-11 07:30:36.000000000 +0000
+++ gateway/nvram/main.c	2009-12-04 10:42:24.000000000 +0000
@@ -56,7 +56,7 @@ int nvram_save(char *file, char *buf)
 int issyspara(char *p)
 {
 	extern struct nvram_tuple router_defaults[];
-	struct nvram_tuple *t, *u;
+	struct nvram_tuple *t;
 
 	for (t = router_defaults; t->name; t++)
 	{
@@ -94,7 +94,7 @@ int nvram_restore(char *file, char *buf)
 
 		if(v!=NULL)
 		{	
-			*v++ = NULL;
+			*v++ = '\0';
 
 			if (issyspara(p))
 			{
