diff -BurN router/pppoe-relay/Makefile gateway/pppoe-relay/Makefile
--- router/pppoe-relay/Makefile	2004-11-11 10:31:34.000000000 +0300
+++ gateway/pppoe-relay/Makefile	2008-12-24 14:49:39.000000000 +0300
@@ -34,7 +34,7 @@
 
 # Include ServPoET version if we're building for ServPoET
 
-CFLAGS= -s -O2 -I. -DVERSION=3.5
+CFLAGS= -s -O2 -I. -DVERSION=3.5 $(EXTRACFLAGS)
 TARGET=pppoe-relay
 
 all: $(TARGET)
