diff -Bup ppp.orig/pppd/plugins/pppol2tp/Makefile.linux ppp/pppd/plugins/pppol2tp/Makefile.linux
--- ppp.orig/pppd/plugins/pppol2tp/Makefile.linux.orig	2008-11-26 13:02:29.000000000 +0300
+++ ppp/pppd/plugins/pppol2tp/Makefile.linux	2009-10-08 10:00:57.000000000 +0400
@@ -1,6 +1,6 @@
 #CC	= gcc
 COPTS	= -O2 -g
-CFLAGS	= $(COPTS) -I. -I../.. -I../../../include -fPIC
+CFLAGS	= $(COPTS) -I. -I../.. -I$(LINUXDIR)/include -fPIC
 LDFLAGS	= -shared
 INSTALL	= install
