diff -BurpN xl2tpd.orig/l2tp.h xl2tpd/l2tp.h
--- xl2tpd.orig/l2tp.h	2012-05-25 14:19:39.000000000 +0000
+++ xl2tpd/l2tp.h	2012-07-19 12:07:15.740575002 +0000
@@ -100,11 +100,11 @@ struct payload_hdr
 #define DEFAULT_MAX_RETRIES 95   /* give us more time to debug */
 #endif
 
-#define DEFAULT_RWS_SIZE   4    /* Default max outstanding 
+#define DEFAULT_RWS_SIZE   8    /* Default max outstanding 
                                    control packets in queue */
-#define DEFAULT_TX_BPS		10000000        /* For outgoing calls, report this speed */
-#define DEFAULT_RX_BPS		10000000
-#define DEFAULT_MAX_BPS		10000000        /* jz: outgoing calls max bps */
+#define DEFAULT_TX_BPS		100000000        /* For outgoing calls, report this speed */
+#define DEFAULT_RX_BPS		100000000
+#define DEFAULT_MAX_BPS		100000000        /* jz: outgoing calls max bps */
 #define DEFAULT_MIN_BPS		10000   /* jz: outgoing calls min bps */
 #define PAYLOAD_FUDGE		2       /* How many packets we're willing to drop */
 #define MIN_PAYLOAD_HDR_LEN 6
