diff -Bup gateway/rc/firewall_ex.c.old gateway/rc/firewall_ex.c
--- gateway/rc/firewall_ex.c.old	2009-02-25 08:05:20.000000000 +0100
+++ gateway/rc/firewall_ex.c	2009-04-12 21:09:13.000000000 +0200
@@ -774,6 +774,10 @@ filter_setting(char *wan_if, char *wan_i
 
 		if (nvram_invmatch("http_lanport", "80"))
 			fprintf(fp, "-A INPUT -p tcp -m tcp -d %s --dport 80 -j %s\n", nvram_safe_get("lan_ipaddr"), logaccept);
+
+		if (nvram_invmatch("udpxy_enable_x", "0") && nvram_invmatch("udpxy_wan_x", "0"))
+			fprintf(fp, "-A INPUT -p tcp -m tcp --dport %s -j %s\n", nvram_safe_get("udpxy_enable_x"), logaccept);
+
 		if (nvram_match("misc_http_x", "1"))
 		{
             		fprintf(fp, "-A INPUT -p tcp -m tcp -d %s --dport %s -j %s\n", nvram_safe_get("lan_ipaddr"), nvram_safe_get("http_lanport"), logaccept);  
diff -Bup gateway/rc/network.c.old gateway/rc/network.c
--- gateway/rc/network.c.old	2009-02-21 21:09:16.000000000 +0100
+++ gateway/rc/network.c	2009-04-12 21:14:12.000000000 +0200
@@ -203,9 +203,14 @@ start_igmpproxy(char *wan_ifname)
 	struct stat	st_buf;
 	FILE 		*fp;
 
-	if (atoi(nvram_safe_get("udpxy_enable_x")))
-		eval("/usr/sbin/udpxy", "-a", nvram_get("lan_ifname") ? : "br0",
-			"-m", wan_ifname, "-p", nvram_get("udpxy_enable_x"));
+	if (atoi(nvram_safe_get("udpxy_enable_x"))) {
+		if (nvram_invmatch("udpxy_wan_x", "0"))
+			eval("/usr/sbin/udpxy",
+				"-m", wan_ifname, "-p", nvram_get("udpxy_enable_x"));
+		else
+			eval("/usr/sbin/udpxy", "-a", nvram_get("lan_ifname") ? : "br0",
+				"-m", wan_ifname, "-p", nvram_get("udpxy_enable_x"));
+	}
 	
 	if (!nvram_match("mr_enable_x", "1"))
 		return;
