diff -BurpN router/rc/services.c gateway/rc/services.c
--- router/rc/services.c	2009-02-19 14:01:53.000000000 +0000
+++ gateway/rc/services.c	2009-04-30 08:00:57.000000000 +0000
@@ -385,6 +385,8 @@ start_services(void)
 int
 stop_services(void)
 {
+	preshutdown_system();
+
 #ifdef ASUS_EXT
 	stop_usb();
 #endif
diff -BurpN router/rc/services_ex.c gateway/rc/services_ex.c
--- router/rc/services_ex.c	2009-02-25 20:18:31.000000000 +0000
+++ gateway/rc/services_ex.c	2009-05-28 18:01:02.000000000 +0000
@@ -1946,9 +1946,9 @@ usbhandler:	
 int
 stop_service_main(int argc, char *argv[])
 {
-	stop_snmpd();
 	stop_misc();
-	stop_logger();
+
+	preshutdown_system();
 	stop_usb();
 
 	/* nas is still needed for upgrade over WiFI with WPA enabled */
@@ -1953,7 +1953,11 @@ stop_service_main(int argc, char *argv[]
 
 	/* nas is still needed for upgrade over WiFI with WPA enabled */
 	/* stop_nas();*/
+
 	stop_upnp();
+	stop_snmpd();
+	stop_logger();
+
 	stop_dhcpd();
 	stop_dns();
 
