diff -BurpN rp-pppoe-3.11.orig/src/plugin.c rp-pppoe-3.11/src/plugin.c
--- rp-pppoe-3.11.orig/src/plugin.c	2012-08-18 00:31:25.000000000 +0600
+++ rp-pppoe-3.11/src/plugin.c	2012-11-03 19:04:34.078752509 +0600
@@ -58,7 +58,7 @@ static char const RCSID[] =
 #define _ROOT_PATH ""
 #endif
 
-#define _PATH_ETHOPT         _ROOT_PATH "/etc/ppp/options."
+#define _PATH_ETHOPT         _ROOT_PATH "/ppp/options."
 
 char pppd_version[] = VERSION;
 
