diff -BurpN rp-pppoe-3.11.orig/src/relay.c rp-pppoe-3.11/src/relay.c
--- rp-pppoe-3.11.orig/src/relay.c	2012-11-03 19:05:25.000000000 +0600
+++ rp-pppoe-3.11/src/relay.c	2012-11-03 19:33:39.482729000 +0600
@@ -1320,7 +1320,6 @@ relayHandlePADS(PPPoEInterface const *if
     PPPoETag tag;
     unsigned char *loc;
     int ifIndex;
-    int acIndex;
     PPPoESession *ses = NULL;
     SessionHash *sh;
 
@@ -1338,8 +1337,6 @@ relayHandlePADS(PPPoEInterface const *if
 	return;
     }
 
-    acIndex = iface - Interfaces;
-
     /* Source address must be unicast */
     if (NOT_UNICAST(packet->ethHdr.h_source)) {
 	syslog(LOG_ERR,
