
---
 dpkt.c  |    5 ++++-
 udpxy.c |   32 +++++++++++++++++++++++++-------
 2 files changed, 29 insertions(+), 8 deletions(-)

diff --git a/dpkt.c b/dpkt.c
--- a/dpkt.c
+++ b/dpkt.c
@@ -617,12 +617,15 @@ read_packet( struct dstream_ctx* spc, int fd, char* buf, size_t len )
     if( (spc->stype == UPXDT_RTP_TS) || (spc->flags & F_CHECK_FMT) )
         chunk_len = (len > spc->mtu) ? spc->mtu : len;
 
+#ifdef UDPXY_FILEIO
     if( spc->flags & F_FILE_INPUT ) {
         assert( !buf_overrun( buf, len, 0, chunk_len, g_flog ) );
         n = read_frecord( fd, buf, chunk_len, &(spc->stype), g_flog );
         if( n <= 0 ) return n;
     }
-    else {
+    else
+#endif /* UDPXY_FILEIO */
+    {
         assert( !buf_overrun(buf, len, 0, chunk_len, g_flog) );
         n = read_buf( fd, buf, chunk_len, g_flog );
         if( n <= 0 ) return n;
diff --git a/udpxy.c b/udpxy.c
--- a/udpxy.c
+++ b/udpxy.c
@@ -499,13 +499,19 @@ sync_dsockbuf_len( int ssockfd, int dsockfd )
  */
 static int
 relay_traffic( int ssockfd, int dsockfd, struct server_ctx* ctx,
-               int dfilefd, const struct ip_mreq* mreq )
+#ifdef UDPXY_FILEIO
+               int dfilefd,
+#endif /* UDPXY_FILEIO */
+               const struct ip_mreq* mreq )
 {
     volatile sig_atomic_t quit = 0;
 
     int rc = 0;
     ssize_t nmsgs = -1;
-    ssize_t nrcv = 0, nsent = 0, nwr = 0,
+    ssize_t nrcv = 0, nsent = 0,
+#ifdef UDPXY_FILEIO
+            nwr = 0,
+#endif /* UDPXY_FILEIO */
             lrcv = 0, lsent = 0;
     char*  data = NULL;
     size_t data_len = g_uopt.rbuf_len;
@@ -628,6 +634,7 @@ relay_traffic( int ssockfd, int dsockfd, struct server_ctx* ctx,
             lsent = nsent;
         }
 
+#ifdef UDPXY_FILEIO
         if( (dfilefd > 0) && (nrcv > 0) ) {
             nwr = write_data( &ds, data, nrcv, dfilefd );
             if( -1 == nwr )
@@ -636,6 +643,7 @@ relay_traffic( int ssockfd, int dsockfd, struct server_ctx* ctx,
                     nrcv, lsent, nwr, t_delta, g_flog ) );
             lsent = nwr;
         }
+#endif /* UDPXY_FILEIO */
 
         if( ds.flags & F_SCATTERED ) reset_pkt_registry( &ds );
 
@@ -674,9 +682,11 @@ udp_relay( int sockfd, struct server_ctx* ctx )
     uint16_t    port;
     pid_t       new_pid;
     int         rc = 0, flags; 
-    int         msockfd = -1, sfilefd = -1,
-                dfilefd = -1, srcfd = -1;
+    int         msockfd = -1, srcfd = -1;
+#ifdef UDPXY_FILEIO
+    int         sfilefd = -1, dfilefd = -1;
     char        dfile_name[ MAXPATHLEN ];
+#endif /* UDPXY_FILEIO */
     size_t      rcvbuf_len = 0;
 
     assert( (sockfd > 0) && ctx );
@@ -749,6 +759,7 @@ udp_relay( int sockfd, struct server_ctx* ctx )
             break;
         }
 
+#ifdef UDPXY_FILEIO
         if( NULL != g_uopt.dstfile ) {
             (void) snprintf( dfile_name, MAXPATHLEN - 1,
                     "%s.%d", g_uopt.dstfile, getpid() );
@@ -777,7 +788,9 @@ udp_relay( int sockfd, struct server_ctx* ctx )
                 srcfd = sfilefd;
             }
         }
-        else {
+        else
+#endif /* UDPXY_FILEIO */
+        {
             rc = calc_buf_settings( NULL, &rcvbuf_len );
             if (0 == rc ) {
                 rc = setup_mcast_listener( &addr, &mreq, &msockfd,
@@ -787,7 +800,11 @@ udp_relay( int sockfd, struct server_ctx* ctx )
         }
         if( 0 != rc ) break;
 
-        rc = relay_traffic( srcfd, sockfd, ctx, dfilefd, &mreq );
+        rc = relay_traffic( srcfd, sockfd, ctx,
+#ifdef UDPXY_FILEIO
+                            dfilefd,
+#endif /* UDPXY_FILEIO */
+                            &mreq );
         if( 0 != rc ) break;
 
     } while(0);
@@ -795,6 +812,7 @@ udp_relay( int sockfd, struct server_ctx* ctx )
     if( msockfd > 0 ) {
         close_mcast_listener( msockfd, &mreq );
     }
+#ifdef UDPXY_FILEIO
     if( sfilefd > 0 ) {
        (void) close( sfilefd );
        TRACE( (void) tmfprintf( g_flog, "Source file [%s] closed\n",
@@ -805,7 +823,7 @@ udp_relay( int sockfd, struct server_ctx* ctx )
        TRACE( (void) tmfprintf( g_flog, "Dest file [%s] closed\n",
                             dfile_name ) );
     }
-
+#endif /* UDPXY_FILEIO */
     if( 0 != rc ) {
         (void) send_http_response( sockfd, 500, "Service error" );
     }
-- 
1.7.2.5

