diff -BurpN rp-l2tp.orig/tunnel.c rp-l2tp/tunnel.c
--- rp-l2tp.orig/tunnel.c	2009-09-02 17:00:06.000000000 +0000
+++ rp-l2tp/tunnel.c	2010-12-27 09:56:48.505982002 +0000
@@ -26,6 +26,8 @@ static char const RCSID[] =
 #include <errno.h>
 #include <sys/ioctl.h>
 #include <netdb.h>
+#include <features.h>
+#include <resolv.h>
 
 /* Hash tables of all tunnels */
 static hash_table tunnels_by_my_id;
@@ -517,6 +519,12 @@ tunnel_establish(l2tp_peer *peer, EventS
 
     /* check peer_addr and resolv it based on the peername if needed */
     if (peer_addr.sin_addr.s_addr == INADDR_ANY) {
+#if !defined(__UCLIBC__) \
+ || (__UCLIBC_MAJOR__ == 0 \
+ && (__UCLIBC_MINOR__ < 9 || (__UCLIBC_MINOR__ == 9 && __UCLIBC_SUBLEVEL__ < 31)))
+	/* force ns refresh from resolv.conf with uClibc pre-0.9.31 */
+	res_init();
+#endif
 	he = gethostbyname(peer->peername);
 	if (!he) {
             l2tp_set_errmsg("tunnel_establish: gethostbyname failed for '%s'", peer->peername);
