--- a/src/config.h
+++ b/src/config.h
@@ -133,11 +133,11 @@ RESOLVFILE
 #define HAVE_DHCP
 #define HAVE_DHCP6 
 #define HAVE_TFTP
-#define HAVE_SCRIPT
-#define HAVE_AUTH
-#define HAVE_IPSET 
+/* #define HAVE_SCRIPT */
+/* #define HAVE_AUTH */
+/* #define HAVE_IPSET  */
 /* #define HAVE_LUASCRIPT */
-/* #define HAVE_BROKEN_RTC */
+#define HAVE_BROKEN_RTC
 /* #define HAVE_DBUS */
 /* #define HAVE_IDN */
 /* #define HAVE_CONNTRACK */
@@ -226,6 +227,8 @@ HAVE_SOCKADDR_SA_LEN
 #  ifndef IPV6_V6ONLY
 #    define IPV6_V6ONLY 26
 #  endif
+#elif !defined(NO_IPV6)
+#  define NO_IPV6
 #endif
 
 /* This is for glibc 2.x */
