From 8db957dfbfb17f662973cfa431107b34a95b9c46 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Tue, 17 Dec 2013 15:47:10 +0000
Subject: [PATCH 05/10] Fix endless loop with some bogu-nxdomain. Another F_CONFIG botch.

---
 CHANGELOG     |    3 +++
 src/rfc1035.c |    2 +-
 2 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/CHANGELOG b/CHANGELOG
index 39dbe17..cbb9420 100644
--- a/CHANGELOG
+++ b/CHANGELOG
@@ -4,6 +4,9 @@ version 2.69
 	    on the BSD platform. Thanks to Matthias Andree for
 	    valuable research on how to implement this.
 
+	    Fix infinite loop associated with some --bogus-nxdomain
+	    configs. Thanks forgobogo for the bug report.
+
 
 version 2.68
             Use random addresses for DHCPv6 temporary address
diff --git a/src/rfc1035.c b/src/rfc1035.c
index 9fdba05..c120d55 100644
--- a/src/rfc1035.c
+++ b/src/rfc1035.c
@@ -1283,7 +1283,7 @@ int check_for_bogus_wildcard(struct dns_header *header, size_t qlen, char *name,
 		/* Found a bogus address. Insert that info here, since there no SOA record
 		   to get the ttl from in the normal processing */
 		cache_start_insert();
-		cache_insert(name, NULL, now, ttl, F_IPV4 | F_FORWARD | F_NEG | F_NXDOMAIN | F_CONFIG);
+		cache_insert(name, NULL, now, ttl, F_IPV4 | F_FORWARD | F_NEG | F_NXDOMAIN);
 		cache_end_insert();
 		
 		return 1;
-- 
1.7.2.5

