From 875b8160f6a972a427969738667bc80d424fe6ed Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Tue, 17 Dec 2013 17:40:32 +0000
Subject: [PATCH 07/10] Remove unused code.

---
 src/dnsmasq.h |    2 +-
 src/slaac.c   |    1 -
 2 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/src/dnsmasq.h b/src/dnsmasq.h
index 8bd3ddf..724efde 100644
--- a/src/dnsmasq.h
+++ b/src/dnsmasq.h
@@ -561,7 +561,7 @@ struct dhcp_lease {
   struct in6_addr addr6;
   int iaid;
   struct slaac_address {
-    struct in6_addr addr, local;
+    struct in6_addr addr;
     time_t ping_time;
     int backoff; /* zero -> confirmed */
     struct slaac_address *next;
diff --git a/src/slaac.c b/src/slaac.c
index 43c2c38..7eb4236 100644
--- a/src/slaac.c
+++ b/src/slaac.c
@@ -93,7 +93,6 @@ void slaac_add_addrs(struct dhcp_lease *lease, time_t now, int force)
 	    slaac->ping_time = now;
 	    slaac->backoff = 1;
 	    slaac->addr = addr;
-	    slaac->local = context->local6;
 	    /* Do RA's to prod it */
 	    ra_start_unsolicted(now, context);
 	  }
-- 
1.7.2.5

