From 0d6eb134f50e4fcf08b4ec426ddf5b58a4bfd019 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Tue, 26 Nov 2013 13:30:12 +0000
Subject: [PATCH] Do immediate RA when a prefix goes from old->current.

---
 src/dhcp6.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/src/dhcp6.c b/src/dhcp6.c
index 5da2d0f..ca7014f 100644
--- a/src/dhcp6.c
+++ b/src/dhcp6.c
@@ -661,6 +661,7 @@ static int construct_worker(struct in6_addr *local, int prefix,
 		  log_context(AF_INET6, context); 
 		  /* fast RAs for a while */
 		  ra_start_unsolicted(param->now, context);
+		  param->newone = 1; 
 		  /* Add address to name again */
 		  if (context->flags & CONTEXT_RA_NAME)
 		    param->newname = 1;
-- 
1.7.2.5

