From 3dffbc3ebf3602135310abf3fd98cb27f92446b9 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Mon, 2 Dec 2013 13:22:37 +0000
Subject: [PATCH] Don't overwrite errno before generating message.

---
 src/network.c |    6 +++++-
 1 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/src/network.c b/src/network.c
index b9d7247..2ebcbcb 100644
--- a/src/network.c
+++ b/src/network.c
@@ -630,7 +630,11 @@ static int make_sock(union mysockaddr *addr, int type, int dienow)
       s = _("failed to create listening socket for %s: %s");
       
       if (fd != -1)
-	close (fd);
+	{
+	  int errsav = errno;
+	  close (fd);
+	  errno = errsav;
+	}
       
       if (dienow)
 	{
-- 
1.7.2.5

