From 57310500622f57315b401c405a07e579ced6417c Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Mon, 21 Oct 2013 18:26:20 +0100
Subject: [PATCH] Get NXDOMAIN right for local queries to auth zones.

---
 src/auth.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/auth.c b/src/auth.c
index 50fb61d..1c1f12a 100644
--- a/src/auth.c
+++ b/src/auth.c
@@ -793,7 +793,7 @@ size_t answer_auth(struct dns_header *header, char *limit, size_t qlen, time_t n
   if (trunc)
     header->hb3 |= HB3_TC;
   
-  if (anscount == 0 && auth && nxdomain)
+  if ((auth || local_query) && nxdomain)
     SET_RCODE(header, NXDOMAIN);
   else
     SET_RCODE(header, NOERROR); /* no error */
-- 
1.7.2.5

