From 4d0f5b4c44e66eb6c426180186b4619bd2f43046 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Wed, 5 Sep 2012 23:29:30 +0100
Subject: [PATCH 10/11] Fix BOOTP option processing.

---
 CHANGELOG     |    3 +++
 src/rfc2131.c |    3 ++-
 2 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/CHANGELOG b/CHANGELOG
index ded2d26..4f8777d 100644
--- a/CHANGELOG
+++ b/CHANGELOG
@@ -10,6 +10,9 @@ version 2.64
 	    also a match. This allows multiple addresses for a name in 
 	    /etc/hosts with one of them assigned via DHCP.
 
+	    Fix broken vendor-option processing for BOOTP. Thanks to
+	    Hans-Joachim Baader for the bug report.
+
 
 version 2.63
             Do duplicate dhcp-host address check in --test mode.
diff --git a/src/rfc2131.c b/src/rfc2131.c
index 207f450..250b1ab 100644
--- a/src/rfc2131.c
+++ b/src/rfc2131.c
@@ -493,8 +493,9 @@ size_t dhcp_reply(struct dhcp_context *context, char *iface_name, int int_index,
 	      lease_set_interface(lease, int_index, now);
 	      
 	      clear_packet(mess, end);
+	      match_vendor_opts(NULL, daemon->dhcp_opts); /* clear flags */
 	      do_options(context, mess, end, NULL, hostname, get_domain(mess->yiaddr), 
-			 netid, subnet_addr, 0, 0, 0, NULL, 0, now);
+			 netid, subnet_addr, 0, 0, -1, NULL, 0, now);
 	    }
 	}
       
-- 
1.7.2.5

