diff -urBp inadyn.orig/src/dyndns.c inadyn/src/dyndns.c
--- inadyn.orig/src/dyndns.c	2010-07-11 14:03:49.000000000 +0400
+++ inadyn/src/dyndns.c	2010-07-11 16:52:12.000000000 +0400
@@ -34,7 +34,7 @@ Author: Narcis Ilisei
         - April 2004 - freedns.afraid.org system added.
         - October 2004 - Unix syslog capability added.
 */
-#define MODULE_TAG      "INADYN: "  
+#define MODULE_TAG 
 #include <stdlib.h>
 #include <string.h>
 #include <time.h>
@@ -969,13 +969,13 @@ RC_TYPE dyn_dns_update_ip(DYN_DNS_CLIENT
 		rc = do_ip_server_transaction(p_self, servernum);
 		if (rc != RC_OK)
 		{
-			DBG_PRINTF((LOG_WARNING,"W: DYNDNS: Error '%s' (0x%x) when talking to IP server\n",
+			DBG_PRINTF((LOG_WARNING,"W: Error '%s' (0x%x) when talking to IP server\n",
 				errorcode_get_name(rc), rc));
 			break;
 		}
 		if (p_self->dbg.level > 1)
 		{
-			DBG_PRINTF((LOG_DEBUG,"DYNDNS: IP server response: %s\n", p_self->p_work_buffer));		
+			DBG_PRINTF((LOG_DEBUG,"IP server response: %s\n", p_self->p_work_buffer));		
 		}
 
 		/*extract my IP, check if different than previous one*/
@@ -987,7 +987,7 @@ RC_TYPE dyn_dns_update_ip(DYN_DNS_CLIENT
 		
 		if (p_self->dbg.level > 1)
 		{
-			DBG_PRINTF((LOG_WARNING,"W: DYNDNS: My IP address: %s\n", p_self->info[servernum].my_ip_address.name));		
+			DBG_PRINTF((LOG_WARNING,"W: My IP address: %s\n", p_self->info[servernum].my_ip_address.name));		
 		}
 
 		/*step through aliases list, resolve them and check if they point to my IP*/
@@ -1113,7 +1113,7 @@ int dyn_dns_main(DYN_DNS_CLIENT *p_dyndn
     if ((fp=fopen(p_dyndns->ip_cache, "r")))
     {
 	if (!fgets (p_dyndns->info[0].my_ip_address.name, sizeof (p_dyndns->info[0].my_ip_address.name),fp)) {
-		DBG_PRINTF((LOG_WARNING,"DYNDNS: Error reading IP from cache\n"));
+		DBG_PRINTF((LOG_WARNING,"Error reading IP from cache\n"));
 	}
 	fclose(fp);
 	DBG_PRINTF((LOG_INFO, MODULE_TAG "IP read from cache file is '%s'. No update required.\n", p_dyndns->info[0].my_ip_address.name));
@@ -1142,7 +1142,7 @@ int dyn_dns_main(DYN_DNS_CLIENT *p_dyndn
 			rc = os_install_signal_handler(p_dyndns);
 			if (rc != RC_OK)
 			{
-				DBG_PRINTF((LOG_WARNING,"DYNDNS: Error '%s' (0x%x) installing OS signal handler\n",
+				DBG_PRINTF((LOG_WARNING,"Error '%s' (0x%x) installing OS signal handler\n",
 					errorcode_get_name(rc), rc));
 				break;
 			}
diff -urBp inadyn.orig/src/get_cmd.c inadyn/src/get_cmd.c
--- inadyn.orig/src/get_cmd.c	2010-07-11 03:31:21.000000000 +0400
+++ inadyn/src/get_cmd.c	2010-07-11 20:03:13.000000000 +0400
@@ -24,7 +24,7 @@ Foundation, Inc., 59 Temple Place - Suit
 			Basic, table driven option line parsing.
 */ 
 
-#define MODULE_TAG "INADYN:CMD_LINE_OPT: "
+#define MODULE_TAG "CMD_LINE_OPT: "
 #include <string.h>
 #include <stdlib.h>
 #include "get_cmd.h"
diff -urBp inadyn.orig/src/ip.c inadyn/src/ip.c
--- inadyn.orig/src/ip.c	2010-07-11 03:31:21.000000000 +0400
+++ inadyn/src/ip.c	2010-07-11 20:02:44.000000000 +0400
@@ -15,7 +15,7 @@ You should have received a copy of the G
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
-#define MODULE_TAG "INADYN:IP: "
+#define MODULE_TAG "IP: "
 #include <stdlib.h>
 #include <string.h>
 
diff -urBp inadyn.orig/src/main.c inadyn/src/main.c
--- inadyn.orig/src/main.c	2010-07-11 03:31:21.000000000 +0400
+++ inadyn/src/main.c	2010-07-11 20:02:21.000000000 +0400
@@ -36,7 +36,7 @@ Foundation, Inc., 59 Temple Place - Suit
             - DBG print modified  
 		History not updated anymore. See readme.  		
 */
-#define MODULE_TAG "INADYN: "
+#define MODULE_TAG 
 #include "debug_if.h"
 #include <stdlib.h>
 #include "errorcode.h"
diff -urBp inadyn.orig/src/os.c inadyn/src/os.c
--- inadyn.orig/src/os.c	2010-07-11 03:31:21.000000000 +0400
+++ inadyn/src/os.c	2010-07-11 19:15:18.000000000 +0400
@@ -128,7 +128,7 @@ void os_printf(int prio, char *fmt, ...
 	#ifdef HAVE_OS_SYSLOG
     if (get_dbg_dest() == DBG_SYS_LOG)
     {
-        syslog(prio, "%s %s", print_time(), message);
+        syslog(prio, "%s", message);
     }
     else
 	#endif
diff -urBp inadyn.orig/src/os_unix.c inadyn/src/os_unix.c
--- inadyn.orig/src/os_unix.c	2010-07-11 03:31:21.000000000 +0400
+++ inadyn/src/os_unix.c	2010-07-11 20:08:51.000000000 +0400
@@ -127,7 +127,7 @@ RC_TYPE os_install_signal_handler(void *
              sigaction(SIGQUIT, &newact, NULL);
  	if (rc != RC_OK)
  	{
- 		DBG_PRINTF((LOG_WARNING,"DYNDNS: Error '%s' (0x%x) installing OS signal handler\n", rc));
+ 		DBG_PRINTF((LOG_WARNING,"Error '%s' (0x%x) installing OS signal handler\n", rc));
 	}
 	else
 	{
