diff -BurpN inadyn.orig/src/inadyn_cmd.c inadyn/src/inadyn_cmd.c
--- inadyn.orig/src/inadyn_cmd.c	2010-07-14 18:17:51.000000000 +0000
+++ inadyn/src/inadyn_cmd.c	2011-10-27 18:33:01.206624000 +0000
@@ -111,7 +111,7 @@ static CMD_DESCRIPTION_TYPE cmd_options_
 
 	{"--dyndns_system",	1,	{get_dyndns_system_handler, NULL},	
             "[NAME] - optional DYNDNS service type. SHOULD be one of the following: \n"
-            "\t\t-For dyndns.org: dyndns@dyndns.org OR statdns@dyndns.org OR customdns@dyndns.org.\n"
+            "\t\t-For dyndns.org: dyndns@dyndns.org OR statdns@dyndns.org OR custom@dyndns.org.\n"
             "\t\t-For freedns.afraid.org: default@freedns.afraid.org\n"
             "\t\t-For zoneedit.com: default@zoneedit.com\n"
             "\t\t-For no-ip.com: default@no-ip.com\n"
