 uClibc 0.9.29 compatibility patch due to unconfigurable separate libgcc
  
diff -BurN router/lib/Makefile gateway/lib/Makefile
--- router/lib/Makefile	2004-10-14 23:11:28.000000000 +0400
+++ gateway/lib/Makefile	2008-06-28 16:30:50.000000000 +0400
@@ -22,6 +22,7 @@
 lib-$(CONFIG_LIBPTHREAD) += libpthread
 lib-$(CONFIG_LIBRESOLV) += libresolv
 lib-$(CONFIG_LIBUTIL) += libutil
+lib-$(CONFIG_UCLIBC) += libgcc_s
 
 ifeq ($(CONFIG_UCLIBC),y)
 
