diff -BurpN '-x*.o' router/nvram/Makefile gateway/nvram/Makefile
--- router/nvram/Makefile	2004-11-11 10:30:36.000000000 +0300
+++ gateway/nvram/Makefile	2009-03-31 23:23:16.000000000 +0400
@@ -14,7 +14,7 @@
 
 CFLAGS	+= -I. -I$(SRCBASE)/include -Wall
 #CFLAGS	+= -g -DDEBUG
-CFLAGS	+= -s -O2 -DASUS
+CFLAGS	+= -s -O2 -DASUS $(EXTRACFLAGS)
 LDFLAGS += -L.
 LDFLAGS	+= -L$(TOP)/shared -L$(INSTALLDIR)/shared/usr/lib -lshared
 
