diff -uBp p910nd-0.93/p910nd.c p910nd/p910nd.c
--- p910nd-0.93/p910nd.c	2009-02-08 23:46:56.000000000 +0300
+++ p910nd/p910nd.c	2009-08-27 20:05:01.000000000 +0400
@@ -167,8 +167,7 @@ int open_printer(int lpnumber)
 		device = lpname;
 	if ((lp = open(device, bidir ? O_RDWR : O_WRONLY)) == -1) {
 		if (errno != EBUSY)
-			syslog(LOGOPTS, "%s: %m\n", device);
-		syslog(LOGOPTS, "%s: %m, will try opening later\n", device);
+			syslog(LOGOPTS, "%s: %m, will try opening later\n", device);
 	}
 	return (lp);
 }
@@ -335,8 +335,12 @@ int copy_stream(int fd, int lp)
 				if ((now.tv_sec > then.tv_sec) || (now.tv_sec == then.tv_sec && now.tv_usec > then.tv_usec)) {
 					timer = 0;
 				} else {
-					timeout.tv_sec = then.tv_sec;
-					timeout.tv_usec = then.tv_usec;
+					timeout.tv_sec = then.tv_sec - now.tv_sec;
+					timeout.tv_usec = then.tv_usec - now.tv_usec;
+					if (timeout.tv_usec < 0) {
+						timeout.tv_usec += 1000000;
+						timeout.tv_sec--;
+					}
 					FD_CLR(lp, &readfds);
 				}
 			}
@@ -398,7 +397,7 @@ void one_job(int lpnumber)
 
 void one_job(int lpnumber)
 {
-	int lp;
+	int lp, open_sleep = 10;
 	struct sockaddr_in client;
 	socklen_t clientlen = sizeof(client);
 
@@ -407,8 +406,11 @@ void one_job(int lpnumber)
 	if (get_lock(lpnumber) == 0)
 		return;
 	/* Make sure lp device is open... */
-	while ((lp = open_printer(lpnumber)) == -1)
-		sleep(10);
+	while ((lp = open_printer(lpnumber)) == -1) {
+		sleep(open_sleep);
+		if (open_sleep < 320) /* ~5 min interval to avoid spam in syslog */
+			open_sleep *= 2;
+	}
 	if (copy_stream(0, lp) < 0)
 		syslog(LOGOPTS, "copy_stream: %m\n");
 	close(lp);
@@ -422,6 +422,7 @@ void server(int lpnumber)
 	struct protoent *proto;
 #endif
 	int netfd, fd, lp, one = 1;
+	int open_sleep = 10;
 	socklen_t clientlen;
 	struct sockaddr_in netaddr, client;
 	char pidfilename[sizeof(PIDFILE)];
@@ -521,8 +522,12 @@ void server(int lpnumber)
 		/*write(fd, "Printing", 8); */
 
 		/* Make sure lp device is open... */
-		while ((lp = open_printer(lpnumber)) == -1)
-			sleep(10);
+		while ((lp = open_printer(lpnumber)) == -1) {
+			sleep(open_sleep);
+			if (open_sleep < 320) /* ~5 min interval to avoid spam in syslog */
+				open_sleep *= 2;
+		}
+		open_sleep = 10;
 
 		if (copy_stream(fd, lp) < 0)
 			syslog(LOGOPTS, "copy_stream: %m\n");
