diff -BurpN scsi-idle.orig/Makefile scsi-idle/Makefile
--- scsi-idle.orig/Makefile	2003-12-15 22:44:13.000000000 +0000
+++ scsi-idle/Makefile	2009-06-08 17:42:55.000000000 +0000
@@ -1,6 +1,6 @@
-CFLAGS=-Wall -O2 -I/lib/modules/`uname -r`/build/include
+CFLAGS=-Wall -O2 -I$(LINUXDIR)/include $(EXTRACFLAGS)
 
-TARGETS = scsi-start scsi-stop scsi-idle
+TARGETS = scsi-start scsi-stop
 
 all: $(TARGETS)
 
