diff -BurpN '-x*.o' router/sendmail/Makefile gateway/sendmail/Makefile
--- router/sendmail/Makefile	2004-11-11 10:32:18.000000000 +0300
+++ gateway/sendmail/Makefile	2009-03-31 23:08:51.000000000 +0400
@@ -9,7 +9,7 @@
 #CC =		mipsel-linux-gcc
 #CC =		gcc
 #STRIP = 	mipsel-linux-strip
-CFLAGS +=	-DWL600
+CFLAGS +=	-DWL600 -O2 $(EXTRACFLAGS)
 #CFLAGS =	-g
 #LDFLAGS =	-s -static
 #LDFLAGS =	-g -static
diff -BurpN '-x*.o' router/sendmail/mini_sendmail.c gateway/sendmail/mini_sendmail.c
--- router/sendmail/mini_sendmail.c	2004-11-11 10:32:18.000000000 +0300
+++ gateway/sendmail/mini_sendmail.c	2006-10-21 22:59:06.000000000 +0400
@@ -134,12 +134,7 @@ main( int argc, char** argv )
 	++argn;
 	}
 
-#ifndef WL600
     username = getlogin();
-#else
-    username = "WL600";
-    //fprintf(stderr, "%s\n", username);
-#endif
     if ( username == (char*) 0 )
 	{
 #ifdef DO_GETPWUID
@@ -156,15 +151,15 @@ main( int argc, char** argv )
 #endif /* DO_GETPWUID */
 	}
 
-#ifndef WL600
     if ( gethostname( hostname, sizeof(hostname) - 1 ) < 0 )
 	show_error( "gethostname" );
-#else
-	strcpy(hostname,"MAILER");
-#endif
 
     if ( fake_from == (char*) 0 )
+#ifdef WL600
+	(void) snprintf( from, sizeof(from), "%s", argv[argn]);
+#else
 	(void) snprintf( from, sizeof(from), "%s@%s", username, hostname );
+#endif
     else
 	if ( strchr( fake_from, '@' ) == (char*) 0 )
 	    (void) snprintf( from, sizeof(from), "%s@%s", fake_from, hostname );
@@ -212,11 +207,7 @@ main( int argc, char** argv )
 	exit( 1 );
 	}
 
-#ifdef WL600
-	(void) snprintf( buf, sizeof(buf), "MAIL FROM:<%s>", argv[argn]);
-#else
     (void) snprintf( buf, sizeof(buf), "MAIL FROM:<%s>", from );
-#endif
     send_command( buf );
     status = read_response();
     if ( status != 250 )
