diff -urbB iptables-1.3.8/extensions/libipt_set.h iptables/extensions/libipt_set.h
--- iptables-1.3.8/extensions/libipt_set.h	2007-01-23 15:50:00.000000000 +0300
+++ iptables/extensions/libipt_set.h	2008-12-19 12:15:51.000000000 +0300
@@ -66,7 +66,7 @@
 	if (size != sizeof(struct ip_set_req_get_set))
 		exit_error(OTHER_PROBLEM,
 			   "Incorrect return size from kernel during ipset lookup, "
-			   "(want %ld, got %ld)\n",
+			   "(want %d, got %d)\n",
 			   sizeof(struct ip_set_req_get_set), size);
 	if (req.set.index == IP_SET_INVALID_ID)
 		exit_error(PARAMETER_PROBLEM,
@@ -92,7 +92,7 @@
 	if (size != sizeof(struct ip_set_req_get_set))
 		exit_error(OTHER_PROBLEM,
 			   "Incorrect return size from kernel during ipset lookup, "
-			   "(want %ld, got %ld)\n",
+			   "(want %d, got %d)\n",
 			   sizeof(struct ip_set_req_get_set), size);
 	if (req.set.name[0] == '\0')
 		exit_error(PARAMETER_PROBLEM,
