--- src/shared/siutils.c.orig	2007-04-17 00:13:19.405953208 +0400
+++ src/shared/siutils.c	2007-04-17 00:16:20.324449416 +0400
@@ -1164,8 +1164,13 @@ BCMINITFN(si_clock)(si_t *sih)
 	else
 		m = R_REG(sii->osh, &cc->clockcontrol_sb);
 
+	if (sih->chip == BCM5365_CHIP_ID)
+	{
+		rate = 200000000; /* PLL_TYPE3 */
+	} else {
 	/* calculate rate */
 	rate = si_clock_rate(pll_type, n, m);
+	}
 
 	if (pll_type == PLL_TYPE3)
 		rate = rate / 2;
--- src/shared/hndmips.c.orig	2007-04-17 00:00:33.073453456 +0400
+++ src/shared/hndmips.c	2007-04-17 00:04:15.790595304 +0400
@@ -234,10 +234,12 @@
 	tmp = CEIL(10, ns) << FW_W3_SHIFT;	/* W3 = 10nS */
 	tmp |= CEIL(10, ns) << FW_W1_SHIFT;	/* W1 = 10nS */
 	tmp |= CEIL(120, ns);			/* W0 = 120nS */
-	if (sih->ccrev < 9)
+	if ((sih->ccrev < 9) ||
+	    (sih->chip == BCM5365_CHIP_ID))
 		W_REG(osh, &cc->flash_waitcount, tmp);
 
 	if ((sih->ccrev < 9) ||
+	    (sih->chip == BCM5365_CHIP_ID) ||
 	    ((sih->chip == BCM5350_CHIP_ID) && sih->chiprev == 0)) {
 		W_REG(osh, &cc->pcmcia_memwait, tmp);
 	}
