By rss @ wl500g.info

diff -urBp madwimax/src/wimax.c b/src/wimax.c
--- madwimax/src/wimax.c	2010-02-21 03:00:00.000000000 +0300
+++ b/src/wimax.c	2010-03-07 10:13:20.000000000 +0300
@@ -653,6 +653,7 @@ void usage(const char *progname)
 	printf("  -e, --event-script=FILE     specify path to the event script\n");
 	printf("  -p, --pid-file=FILE         specify path to the pid-file\n");
 	printf("  -s, --stats-file=FILE       specify path to the statistics file\n");
+	printf("  -i, --interface=NAME        specify wimax interface name (by default wmx...)\n");
 	printf("  -h, --help                  display this help\n");
 }
 
@@ -685,10 +686,11 @@ static void parse_args(int argc, char **
 			{"pid-file",		required_argument,	0, 'p'},
 			{"stats-file",		required_argument,	0, 's'},
 			{"help",		no_argument,		0, 'h'},
+			{"interface",		required_argument,	0, 'i'},
 			{0, 0, 0, 0}
 		};
 
-		c = getopt_long(argc, argv, "vqdl:ofVe:p:s:h", long_options, &option_index);
+		c = getopt_long(argc, argv, "vqdl:ofVe:p:s:h:i:", long_options, &option_index);
 
 		/* detect the end of the options. */
 		if (c == -1)
@@ -803,6 +805,14 @@ static void parse_args(int argc, char **
 					exit(1);
 					break;
 				}
+			case 'i': {
+					if( strlen(optarg)>=sizeof(tap_dev) ){
+						fprintf(stderr, "Too long interface name.\n");
+						exit(1);
+					}
+					strcpy( tap_dev, optarg );
+					break;
+			}
 			default: {
 					exit(1);
 				}
@@ -883,6 +893,11 @@ static void sighandler_stats(int signum)
 				wd_status.bsid[0], wd_status.bsid[1], wd_status.bsid[2],
 				wd_status.bsid[3], wd_status.bsid[4], wd_status.bsid[5]);
 		}
+		fprintf( fd,
+			"-----------------------\n"
+			"Chip info: %s\n" "Firmware info: %s", 
+			wd_status.chip, wd_status.firmware );
+
 		fclose(fd);
 	}
 }
