Support of ls -v even for our uClibc 0.9.30.1

diff -uBp a/include/platform.h b/include/platform.h
--- a/include/platform.h
+++ b/include/platform.h
@@ -366,7 +366,7 @@ typedef unsigned smalluint;
 #if defined(__UCLIBC_MAJOR__)
 # if __UCLIBC_MAJOR__ == 0 \
   && (   __UCLIBC_MINOR__ < 9 \
-     || (__UCLIBC_MINOR__ == 9 && __UCLIBC_SUBLEVEL__ < 32) \
+     || (__UCLIBC_MINOR__ == 9 && __UCLIBC_SUBLEVEL__ < 30) \
      )
 #  undef HAVE_STRVERSCMP
 # endif
