--- src.orig/shared/bcmrobo.c	2012-02-12 17:10:20.924343000 +0000
+++ src/shared/bcmrobo.c	2012-02-11 22:01:38.000000000 +0000
@@ -703,11 +703,12 @@ bcm_robo_attach(si_t *sih, void *h, char
 	/* Enable switch leds */
 	if (sih->chip == BCM5356_CHIP_ID) {
 		si_pmu_chipcontrol(sih, 2, (1 << 25), (1 << 25));
-	} else if (sih->chip == BCM5357_CHIP_ID) {
+	} else if ((sih->chip == BCM5357_CHIP_ID) || (sih->chip == BCM53572_CHIP_ID)) {
+
 		uint32 led_gpios = 0;
 		char *var;
 
-		if (sih->chippkg != BCM47186_PKG_ID)
+		if ((sih->chippkg != BCM47186_PKG_ID) && (sih->chippkg != BCM47188_PKG_ID))
 			led_gpios = 0x1f;
 		var = getvar(vars, "et_swleds");
 		if (var)
@@ -759,7 +760,8 @@ bcm_robo_attach(si_t *sih, void *h, char
 	       (robo->devid == DEVID5395) ||
 	       (robo->devid == DEVID5397) ||
 	       (robo->devid == DEVID5398) ||
-	       (robo->devid == DEVID53115));
+	       (robo->devid == DEVID53115) ||
+	       (robo->devid == DEVID53125));
 
 	/* nvram variable switch_mode controls the power save mode on the switch
 	 * set the default value in the beginning
@@ -1136,7 +1138,9 @@ vlan_setup:
 		} else {
 			uint8 vtble, vtbli, vtbla;
 
-			if ((robo->devid == DEVID5395) || (robo->devid == DEVID53115)) {
+			if ((robo->devid == DEVID5395) ||
+				(robo->devid == DEVID53115) ||
+				(robo->devid == DEVID53125)) {
 				vtble = REG_VTBL_ENTRY_5395;
 				vtbli = REG_VTBL_INDX_5395;
 				vtbla = REG_VTBL_ACCESS_5395;
@@ -1205,24 +1209,25 @@ vlan_setup:
 			 (3 << 12) |	/* Pri 6 mapped to TXQ 3 */
 			 (3 << 14));	/* Pri 7 mapped to TXQ 3 */
 		robo->ops->write_reg(robo, 0x30, 0x62, &val16, sizeof(val16));
+	}
 
-		/* Drop reserved bit, if any */
-		robo->ops->read_reg(robo, PAGE_CTRL, 0x2f, &val8, sizeof(val8));
-		if (val8 & (1 << 1)) {
-			val8 &= ~(1 << 1);
-			robo->ops->write_reg(robo, PAGE_CTRL, 0x2f, &val8, sizeof(val8));
-		}
+	/* Drop reserved bit, if any */
+	robo->ops->read_reg(robo, PAGE_CTRL, 0x2f, &val8, sizeof(val8));
+	if (/*((robo->devid == DEVID53115) || (robo->devid == DEVID53125)) &&*/
+	    val8 & (1 << 1)) {
+		val8 &= ~(1 << 1);
+		robo->ops->write_reg(robo, PAGE_CTRL, 0x2f, &val8, sizeof(val8));
+	}
 
-		/* Jumbo Frame control refister (Page 0x40, Address 0x01 */
+	/* Jumbo Frame control register (Page 0x40, Address 0x01 */
+	if ((robo->devid == DEVID53115) || (robo->devid == DEVID53125)) {
 		if (nvram_match("jumbo_frame_enable", "1")) {
 			val16 = (( 1 << 0) |
 				 ( 1 << 1) |
 				 ( 1 << 2) |
 				 ( 1 << 3) |
-				 ( 1 << 4) 
-				);
-		}
-		else
+				 ( 1 << 4));
+		} else
 			val16 = 0;
 		robo->ops->write_reg(robo, PAGE_JUMBO, REG_JUMBO_CTRL, &val16, sizeof(val16));
 	}
@@ -1266,7 +1271,8 @@ bcm_robo_enable_switch(robo_info_t *robo
 
 		/* No spanning tree for unmanaged mode */
 		val8 = 0;
-		max_port_ind = ((robo->devid == DEVID5398) ? REG_CTRL_PORT7 : REG_CTRL_PORT4);
+		max_port_ind = ((robo->devid == DEVID5398) ? REG_CTRL_PORT7 :
+		                (robo->devid == DEVID53115) ? REG_CTRL_PORT5 : REG_CTRL_PORT4);
 		for (i = REG_CTRL_PORT0; i <= max_port_ind; i++) {
 			robo->ops->write_reg(robo, PAGE_CTRL, i, &val8, sizeof(val8));
 		}
@@ -1280,6 +1286,13 @@ bcm_robo_enable_switch(robo_info_t *robo
 	{
 		val8 = 0;
 		robo->ops->write_reg(robo, PAGE_CTRL, REG_CTRL_PORT0, &val8, sizeof(val8));
+	} else
+	if (robo->devid == DEVID53125) {
+		/* Over ride IMP port status to make it link by default */
+		val8 = 0;
+		robo->ops->read_reg(robo, PAGE_CTRL, REG_CTRL_MIIPO, &val8, sizeof(val8));
+		val8 |= 0x81;	/* Make Link pass and override it. */
+		robo->ops->write_reg(robo, PAGE_CTRL, REG_CTRL_MIIPO, &val8, sizeof(val8));
 	}
 
 	/* Disable management interface access */
@@ -1676,7 +1689,8 @@ robo_power_save_mode_manual(robo_info_t
 		return 0;
 
 	/* If the switch supports manual power down enable that */  
-	if (robo->devid ==  DEVID53115) {
+	if ((robo->devid ==  DEVID53115) ||
+	    (robo->sih->chip == BCM5356_CHIP_ID)) {
 		/* For 53115 0x0 is the MII control register bit 11 is the
 		 * power down mode bit 
 		 */
@@ -1722,7 +1736,8 @@ robo_power_save_mode_wlonly(robo_info_t
 	uint16 val16;
 	int32 phy;
 
-	if (robo->devid ==  DEVID53115) {
+	if ((robo->devid ==  DEVID53115) ||
+	    (robo->devid ==  DEVID53125)) {
 		/* For 53115 0x0 is the MII control register bit 11 is the
 		 * power down mode bit 
 		 */
