Reduce logging spam patches from Tomato-RT

diff -urBp a/natpmp.c b/natpmp.c
--- a/natpmp.c	2010-05-11 20:20:02.000000000 +0400
+++ b/natpmp.c	2011-05-19 21:05:53.000000000 +0400
@@ -209,8 +209,7 @@ void ProcessIncomingNATPMPPacket(int s)
 					r = _upnp_delete_redir(eport, proto);
 					/*syslog(LOG_DEBUG, "%hu %d r=%d", eport, proto, r);*/
 					if(r<0) {
-						syslog(LOG_ERR, "Failed to remove NAT-PMP mapping eport %hu, protocol %s",
-						       eport, (proto==IPPROTO_TCP)?"TCP":"UDP");
+						//syslog(LOG_ERR, "Failed to remove NAT-PMP mapping eport %hu, protocol %s", eport, (proto==IPPROTO_TCP)?"TCP":"UDP");
 						resp[3] = 2;	/* Not Authorized/Refused */
 					}
 				}
@@ -328,7 +328,7 @@ int CleanExpiredNATPMP()
 	/* remove redirection then search for next one:) */
 	if(_upnp_delete_redir(nextnatpmptoclean_eport, nextnatpmptoclean_proto)<0)
 		return -1;
-	syslog(LOG_INFO, "Expired NAT-PMP mapping port %hu %s removed",
+	syslog(LOG_NOTICE, "Expired NAT-PMP mapping port %hu %s removed",
 	       nextnatpmptoclean_eport,
 	       nextnatpmptoclean_proto==IPPROTO_TCP?"TCP":"UDP");
 	return ScanNATPMPforExpiration();
diff -urBp a/upnphttp.c b/upnphttp.c
--- a/upnphttp.c	2009-02-13 22:38:34.000000000 +0300
+++ b/upnphttp.c	2011-05-19 21:05:47.000000000 +0400
@@ -431,7 +431,7 @@ ProcessHttpQuery_upnphttp(struct upnphtt
 #endif
 		else
 		{
-			syslog(LOG_NOTICE, "%s not found, responding ERROR 404", HttpUrl);
+			syslog(LOG_INFO, "%s not found, responding ERROR 404", HttpUrl);
 			Send404(h);
 		}
 	}
