--- src/shared/sbutils.c.orig	2007-04-17 00:13:19.405953208 +0400
+++ src/shared/sbutils.c	2007-04-17 00:16:20.324449416 +0400
@@ -1943,8 +1943,13 @@
 		return 0;
 	}
 
+	if (sb_chip(sbh) == BCM5365_CHIP_ID)
+	{
+		rate = 200000000; /* PLL_TYPE3 */
+	} else {
 		/* calculate rate */
 		rate = sb_clock_rate(pll_type, n, m);
+	}
 
 	if (pll_type == PLL_TYPE3)
 		rate = rate / 2;
--- src/shared/hndmips.c.orig	2007-04-17 00:00:33.073453456 +0400
+++ src/shared/hndmips.c	2007-04-17 00:04:15.790595304 +0400
@@ -193,10 +193,12 @@
 	tmp = CEIL(10, ns) << FW_W3_SHIFT;	/* W3 = 10nS */
 	tmp |= CEIL(10, ns) << FW_W1_SHIFT;	/* W1 = 10nS */
 	tmp |= CEIL(120, ns);			/* W0 = 120nS */
-	if (sbh->ccrev < 9)
+	if ((sbh->ccrev < 9) ||
+	    (sb_chip(sbh) == BCM5365_CHIP_ID))
 		W_REG(osh, &cc->flash_waitcount, tmp);
 
 	if ((sbh->ccrev < 9) ||
+	    (sb_chip(sbh) == BCM5365_CHIP_ID) ||
 	    ((sb_chip(sbh) == BCM5350_CHIP_ID) && sb_chiprev(sbh) == 0)) {
 		W_REG(osh, &cc->pcmcia_memwait, tmp);
 	}
