Fix uninitialized vars.

--- rp-l2tp-0.4/dgram.c	2009-12-04 14:30:24.000000000 +0300
+++ rp-l2tp/dgram.c	2009-12-04 14:30:34.000000000 +0300
@@ -316,7 +316,7 @@ l2tp_dgram_take_from_wire(struct sockadd
     unsigned char *payload;
     unsigned char *tidptr;
     uint16_t tid, sid;
-    uint16_t cache_tid, cache_sid;
+    uint16_t cache_tid = 0, cache_sid = 0;
     l2tp_tunnel *tunnel;
     l2tp_session *ses = NULL;
     int mandatory, hidden, err;
