diff -BurpN rp-l2tp-0.4/tunnel.c rp-l2tp/tunnel.c
--- rp-l2tp-0.4/tunnel.c	2009-08-31 06:00:47.000000000 +0000
+++ rp-l2tp/tunnel.c	2009-08-31 06:01:57.000000000 +0000
@@ -539,6 +539,10 @@ tunnel_establish(l2tp_peer *peer, EventS
     tunnel->peer = peer;
     tunnel->peer_addr = peer_addr;
 
+    DBG(l2tp_db(DBG_TUNNEL, "tunnel_establish(%s) -> %s (%s)\n",
+	    l2tp_debug_tunnel_to_str(tunnel),
+	    inet_ntoa(peer_addr.sin_addr), peer->peername));
+
     memset(&tunnel->rt, 0, sizeof(tunnel->rt));
     route_add(tunnel->peer_addr.sin_addr, &tunnel->rt);
 
@@ -665,8 +669,8 @@ l2tp_tunnel_handle_received_control_data
 
     if (!tunnel) {
 	/* TODO: Send error message back? */
-	l2tp_set_errmsg("Invalid control message - unknown tunnel ID %d",
-		   (int) dgram->tid);
+	DBG(l2tp_db(DBG_TUNNEL, "Invalid control message - unknown tunnel ID %d",
+		   (int) dgram->tid));
 	return;
     }
 
@@ -981,6 +985,8 @@ tunnel_handle_timeout(EventSelector *es,
 
     /* Timeout handler has fired */
     tunnel->timeout_handler = NULL;
+    DBG(l2tp_db(DBG_FLOW, "tunnel_handle_timeout(%s)\n",
+	   l2tp_debug_tunnel_to_str(tunnel)));
 
     /* Reset xmit_new_dgrams */
     tunnel->xmit_new_dgrams = tunnel->xmit_queue_head;
