--- rp-l2tp/main.c.orig	2008-02-02 22:59:23.000000000 +0300
+++ rp-l2tp/main.c	2008-02-02 23:03:27.000000000 +0300
@@ -39,6 +39,13 @@
     exit(exitcode);
 }
 
+static void
+sighandler(int signum)
+{
+    l2tp_cleanup();
+    exit(EXIT_FAILURE);
+}
+
 int
 main(int argc, char *argv[])
 {
@@ -114,6 +121,9 @@
 	}
     }
 
+    signal(SIGTERM, sighandler);
+    signal(SIGINT, sighandler);
+
     while(1) {
 	i = Event_HandleEvent(es);
 	if (i < 0) {
