diff -urBp fuse-2.5.3/Makefile.in fuse/Makefile.in
--- fuse-2.5.3/Makefile.in	2006-02-02 20:04:52.000000000 +0300
+++ fuse/Makefile.in	2009-06-09 22:15:11.000000000 +0400
@@ -3,6 +3,7 @@
 SHELL = /bin/sh
 INSTALL = @INSTALL@
 mkdir_p = mkdir -p
+DEPMOD = /sbin/depmod -a
 majver = @majver@
 VERSION = @PACKAGE_VERSION@
 
@@ -10,7 +11,7 @@ DISTFILES = Makefile.in configure.ac con
 	dev.c dir.c file.c inode.c fuse_i.h fuse_kernel.h
 COMPATDISTFILES = compat/parser.c compat/parser.h
 
-fusemoduledir = @kmoduledir@/kernel/fs/fuse
+fusemoduledir = $(INSTALL_MOD_PATH)@kmoduledir@/fs/fuse
 
 ifeq ($(majver), 2.4)
 fusemodule := fuse.o
@@ -18,7 +19,10 @@ else
 fusemodule := fuse.ko
 endif
 
+ifndef MAKING_MODULES
 all: all-@ENABLE_FUSE_MODULE@
+all-y: all-spec
+endif
 install: install-@ENABLE_FUSE_MODULE@
 uninstall: uninstall-@ENABLE_FUSE_MODULE@
 
@@ -26,16 +30,14 @@ all-n:
 install-n:
 uninstall-n:
 
-all-y: all-spec
-
 install-y: all
-	$(mkdir_p) $(DESTDIR)$(fusemoduledir)
-	$(INSTALL) -m 644 $(fusemodule) $(DESTDIR)$(fusemoduledir)/$(fusemodule)
-	-/sbin/depmod -a
+	$(mkdir_p) $(fusemoduledir)
+	$(INSTALL) -m 644 $(fusemodule) $(fusemoduledir)/$(fusemodule)
+	-$(DEPMOD)
 
 uninstall-y:
-	rm -f $(DESTDIR)$(fusemoduledir)/$(fusemodule)
-	-/sbin/depmod -a
+	rm -f $(fusemoduledir)/$(fusemodule)
+	-$(DEPMOD)
 
 clean:
 	-rm -f $(fusemodule) *.o .*.cmd *.mod.c *.ko *.s */*.o
@@ -52,24 +54,18 @@ distdir: $(DISTFILES) $(COMPATDISTFILES)
 	mkdir $(distdir)/compat
 	cp -p $(COMPATDISTFILES) $(distdir)/compat
 
-ifeq ($(majver), 2.4)
-
-CC = gcc
-LD = ld
-CFLAGS = -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -pipe @KERNELCFLAGS@
-CPPFLAGS = -I@kernelsrc@/include -I. -D__KERNEL__ -DMODULE -D_LOOSE_KERNEL_NAMES -DFUSE_VERSION=\"$(VERSION)\" @KERNELCPPFLAGS@
 
-fuse_objs = dev.o dir.o file.o inode.o compat/parser.o
+EXTRA_CFLAGS += -DFUSE_VERSION=\"$(VERSION)\"
+KERNELDIR := @kernelsrc@
 
-SUFFIXES = .c .o .s
+ifeq ($(majver), 2.4)
 
-all-spec: fuse.o
+fuse-objs := dev.o dir.o file.o inode.o compat/parser.o
 
-.c.o:
-	$(CC) $(CFLAGS) $(CPPFLAGS) -c $< -o $@
+O_TARGET := $(fusemodule)
 
-fuse.o: $(fuse_objs)
-	$(LD) -r -o fuse.o $(fuse_objs)
+obj-y := $(fuse-objs)
+obj-m := $(O_TARGET)
 
 fuse_headers = fuse_i.h fuse_kernel.h
 
@@ -78,13 +74,18 @@ dir.o: $(fuse_headers)
 file.o: $(fuse_headers)
 inode.o: $(fuse_headers)
+compat/parser.o: compat/parser.h
 
-else
+IGNORE_FLAGS_OBJS=compat/parser.o
+TOPDIR := $(KERNELDIR)
 
-EXTRA_CFLAGS += -DFUSE_VERSION=\"$(VERSION)\"
+include $(TOPDIR)/Rules.make
+
+else
 
 obj-m := fuse.o
 fuse-objs := dev.o dir.o file.o inode.o
 
-all-spec:
-	$(MAKE) -C @kernelsrc@ SUBDIRS=$(PWD) @KERNELMAKE_PARAMS@ modules
 endif
+
+all-spec:
+	$(MAKE) -C $(KERNELDIR) SUBDIRS=$(PWD) @KERNELMAKE_PARAMS@ CC=@CC@ modules
diff -urBp fuse-2.5.3/configure fuse/configure
--- fuse-2.5.3/configure	2006-04-10 12:45:53.000000000 +0400
+++ fuse/configure	2009-06-09 19:08:32.000000000 +0400
@@ -271,7 +271,7 @@ PACKAGE_VERSION='2.5.3'
 PACKAGE_STRING='fuse-kernel 2.5.3'
 PACKAGE_BUGREPORT=''
 
-ac_subst_vars='SHELL PATH_SEPARATOR PACKAGE_NAME PACKAGE_TARNAME PACKAGE_VERSION PACKAGE_STRING PACKAGE_BUGREPORT exec_prefix prefix program_transform_name bindir sbindir libexecdir datadir sysconfdir sharedstatedir localstatedir libdir includedir oldincludedir infodir mandir build_alias host_alias target_alias DEFS ECHO_C ECHO_N ECHO_T LIBS INSTALL_PROGRAM INSTALL_SCRIPT INSTALL_DATA kernelsrc majver kmoduledir CC CFLAGS LDFLAGS CPPFLAGS ac_ct_CC EXEEXT OBJEXT ENABLE_FUSE_MODULE KERNELMAKE_PARAMS KERNELCPPFLAGS KERNELCFLAGS LIBOBJS LTLIBOBJS'
+ac_subst_vars='SHELL PATH_SEPARATOR PACKAGE_NAME PACKAGE_TARNAME PACKAGE_VERSION PACKAGE_STRING PACKAGE_BUGREPORT exec_prefix prefix program_transform_name bindir sbindir libexecdir datadir sysconfdir sharedstatedir localstatedir libdir includedir oldincludedir infodir mandir build_alias host_alias target_alias DEFS ECHO_C ECHO_N ECHO_T LIBS INSTALL_PROGRAM INSTALL_SCRIPT INSTALL_DATA kernelsrc majver kmoduledir CC CFLAGS LDFLAGS CPPFLAGS ac_ct_CC EXEEXT OBJEXT ENABLE_FUSE_MODULE KERNELMAKE_PARAMS KERNELCPPFLAGS KERNELCFLAGS LD LIBOBJS LTLIBOBJS'
 ac_subst_files=''
 
 # Initialize some variables set by options.
@@ -907,7 +907,7 @@ esac
     else
       echo "$as_me: WARNING: no configuration information is in $ac_dir" >&2
     fi
-    cd "$ac_popdir"
+    cd $ac_popdir
   done
 fi
 
@@ -1374,7 +1374,6 @@ test -z "$INSTALL_DATA" && INSTALL_DATA=
 
 runver=`uname -r`
 ENABLE_FUSE_MODULE=y
-KERNELCFLAGS=
 
 kernelsrc=
 kernelbuild=
@@ -1491,7 +1490,7 @@ echo "$as_me: error:
 	echo "$as_me:$LINENO: result: $kernsrcver" >&5
 echo "${ECHO_T}$kernsrcver" >&6
 	majver=`echo "$kernsrcver" | cut -f-2 -d.`
-	kmoduledir=${INSTALL_MOD_PATH}/lib/modules/$kernsrcver
+	kmoduledir=/lib/modules/$kernsrcver/kernel
 
 
 
@@ -2067,7 +2066,8 @@ if { (eval echo "$as_me:$LINENO: \"$ac_c
   cat conftest.err >&5
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); } &&
-	 { ac_try='test -z "$ac_c_werror_flag"			 || test ! -s conftest.err'
+	 { ac_try='test -z "$ac_c_werror_flag"
+			 || test ! -s conftest.err'
   { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
   (eval $ac_try) 2>&5
   ac_status=$?
@@ -2125,7 +2125,8 @@ if { (eval echo "$as_me:$LINENO: \"$ac_c
   cat conftest.err >&5
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); } &&
-	 { ac_try='test -z "$ac_c_werror_flag"			 || test ! -s conftest.err'
+	 { ac_try='test -z "$ac_c_werror_flag"
+			 || test ! -s conftest.err'
   { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
   (eval $ac_try) 2>&5
   ac_status=$?
@@ -2241,7 +2242,8 @@ if { (eval echo "$as_me:$LINENO: \"$ac_c
   cat conftest.err >&5
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); } &&
-	 { ac_try='test -z "$ac_c_werror_flag"			 || test ! -s conftest.err'
+	 { ac_try='test -z "$ac_c_werror_flag"
+			 || test ! -s conftest.err'
   { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
   (eval $ac_try) 2>&5
   ac_status=$?
@@ -2295,7 +2297,8 @@ if { (eval echo "$as_me:$LINENO: \"$ac_c
   cat conftest.err >&5
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); } &&
-	 { ac_try='test -z "$ac_c_werror_flag"			 || test ! -s conftest.err'
+	 { ac_try='test -z "$ac_c_werror_flag"
+			 || test ! -s conftest.err'
   { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
   (eval $ac_try) 2>&5
   ac_status=$?
@@ -2340,7 +2343,8 @@ if { (eval echo "$as_me:$LINENO: \"$ac_c
   cat conftest.err >&5
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); } &&
-	 { ac_try='test -z "$ac_c_werror_flag"			 || test ! -s conftest.err'
+	 { ac_try='test -z "$ac_c_werror_flag"
+			 || test ! -s conftest.err'
   { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
   (eval $ac_try) 2>&5
   ac_status=$?
@@ -2384,7 +2388,8 @@ if { (eval echo "$as_me:$LINENO: \"$ac_c
   cat conftest.err >&5
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); } &&
-	 { ac_try='test -z "$ac_c_werror_flag"			 || test ! -s conftest.err'
+	 { ac_try='test -z "$ac_c_werror_flag"
+			 || test ! -s conftest.err'
   { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
   (eval $ac_try) 2>&5
   ac_status=$?
@@ -2457,7 +2462,8 @@ if { (eval echo "$as_me:$LINENO: \"$ac_c
   cat conftest.err >&5
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); } &&
-	 { ac_try='test -z "$ac_c_werror_flag"			 || test ! -s conftest.err'
+	 { ac_try='test -z "$ac_c_werror_flag"
+			 || test ! -s conftest.err'
   { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
   (eval $ac_try) 2>&5
   ac_status=$?
@@ -2521,7 +2527,8 @@ if { (eval echo "$as_me:$LINENO: \"$ac_c
   cat conftest.err >&5
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); } &&
-	 { ac_try='test -z "$ac_c_werror_flag"			 || test ! -s conftest.err'
+	 { ac_try='test -z "$ac_c_werror_flag"
+			 || test ! -s conftest.err'
   { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
   (eval $ac_try) 2>&5
   ac_status=$?
@@ -2580,7 +2587,8 @@ if { (eval echo "$as_me:$LINENO: \"$ac_c
   cat conftest.err >&5
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); } &&
-	 { ac_try='test -z "$ac_c_werror_flag"			 || test ! -s conftest.err'
+	 { ac_try='test -z "$ac_c_werror_flag"
+			 || test ! -s conftest.err'
   { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
   (eval $ac_try) 2>&5
   ac_status=$?
@@ -2678,8 +2686,8 @@ echo "${ECHO_T}no" >&6
 	fi
 
 	isuml=no
+	KERNELCPPFLAGS=$CPPFLAGS
 	KERNELMAKE_PARAMS=
-	KERNELCPPFLAGS=
 	echo "$as_me:$LINENO: checking if this is user mode linux" >&5
 echo $ECHO_N "checking if this is user mode linux... $ECHO_C" >&6
 	if test -f $kernelbuild/include/linux/autoconf.h && egrep -q "^#define CONFIG_(USERMODE|UML) 1" $kernelbuild/include/linux/autoconf.h; then
@@ -3327,6 +3335,7 @@ s,@ENABLE_FUSE_MODULE@,$ENABLE_FUSE_MODU
 s,@KERNELMAKE_PARAMS@,$KERNELMAKE_PARAMS,;t t
 s,@KERNELCPPFLAGS@,$KERNELCPPFLAGS,;t t
 s,@KERNELCFLAGS@,$KERNELCFLAGS,;t t
+s,@LD@,$LD,;t t
 s,@LIBOBJS@,$LIBOBJS,;t t
 s,@LTLIBOBJS@,$LTLIBOBJS,;t t
 CEOF
@@ -3495,6 +3504,11 @@ esac
   *) ac_INSTALL=$ac_top_builddir$INSTALL ;;
   esac
 
+  if test x"$ac_file" != x-; then
+    { echo "$as_me:$LINENO: creating $ac_file" >&5
+echo "$as_me: creating $ac_file" >&6;}
+    rm -f "$ac_file"
+  fi
   # Let's still pretend it is `configure' which instantiates (i.e., don't
   # use $as_me), people would be surprised to read:
   #    /* config.h.  Generated by config.status.  */
@@ -3533,12 +3547,6 @@ echo "$as_me: error: cannot find input f
 	 fi;;
       esac
     done` || { (exit 1); exit 1; }
-
-  if test x"$ac_file" != x-; then
-    { echo "$as_me:$LINENO: creating $ac_file" >&5
-echo "$as_me: creating $ac_file" >&6;}
-    rm -f "$ac_file"
-  fi
 _ACEOF
 cat >>$CONFIG_STATUS <<_ACEOF
   sed "$ac_vpsub
diff -urBp fuse-2.5.3/configure.ac fuse/configure.ac
--- fuse-2.5.3/configure.ac	2006-04-10 12:43:52.000000000 +0400
+++ fuse/configure.ac	2009-06-09 19:06:05.000000000 +0400
@@ -5,7 +5,6 @@ AC_PROG_INSTALL
 
 runver=`uname -r`
 ENABLE_FUSE_MODULE=y
-KERNELCFLAGS=
 
 kernelsrc=
 kernelbuild=
@@ -87,7 +86,7 @@ if test "$ENABLE_FUSE_MODULE" = y; then
 	fi
 	AC_MSG_RESULT([$kernsrcver])
 	majver=`echo "$kernsrcver" | cut -f-2 -d.`
-	kmoduledir=${INSTALL_MOD_PATH}/lib/modules/$kernsrcver
+	kmoduledir=/lib/modules/$kernsrcver/kernel
 	AC_SUBST(kernelsrc)
 	AC_SUBST(majver)
 	AC_SUBST(kmoduledir)
@@ -164,8 +163,8 @@ if test "$ENABLE_FUSE_MODULE" = y; then
 	fi
 
 	isuml=no
+	KERNELCPPFLAGS=$CPPFLAGS
 	KERNELMAKE_PARAMS=
-	KERNELCPPFLAGS=
 	AC_MSG_CHECKING([if this is user mode linux])
 	if test -f $kernelbuild/include/linux/autoconf.h && egrep -q "^#define CONFIG_(USERMODE|UML) 1" $kernelbuild/include/linux/autoconf.h; then
 		isuml=yes
@@ -181,5 +180,7 @@ if test "$ENABLE_FUSE_MODULE" = y; then
 	AC_SUBST(KERNELCFLAGS)
 fi
 
+AC_SUBST([LD])
+
 AC_CONFIG_FILES([Makefile])
 AC_OUTPUT
diff -urBp fuse-2.5.3/dev.c fuse/dev.c
--- fuse-2.5.3/dev.c	2006-02-02 20:04:52.000000000 +0300
+++ fuse/dev.c	2009-06-04 14:48:11.000000000 +0400
@@ -21,6 +21,11 @@
 MODULE_ALIAS_MISCDEV(FUSE_MINOR);
 #endif
 
+#ifdef DCACHE_BUG
+extern void r4k_flush_cache_page(struct vm_area_struct *vma,
+                                        unsigned long page);
+#endif
+
 static kmem_cache_t *fuse_req_cachep;
 
 static inline struct fuse_conn *fuse_get_conn(struct file *file)
@@ -512,6 +517,9 @@ static int fuse_copy_fill(struct fuse_co
 {
 	unsigned long offset;
 	int err;
+#ifdef DCACHE_BUG
+	struct vm_area_struct *vma;
+#endif
 
 	unlock_request(cs->req);
 	fuse_copy_finish(cs);
@@ -523,14 +531,22 @@ static int fuse_copy_fill(struct fuse_co
 		cs->nr_segs --;
 	}
 	down_read(&current->mm->mmap_sem);
+#ifndef DCACHE_BUG
 	err = get_user_pages(current, current->mm, cs->addr, 1, cs->write, 0,
 			     &cs->pg, NULL);
+#else
+	err = get_user_pages(current, current->mm, cs->addr, 1, cs->write, 0,
+			     &cs->pg, &vma);
+#endif
 	up_read(&current->mm->mmap_sem);
 	if (err < 0)
 		return err;
 	BUG_ON(err != 1);
 	offset = cs->addr % PAGE_SIZE;
 	cs->mapaddr = kmap_atomic(cs->pg, KM_USER0);
+#ifdef DCACHE_BUG
+	r4k_flush_cache_page(vma, cs->addr); 
+#endif
 	cs->buf = cs->mapaddr + offset;
 	cs->len = min(PAGE_SIZE - offset, cs->seglen);
 	cs->seglen -= cs->len;
@@ -545,6 +561,11 @@ static inline int fuse_copy_do(struct fu
 {
 	unsigned ncpy = min(*size, cs->len);
 	if (val) {
+#ifdef DCACHE_BUG
+		// patch from mailing list, it is very important, otherwise,
+		// can't mount, or ls mount point will hang
+		flush_cache_all();
+#endif
 		if (cs->write)
 			memcpy(cs->buf, *val, ncpy);
 		else
diff -urBp fuse-2.5.3/fuse_i.h fuse/fuse_i.h
--- fuse-2.5.3/fuse_i.h	2006-02-02 20:04:52.000000000 +0300
+++ fuse/fuse_i.h	2008-10-22 19:03:50.000000000 +0400
@@ -45,6 +45,10 @@
 #  endif
 #endif
 
+#if (defined(__mips__) || defined(__arm__)) && LINUX_VERSION_CODE < KERNEL_VERSION(2,6,16)
+#define DCACHE_BUG
+#endif
+
 #include "config.h"
 #ifndef KERNEL_2_6
 #  include <linux/config.h>
