diff -urbB dnsmasq-2.46/Makefile dnsmasq/Makefile
--- dnsmasq-2.46/Makefile	2008-11-03 19:29:41.000000000 +0300
+++ dnsmasq/Makefile	2008-11-17 11:26:48.000000000 +0300
@@ -23,7 +23,7 @@
 MAN = man
 
 PKG_CONFIG = pkg-config
-AWK = nawk
+AWK = awk
 INSTALL = install
 
 DBUS_MINOR=" `echo $(COPTS) | ../bld/pkg-wrapper $(PKG_CONFIG) --modversion dbus-1 | $(AWK) -F . -- '{ if ($$(NF-1)) print \"-DDBUS_MINOR=\"$$(NF-1) }'`" 
diff -urbB dnsmasq-2.46/src/config.h dnsmasq/src/config.h
--- dnsmasq-2.46/src/config.h	2008-11-11 00:47:46.000000000 +0300
+++ dnsmasq/src/config.h	2008-11-17 11:27:12.000000000 +0300
@@ -157,8 +157,8 @@
 */
 
 /* platform independent options- uncomment to enable */
-#define HAVE_TFTP
-/* #define HAVE_BROKEN_RTC */
+/* #define HAVE_TFTP */
+#define HAVE_BROKEN_RTC
 /* #define HAVE_DBUS */
 
 /* Allow TFTP to be disabled with COPTS=-DNO_TFTP */
@@ -199,10 +199,13 @@
 #if !defined(__ARCH_HAS_MMU__) && !defined(__UCLIBC_HAS_MMU__)
 #  define NO_FORK
 #endif
-#if defined(__UCLIBC_HAS_IPV6__)
+#if defined(__UCLIBC_HAS_IPV6__) && defined(CONFIG_IPV6)
 #  ifndef IPV6_V6ONLY
 #    define IPV6_V6ONLY 26
 #  endif
+#else
+#  undef HAVE_IPV6
+#  define NO_IPV6
 #endif
 
 /* This is for glibc 2.x */
diff -ruNp org/dnsmasq-2.47/src/network.c dnsmasq-2.47/src/network.c
--- org/dnsmasq-2.47/src/network.c	2009-02-02 01:00:58.000000000 +0900
+++ dnsmasq-2.47/src/network.c	2009-02-14 21:36:29.000000000 +0900
@@ -529,7 +529,7 @@ int local_bind(int fd, union mysockaddr 
     
 #if defined(SO_BINDTODEVICE)
   if (intname[0] != 0 &&
-      setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, intname, sizeof(intname)) == -1)
+      setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, intname, strlen(intname)) == -1)
     return 0;
 #endif
 
