diff -BurpN dnsmasq-2.51/src/config.h dnsmasq/src/config.h
--- dnsmasq-2.50/src/config.h	2009-08-18 15:28:01.000000000 +0000
+++ dnsmasq/src/config.h	2009-08-29 07:09:19.000000000 +0000
@@ -166,9 +166,9 @@ NOTES:
 
 /* platform independent options- uncomment to enable */
 #define HAVE_DHCP
-#define HAVE_TFTP
-#define HAVE_SCRIPT
-/* #define HAVE_BROKEN_RTC */
+/* #define HAVE_TFTP */
+/* #define HAVE_SCRIPT */
+#define HAVE_BROKEN_RTC
 /* #define HAVE_DBUS */
 
 /* Allow TFTP to be disabled with COPTS=-DNO_TFTP */
@@ -213,10 +213,13 @@ NOTES:
 #if !defined(__ARCH_HAS_MMU__) && !defined(__UCLIBC_HAS_MMU__)
 #  define NO_FORK
 #endif
-#if defined(__UCLIBC_HAS_IPV6__)
+#if defined(__UCLIBC_HAS_IPV6__) && defined(USE_IPV6)
 #  ifndef IPV6_V6ONLY
 #    define IPV6_V6ONLY 26
 #  endif
+#else
+#  undef HAVE_IPV6
+#  define NO_IPV6
 #endif
 
 /* This is for glibc 2.x */
--- dnsmasq-2.56test20/src/cache.c	2010-12-17 21:15:48.000000000 +0000
+++ dnsmasq/src/cache.c	2011-02-02 11:27:26.497845003 +0000
@@ -1109,7 +1109,7 @@ void dump_cache(time_t now)
 			 cache->flags & F_NXDOMAIN ? "X" : " ",
 			 cache->flags & F_HOSTS ? "H" : " ");
 #ifdef HAVE_BROKEN_RTC
-f	    p += sprintf(p, "%lu", cache->flags & F_IMMORTAL ? 0: (unsigned long)(cache->ttd - now));
+	    p += sprintf(p, "%lu", cache->flags & F_IMMORTAL ? 0: (unsigned long)(cache->ttd - now));
 #else
 	    p += sprintf(p, "%s", cache->flags & F_IMMORTAL ? "\n" : ctime(&(cache->ttd)));
 	    /* ctime includes trailing \n - eat it */
