diff -BurpN dnsmasq-2.51/Makefile dnsmasq/Makefile
--- dnsmasq-2.50/Makefile	2009-08-18 15:28:01.000000000 +0000
+++ dnsmasq/Makefile	2009-08-29 07:04:36.000000000 +0000
@@ -34,6 +34,11 @@ DNSMASQ_CFLAGS=`echo $(COPTS) | ../bld/p
 DNSMASQ_LIBS=  `echo $(COPTS) | ../bld/pkg-wrapper HAVE_DBUS $(PKG_CONFIG) --libs dbus-1` 
 SUNOS_LIBS= `if uname | grep SunOS 2>&1 >/dev/null; then echo -lsocket -lnsl -lposix4; fi`
 
+ifeq ($(CONFIG_IPV6),y)
+ COPTS += -DUSE_IPV6
+endif
+export COPTS
+
 all :   dnsmasq
 
 dnsmasq :
diff -BurpN dnsmasq-2.51/src/config.h dnsmasq/src/config.h
--- dnsmasq-2.50/src/config.h	2009-08-18 15:28:01.000000000 +0000
+++ dnsmasq/src/config.h	2009-08-29 07:09:19.000000000 +0000
@@ -166,9 +166,9 @@ NOTES:
 
 /* platform independent options- uncomment to enable */
 #define HAVE_DHCP
-#define HAVE_TFTP
-#define HAVE_SCRIPT
-/* #define HAVE_BROKEN_RTC */
+/* #define HAVE_TFTP */
+/* #define HAVE_SCRIPT */
+#define HAVE_BROKEN_RTC
 /* #define HAVE_DBUS */
 
 /* Allow TFTP to be disabled with COPTS=-DNO_TFTP */
@@ -213,10 +213,13 @@ NOTES:
 #if !defined(__ARCH_HAS_MMU__) && !defined(__UCLIBC_HAS_MMU__)
 #  define NO_FORK
 #endif
-#if defined(__UCLIBC_HAS_IPV6__)
+#if defined(__UCLIBC_HAS_IPV6__) && defined(USE_IPV6)
 #  ifndef IPV6_V6ONLY
 #    define IPV6_V6ONLY 26
 #  endif
+#else
+#  undef HAVE_IPV6
+#  define NO_IPV6
 #endif
 
 /* This is for glibc 2.x */
