diff -BurN router/others/Makefile gateway/others/Makefile
--- router/others/Makefile	2004-11-15 11:51:39.000000000 +0300
+++ gateway/others/Makefile	2008-08-25 12:46:06.000000000 +0400
@@ -3,34 +3,43 @@
 endif
 
 all :
+	rm -rf $(INSTALLDIR)
 
 install: all
-	install -d $(INSTALLDIR)/etc
+	install -d $(INSTALLDIR)/usr/etc
 	install -d $(INSTALLDIR)/sbin
 	install -d $(INSTALLDIR)/usr/sbin
+	install -d $(INSTALLDIR)/usr/bin
+	install -d $(INSTALLDIR)/usr/share/empty
 	install -d $(INSTALLDIR)/lib
 	install -d $(INSTALLDIR)/www
-	install -d $(INSTALLDIR)/lib/modules/2.4.20/parport
 
-	install services $(INSTALLDIR)/etc
+	install -m 644 services $(INSTALLDIR)/usr/etc
+	install -m 644 protocols $(INSTALLDIR)/usr/etc
+	install -m 644 rpc $(INSTALLDIR)/usr/etc
+	install -m 644 profile $(INSTALLDIR)/usr/etc
+	install -m 644 ipkg.conf $(INSTALLDIR)/usr/etc
 	install netcam_mfc_activeX.cab $(INSTALLDIR)/www
-	install lp.o $(INSTALLDIR)/lib/modules/2.4.20/parport
-	install parport.o $(INSTALLDIR)/lib/modules/2.4.20/parport
-	install parport_splink.o $(INSTALLDIR)/lib/modules/2.4.20/parport
 	#install waveserver $(INSTALLDIR)/usr/sbin/waveserver
 	
-	ln -sf /tmp/resolv.conf $(INSTALLDIR)/etc/resolv.conf
-	ln -sf /tmp/hosts $(INSTALLDIR)/etc/hosts
-	ln -sf /tmp/settings $(INSTALLDIR)/www/WL500g.Deluxe.CFG
-	ln -sf /tmp/settings $(INSTALLDIR)/www/WL500g.CFG
-	ln -sf /tmp/settings $(INSTALLDIR)/www/WL500b.CFG
-	ln -sf /tmp/settings $(INSTALLDIR)/www/WL500bv2.CFG
-	ln -sf /tmp/settings $(INSTALLDIR)/www/WL300g.CFG
+	ln -sf /tmp/settings $(INSTALLDIR)/www/WL500g.Premium.CFG
+	ln -sf /tmp/settings $(INSTALLDIR)/www/WL700g.CFG
+	ln -sf /tmp/settings $(INSTALLDIR)/www/WL500W.CFG
+	ln -sf /tmp/settings $(INSTALLDIR)/www/WL550gE.CFG
+	ln -sf /tmp/flash.tar.gz $(INSTALLDIR)/www/WL500g.Premium.tar.gz
+	ln -sf /tmp/flash.tar.gz $(INSTALLDIR)/www/WL500W.tar.gz
+	ln -sf /tmp/flash.tar.gz $(INSTALLDIR)/www/WL550gE.tar.gz
 	ln -sf /var/tmp/display.jpg $(INSTALLDIR)/www/display.jpg
 
 ifeq ($(SYSLOG_SUPPORT),1)
 	install -m 755 rotatelog.sh $(INSTALLDIR)/usr/sbin
 endif
 
+	install -m 755 linuxrc.sh $(INSTALLDIR)/linuxrc
+	install -m 755 flashfs.sh $(INSTALLDIR)/sbin/flashfs
+	install -m 755 wshaper.sh $(INSTALLDIR)/sbin/wshaper
+	install -m 755 ipkg.sh $(INSTALLDIR)/usr/bin/ipkg.sh
+	install -m 755 dropbearstart.sh $(INSTALLDIR)/usr/sbin/dropbearstart
+
 clean:
 
