diff -uBp udpxy/udpxy.c udpxy.new/udpxy.c
--- udpxy/udpxy.c	2010-03-21 16:29:57.000000000 +0300
+++ udpxy.new/udpxy.c	2010-05-13 16:04:48.000000000 +0400
@@ -268,7 +268,7 @@ send_http_response( int sockfd, int code
 
     assert( (sockfd > 0) && code && reason );
 
-    msglen = snprintf( msg, sizeof(msg) - 1, "HTTP/1.1 %d %s \n\n",
+    msglen = snprintf( msg, sizeof(msg) - 1, "HTTP/1.1 %d %s.\nContent-Type:application/octet-stream.\n\n",
               code, reason );
     if( msglen <= 0 ) return ERR_INTERNAL;
 
