From d1516c4d628dbded9e1abefaacc1535aaecef418 Mon Sep 17 00:00:00 2001
From: Denys Vlasenko <vda.linux@googlemail.com>
Date: Sat, 17 Jul 2010 20:43:42 +0000
Subject: dos2unix: fix the case w/o parameters

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/coreutils/dos2unix.c b/coreutils/dos2unix.c
index 1911f53..e06ecc4 100644
--- a/coreutils/dos2unix.c
+++ b/coreutils/dos2unix.c
@@ -92,7 +92,7 @@ int dos2unix_main(int argc UNUSED_PARAM, char **argv)
 	do {
 		/* might be convert(NULL) if there is no filename given */
 		convert(*argv, conv_type);
-	} while (*++argv);
+	} while (*argv && *++argv);
 
 	return 0;
 }
--
cgit v0.8.2.1
