From c3c1fb676b900ed39c0cfa984059c1d40197b601
From: Chris Metcalf
Date: Fri, 08 Jan 2010 12:18:06 +0000
Subject: ash: fix mishandling of bash-style redirects

Signed-off-by: Chris Metcalf <cmetcalf@tilera.com>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/shell/ash.c b/shell/ash.c
index 8d7c4ca..2c1779f 100644
--- a/shell/ash.c
+++ b/shell/ash.c
@@ -5111,7 +5111,7 @@ redirect(union node *redir, int flags)
 		do {
 			sv_pos++;
 #if ENABLE_ASH_BASH_COMPAT
-			if (redir->nfile.type == NTO2)
+			if (tmp->nfile.type == NTO2)
 				sv_pos++;
 #endif
 			tmp = tmp->nfile.next;
--
cgit v0.8.2.1
