From 153fcaa6c17124ebbf0d50df5da95f3564c7402b
From: Denys Vlasenko
Date: Sun, 21 Feb 2010 04:17:41 +0000
Subject: Replace __uClinux__ define with !BB_MMU

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/procps/free.c b/procps/free.c
index b138591..e8bea50 100644
--- a/procps/free.c
+++ b/procps/free.c
@@ -27,7 +27,7 @@ int free_main(int argc UNUSED_PARAM, char **argv)
 		/* TODO:  Make all this stuff not overflow when mem >= 4 Gib */
 		info.totalram/=info.mem_unit;
 		info.freeram/=info.mem_unit;
-#ifndef __uClinux__
+#if BB_MMU
 		info.totalswap/=info.mem_unit;
 		info.freeswap/=info.mem_unit;
 #endif
@@ -38,7 +38,7 @@ int free_main(int argc UNUSED_PARAM, char **argv)
 		/* TODO:  Make all this stuff not overflow when mem >= 4 Gib */
 		info.totalram*=info.mem_unit;
 		info.freeram*=info.mem_unit;
-#ifndef __uClinux__
+#if BB_MMU
 		info.totalswap*=info.mem_unit;
 		info.freeswap*=info.mem_unit;
 #endif
@@ -56,7 +56,7 @@ int free_main(int argc UNUSED_PARAM, char **argv)
 			info.totalram-info.freeram, info.freeram,
 			info.sharedram, info.bufferram);
 
-#ifndef __uClinux__
+#if BB_MMU
 	printf("%6s%13ld%13ld%13ld\n", "Swap:", info.totalswap,
 			info.totalswap-info.freeswap, info.freeswap);
 
diff --git a/shell/ash.c b/shell/ash.c
index 7f14bed..03904bb 100644
--- a/shell/ash.c
+++ b/shell/ash.c
@@ -73,7 +73,7 @@
 # define PIPE_BUF 4096           /* amount of buffering in a pipe */
 #endif
 
-#if defined(__uClinux__)
+#if !BB_MMU
 # error "Do not even bother, ash will not run on NOMMU machine"
 #endif
 
--
cgit v0.8.2.1
