From 6eaeb7737d95661ca31b162977ac443ffeb7b0b3
From: Denys Vlasenko
Date: Fri, 12 Mar 2010 21:16:25 +0000
Subject: tail: fix error detected by testsuite

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/coreutils/tail.c b/coreutils/tail.c
index 83768d4..8fc22f5 100644
--- a/coreutils/tail.c
+++ b/coreutils/tail.c
@@ -241,7 +241,8 @@ int tail_main(int argc, char **argv)
 						} while (nwrite);
 					}
 				}
-				xwrite(STDOUT_FILENO, buf + nread - nwrite, nwrite);
+				if (nwrite > 0)
+					xwrite(STDOUT_FILENO, buf + nread - nwrite, nwrite);
 			} else if (count) {
 				if (COUNT_BYTES) {
 					taillen += nread;
--
cgit v0.8.2.1
