Add workaround for IPV6_RECVPKTINFO in recv_from_to()

diff -BurpN busybox-1.17.1/libbb/udp_io.c busybox/libbb/udp_io.c
--- busybox-1.17.1/libbb/udp_io.c	2010-07-06 02:25:54.000000000 +0000
+++ busybox/libbb/udp_io.c	2010-07-25 19:18:28.000000000 +0000
@@ -162,8 +162,11 @@ recv_from_to(int fd, void *buf, size_t l
 		}
 # if ENABLE_FEATURE_IPV6 && defined(IPV6_PKTINFO)
 		if (cmsgptr->cmsg_level == IPPROTO_IPV6
-		 && cmsgptr->cmsg_type == IPV6_PKTINFO
-		) {
+		 && (cmsgptr->cmsg_type == IPV6_PKTINFO
+#if defined(IPV6_2292PKTINFO) && defined(IPV6_RECVPKTINFO)
+            		 || cmsgptr->cmsg_type == IPV6_2292PKTINFO
+#endif
+		)) {
 			const int IPI6_ADDR_OFF = offsetof(struct in6_pktinfo, ipi6_addr);
 			to->sa_family = AF_INET6;
 			/*#  define pktinfo(cmsgptr) ( (struct in6_pktinfo*)(CMSG_DATA(cmsgptr)) )*/
