diff -BurpN busybox-1.14.2/editors/vi.c busybox/editors/vi.c
--- busybox-1.14.2/editors/vi.c	2009-06-21 22:40:29.000000000 +0000
+++ busybox/editors/vi.c	2009-07-22 18:34:31.000000000 +0000
@@ -3226,7 +3226,7 @@ static void do_cmd(int c)
 		buf[1] = '\0';
 		q = get_input_line(buf);	// get input line- use "status line"
 		if (q[0] && !q[1]) {
-			if (last_search_pattern[0])
+			if (last_search_pattern)
 				last_search_pattern[0] = c;
 			goto dc3; // if no pat re-use old pat
 		}
@@ -3242,6 +3242,10 @@ static void do_cmd(int c)
 		if (cmdcnt-- > 1) {
 			do_cmd(c);
 		}				// repeat cnt
+		if (last_search_pattern == 0) {
+			msg = "No previous regular expression";
+			goto dc2;
+		}
 		dir = BACK;		// assume BACKWARD search
 		p = dot - 1;
 		if (last_search_pattern[0] == '?') {
