diff -urBp iproute2-ss020116/Makefile iproute2.new/Makefile
--- iproute2-ss020116/Makefile	2002-01-16 02:30:32.000000000 +0300
+++ iproute2.new/Makefile	2009-07-09 20:07:49.000000000 +0400
@@ -4,17 +4,16 @@ SBINDIR=/sbin
 CONFDIR=/etc/iproute2
 DOCDIR=/usr/doc/iproute2
 
-KERNEL_INCLUDE=/usr/src/linux/include
-LIBC_INCLUDE=/usr/include
+#KERNEL_INCLUDE=/usr/src/linux/include
 
 DEFINES= -DRESOLVE_HOSTNAMES
 
 #options if you have a bind>=4.9.4 libresolv (or, maybe, glibc)
-LDLIBS=-lresolv
+#LDLIBS=-lresolv
 ADDLIB=
 
 #options if you compile with libc5, and without a bind>=4.9.4 libresolv
-#LDLIBS=
+LDLIBS=
 #ADDLIB=inet_ntop.o inet_pton.o
 
 #options for decnet
@@ -23,19 +22,15 @@ ADDLIB+=dnet_ntop.o dnet_pton.o
 #options for ipx
 ADDLIB+=ipx_ntop.o ipx_pton.o
 
-ifeq ($(LIBC_INCLUDE)/socketbits.h,$(wildcard $(LIBC_INCLUDE)/socketbits.h))
-  ifeq ($(LIBC_INCLUDE)/net/if_packet.h,$(wildcard $(LIBC_INCLUDE)/net/if_packet.h))
-    GLIBCFIX=-I../include-glibc -include ../include-glibc/glibc-bugs.h
-  endif
-endif
-ifeq ($(LIBC_INCLUDE)/bits/socket.h,$(wildcard $(LIBC_INCLUDE)/bits/socket.h))
-  GLIBCFIX=-I../include-glibc -I/usr/include/db3 -include ../include-glibc/glibc-bugs.h
-endif
+GLIBCFIX=-I../include-glibc -include ../include-glibc/glibc-bugs.h
 
+check_gcc = $(shell if $(CC) $(1) -S -o /dev/null -xc /dev/null > /dev/null 2>&1; then echo "$(1)"; else echo "$(2)"; fi)
 
-CC = gcc
+#CC = gcc
 CCOPTS = -D_GNU_SOURCE -O2 -Wstrict-prototypes -Wall -g
-CFLAGS = $(CCOPTS) $(GLIBCFIX) -I$(KERNEL_INCLUDE) -I../include $(DEFINES)
+# disable pointer signedness warnings in gcc 4.x
+CCOPTS += $(call check_gcc,-Wno-pointer-sign,)
+CFLAGS = $(CCOPTS) $(GLIBCFIX) -I$(KERNEL_INCLUDE) -I../include $(DEFINES) $(EXTRACFLAGS)
 
 LDLIBS += -L../lib -lnetlink -lutil
 
diff -urBp iproute2-ss020116/misc/Makefile iproute2.new/misc/Makefile
--- iproute2-ss020116/misc/Makefile	2002-01-16 02:30:32.000000000 +0300
+++ iproute2.new/misc/Makefile	2009-07-09 20:01:56.000000000 +0400
@@ -6,7 +6,7 @@
 RTSTATOBJ=rtstat.o
 
 ALLOBJ=$(SSOBJ) $(NSTATOBJ) $(IFSTATOBJ) $(RTACCTOBJ) $(ARPDOBJ) $(RTSTATOBJ)
-TARGETS=ss nstat ifstat rtacct arpd rtstat
+TARGETS=ss nstat ifstat rtacct rtstat
 
 all: $(TARGETS)
 
