diff -BurpN linux-2.6.22.19.orig/arch/mips/brcm-boards/bcm947xx/setup.c linux-2.6/arch/mips/brcm-boards/bcm947xx/setup.c
--- linux-2.6.22.19.orig/arch/mips/brcm-boards/bcm947xx/setup.c	2009-02-23 14:14:55.000000000 +0000
+++ linux-2.6/arch/mips/brcm-boards/bcm947xx/setup.c	2010-11-11 18:08:58.000000000 +0300
@@ -101,23 +101,25 @@ bcm947xx_machine_halt(void)
 
 #ifdef CONFIG_SERIAL_CORE
 
-static struct uart_port rs = {
-	line: 0,
-	flags: ASYNC_BOOT_AUTOCONF,
-	iotype: SERIAL_IO_MEM,
-};
+static int num_ports = 0;
 
 static void __init
 serial_add(void *regs, uint irq, uint baud_base, uint reg_shift)
 {
+	struct uart_port rs;
+
+	memset(&rs, 0, sizeof(rs));
+	rs.line = num_ports++;
+	rs.flags = UPF_BOOT_AUTOCONF | UPF_SHARE_IRQ;
+	rs.iotype = UPIO_MEM;
+
+	rs.mapbase = (unsigned int) regs;
 	rs.membase = regs;
 	rs.irq = irq + 2;
 	rs.uartclk = baud_base;
 	rs.regshift = reg_shift;
 
 	early_serial_setup(&rs);
-
-	rs.line++;
 }
 
 static void __init
