CIFS 1.49 fix for CVE-2009-1439.

diff --git a/fs/cifs/connect.c b/fs/cifs/connect.c
index f4e9266..784912f 100644
--- a/fs/cifs/connect.c
+++ b/fs/cifs/connect.c
@@ -3275,14 +3275,11 @@ CIFSTCon(unsigned int xid, struct cifsSesInfo *ses,
 			    BCC(smb_buffer_response)) {
 				kfree(tcon->nativeFileSystem);
 				tcon->nativeFileSystem =
-				    kzalloc(length + 2, GFP_KERNEL);
+				    kzalloc((4 * length) + 2, GFP_KERNEL);
 				cifs_strfromUCS_le(tcon->nativeFileSystem,
 						   (__le16 *) bcc_ptr,
 						   length, nls_codepage);
-				bcc_ptr += 2 * length;
-				bcc_ptr[0] = 0;	/* null terminate the string */
-				bcc_ptr[1] = 0;
-				bcc_ptr += 2;
+				bcc_ptr += (2 * length) + 2;
 			}
 			/* else do not bother copying these informational fields */
 		} else {
