--- linux-2.6.22.19.orig/include/linux/netdevice.h	2010-09-26 06:00:00.000000000 +0600
+++ linux-2.6/include/linux/netdevice.h	2011-02-12 15:01:12.072913000 +0500
@@ -526,12 +526,6 @@ struct net_device
 	int			(*hard_header_parse)(struct sk_buff *skb,
 						     unsigned char *haddr);
 	int			(*neigh_setup)(struct net_device *dev, struct neigh_parms *);
-#ifdef CONFIG_NETPOLL
-	struct netpoll_info	*npinfo;
-#endif
-#ifdef CONFIG_NET_POLL_CONTROLLER
-	void                    (*poll_controller)(struct net_device *dev);
-#endif
 
 	/* bridge stuff */
 	struct net_bridge_port	*br_port;
@@ -543,6 +537,13 @@ struct net_device
 
 	/* rtnetlink link ops */
 	const struct rtnl_link_ops *rtnl_link_ops;
+
+#ifdef CONFIG_NETPOLL
+	struct netpoll_info	*npinfo;
+#endif
+#ifdef CONFIG_NET_POLL_CONTROLLER
+	void                    (*poll_controller)(struct net_device *dev);
+#endif
 };
 #define to_net_dev(d) container_of(d, struct net_device, dev)
 
