--- libusb10,orig/libusb/Makefile.am	2010-04-21 21:44:30.000000000 +0000
+++ libusb10/libusb/Makefile.am	2011-01-13 15:10:50.000000000 +0000
@@ -15,6 +15,7 @@ AM_CFLAGS_EXT = -no-cpp-precomp
 endif
 
 libusb_1_0_la_CFLAGS = $(VISIBILITY_CFLAGS) $(AM_CFLAGS) -pthread
+libusb_1_0_la_LDFLAGS = -avoid-version
 libusb_1_0_la_SOURCES = libusbi.h core.c descriptor.c io.c sync.c $(OS_SRC)
 
 hdrdir = $(includedir)/libusb-1.0
--- libusb10.orig/libusb/Makefile.in	2010-05-04 22:02:16.000000000 +0000
+++ libusb10/libusb/Makefile.in	2011-01-13 15:19:14.000000000 +0000
@@ -247,6 +247,7 @@ EXTRA_DIST = $(LINUX_USBFS_SRC) $(DARWIN
 @OS_LINUX_TRUE@OS_SRC = $(LINUX_USBFS_SRC)
 @OS_DARWIN_TRUE@AM_CFLAGS_EXT = -no-cpp-precomp
 libusb_1_0_la_CFLAGS = $(VISIBILITY_CFLAGS) $(AM_CFLAGS) -pthread
+libusb_1_0_la_LDFLAGS = -avoid-version
 libusb_1_0_la_SOURCES = libusbi.h core.c descriptor.c io.c sync.c $(OS_SRC)
 hdrdir = $(includedir)/libusb-1.0
 hdr_HEADERS = libusb.h
@@ -316,7 +317,7 @@ clean-libLTLIBRARIES:
 	  rm -f "$${dir}/so_locations"; \
 	done
 libusb-1.0.la: $(libusb_1_0_la_OBJECTS) $(libusb_1_0_la_DEPENDENCIES) 
-	$(AM_V_CCLD)$(libusb_1_0_la_LINK) -rpath $(libdir) $(libusb_1_0_la_OBJECTS) $(libusb_1_0_la_LIBADD) $(LIBS)
+	$(AM_V_CCLD)$(libusb_1_0_la_LINK) -rpath $(libdir) $(libusb_1_0_la_LDFLAGS) $(libusb_1_0_la_OBJECTS) $(libusb_1_0_la_LIBADD) $(LIBS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
