Index: linux-2.4.35.4/drivers/usb/serial/usbserial.c
===================================================================
--- linux-2.4.35.4.orig/drivers/usb/serial/usbserial.c
+++ linux-2.4.35.4/drivers/usb/serial/usbserial.c
@@ -333,7 +333,7 @@
 static __u16	product	= 0xffff;
 static int	maxSize = 0;
 
-static struct usb_device_id generic_device_ids[2]; /* Initially all zeroes. */
+static struct usb_device_id generic_device_ids[3]; /* Initially all zeroes. */
 
 /* All of the device info needed for the Generic Serial Converter */
 static struct usb_serial_device_type generic_device = {
@@ -1836,6 +1846,10 @@ static int __init usb_serial_init(void)
 	generic_device_ids[0].idVendor = vendor;
 	generic_device_ids[0].idProduct = product;
 	generic_device_ids[0].match_flags = USB_DEVICE_ID_MATCH_VENDOR | USB_DEVICE_ID_MATCH_PRODUCT;
+	/* Audiovox Aircard */
+	generic_device_ids[1].idVendor = 0x0f3d;
+	generic_device_ids[1].idProduct = 0x0112;
+	generic_device_ids[1].match_flags = USB_DEVICE_ID_MATCH_VENDOR | USB_DEVICE_ID_MATCH_PRODUCT;
 	/* register our generic driver with ourselves */
 	usb_serial_register (&generic_device);
 #endif
