--- linuz/drivers/usb/printer.c.orig	2008-01-26 12:58:05.000000000 +0300
+++ linux/drivers/usb/printer.c	2008-01-26 12:58:50.000000000 +0300
@@ -378,7 +378,7 @@
 
 	prn_info= &prn_info_tmp; // Added by JYWeng 20031212:
 
-	
+	usblp->device_id_string[0] = usblp->device_id_string[1] = '\0';
 	err = usblp_get_id(usblp, 0, usblp->device_id_string, DEVICE_ID_SIZE - 1);
 	
 	if (err < 0) {
@@ -833,6 +833,7 @@
 		switch (cmd) {
 			/*=================================================================================== PaN */
 			case LPGETID: /* get the DEVICE_ID string */
+				usblp->device_id_string[0] = usblp->device_id_string[1] = '\0';
 				err = usblp_get_id(usblp, 0, usblp->device_id_string, DEVICE_ID_SIZE - 1);
 				if (err < 0) {
 					dbg ("usblp%d: error = %d reading IEEE-1284 Device ID string",
@@ -1557,6 +1558,7 @@
 {
 	int err, length;
 
+	usblp->device_id_string[0] = usblp->device_id_string[1] = '\0';
 	err = usblp_get_id(usblp, 0, usblp->device_id_string, USBLP_DEVICE_ID_SIZE - 1);
 	if (err < 0) {
 		dbg("usblp%d: error = %d reading IEEE-1284 Device ID string",
