From c0ca2fbedfd83b111da89d7a049aec17b62737a6 
From: Fedor <fedork@ubuntu.(none)>
Date: Thu, 23 Sep 2010 19:52:13 -0400
Subject: [PATCH] bridge: allow multiple interfaces with same address (necessary for VLANs)

---
 net/bridge/br_fdb.c |   11 +++++------
 1 files changed, 5 insertions(+), 6 deletions(-)

diff --git a/net/bridge/br_fdb.c b/net/bridge/br_fdb.c
index 56a7c3c..715d818 100644
--- a/net/bridge/br_fdb.c
+++ b/net/bridge/br_fdb.c
@@ -298,12 +298,11 @@ void br_fdb_insert(struct net_bridge *br,
 		if (!memcmp(fdb->addr.addr, addr, ETH_ALEN)) {
 			/* attempt to update an entry for a local interface */
 			if (fdb->is_local) {
-				if (is_local) 
-					printk(KERN_INFO "%s: attempt to add"
-					       " interface with same source address.\n",
-					       source->dev->name);
-				else if (net_ratelimit()) 
-					printk(KERN_WARNING "%s: received packet with "
+				/* it is okay to have multiple ports with same 
+				 * address, just don't allow to be spoofed.
+				 */
+				if (!is_local && net_ratelimit())
+					printk(KERN_WARNING "%s: received packet with"
 					       " own address as source address\n",
 					       source->dev->name);
 				goto out;
-- 
1.6.5.GIT
