diff -urbB linux-2.4.37/net/ipv4/netfilter/ip_conntrack_core.c linux.cdma/net/ipv4/netfilter/ip_conntrack_core.c
--- linux-2.4.37/net/ipv4/netfilter/ip_conntrack_core.c	2008-12-02 18:02:55.000000000 +0300
+++ linux.cdma/net/ipv4/netfilter/ip_conntrack_core.c	2008-12-03 16:06:47.000000000 +0300
@@ -61,6 +61,8 @@
 static LIST_HEAD(helpers);
 unsigned int ip_conntrack_htable_size = 0;
 int ip_conntrack_max = 0;
+int ip_conntrack_clear = 0;
+static int kill_all(struct ip_conntrack *i, void *data);
 static atomic_t ip_conntrack_count = ATOMIC_INIT(0);
 struct list_head *ip_conntrack_hash;
 static kmem_cache_t *ip_conntrack_cachep;
@@ -838,6 +840,12 @@
 	int set_reply;
 	int ret;
 
+	if (ip_conntrack_clear != 0)
+	{
+		ip_ct_iterate_cleanup(kill_all, NULL);
+		ip_conntrack_clear = 0;
+	}
+
 	/* FIXME: Do this right please. --RR */
 	(*pskb)->nfcache |= NFC_UNKNOWN;
 
diff -urbB linux-2.4.37/net/ipv4/netfilter/ip_conntrack_standalone.c linux.cdma/net/ipv4/netfilter/ip_conntrack_standalone.c
--- linux-2.4.37/net/ipv4/netfilter/ip_conntrack_standalone.c	2008-12-02 18:02:55.000000000 +0300
+++ linux.cdma/net/ipv4/netfilter/ip_conntrack_standalone.c	2008-12-03 16:04:43.000000000 +0300
@@ -277,6 +277,7 @@
 
 /* From ip_conntrack_core.c */
 extern int ip_conntrack_max;
+extern int ip_conntrack_clear;
 extern unsigned int ip_conntrack_htable_size;
 
 /* From ip_conntrack_proto_tcp.c */
@@ -348,12 +349,16 @@
 };
 
 #define NET_IP_CONNTRACK_MAX 2089
+#define NET_IP_CONNTRACK_CLEAR 2090
 
 static ctl_table ip_ct_netfilter_table[] = {
 	{NET_IPV4_NETFILTER, "netfilter", NULL, 0, 0555, ip_ct_sysctl_table, 0, 0, 0, 0, 0},
 	{NET_IP_CONNTRACK_MAX, "ip_conntrack_max",
 	 &ip_conntrack_max, sizeof(int), 0644, NULL,
 	 &proc_dointvec},
+	{NET_IP_CONNTRACK_CLEAR, "ip_conntrack_clear",
+	 &ip_conntrack_clear, sizeof(ip_conntrack_clear), 0644,  NULL,
+	 &proc_dointvec},
 	{0}
 };
 
