diff -BurpN xl2tpd-1.2.4/file.h xl2tpd/file.h
--- xl2tpd-1.2.4/file.h	2009-03-08 23:25:30.000000000 +0000
+++ xl2tpd/file.h	2009-09-17 20:54:38.000000000 +0000
@@ -50,13 +50,13 @@ struct host
 #define SENSE_DENY 0
 
 #ifndef DEFAULT_AUTH_FILE
-#define DEFAULT_AUTH_FILE "/etc/xl2tpd/l2tp-secrets"
+#define DEFAULT_AUTH_FILE "/etc/l2tp-secrets"
 #endif
 #ifndef DEFAULT_CONFIG_FILE
-#define DEFAULT_CONFIG_FILE "/etc/xl2tpd/xl2tpd.conf"
+#define DEFAULT_CONFIG_FILE "/etc/xl2tpd.conf"
 #endif
-#define ALT_DEFAULT_AUTH_FILE "/etc/l2tpd/l2tp-secrets"
-#define ALT_DEFAULT_CONFIG_FILE "/etc/l2tp/l2tpd.conf"
+#define ALT_DEFAULT_AUTH_FILE ""
+#define ALT_DEFAULT_CONFIG_FILE ""
 #define DEFAULT_PID_FILE "/var/run/xl2tpd.pid"
 
 /* Definition of an LNS */
diff -BurpN xl2tpd-1.2.4/l2tp.h xl2tpd/l2tp.h
--- xl2tpd-1.2.4/l2tp.h	2009-03-08 23:25:30.000000000 +0000
+++ xl2tpd/l2tp.h	2009-09-17 20:56:30.000000000 +0000
@@ -37,7 +37,7 @@ typedef unsigned long long _u64;
 #include "common.h"
 #include "ipsecmast.h"
 
-#define CONTROL_PIPE "/var/run/xl2tpd/l2tp-control"
+#define CONTROL_PIPE "/var/run/l2tp-control"
 #define CONTROL_PIPE_MESSAGE_SIZE 1024
 
 #define BINARY "xl2tpd"
