--- linuz/arch/mips/bcm947xx/setup.c	2004-12-03 05:56:20.000000000 +0300
+++ linux/arch/mips/bcm947xx/setup.c	2006-09-03 23:01:21.467100904 +0400
@@ -53,6 +53,14 @@
 #include <hndmips.h>
 #include <trxhdr.h>
 #include "bcm947xx.h"
+#include <linux/mtd/mtd.h>
+#ifdef CONFIG_MTD_PARTITIONS
+#include <linux/mtd/partitions.h>
+#endif
+#include <linux/romfs_fs.h>
+#include <linux/cramfs_fs.h>
+#include <linux/minix_fs.h>
+#include <linux/ext2_fs.h>
 
 /* Global SB handle */
 sb_t *bcm947xx_sbh = NULL;
@@ -183,6 +183,7 @@
 	{ name: "linux", offset: 0, size: 0, },
 	{ name: "rootfs", offset: 0, size: 0, mask_flags: MTD_WRITEABLE, },
 	{ name: "nvram", offset: 0, size: 0, },
+	{ name: "flashfs", offset: 0, size: 0, },
 	{ name: NULL, },
 };
 
@@ -192,15 +193,17 @@
 	struct minix_super_block *minixsb;
 	struct ext2_super_block *ext2sb;
 	struct romfs_super_block *romfsb;
+	struct squashfs_super_block *squashfsb;
 	struct cramfs_super *cramfsb;
 	struct trx_header *trx;
 	unsigned char buf[512];
-	int off;
+	int off, trx_len = 0;
 	size_t len;
 
 	minixsb = (struct minix_super_block *) buf;
 	ext2sb = (struct ext2_super_block *) buf;
 	romfsb = (struct romfs_super_block *) buf;
+	squashfsb = (struct squashfs_super_block *) buf;
 	cramfsb = (struct cramfs_super *) buf;
 	trx = (struct trx_header *) buf;
 
@@ -218,7 +221,10 @@
 		/* Try looking at TRX header for rootfs offset */
 		if (le32_to_cpu(trx->magic) == TRX_MAGIC) {
 			bcm947xx_parts[1].offset = off;
-			if (le32_to_cpu(trx->offsets[1]) > off)
+			trx_len = trx->len;
+			if (le32_to_cpu(trx->offsets[2]) > off)
+				off = le32_to_cpu(trx->offsets[2]);
+			else if (le32_to_cpu(trx->offsets[1]) > off)
 				off = le32_to_cpu(trx->offsets[1]);
 			continue;
 		}
@@ -232,6 +238,14 @@
 			goto done;
 		}
 
+		/* squashfs is at block zero too */
+		if (squashfsb->s_magic == SQUASHFS_MAGIC) {
+			printk(KERN_NOTICE
+			       "%s: squashfs filesystem found at block %d\n",
+			       mtd->name, off / BLOCK_SIZE);
+			goto done;
+		}
+
 		/* so is cramfs */
 		if (cramfsb->magic == CRAMFS_MAGIC) {
 			printk(KERN_NOTICE
@@ -286,6 +300,23 @@
 	/* Size pmon */
 	bcm947xx_parts[0].size = bcm947xx_parts[1].offset - bcm947xx_parts[0].offset;
 
+	/* Find and size flashfs -- nvram reserved + pmon */
+	bcm947xx_parts[4].size = (128*1024 - bcm947xx_parts[3].size) + 
+		(256*1024 - bcm947xx_parts[0].size);
+	/* ... add unused space above 4MB */
+	if (size > 0x400000) {
+		if (trx_len <= 0x3a0000) // Small firmware - fixed amount
+			bcm947xx_parts[4].size += size - 0x400000;
+		else {
+			bcm947xx_parts[4].size += size - (trx_len + 128*1024 + 256*1024);
+			bcm947xx_parts[4].size &= (~0x3FFFFUL); // Round down 256K
+		}
+	}
+	bcm947xx_parts[4].offset = bcm947xx_parts[3].offset - bcm947xx_parts[4].size;
+	if (bcm947xx_parts[4].size == 0) {
+		bcm947xx_parts[4].name = NULL;
+	}
+	
 	return bcm947xx_parts;
 }
 
