Fix build IDE drivers as modules

--- a/drivers/ide/Makefile
+++ b/drivers/ide/Makefile
@@ -22,7 +22,7 @@
 
 # First come modules that register themselves with the core
 
-ifeq ($(CONFIG_BLK_DEV_IDE),y)
+ifneq ($(CONFIG_BLK_DEV_IDE),)
   obj-y		+= pci/idedriver-pci.o
 endif
 
@@ -51,7 +51,7 @@
 
 obj-$(CONFIG_BLK_DEV_IDE)		+= ide-core.o
 
-ifeq ($(CONFIG_BLK_DEV_IDE),y)
+ifneq ($(CONFIG_BLK_DEV_IDE),)
   obj-y		+= legacy/idedriver-legacy.o
   obj-y		+= ppc/idedriver-ppc.o
   obj-y		+= arm/idedriver-arm.o
@@ -71,7 +71,7 @@
 
 obj-$(CONFIG_BLK_DEV_IDE) += ide-detect.o
 
-ifeq ($(CONFIG_BLK_DEV_IDE),y)
+ifneq ($(CONFIG_BLK_DEV_ATARAID),)
 # RAID must be last of all
   obj-y		+= raid/idedriver-raid.o
 endif
--- a/drivers/ide/pci/Makefile
+++ b/drivers/ide/pci/Makefile
@@ -1,43 +1,42 @@
 
 O_TARGET := idedriver-pci.o
 
-obj-y		:=
-obj-m		:=
-
-obj-$(CONFIG_BLK_DEV_ADMA100)		+= adma100.o
-obj-$(CONFIG_BLK_DEV_AEC62XX)		+= aec62xx.o
-obj-$(CONFIG_BLK_DEV_ALI15X3)		+= alim15x3.o
-obj-$(CONFIG_BLK_DEV_AMD74XX)		+= amd74xx.o
-obj-$(CONFIG_BLK_DEV_ATIIXP)		+= atiixp.o
-obj-$(CONFIG_BLK_DEV_CMD640)		+= cmd640.o
-obj-$(CONFIG_BLK_DEV_CMD64X)		+= cmd64x.o
-obj-$(CONFIG_BLK_DEV_CS5530)		+= cs5530.o
-obj-$(CONFIG_BLK_DEV_CY82C693)		+= cy82c693.o
-obj-$(CONFIG_BLK_DEV_HPT34X)		+= hpt34x.o
-obj-$(CONFIG_BLK_DEV_HPT366)		+= hpt366.o
-#obj-$(CONFIG_BLK_DEV_HPT37X)		+= hpt37x.o
-obj-$(CONFIG_BLK_DEV_IDE_ICSIDE)	+= icside.o
-obj-$(CONFIG_BLK_DEV_IT8172)		+= it8172.o
-obj-$(CONFIG_BLK_DEV_NS87415)		+= ns87415.o
-obj-$(CONFIG_BLK_DEV_OPTI621)		+= opti621.o
-obj-$(CONFIG_BLK_DEV_PDC202XX_OLD)	+= pdc202xx_old.o
-obj-$(CONFIG_BLK_DEV_PDC202XX_NEW)	+= pdc202xx_new.o
-obj-$(CONFIG_BLK_DEV_PIIX)		+= piix.o
-obj-$(CONFIG_BLK_DEV_RZ1000)		+= rz1000.o
-obj-$(CONFIG_BLK_DEV_SVWKS)		+= serverworks.o
-obj-$(CONFIG_BLK_DEV_SC1200)		+= sc1200.o
-obj-$(CONFIG_BLK_DEV_SGIIOC4)		+= sgiioc4.o
-obj-$(CONFIG_BLK_DEV_SIIMAGE)		+= siimage.o
-obj-$(CONFIG_BLK_DEV_SIS5513)		+= sis5513.o
-obj-$(CONFIG_BLK_DEV_SL82C105)		+= sl82c105.o
-obj-$(CONFIG_BLK_DEV_SLC90E66)		+= slc90e66.o
-obj-$(CONFIG_BLK_DEV_TRM290)		+= trm290.o
-obj-$(CONFIG_BLK_DEV_VIA82CXXX)		+= via82cxxx.o
-obj-$(CONFIG_BLK_DEV_TRIFLEX)		+= triflex.o
-obj-$(CONFIG_BLK_DEV_DELKIN)		+= delkin_cb.o
+ide-obj-$(CONFIG_BLK_DEV_ADMA100)		+= adma100.o
+ide-obj-$(CONFIG_BLK_DEV_AEC62XX)		+= aec62xx.o
+ide-obj-$(CONFIG_BLK_DEV_ALI15X3)		+= alim15x3.o
+ide-obj-$(CONFIG_BLK_DEV_AMD74XX)		+= amd74xx.o
+ide-obj-$(CONFIG_BLK_DEV_ATIIXP)		+= atiixp.o
+ide-obj-$(CONFIG_BLK_DEV_CMD640)		+= cmd640.o
+ide-obj-$(CONFIG_BLK_DEV_CMD64X)		+= cmd64x.o
+ide-obj-$(CONFIG_BLK_DEV_CS5530)		+= cs5530.o
+ide-obj-$(CONFIG_BLK_DEV_CY82C693)		+= cy82c693.o
+ide-obj-$(CONFIG_BLK_DEV_HPT34X)		+= hpt34x.o
+ide-obj-$(CONFIG_BLK_DEV_HPT366)		+= hpt366.o
+#ide-obj-$(CONFIG_BLK_DEV_HPT37X)		+= hpt37x.o
+ide-obj-$(CONFIG_BLK_DEV_IDE_ICSIDE)		+= icside.o
+ide-obj-$(CONFIG_BLK_DEV_IT8172)		+= it8172.o
+ide-obj-$(CONFIG_BLK_DEV_NS87415)		+= ns87415.o
+ide-obj-$(CONFIG_BLK_DEV_OPTI621)		+= opti621.o
+ide-obj-$(CONFIG_BLK_DEV_PDC202XX_OLD)		+= pdc202xx_old.o
+ide-obj-$(CONFIG_BLK_DEV_PDC202XX_NEW)		+= pdc202xx_new.o
+ide-obj-$(CONFIG_BLK_DEV_PIIX)			+= piix.o
+ide-obj-$(CONFIG_BLK_DEV_RZ1000)		+= rz1000.o
+ide-obj-$(CONFIG_BLK_DEV_SVWKS)			+= serverworks.o
+ide-obj-$(CONFIG_BLK_DEV_SC1200)		+= sc1200.o
+ide-obj-$(CONFIG_BLK_DEV_SGIIOC4)		+= sgiioc4.o
+ide-obj-$(CONFIG_BLK_DEV_SIIMAGE)		+= siimage.o
+ide-obj-$(CONFIG_BLK_DEV_SIS5513)		+= sis5513.o
+ide-obj-$(CONFIG_BLK_DEV_SL82C105)		+= sl82c105.o
+ide-obj-$(CONFIG_BLK_DEV_SLC90E66)		+= slc90e66.o
+ide-obj-$(CONFIG_BLK_DEV_TRM290)		+= trm290.o
+ide-obj-$(CONFIG_BLK_DEV_VIA82CXXX)		+= via82cxxx.o
+ide-obj-$(CONFIG_BLK_DEV_TRIFLEX)		+= triflex.o
+ide-obj-$(CONFIG_BLK_DEV_DELKIN)		+= delkin_cb.o
 
 # Must appear at the end of the block
-obj-$(CONFIG_BLK_DEV_GENERIC)		+= generic.o
+ide-obj-$(CONFIG_BLK_DEV_GENERIC)		+= generic.o
+
+obj-$(CONFIG_BLK_DEV_IDE) += $(ide-obj-y)
 
 EXTRA_CFLAGS	:= -I../
 
-- 
