Fix build warnings

diff -urB linux-2.4.37.orig/fs/xfs/linux-2.4/xfs_linux.h linux/fs/xfs/linux-2.4/xfs_linux.h
--- linux-2.4.37.orig/fs/xfs/linux-2.4/xfs_linux.h	2009-04-29 19:48:17.000000000 +0400
+++ linux/fs/xfs/linux-2.4/xfs_linux.h	2009-04-29 19:47:13.000000000 +0400
@@ -252,7 +252,9 @@
 #define MIN(a,b)	(min(a,b))
 #define MAX(a,b)	(max(a,b))
 #define howmany(x, y)	(((x)+((y)-1))/(y))
-#define roundup(x, y)	((((x)+((y)-1))/(y))*(y))
+#ifndef roundup
+ #define roundup(x, y)	((((x)+((y)-1))/(y))*(y))
+#endif
 
 /*
  * Juggle IRIX device numbers - still used in ondisk structures
