--- linux/drivers/usb/printer.c~	2004-12-17 22:38:57.942206280 +0300
+++ linux/drivers/usb/printer.c	2004-12-17 22:40:52.885732208 +0300
@@ -568,7 +568,6 @@
 	int minor = MINOR(inode->i_rdev) - USBLP_MINOR_BASE;
 	struct usblp *usblp;
 	int retval;
-	unsigned int arg = NULL, ioctl_retval; //Added by PaN
 
 	if (minor < 0 || minor >= USBLP_MINORS)
 		return -ENODEV;
@@ -616,11 +615,6 @@
 		}
 	}
 
-	/* Added by PaN */
-	if ((ioctl_retval=usblp_ioctl(inode, file, LPGETID, arg)) <0)
-	{
-		// Update device id failed
-	}	
 out:
 	unlock_kernel();
 	return retval;
@@ -634,12 +627,7 @@
 {
 	devfs_unregister (usblp->devfs);
 	usblp_table [usblp->minor] = NULL;
-	//info("usblp%d: removed", usblp->minor);
-
-	/* Added by PaN */
-	remove_proc_entry("usblpid", usblp_dir);
-	remove_proc_entry(MODULE_NAME, NULL);
-	/* End PaN */
+	info("usblp%d: removed", usblp->minor);
 
 	kfree (usblp->writebuf);
 	kfree (usblp->readbuf);
@@ -1145,9 +1133,8 @@
 			/* if the fault was due to disconnect, let khubd's
 			 * call to usblp_disconnect() grab usblp->sem ...
 			 */
-			return writecount; 	//Added by PaN
-			//schedule (); 		//Removed by PaN
-			//continue;	
+			schedule ();
+			continue;
 		}
 
 		/* We must increment writecount here, and not at the
