--- linuz/drivers/usb/printer.c	2008-01-25 19:31:47.000000000 +0300
+++ linux/drivers/usb/printer.c	2008-01-25 19:36:57.000000000 +0300
@@ -84,13 +84,6 @@
 	char model[MAX_MODEL];
 	char description[MAX_DESCRIPT];
 };
-static char *usblp_status_type[MAX_STATUS_TYPE]={ "Lexmark", "Canon", "Hp", "Epson", "EPSON", NULL};
-static int usblp_status_maping[MAX_STATUS_TYPE][4]={ {0,0,0,0},
-				       		     {0, LP_POUTPA, LP_PERRORP, LP_PBUSY},
-				       		     {0,0,0,0},
-				       		     {0,0,0,0},
-				       		     {0,0,0,0},
-				       		     {0,0,0,0}};
 			       	       	       
 static struct parport_splink_device_info usblpid_info;
 struct parport_splink_device_info prn_info_tmp, *prn_info; // Added by JYWeng 20031212:
@@ -991,48 +991,16 @@
 				break;
 
 			case LPGETSTATUS:
-				/* OLD USB Code Removed by PaN for Printer Server 
-				if (usblp_read_status(usblp, &status)) {
+				if (usblp_read_status(usblp, usblp->statusbuf)) {
 					err("usblp%d: failed reading printer status", usblp->minor);
 					retval = -EIO;
 					goto done;
 				}
-				if (copy_to_user ((int *)arg, &status, 2))
-					retval = -EFAULT;
-				*/
-                                status = usblp_check_status(usblp, 0);
-#if 0
-				info("start=%s", usblpid_info.mfr);
-				for (i=0; i< MAX_STATUS_TYPE; i++) {
-				info("compare=%s", usblp_status_type[i]);
-					if ( !( strcmp(usblpid_info.mfr, usblp_status_type[i]) ) )
-						break;
-				}
-				info("%d=%s", i, usblp_status_type[i]);
-				status=usblp_status_maping[i][status];
-				info("STATUS=%x", status);
-#endif
-				status=0;
+				status = *usblp->statusbuf;
 				if (copy_to_user ((int *)arg, &status, 2))
 					retval = -EFAULT;
 				break;
 
-/*=================================================================== PaN for Printer Server */
-
-/* Marked by JY 20031118*/
-#if 0
-			case LPGETSTATUS:
-				if (usblp_read_status(usblp, &lpstatus)) {
-					err("usblp%d: failed reading printer status", usblp->minor);
-					retval = -EIO;
-					goto done;
-				}
-				status = lpstatus;
-				if (copy_to_user ((int *)arg, &status, sizeof(int)))
-					retval = -EFAULT;
-				break;
-#endif
-/* Marked by JY 20031118*/			
 			default:
 				retval = -ENOTTY;
 		}
