diff -BurpN -x'*.o*' linux-2.4.37.orig/drivers/scsi/hosts.h linux/drivers/scsi/hosts.h
--- linux-2.4.37.orig/drivers/scsi/hosts.h	2008-12-02 13:01:34.000000000 +0500
+++ linux/drivers/scsi/hosts.h	2009-02-08 03:12:16.000000000 +0500
@@ -300,6 +300,19 @@ typedef struct	SHT
      * Name of proc directory
      */
     char *proc_name;
+    
+    /*
+     * Once the device has responded to an INQUIRY and we know the
+     * device is online, we call into the low level driver with the
+     * struct scsi_device *.  If the low level device driver implements
+     * this function, it *must* perform the task of setting the queue
+     * depth on the device.  All other tasks are optional and depend
+     * on what the driver supports and various implementation details.
+     *
+     * Status: OPTIONAL
+     */
+    int (* slave_configure)(Scsi_Device *dev);
+
 
 } Scsi_Host_Template;
 
diff -BurpN -x'*.o*' linux-2.4.37.orig/drivers/scsi/scsi_error.c linux/drivers/scsi/scsi_error.c
--- linux-2.4.37.orig/drivers/scsi/scsi_error.c	2008-12-02 13:01:34.000000000 +0500
+++ linux/drivers/scsi/scsi_error.c	2009-02-08 05:02:15.000000000 +0500
@@ -57,6 +57,7 @@
 #define SENSE_TIMEOUT (10*HZ)
 #define RESET_TIMEOUT (2*HZ)
 #define ABORT_TIMEOUT (15*HZ)
+#define START_STOP_TIMEOUT (60*HZ)
 #endif
 
 #define STATIC
@@ -75,6 +76,7 @@ STATIC int scsi_check_sense(Scsi_Cmnd * 
 STATIC int scsi_request_sense(Scsi_Cmnd *);
 STATIC void scsi_send_eh_cmnd(Scsi_Cmnd * SCpnt, int timeout);
 STATIC int scsi_try_to_abort_command(Scsi_Cmnd *, int);
+STATIC int scsi_try_start_unit(Scsi_Cmnd * SCpnt);
 STATIC int scsi_test_unit_ready(Scsi_Cmnd *);
 STATIC int scsi_try_bus_device_reset(Scsi_Cmnd *, int timeout);
 STATIC int scsi_try_bus_reset(Scsi_Cmnd *);
@@ -535,6 +537,63 @@ STATIC int scsi_test_unit_ready(Scsi_Cmn
 }
 
 /*
+ * Function:  scsi_send_stu()
+ *
+ * Purpose:     Send start unit command.
+ *
+ */
+STATIC int scsi_send_stu(Scsi_Cmnd * SCpnt)
+{
+	static unsigned char stu_command[6] =
+	{START_STOP, 0, 0, 0, 1, 0};
+	int saved_resid;
+
+	memcpy((void *) SCpnt->cmnd, (void *) stu_command,
+	       sizeof(stu_command));
+
+	if (SCpnt->device->scsi_level <= SCSI_2)
+		SCpnt->cmnd[1] = SCpnt->lun << 5;
+
+	/*
+	 * Zero the sense buffer.  The SCSI spec mandates that any
+	 * untransferred sense data should be interpreted as being zero.
+	 */
+	memset((void *) SCpnt->sense_buffer, 0, sizeof(SCpnt->sense_buffer));
+
+	saved_resid = SCpnt->resid;
+	SCpnt->request_buffer = NULL;
+	SCpnt->request_bufflen = 0;
+	SCpnt->use_sg = 0;
+	SCpnt->cmd_len = COMMAND_SIZE(SCpnt->cmnd[0]);
+	SCpnt->underflow = 0;
+	SCpnt->sc_data_direction = SCSI_DATA_NONE;
+
+	scsi_send_eh_cmnd(SCpnt, START_STOP_TIMEOUT);
+
+	/*
+	 * When we eventually call scsi_finish, we really wish to complete
+	 * the original request, so let's restore the original data. (DB)
+	 */
+	memcpy((void *) SCpnt->cmnd, (void *) SCpnt->data_cmnd,
+	       sizeof(SCpnt->data_cmnd));
+	SCpnt->resid = saved_resid;
+	SCpnt->request_buffer = SCpnt->buffer;
+	SCpnt->request_bufflen = SCpnt->bufflen;
+	SCpnt->use_sg = SCpnt->old_use_sg;
+	SCpnt->cmd_len = SCpnt->old_cmd_len;
+	SCpnt->sc_data_direction = SCpnt->sc_old_data_direction;
+	SCpnt->underflow = SCpnt->old_underflow;
+
+	/*
+	 * Hey, we are done.  Let's look to see what happened.
+	 */
+	SCSI_LOG_ERROR_RECOVERY(3,
+		printk("scsi_send_stu: SCpnt %p eh_state %x\n",
+		SCpnt, SCpnt->eh_state));
+	return SCpnt->eh_state;
+}
+
+/*
  * This would normally need to get the IO request lock,
  * but as it doesn't actually touch anything that needs
  * to be locked we can avoid the lock here..
@@ -780,6 +839,32 @@ STATIC int scsi_try_to_abort_command(Scs
 }
 
 /*
+ * Function:  scsi_try_start_unit
+ *
+ * Purpose:     Try to start device.
+ *
+ * Returns:     FAILED          Operation failed or not supported.
+ *              SUCCESS         Succeeded.
+ */
+
+
+STATIC int scsi_try_start_unit(Scsi_Cmnd * SCpnt)
+{
+	unsigned long flags;
+	int rtn;
+
+	if (SCpnt->device->allow_restart == 0)
+		return FAILED;
+
+	rtn = scsi_send_stu(SCpnt);
+
+	if (rtn == SUCCESS)
+		SCpnt->eh_state = SUCCESS;
+
+	return SCpnt->eh_state;
+}
+
+/*
  * Function:  scsi_try_bus_device_reset
  *
  * Purpose:     Ask host adapter to perform a bus device reset for a given
@@ -1195,6 +1280,15 @@ STATIC int scsi_check_sense(Scsi_Cmnd * 
 			(SCpnt->sense_buffer[13] == 0x01)) {
 			return NEEDS_RETRY;
 		}
+		/*
+		 * if the device is not started, we need to wake
+		 * the error handler to start the motor
+		 */
+		if (SCpnt->device->allow_restart &&
+			(SCpnt->sense_buffer[12] == 0x04) &&
+			(SCpnt->sense_buffer[13] == 0x02)) {
+			return FAILED;
+		}
 		return SUCCESS;
 
 		/* these three are not supported */
@@ -1204,12 +1298,20 @@ STATIC int scsi_check_sense(Scsi_Cmnd * 
 		return SUCCESS;
 
 	case MEDIUM_ERROR:
+		if (SCpnt->sense_buffer[12] == 0x11 || /* UNRECOVERED READ END */
+		    SCpnt->sense_buffer[12] == 0x13 || /* AMNF DATA FILED */
+		    SCpnt->sense_buffer[12] == 0x14) { /* RECORD NOT FOUND */
+			return SUCCESS;
+		}
 		return NEEDS_RETRY;
+		
+	case HARDWARE_ERROR:
+		if (SCpnt->device->retry_hwerror)
+			return ADD_TO_MLQUEUE;
 
 	case ILLEGAL_REQUEST:
 	case BLANK_CHECK:
 	case DATA_PROTECT:
-	case HARDWARE_ERROR:
 	default:
 		return SUCCESS;
 	}
@@ -1494,6 +1596,45 @@ STATIC int scsi_unjam_host(struct Scsi_H
 		ourrtn = TRUE;
 		goto leave;
 	}
+
+	/*
+	 *  If commands are failing due to not ready, initializing command required,
+	 *  try revalidating the device, which will end up sending a start unit.
+	 */
+	SCSI_LOG_ERROR_RECOVERY(3, printk("scsi_unjam_host: Checking to see if we want to try start unit\n"));
+
+	for (SDpnt = host->host_queue; SDpnt; SDpnt = SDpnt->next) {
+		for (SCloop = SDpnt->device_queue; SCloop; SCloop = SCloop->next) {
+			if (SCloop->state == SCSI_STATE_FAILED) {
+				break;
+			}
+		}
+
+		if (SCloop == NULL) {
+			continue;
+		}
+
+		rtn = scsi_try_start_unit(SCloop);
+
+		if (rtn == SUCCESS) {
+			rtn = scsi_test_unit_ready(SCloop);
+
+			if (rtn == SUCCESS && scsi_unit_is_ready(SCloop)) {
+				rtn = scsi_eh_retry_command(SCloop);
+
+				if (rtn == SUCCESS) {
+					SCloop->host->host_failed--;
+					scsi_eh_finish_command(&SCdone, SCloop);
+				}
+			}
+		}
+	}
+
+	if (host->host_failed == 0) {
+		ourrtn = TRUE;
+		goto leave;
+	}
+
 	/*
 	 * Either the abort wasn't appropriate, or it didn't succeed.
 	 * Now try a bus device reset.  Still, look to see whether we have
@@ -1714,7 +1855,12 @@ STATIC int scsi_unjam_host(struct Scsi_H
 						    && SCloop->state != SCSI_STATE_TIMEOUT) {
 							continue;
 						}
-						rtn = scsi_test_unit_ready(SCloop);
+
+						if (scsi_try_start_unit(SCloop) == SUCCESS &&
+						    scsi_test_unit_ready(SCloop) == SUCCESS)
+							rtn = SUCCESS;
+						else
+							rtn = scsi_test_unit_ready(SCloop);
 
 						if (rtn == SUCCESS && scsi_unit_is_ready(SCloop)) {
 							rtn = scsi_eh_retry_command(SCloop);
diff -BurpN -x'*.o*' linux-2.4.37.orig/drivers/scsi/scsi.h linux/drivers/scsi/scsi.h
--- linux-2.4.37.orig/drivers/scsi/scsi.h	2008-12-02 13:01:34.000000000 +0500
+++ linux/drivers/scsi/scsi.h	2009-02-08 03:12:16.000000000 +0500
@@ -609,6 +609,8 @@ struct scsi_device {
 	unsigned starved:1;	/* unable to process commands because
 				   host busy */
 	unsigned no_start_on_add:1;	/* do not issue start on add */
+	unsigned allow_restart:1; /* issue START_UNIT in error handler */
+	unsigned retry_hwerror:1; /* retry HARDWARE_ERROR */
 
 	// Flag to allow revalidate to succeed in sd_open
 	int allow_revalidate;
diff -BurpN -x'*.o*' linux-2.4.37.orig/drivers/scsi/scsi_scan.c linux/drivers/scsi/scsi_scan.c
--- linux-2.4.37.orig/drivers/scsi/scsi_scan.c	2008-12-02 13:01:34.000000000 +0500
+++ linux/drivers/scsi/scsi_scan.c	2009-02-02 23:50:41.000000000 +0500
@@ -38,7 +38,7 @@
 #define BLIST_ISROM     	0x200	/* Treat as (removable) CD-ROM */
 #define BLIST_LARGELUN		0x400	/* LUNs larger than 7 despite reporting as SCSI 2 */
 #define BLIST_NOSTARTONADD	0x1000	/* do not do automatic start on add */
-
+#define BLIST_RETRY_HWERROR	0x400000/* retry HARDWARE ERROR */
 
 static void print_inquiry(unsigned char *data);
 static int scan_scsis_single(unsigned int channel, unsigned int dev,
@@ -815,6 +815,14 @@ static int scan_scsis_single(unsigned in
 	if (bflags & BLIST_SINGLELUN)
 		SDpnt->single_lun = 1;
 
+	/* configure device if needed */
+	if (SDpnt->host->hostt->slave_configure) {
+		if (!SDpnt->host->hostt->slave_configure(SDpnt)) {
+			printk("scsi: scan_scsis_single: Cannot configure device\n");
+			return 0;
+		}
+	}
+
 	/*
 	 * These devices need this "key" to unlock the devices so we can use it
 	 */
diff -BurpN -x'*.o*' linux-2.4.37.orig/drivers/usb/storage/scsiglue.c linux/drivers/usb/storage/scsiglue.c
--- linux-2.4.37.orig/drivers/usb/storage/scsiglue.c	2007-12-17 00:34:58.000000000 +0500
+++ linux/drivers/usb/storage/scsiglue.c	2009-02-03 00:40:03.000000000 +0500
@@ -335,6 +335,28 @@ static int host_reset( Scsi_Cmnd *srb )
 	return FAILED;
 }
 
+static int slave_configure( Scsi_Device *dev )
+{
+	US_DEBUGP("slave_configure() called\n" );
+
+	if (dev->type == TYPE_DISK) {
+
+		/* USB-IDE bridges tend to report SK = 0x04 (Non-recoverable
+		 * Hardware Error) when any low-level error occurs,
+	         * recoverable or not.  Setting this flag tells the SCSI
+		 * midlayer to retry such commands, which frequently will
+		 * succeed and fix the error.  The worst this can lead to
+		 * is an occasional series of retries that will all fail. */
+		dev->retry_hwerror = 1;
+
+		/* USB disks should allow restart. Some drives spin down
+		 * automatically, requiring a START-STOP UNIT command. */
+		dev->allow_restart = 1;
+	}
+
+	return 1;
+}
+
 /***********************************************************************
  * /proc/scsi/ functions
  ***********************************************************************/
@@ -429,6 +451,8 @@ Scsi_Host_Template usb_stor_host_templat
 	eh_bus_reset_handler:	bus_reset,
 	eh_host_reset_handler:	host_reset,
 
+	slave_configure:	slave_configure,
+
 	can_queue:		1,
 	this_id:		-1,
 
