--- linuz/net/ipv4/netfilter/ip_conntrack_proto_tcp.c.orig	2007-10-24 21:57:45.051904512 +0400
+++ linux/net/ipv4/netfilter/ip_conntrack_proto_tcp.c	2007-10-24 21:58:03.849046912 +0400
@@ -51,7 +51,7 @@
 
 unsigned long ip_ct_tcp_timeout_syn_sent =      2 MINS;
 unsigned long ip_ct_tcp_timeout_syn_recv =     60 SECS;
-unsigned long ip_ct_tcp_timeout_established =   5 DAYS;
+unsigned long ip_ct_tcp_timeout_established =   6 HOURS;
 unsigned long ip_ct_tcp_timeout_fin_wait =      2 MINS;
 unsigned long ip_ct_tcp_timeout_close_wait =   60 SECS;
 unsigned long ip_ct_tcp_timeout_last_ack =     30 SECS;
