--- linux/net/8021q/vlan_dev.c.orig	2007-10-27 22:58:30.474952912 +0400
+++ linux/net/8021q/vlan_dev.c	2007-10-27 22:59:19.586486824 +0400
@@ -802,7 +802,7 @@
 		/* compare the current promiscuity to the last promisc we had.. */
 		inc = vlan_dev->promiscuity - VLAN_DEV_INFO(vlan_dev)->old_promiscuity;
 		if (inc) {
-			printk(KERN_INFO "%s: dev_set_promiscuity(master, %d)\n",
+			printk(KERN_DEBUG "%s: dev_set_promiscuity(master, %d)\n",
 			       vlan_dev->name, inc);
 			dev_set_promiscuity(real_dev, inc); /* found in dev.c */
 			VLAN_DEV_INFO(vlan_dev)->old_promiscuity = vlan_dev->promiscuity;
@@ -810,7 +810,7 @@
 
 		inc = vlan_dev->allmulti - VLAN_DEV_INFO(vlan_dev)->old_allmulti;
 		if (inc) {
-			printk(KERN_INFO "%s: dev_set_allmulti(master, %d)\n",
+			printk(KERN_DEBUG "%s: dev_set_allmulti(master, %d)\n",
 			       vlan_dev->name, inc);
 			dev_set_allmulti(real_dev, inc); /* dev.c */
 			VLAN_DEV_INFO(vlan_dev)->old_allmulti = vlan_dev->allmulti;
