diff -BurpN linux-2.4.37/net/ipv4/netfilter/ipt_MASQUERADE.c linux/net/ipv4/netfilter/ipt_MASQUERADE.c
--- linux-2.4.37/net/ipv4/netfilter/ipt_MASQUERADE.c	2005-01-20 12:55:47.000000000 +0500
+++ linux/net/ipv4/netfilter/ipt_MASQUERADE.c	2009-02-24 01:04:09.000000000 +0500
@@ -140,6 +140,9 @@ static int masq_device_event(struct noti
 		/* Device was downed.  Search entire table for
 		   conntracks which were associated with that device,
 		   and forget them. */
+		/* IP address was deleted.  Search entire table for
+		   conntracks which were associated with that device,
+		   and forget them. */
 		IP_NF_ASSERT(dev->ifindex != 0);
  
 		ip_ct_iterate_cleanup(device_cmp, (void *)(long)dev->ifindex);
@@ -154,17 +157,7 @@ static int masq_inet_event(struct notifi
 			   void *ptr)
 {
 	struct net_device *dev = ((struct in_ifaddr *)ptr)->ifa_dev->dev;
-
-	if (event == NETDEV_DOWN) {
-		/* IP address was deleted.  Search entire table for
-		   conntracks which were associated with that device,
-		   and forget them. */
-		IP_NF_ASSERT(dev->ifindex != 0);
-
-		ip_ct_iterate_cleanup(device_cmp, (void *)(long)dev->ifindex);
-	}
-
-	return NOTIFY_DONE;
+	return masq_device_event(this, event, dev);
 }
 
 static struct notifier_block masq_dev_notifier = {
