diff -BurpN linux-2.4.37.2/net/ipv4/netfilter/ipt_MASQUERADE.c linux/net/ipv4/netfilter/ipt_MASQUERADE.c
--- linux-2.4.37.2/net/ipv4/netfilter/ipt_MASQUERADE.c	2009-06-07 19:50:24.000000000 +0000
+++ linux/net/ipv4/netfilter/ipt_MASQUERADE.c	2009-06-20 21:43:35.000000000 +0000
@@ -18,9 +18,6 @@
 #define DEBUGP(format, args...)
 #endif
 
-/* Lock protects masq region inside conntrack */
-static DECLARE_RWLOCK(masq_lock);
-
 /* FIXME: Multiple targets. --RR */
 static int
 masquerade_check(const char *tablename,
@@ -104,9 +101,7 @@ masquerade_target(struct sk_buff **pskb,
 	DEBUGP("newsrc = %u.%u.%u.%u\n", NIPQUAD(newsrc));
 	ip_rt_put(rt);
 
-	WRITE_LOCK(&masq_lock);
 	ct->nat.masq_index = out->ifindex;
-	WRITE_UNLOCK(&masq_lock);
 
 	/* Transfer from original range. */
 	newrange = ((struct ip_nat_multi_range)
@@ -121,13 +116,7 @@ masquerade_target(struct sk_buff **pskb,
 static inline int
 device_cmp(struct ip_conntrack *i, void *ifindex)
 {
-	int ret;
-
-	READ_LOCK(&masq_lock);
-	ret = (i->nat.masq_index == (int)(long)ifindex);
-	READ_UNLOCK(&masq_lock);
-
-	return ret;
+	return (i->nat.masq_index == (int)(long)ifindex);
 }
 
 static int masq_device_event(struct notifier_block *this,
@@ -140,6 +129,9 @@ static int masq_device_event(struct noti
 		/* Device was downed.  Search entire table for
 		   conntracks which were associated with that device,
 		   and forget them. */
+		/* IP address was deleted.  Search entire table for
+		   conntracks which were associated with that device,
+		   and forget them. */
 		IP_NF_ASSERT(dev->ifindex != 0);
  
 		ip_ct_iterate_cleanup(device_cmp, (void *)(long)dev->ifindex);
@@ -154,17 +146,7 @@ static int masq_inet_event(struct notifi
 			   void *ptr)
 {
 	struct net_device *dev = ((struct in_ifaddr *)ptr)->ifa_dev->dev;
-
-	if (event == NETDEV_DOWN) {
-		/* IP address was deleted.  Search entire table for
-		   conntracks which were associated with that device,
-		   and forget them. */
-		IP_NF_ASSERT(dev->ifindex != 0);
-
-		ip_ct_iterate_cleanup(device_cmp, (void *)(long)dev->ifindex);
-	}
-
-	return NOTIFY_DONE;
+	return masq_device_event(this, event, dev);
 }
 
 static struct notifier_block masq_dev_notifier = {
