--- linuz/Makefile	2005-03-18 18:31:27.000000000 +0300
+++ linux/Makefile	2008-12-13 12:27:26.173106151 +0300
@@ -5,7 +5,7 @@
 
 KERNELRELEASE=$(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)
 
-ARCH := $(shell uname -m | sed -e s/i.86/i386/ -e s/sun4u/sparc64/ -e s/arm.*/arm/ -e s/sa110/arm/)
+ARCH := mips
 KERNELPATH=kernel-$(shell echo $(KERNELRELEASE) | sed -e "s/-//g")
 
 CONFIG_SHELL := $(shell if [ -x "$$BASH" ]; then echo $$BASH; \
@@ -12,6 +12,8 @@
 	  else if [ -x /bin/bash ]; then echo /bin/bash; \
 	  else echo sh; fi ; fi)
 TOPDIR	:= $(shell /bin/pwd)
+SRCBASE ?= $(TOPDIR)/../..
+export SRCBASE
 
 HPATH   	= $(TOPDIR)/include
 FINDHPATH	= $(HPATH)/asm $(HPATH)/linux $(HPATH)/scsi $(HPATH)/net $(HPATH)/math-emu
@@ -100,13 +100,7 @@
 AFLAGS := -D__ASSEMBLY__ $(CPPFLAGS)
 
 check_gcc = $(shell if $(CC) $(1) -S -o /dev/null -xc /dev/null > /dev/null 2>&1; then echo "$(1)"; else echo "$(2)"; fi)
 if_gcc4 = $(shell if echo __GNUC__ | $(CC) -E -xc - | grep -q '^4$$' > /dev/null 2>&1; then echo "$(1)"; else echo "$(2)"; fi)
-
-# disable pointer signedness warnings in gcc 4.0
-CFLAGS += $(call check_gcc,-Wno-pointer-sign,)
-
-# disable stupid and dangerous "optimization" in gcc 3.2+
-CFLAGS += $(call check_gcc,-fno-delete-null-pointer-checks,)
 
 #
 # ROOT_DEV specifies the default root-device when making the image.
@@ -274,6 +272,14 @@
 # 'kbuild_2_4_nostdinc :=' or -I/usr/include for kernel code and you are not UML
 # then your code is broken!  KAO.
 
+# disable pointer signedness warnings in gcc 4.x
+CFLAGS += $(call check_gcc,-Wno-pointer-sign,)
+
+# disable stupid and dangerous "optimization" in gcc 3.2+
+CFLAGS += $(call check_gcc,-fno-delete-null-pointer-checks,)
+
+CFLAGS += -I$(SRCBASE)/include -DBCMDRIVER
+
 kbuild_2_4_nostdinc	:= -nostdinc -iwithprefix include
 export kbuild_2_4_nostdinc
 
@@ -347,15 +350,15 @@
 	@echo -n \#`cat .version` > .ver1
 	@if [ -n "$(CONFIG_SMP)" ] ; then echo -n " SMP" >> .ver1; fi
 	@if [ -f .name ]; then  echo -n \-`cat .name` >> .ver1; fi
-	@LANG=C echo ' '`date` >> .ver1
+	@LANG=C echo ' '`date +%F' '%T' '%Z` >> .ver1
 	@echo \#define UTS_VERSION \"`cat .ver1 | $(uts_truncate)`\" > .ver
 	@LANG=C echo \#define LINUX_COMPILE_TIME \"`date +%T`\" >> .ver
-	@echo \#define LINUX_COMPILE_BY \"`whoami`\" >> .ver
-	@echo \#define LINUX_COMPILE_HOST \"`hostname | $(uts_truncate)`\" >> .ver
+	@echo \#define LINUX_COMPILE_BY \"root\" >> .ver
+	@echo \#define LINUX_COMPILE_HOST \"localhost\" >> .ver
 	@([ -x /bin/dnsdomainname ] && /bin/dnsdomainname > .ver1) || \
 	 ([ -x /bin/domainname ] && /bin/domainname > .ver1) || \
 	 echo > .ver1
-	@echo \#define LINUX_COMPILE_DOMAIN \"`cat .ver1 | $(uts_truncate)`\" >> .ver
+	@echo \#define LINUX_COMPILE_DOMAIN \"localdomain\" >> .ver
 	@echo \#define LINUX_COMPILER \"`$(CC) $(CFLAGS) -v 2>&1 | tail -n 1`\" >> .ver
 	@mv -f .ver $@
 	@rm -f .ver1
--- linuz/scripts/Configure	2009-08-14 02:04:00.000000000 +0400
+++ linux/scripts/Configure	2009-10-29 11:39:13.000000000 +0300
@@ -539,7 +539,7 @@
 
 DEFAULTS=arch/$ARCH/defconfig
 if [ -f .config ]; then
-  DEFAULTS=.config
+  DEFAULTS=./.config
 fi
 
 if [ -f $DEFAULTS ]; then
@@ -548,7 +548,7 @@
   echo "#"
   . $DEFAULTS
   sed -e 's/# \(CONFIG_[^ ]*\) is not.*/\1=n/' <$DEFAULTS >.config-is-not.$$
-  . .config-is-not.$$
+  . ./.config-is-not.$$
   rm .config-is-not.$$
 else
   echo "#"
--- linux-2.4.35.5-orig/include/linux/byteorder/generic.h	2007-12-17 02:16:04.000000000 +0300
+++ linux-2.4.35.5/include/linux/byteorder/generic.h	2007-12-20 18:52:54.000000000 +0300
@@ -122,7 +122,7 @@
 #define be16_to_cpus __be16_to_cpus
 #endif
 
-#if defined(__KERNEL__)
+//#if defined(__KERNEL__)
 /*
  * Handle ntohl and suches. These have various compatibility
  * issues - like we want to give the prototype even though we
@@ -165,7 +165,7 @@
 
 #endif /* OPTIMIZE */
 
-#endif /* KERNEL */
+//#endif /* KERNEL */
 
 
 #endif /* _LINUX_BYTEORDER_GENERIC_H */
