diff -BurpN linux-2.4.37.3/drivers/net/pptp.c linux/drivers/net/pptp.c
--- linux-2.4.37.3/drivers/net/pptp.c	2009-08-28 20:49:41.000000000 +0000
+++ linux/drivers/net/pptp.c	2009-09-04 09:43:38.000000000 +0000
@@ -21,7 +21,7 @@
 #include <linux/init.h>
 #include <linux/ppp_channel.h>
 #include <linux/ppp_defs.h>
-#include "if_pppox.h"
+#include <linux/if_pppox.h>
 #include <linux/if_ppp.h>
 #include <linux/notifier.h>
 #include <linux/file.h>
@@ -89,7 +89,9 @@ static inline void *kzalloc(size_t size,
 	memset(p,0,size);
 	return p;
 }
+#endif
 
+#if 0 // already defined for mips32
 /**
  * __ffs - find first bit in word.
  * @word: The word to search
@@ -451,13 +453,16 @@ static int pptp_xmit(struct ppp_channel 
 	#endif
 	iph->version		=	4;
 	iph->ihl		=	sizeof(struct iphdr) >> 2;
-	iph->frag_off		=	0;//df;
+	if (ip_dont_fragment(sk, &rt->u.dst))
+		iph->frag_off	=	htons(IP_DF);
+	else
+		iph->frag_off	=	0;
 	iph->protocol		=	IPPROTO_GRE;
-	iph->tos		=	0;
+	iph->tos		=	sk->protinfo.af_inet.tos;
 	iph->daddr		=	rt->rt_dst;
 	iph->saddr		=	rt->rt_src;
 	#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0)
-	iph->ttl = sysctl_ip_default_ttl;
+	iph->ttl = sk->protinfo.af_inet.ttl;
 	#else
 	iph->ttl = dst_metric(&rt->u.dst, RTAX_HOPLIMIT);
 	#endif
