diff -uBp a/drivers/net/pppoe.c b/drivers/net/pppoe.c
--- a/drivers/net/pppoe.c
+++ b/drivers/net/pppoe.c
@@ -433,6 +433,10 @@ static int pppoe_disc_rcv(struct sk_buff
 	if (ph->code != PADT_CODE)
 		goto abort;
 
+	/* Ignore PADT packets whose destination address isn't ours */
+	if (memcmp(skb->mac.ethernet->h_dest, dev->dev_addr, ETH_ALEN))
+		goto abort;
+
 	po = get_item((unsigned long) ph->sid, skb->mac.ethernet->h_source);
 	if (po) {
 		struct sock *sk = po->sk;
-- 
