--- linux-2.4.37/net/ipv4/netfilter/ip_conntrack_core.c	2008-12-02 11:01:34.000000000 +0300
+++ linux/net/ipv4/netfilter/ip_conntrack_core.c	2009-02-06 12:05:32.000000000 +0300
@@ -1439,7 +1439,8 @@
 	nf_unregister_sockopt(&so_getorigdst);
 }
 
-static int hashsize = 0;
+/* OpenWRT mod */
+static int hashsize = 5953;
 MODULE_PARM(hashsize, "i");
 
 int __init ip_conntrack_init(void)
@@ -1460,7 +1460,7 @@
 		if (ip_conntrack_htable_size < 16)
 			ip_conntrack_htable_size = 16;
 	}
-	ip_conntrack_max = 8 * ip_conntrack_htable_size;
+	ip_conntrack_max = 2 * ip_conntrack_htable_size;
 
 	printk("ip_conntrack version %s (%u buckets, %d max)"
 	       " - %Zd bytes per conntrack\n", IP_CONNTRACK_VERSION,
diff -urbBp linux-2.4.20/net/ipv4/route.c tomato/net/ipv4/route.c
--- linux-2.4.37/net/ipv4/route.c	2008-12-02 11:01:34.000000000 +0300
+++ tomato/net/ipv4/route.c	2009-02-06 12:26:20.000000000 +0300
@@ -114,7 +114,8 @@ int ip_rt_redirect_load		= HZ / 50;
 int ip_rt_redirect_silence	= ((HZ / 50) << (9 + 1));
 int ip_rt_error_cost		= HZ;
 int ip_rt_error_burst		= 5 * HZ;
-int ip_rt_gc_elasticity		= 8;
+/* SpeedMod */
+int ip_rt_gc_elasticity		= 2;
 int ip_rt_mtu_expires		= 10 * 60 * HZ;
 int ip_rt_min_pmtu		= 512 + 20 + 20;
 int ip_rt_min_advmss		= 256;
@@ -2619,7 +2619,10 @@ void __init ip_rt_init(void)
 	if (!ipv4_dst_ops.kmem_cachep)
 		panic("IP: failed to allocate ip_dst_cache\n");
 
-	goal = num_physpages >> (26 - PAGE_SHIFT);
+	/* SpeedMod: goal=8 gives 4096 buckets for 4K page size & 32M memory */
+	goal = num_physpages >> (22 - PAGE_SHIFT);
+	if (goal > 32)
+		goal = 32;
 
 	for (order = 0; (1UL << order) < goal; order++)
 		/* NOTHING */;
@@ -2649,8 +2649,9 @@ void __init ip_rt_init(void)
 		rt_hash_table[i].chain = NULL;
 	}
 
 	ipv4_dst_ops.gc_thresh = (rt_hash_mask + 1);
-	ip_rt_max_size = (rt_hash_mask + 1) * 16;
+	/* SpeedMod */
+	ip_rt_max_size = (rt_hash_mask + 1) * 4;
 
 	devinet_init();
 	ip_fib_init();
