--- miniupnpd.orig/linux/getifstats.c
+++ miniupnpd/linux/getifstats.c
@@ -87,6 +87,7 @@ getifstats(const char * ifname, struct i
 	}
 	fclose(f);
 	/* get interface speed */
+	/* NB! some interfaces, like ppp, don't support speed queries */
 	snprintf(fname, sizeof(fname), "/sys/class/net/%s/speed", ifname);
 	f = fopen(fname, "r");
 	if(f) {
@@ -94,8 +95,6 @@ getifstats(const char * ifname, struct i
 			data->baudrate = 1000000*atoi(line);
 		}
 		fclose(f);
-	} else {
-		syslog(LOG_WARNING, "cannot read %s file : %m", fname);
 	}
 #ifdef ENABLE_GETIFSTATS_CACHING
 	if(r==0 && current_time!=((time_t)-1)) {
