soap: Fix error path memleaks

diff -uBp a/upnpsoap.c b/upnpsoap.c
--- a/upnpsoap.c
+++ b/upnpsoap.c
@@ -923,6 +923,7 @@ http://www.upnp.org/schemas/gw/WANIPConn
 	if(bodylen < 0)
 	{
 		SoapError(h, 501, "ActionFailed");
+		free(body);
 		return;
 	}
 	memcpy(body+bodylen, list_start, sizeof(list_start));
@@ -936,15 +937,19 @@ http://www.upnp.org/schemas/gw/WANIPConn
 		/* have a margin of 1024 bytes to store the new entry */
 		if((unsigned int)bodylen + 1024 > bodyalloc)
 		{
+			char *new_body;
+
 			bodyalloc += 4096;
-			body = realloc(body, bodyalloc);
-			if(!body)
+			new_body = realloc(body, bodyalloc);
+			if(!new_body)
 			{
 				ClearNameValueList(&data);
 				SoapError(h, 501, "ActionFailed");
 				free(port_list);
+				free(body);
 				return;
 			}
+			body = new_body;
 		}
 		rhost[0] = '\0';
 		r = upnp_get_redirection_infos(port_list[i], protocol, &iport,
