--- linux/arch/mips/bcm947xx/time.c.orig	2007-10-28 11:32:14.886198256 +0300
+++ linux/arch/mips/bcm947xx/time.c	2007-10-28 11:35:20.156032960 +0300
@@ -57,8 +57,8 @@
 	if (!(hz = sb_cpu_clock(sbh)))
 		hz = 100000000;
 
-	printk("CPU: BCM%04x rev %d at %d MHz\n", sb_chip(sbh), sb_chiprev(sbh),
-	       (hz + 500000) / 1000000);
+	printk("CPU: BCM%04x rev %d pkg %d at %d MHz\n", sb_chip(sbh),
+	        sb_chiprev(sbh), sb_chippkg(sbh), (hz + 500000) / 1000000);
 
 	/* Set MIPS counter frequency for fixed_rate_gettimeoffset() */
 	mips_counter_frequency = hz / 2;
--- linux/arch/mips/bcm947xx/setup.c.orig	2007-10-28 11:37:15.630478176 +0300
+++ linux/arch/mips/bcm947xx/setup.c	2007-10-28 11:50:33.153236288 +0300
@@ -173,11 +173,11 @@
 const char *
 get_system_type(void)
 {
-	static char s[32];
+	static char s[40];
 
 	if (bcm947xx_sbh) {
-		sprintf(s, "Broadcom BCM%X chip rev %d", sb_chip(bcm947xx_sbh),
-			sb_chiprev(bcm947xx_sbh));
+		sprintf(s, "Broadcom BCM%X chip rev %d pkg %d", sb_chip(bcm947xx_sbh),
+			sb_chiprev(bcm947xx_sbh), sb_chippkg(bcm947xx_sbh));
 		return s;
 	}
 	else
