diff -BurpN linux-2.4.37.10.orig/drivers/net/Makefile linux/drivers/net/Makefile
--- linux-2.4.37.10.orig/drivers/net/Makefile	2004-10-08 18:56:45.000000000 +0000
+++ linux/drivers/net/Makefile	2010-09-12 11:21:24.810000002 +0000
@@ -307,6 +307,13 @@ ifeq ($(CONFIG_WL),y)
   obj-y += wl/wl.o
 endif
 
+ifdef CONFIG_EMF
+subdir-$(CONFIG_EMF) += emf igs
+endif
+ifeq ($(CONFIG_EMF),y)
+  obj-y += emf/emf.o igs/igs.o
+endif
+
 include $(TOPDIR)/Rules.make
 
 clean:
diff -BurpN linux-2.4.37.10.orig/drivers/net/emf/Makefile linux/drivers/net/emf/Makefile
--- linux-2.4.37.10.orig/drivers/net/emf/Makefile	1970-01-01 00:00:00.000000000 +0000
+++ linux/drivers/net/emf/Makefile	2010-09-12 09:01:02.000000000 +0000
@@ -0,0 +1,30 @@
+#
+# Makefile for the Broadcom EMF Layer
+#
+# Copyright (C) 2008, Broadcom Corporation
+# All Rights Reserved.
+# 
+# THIS SOFTWARE IS OFFERED "AS IS", AND BROADCOM GRANTS NO WARRANTIES OF ANY
+# KIND, EXPRESS OR IMPLIED, BY STATUTE, COMMUNICATION OR OTHERWISE. BROADCOM
+# SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
+# FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
+#
+#
+
+O_TARGET	:= emf.o
+EMF_OBJS	:= emfc.o emf_linux.o
+
+# Search for sources under src/emf/emf
+ifneq ($(wildcard $(SRCBASE)/emf/emf),)
+	EXTRA_CFLAGS	+= -I$(SRCBASE)/include/emf/emf -I$(SRCBASE)/include/emf/igs
+	EXTRA_CFLAGS	+= -I$(SRCBASE)/emf/emf -I$(SRCBASE)/emf/igs
+	vpath %.c $(SRCBASE)/emf/emf
+	export-objs	:=
+	obj-y		:= $(EMF_OBJS)
+	obj-m		:= $(O_TARGET)
+else
+	obj-y		:= $(SRCBASE)/emf/emf.o
+	obj-m		:= $(O_TARGET)
+endif
+
+include $(TOPDIR)/Rules.make
diff -BurpN linux-2.4.37.10.orig/drivers/net/hnd/Config.in linux/drivers/net/hnd/Config.in
--- linux-2.4.37.10.orig/drivers/net/hnd/Config.in	2007-03-20 12:21:03.000000000 +0000
+++ linux/drivers/net/hnd/Config.in	2010-09-12 11:14:50.580000002 +0000
@@ -44,5 +44,6 @@ mainmenu_option next_comment
                 apsta_onchip_g	CONFIG_WL_USE_APSTA_ONCHIP_G"
          fi
       fi
+      dep_tristate '  Efficient Multicast Forwarding & IGMP Snooping Layers' CONFIG_EMF $CONFIG_NETFILTER
    fi
 endmenu
diff -BurpN linux-2.4.37.10.orig/drivers/net/igs/Makefile linux/drivers/net/igs/Makefile
--- linux-2.4.37.10.orig/drivers/net/igs/Makefile	1970-01-01 00:00:00.000000000 +0000
+++ linux/drivers/net/igs/Makefile	2010-09-12 09:02:06.000000000 +0000
@@ -0,0 +1,30 @@
+#
+# Makefile for the Broadcom IGS Layer
+#
+# Copyright (C) 2008, Broadcom Corporation
+# All Rights Reserved.
+# 
+# THIS SOFTWARE IS OFFERED "AS IS", AND BROADCOM GRANTS NO WARRANTIES OF ANY
+# KIND, EXPRESS OR IMPLIED, BY STATUTE, COMMUNICATION OR OTHERWISE. BROADCOM
+# SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
+# FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
+#
+#
+
+O_TARGET	:= igs.o
+IGS_OBJS	:= igsc.o igsc_sdb.o igs_linux.o osl_linux.o
+
+# Search for sources under src/emf/igs
+ifneq ($(wildcard $(SRCBASE)/emf/igs),)
+	EXTRA_CFLAGS	+= -I$(SRCBASE)/include/emf/igs -I$(SRCBASE)/include/emf/emf
+	EXTRA_CFLAGS	+= -I$(SRCBASE)/emf/igs -I$(SRCBASE)/emf/emf
+	vpath %.c $(SRCBASE)/emf/igs
+	export-objs	:=
+	obj-y		:= $(IGS_OBJS)
+	obj-m		:= $(O_TARGET)
+else
+	obj-y		:= $(SRCBASE)/emf/igs.o
+	obj-m		:= $(O_TARGET)
+endif
+
+include $(TOPDIR)/Rules.make
