Disable dependence from PAGE_SIZE, force it to 32kB

diff -BurpN linux-2.4.37.1.orig/include/linux/nfsd/const.h linux/include/linux/nfsd/const.h
--- linux-2.4.37.1.orig/include/linux/nfsd/const.h	2009-04-19 18:33:42.000000000 +0000
+++ linux/include/linux/nfsd/const.h	2009-05-04 13:52:23.000000000 +0000
@@ -25,11 +25,14 @@
  * 2) is a multiple of page size 3) can be successfully kmalloc()ed
  * by each nfsd.   
  */
+/* Disable dependence from PAGE_SIZE, force it to 32kB
 #if PAGE_SIZE > (16*1024)
 #define NFSSVC_MAXBLKSIZE	(32*1024)
 #else
 #define NFSSVC_MAXBLKSIZE	(2*PAGE_SIZE)
 #endif
+ */
+#define NFSSVC_MAXBLKSIZE	(32*1024)
 
 #ifdef __KERNEL__
 
