diff -uBp a/fs/ntfs/fs.c b/fs/ntfs/fs.c
--- a/fs/ntfs/fs.c	2010-02-02 00:04:46.000000000 +0300
+++ b/fs/ntfs/fs.c	2010-07-22 12:45:48.000000000 +0400
@@ -165,7 +165,7 @@ struct ntfs_filldir {
 
 static int ntfs_printcb(ntfs_u8 *entry, void *param)
 {
-	unsigned long inum = NTFS_GETU64(entry) & 0xffffffffffff;
+	ino_t inum = NTFS_GETU64(entry) & 0xffffffffffffULL;
 	struct ntfs_filldir *nf = param;
 	u32 flags = NTFS_GETU32(entry + 0x48);
 	char show_sys_files = 0;
