--- linux-2.4.35.3/drivers/usb/storage/scsiglue.c	2007-09-24 02:02:58.000000000 +0400
+++ linux/drivers/usb/storage/scsiglue.c	2007-12-16 22:34:58.000000000 +0300
@@ -389,6 +389,14 @@
 	SPRINTF("         GUID: " GUID_FORMAT "\n", GUID_ARGS(us->guid));
 	SPRINTF("     Attached: %s\n", us->pusb_dev ? "Yes" : "No");
 
+        if (us->pusb_dev && us->pusb_dev->devpath)
+        {
+		SPRINTF("         Port: %s\n", us->pusb_dev->devpath);
+
+		if (us->pusb_dev->bus && us->pusb_dev->bus->bus_name)
+		SPRINTF("          Bus: %s-%s\n", us->pusb_dev->bus->bus_name, us->pusb_dev->devpath);
+        }
+
 	/*
 	 * Calculate start of next buffer, and return value.
 	 */
