Always export INTERFACE for hotplug

--- linux-2.4.37/drivers/usb/usb.c	2009-05-13 20:32:13.000000000 +0400
+++ linux/drivers/usb/usb.c	2009-05-13 20:44:23.000000000 +0400
@@ -813,7 +813,7 @@ int usb_find_interface_driver_for_ifnum(
 static void call_policy_interface (char *verb, struct usb_device *dev, int interface)
 {
 	char *argv [3], **envp, *buf, *scratch;
-	int i = 0, value;
+	int i = 0, value, alt;
 
 	if (!hotplug_path [0])
 		return;
@@ -888,16 +888,15 @@ static void call_policy_interface (char
 			    dev->descriptor.bDeviceClass,
 			    dev->descriptor.bDeviceSubClass,
 			    dev->descriptor.bDeviceProtocol) + 1;
-	if (dev->descriptor.bDeviceClass == 0) {
-		int alt = dev->actconfig->interface [interface].act_altsetting;
 
-		envp [i++] = scratch;
-		scratch += sprintf (scratch, "INTERFACE=%d/%d/%d",
+	alt = dev->actconfig->interface [interface].act_altsetting;
+
+	envp [i++] = scratch;
+	scratch += sprintf (scratch, "INTERFACE=%d/%d/%d",
 			dev->actconfig->interface [interface].altsetting [alt].bInterfaceClass,
 			dev->actconfig->interface [interface].altsetting [alt].bInterfaceSubClass,
 			dev->actconfig->interface [interface].altsetting [alt].bInterfaceProtocol)
 			+ 1;
-	}
 	envp [i++] = 0;
 	/* assert: (scratch - buf) < sizeof buf */
 
