diff -urBp rp-pppoe-3.10/src/configure rp-pppoe/src/configure
--- rp-pppoe-3.10/src/configure	2008-06-30 18:00:42.000000000 +0400
+++ rp-pppoe/src/configure	2009-02-11 16:21:20.000000000 +0300
@@ -6271,8 +6271,12 @@ esac
 
 { echo "$as_me:$LINENO: checking packing order of bit fields" >&5
 echo $ECHO_N "checking packing order of bit fields... $ECHO_C" >&6; }
+if test "${rpppoe_cv_pack_bitfields+set}" = set; then
+  echo $ECHO_N "(cached) $ECHO_C" >&6
+else
+
 if test "$cross_compiling" = yes; then
-  $ECHO "no defaults for cross-compiling"; exit 0
+  $ECHO "no defaults for cross-compiling"; exit 1
 else
   cat >conftest.$ac_ext <<_ACEOF
 /* confdefs.h.  */
@@ -6336,6 +6340,7 @@ fi
 rm -f core *.core core.conftest.* gmon.out bb.out conftest$ac_exeext conftest.$ac_objext conftest.$ac_ext
 fi
 
+fi
 
 
 if test "$rpppoe_cv_pack_bitfields" = "rev" ; then
diff -urBp rp-pppoe-3.10/src/configure.in rp-pppoe/src/configure.in
--- rp-pppoe-3.10/src/configure.in	2008-06-30 18:00:42.000000000 +0400
+++ rp-pppoe/src/configure.in	2009-02-11 16:19:20.000000000 +0300
@@ -231,6 +231,7 @@ esac
 
 dnl Figure out packing order of structures
 AC_MSG_CHECKING([packing order of bit fields])
+AC_CACHE_VAL(rpppoe_cv_pack_bitfields,[
 AC_TRY_RUN([
 union foo {
     struct bar {
@@ -254,7 +255,8 @@ main(void)
 	return 2;
     }
 }], rpppoe_cv_pack_bitfields=normal, rpppoe_cv_pack_bitfields=rev,
-$ECHO "no defaults for cross-compiling"; exit 0)
+[$ECHO "no defaults for cross-compiling"; exit 1])
+])
 
 if test "$rpppoe_cv_pack_bitfields" = "rev" ; then
 	AC_MSG_RESULT(reversed)
