From 3237f5c307c066a008909f35115e21c9473504ab
From: Denys Vlasenko
Date: Thu, 03 Feb 2011 03:00:23 +0000
Subject: tar: support -T - and -X -

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/archival/tar.c b/archival/tar.c
index 972f8d0..e9dc41f 100644
--- a/archival/tar.c
+++ b/archival/tar.c
@@ -655,7 +655,7 @@ static llist_t *append_file_list_to_list(llist_t *list)
 	llist_t *newlist = NULL;
 
 	while (list) {
-		src_stream = xfopen_for_read(llist_pop(&list));
+		src_stream = xfopen_stdin(llist_pop(&list));
 		while ((line = xmalloc_fgetline(src_stream)) != NULL) {
 			/* kill trailing '/' unless the string is just "/" */
 			char *cp = last_char_is(line, '/');
--
cgit v0.8.2.1
