diff -BurpN router/rc/init.c gateway/rc/init.c
--- router/rc/init.c	2008-03-19 17:07:24.000000000 +0000
+++ gateway/rc/init.c	2009-04-26 14:09:49.000000000 +0000
@@ -159,6 +159,13 @@ run_shell(int timeout, int nowait)
 	}
 }
 
+void
+preshutdown_system(void)
+{
+	eval("/usr/local/sbin/pre-shutdown");
+	sleep(2);
+}
+
 static void
 shutdown_system(void)
 {
@@ -171,9 +178,6 @@ shutdown_system(void)
 	/* Disconnect pppd - need this for PPTP/L2TP to finish gracefully */
 	eval("killall", "pppd");
 
-	eval("/usr/local/sbin/pre-shutdown");
-	sleep(2);
-	
 	if (exists("/dev/misc/rtc"))
 		eval("/sbin/hwclock", "-w");
 
@@ -226,7 +230,6 @@ fatal_signal(int sig)
 	case SIGTRAP: message = "Trace trap"; break;
 	case SIGPWR: message = "Power failure"; break;
 	case SIGTERM: message = "Terminated"; break;
-	case SIGUSR1: message = "User-defined signal 1"; break;
 	}
 
 	if (message)
@@ -234,6 +237,14 @@ fatal_signal(int sig)
 	else
 		cprintf("Caught signal %d\n", sig);
 
+	/* Perform pre-shutdown script only with common signals */
+	switch (sig) {
+	case SIGQUIT:
+	case SIGTERM:
+		preshutdown_system();
+		break;
+	}
+
 	shutdown_system();
 	sleep(2);
 
diff -BurpN router/rc/rc.h gateway/rc/rc.h
--- router/rc/rc.h	2008-12-28 08:59:11.000000000 +0000
+++ gateway/rc/rc.h	2009-04-26 13:47:24.000000000 +0000
@@ -56,6 +56,7 @@ int console_init(void);
 pid_t run_shell(int timeout, int nowait);
 void signal_init(void);
 void fatal_signal(int sig);
+void preshutdown_system(void);
 
 /* interface */
 int ifconfig(char *ifname, int flags, char *addr, char *netmask);
