diff -uBp gateway/rc.old/network_ex.c gateway/rc/network_ex.c
--- gateway/rc.old/network_ex.c	2009-09-18 02:20:51.000000000 +0400
+++ gateway/rc/network_ex.c	2009-09-25 11:53:22.000000000 +0400
@@ -128,16 +128,34 @@ int start_pppd(char *prefix)
 	fprintf(fp, "password '%s'\n", 
 		nvram_safe_get(strcat_r(prefix, "pppoe_passwd", tmp)));
        	       	
-       	if (nvram_match(strcat_r(prefix, "proto", tmp), "pptp")) 
-       	{
+       	if (nvram_match(strcat_r(prefix, "proto", tmp), "pptp")) {
+#ifdef __CONFIG_ACCEL_PPTP__
+#ifdef __CONFIG_PPTP__
+	    if (!nvram_match(strcat_r(prefix, "pptp_kernel", tmp), "0"))
+#endif
+	    {
+		fprintf(fp, "plugin pptp.so\n");
+		fprintf(fp, "pptp_server %s\n", 
+   			nvram_invmatch("wan_heartbeat_x", "") ?
+   			nvram_safe_get("wan_heartbeat_x") : 
+   			nvram_safe_get(strcat_r(prefix, "pppoe_gateway", tmp))); 
+ 	    }
+#ifdef __CONFIG_PPTP__
+	    else
+#endif
+#endif
+#ifdef __CONFIG_PPTP__
+	    {
 		fprintf(fp, "connect true\n");    
 		fprintf(fp, "sync pty '/usr/sbin/pptp --idle-wait 0 %s --nolaunchpppd --nobuffer --sync'\n", 
    			nvram_invmatch("wan_heartbeat_x", "") ?
    			nvram_safe_get("wan_heartbeat_x") : 
    			nvram_safe_get(strcat_r(prefix, "pppoe_gateway", tmp))); 
 		fprintf(fp, "lock\n");
-		/* see KB Q189595 -- historyless & mtu */
-		fprintf(fp, "nomppe-stateful %s mtu 1400\n",
+	    }
+#endif
+	    /* see KB Q189595 -- historyless & mtu */
+	    fprintf(fp, "nomppe-stateful %s mtu 1400\n",
 			nvram_safe_get(strcat_r(prefix, "pptp_options_x", tmp)));
 	} else {
 		fprintf(fp, "nomppe nomppc\n");
