--- gateway.orig/rc/services_ex.c	2009-02-26 01:18:14.000000000 +0500
+++ gateway/rc/services_ex.c	2009-02-26 01:18:31.000000000 +0500
@@ -628,6 +628,8 @@ start_ddns(void)
 		strcpy(service, "justlinux");
 	else if (strcmp(server, "WWW.EASYDNS.COM")==0)
 		strcpy(service, "easydns");
+	else if (strcmp(server, "WWW.DNSOMATIC.COM")==0)
+		strcpy(service, "dnsomatic");
 	else strcpy(service, "dyndns");
 			
 	sprintf(usrstr, "%s:%s", user, passwd);
