diff -BurpN gateway/rc.orig/services_ex.c gateway/rc/services_ex.c
--- gateway/rc.orig/services_ex.c	2009-11-14 16:53:35.000000000 +0000
+++ gateway/rc/services_ex.c	2009-11-17 09:41:24.000000000 +0000
@@ -148,7 +148,7 @@ void diag_PaN(void)
    fprintf(stderr, "echo for PaN ::: &&&PaN\r\n");
 }
 
-#if defined(DNSMASQ) || defined(DDNSCONF)
+#if defined(DNSMASQ) || defined(DDNSCONF) || 1
 size_t
 fappend(char *name, FILE *f)
 {
@@ -1305,7 +1305,7 @@ void restart_smbd()
 				fprintf(fp, "\tcomment = %s\n", nvram_get(tmp));
 		}
 	}
-			
+	fappend("/usr/local/etc/smb.conf", fp);
 	fclose(fp);
 
 	eval("/usr/sbin/nmbd", "-D");
