--- libbcmcrypto.orig/Makefile	2004-10-14 23:11:28.000000000 +0400
+++ libbcmcrypto/Makefile	2005-03-27 17:46:48.705385312 +0400
@@ -20,7 +20,7 @@
 LDFLAGS += -L.
 
 vpath %.c $(SRCBASE)/bcmcrypto
-vpath %.o $(SRCBASE)/router/libbcmcrypto/prebuilt
+vpath %.o $(TOP)/libbcmcrypto/prebuilt
 
 OBJS := aes.o aeskeywrap.o rijndael-alg-fst.o dh.o bn.o sha1.o passhash.o prf.o md5.o hmac.o rc4.o
 
@@ -33,9 +33,10 @@
 all: libbcmcrypto.so
 
 install: all
+	rm -rf $(INSTALLDIR)
 	install -d $(INSTALLDIR)/usr/lib
-	install -m 755 libbcmcrypto.so $(INSTALLDIR)/usr/lib
-	$(STRIP) $(INSTALLDIR)/usr/lib/libbcmcrypto.so
+	install -m 755 libbcmcrypto.so $(INSTALLDIR)/usr/lib/libbcmcrypto.so
+	$(STRIP) $(INSTALLDIR)/usr/lib/libbcmcrypto.so
 
 clean:
 	rm -f *.o *.so *.a $(MAP) $(UNR) $(SYM)
@@ -43,8 +44,11 @@
 libbcmcrypto.so: $(OBJS)
 	$(LD) -shared -o $@ $^
 
+libbcmcrypto.a: $(OBJS)
+	$(AR) cr $@ $^
+
 # rule for removing unneeded symbols in the shared library
-optimize: libbcmcrypto.so
+optimize: libbcmcrypto.so libbcmcrypto.a
 	$(NM) -o --defined-only --no-sort libbcmcrypto.so | cut -d' ' -f3 > $(MAP)
 	$(NM) --dynamic -u --no-sort $(BINARIES) | sort -u > $(UNR)
 	rm -rf $(SYM)
@@ -53,9 +57,8 @@
 	fi ; done
 	# if no symbols are needed then delete the shared lib
 	if ls $(SYM) ; then \
-	$(AR) cr libbcmcrypto.a $(OBJS) ; \
-	xargs -t $(LD) -shared -o libbcmcrypto.so libbcmcrypto.a < $(SYM) ; \
-	install libbcmcrypto.so $(TARGETDIR)/usr/lib ; \
-	$(STRIP) $(TARGETDIR)/usr/lib/libbcmcrypto.so ; \
+	xargs -t $(LD) -shared -o lib.so libbcmcrypto.a < $(SYM) ; \
+	install lib.so $(TARGETDIR)/usr/lib/libbcmcrypto.so ; \
+	$(STRIP) $(TARGETDIR)/usr/lib/libbcmcrypto.so ; \
 	else \
	rm $(TARGETDIR)/usr/lib/libbcmcrypto.so ; fi
