--- accel-pptp.orig/pppd_plugin/src/pptp_ctrl.c	2010-08-21 09:07:46.000000000 +0000
+++ accel-pptp/pppd_plugin/src/pptp_ctrl.c	2011-07-21 15:13:19.097301000 +0000
@@ -9,6 +9,7 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <netinet/in.h>
+#include <netinet/tcp.h>
 #include <unistd.h>
 #include <stdlib.h>
 #include <assert.h>
@@ -258,6 +259,7 @@ static void ctrlp_rep( void * buffer, in
 PPTP_CONN * pptp_conn_open(int inet_sock, int isclient, pptp_conn_cb callback)
 {
     PPTP_CONN *conn;
+    int on = 1;
     /* Allocate structure */
     if ((conn = malloc(sizeof(*conn))) == NULL) return NULL;
     if ((conn->call = vector_create()) == NULL) { free(conn); return NULL; }
@@ -282,6 +284,9 @@ PPTP_CONN * pptp_conn_open(int inet_sock
     }
     /* Make this socket non-blocking. */
     fcntl(conn->inet_sock, F_SETFL, O_NONBLOCK);
+    /* Disable nagle */
+    setsockopt(conn->inet_sock, IPPROTO_TCP, TCP_NODELAY, &on, sizeof(on));
+
     /* Request connection from server, if this is a client */
     if (isclient) {
         struct pptp_start_ctrl_conn packet = {
