 
 1. string.h detection
 2. broken configure fix
 
diff -BurpN -x'*.o' router/ez-ipupdate/config.h.in gateway/ez-ipupdate/config.h.in
--- router/ez-ipupdate/config.h.in	2004-11-11 07:26:51.000000000 +0000
+++ gateway/ez-ipupdate/config.h.in	2008-12-20 07:14:17.000000000 +0000
@@ -132,6 +132,9 @@
 /* Define if you have the <stdarg.h> header file.  */
 #undef HAVE_STDARG_H
 
+/* Define if you have the <string.h> header file.  */
+#undef HAVE_STRING_H
+
 /* Define if you have the <sys/socket.h> header file.  */
 #undef HAVE_SYS_SOCKET_H
 
diff -BurpN -x'*.o' router/ez-ipupdate/configure gateway/ez-ipupdate/configure
--- router/ez-ipupdate/configure	2004-11-11 07:26:51.000000000 +0000
+++ gateway/ez-ipupdate/configure	2008-12-20 07:16:10.000000000 +0000
@@ -1505,6 +1505,7 @@ for ac_hdr in arpa/inet.h \
 		  syslog.h \
 		  pwd.h \
 		  stdarg.h \
+		  string.h \
 		  grp.h \
 		  errno.h \
 		  sys/sockio.h \
@@ -1521,6 +1522,8 @@ else
 #line 1522 "configure"
 #include "confdefs.h"
 #include <$ac_hdr>
+
+int main() { return 0; }
 EOF
 ac_try="$ac_cpp conftest.$ac_ext >/dev/null 2>conftest.out"
 { (eval echo configure:1527: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
@@ -1565,6 +1568,8 @@ else
 #line 1566 "configure"
 #include "confdefs.h"
 #include <$ac_hdr>
+
+int main() { return 0; }
 EOF
 ac_try="$ac_cpp conftest.$ac_ext >/dev/null 2>conftest.out"
 { (eval echo configure:1571: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
diff -BurpN -x'*.o' router/ez-ipupdate/configure.in gateway/ez-ipupdate/configure.in
--- router/ez-ipupdate/configure.in	2004-11-11 07:26:51.000000000 +0000
+++ gateway/ez-ipupdate/configure.in	2008-12-20 07:14:17.000000000 +0000
@@ -65,6 +65,7 @@ AC_CHECK_HEADERS( arpa/inet.h \
 		  syslog.h \
 		  pwd.h \
 		  stdarg.h \
+		  string.h \
 		  grp.h \
 		  errno.h \
 		  sys/sockio.h \
diff -BurpN -x'*.o' router/ez-ipupdate/Makefile.in gateway/ez-ipupdate/Makefile.in
--- router/ez-ipupdate/Makefile.in	2004-11-11 07:26:51.000000000 +0000
+++ gateway/ez-ipupdate/Makefile.in	2008-12-20 07:14:17.000000000 +0000
@@ -115,7 +115,7 @@ Makefile: $(srcdir)/Makefile.in  $(top_b
 	cd $(top_builddir) \
 	  && CONFIG_FILES=$@ CONFIG_HEADERS= $(SHELL) ./config.status
 
-$(ACLOCAL_M4):  configure.in 
+$(ACLOCAL_M4):
 	cd $(srcdir) && $(ACLOCAL)
 
 config.status: $(srcdir)/configure $(CONFIG_STATUS_DEPENDENCIES)
@@ -123,11 +123,6 @@ config.status: $(srcdir)/configure $(CON
 $(srcdir)/configure: $(srcdir)/configure.in $(ACLOCAL_M4) $(CONFIGURE_DEPENDENCIES)
 	cd $(srcdir) && $(AUTOCONF)
 
-config.h: stamp-h
-	@if test ! -f $@; then \
-		rm -f stamp-h; \
-		$(MAKE) stamp-h; \
-	else :; fi
 stamp-h: $(srcdir)/config.h.in $(top_builddir)/config.status
 	cd $(top_builddir) \
 	  && CONFIG_FILES= CONFIG_HEADERS=config.h \
@@ -306,7 +301,7 @@ install: install-am
 uninstall-am: uninstall-binPROGRAMS
 uninstall: uninstall-am
 all-am: Makefile $(PROGRAMS) config.h
-all-redirect: all-am
+all-redirect: $(PROGRAMS)
 install-strip:
 	$(MAKE) $(AM_MAKEFLAGS) AM_INSTALL_PROGRAM_FLAGS=-s install
 installdirs:
