From 6500d123d01af895d2b8eae1a36e553a1f75e211
From: Denys Vlasenko
Date: Tue, 17 Apr 2012 15:10:31 +0000
Subject: udhcp: make arpping code resistant to time jumps

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/networking/udhcp/arpping.c b/networking/udhcp/arpping.c
--- a/networking/udhcp/arpping.c
+++ b/networking/udhcp/arpping.c
@@ -118,8 +118,13 @@ int FAST_FUNC arpping(uint32_t test_nip,
 				break;
 			}
 		}
-		timeout_ms -= (unsigned)monotonic_ms() - prevTime;
-	} while (timeout_ms > 0);
+		timeout_ms -= (unsigned)monotonic_ms() - prevTime + 1;
+
+		/* We used to check "timeout_ms > 0", but
+		 * this is more under/overflow-resistant
+		 * (people did see overflows here when system time jumps):
+		 */
+	} while ((unsigned)timeout_ms <= 2000);
 
  ret:
 	close(s);
--
cgit v0.9.0.1-2-gef13
