diff -BurpN busybox-1.17.1/examples/depmod.pl busybox/examples/depmod.pl
--- busybox-1.17.1/examples/depmod.pl	2010-07-06 02:25:53.000000000 +0000
+++ busybox/examples/depmod.pl	2010-07-25 18:51:49.000000000 +0000
@@ -118,6 +118,8 @@ warn "**** Finished locating modules\n" 
 foreach my $obj ( @liblist ){
     # turn the input file name into a target tag name
     my ($tgtname) = $obj =~ m-(/lib/modules/.*)$-;
+    
+    $tgtname = "/opt" . $tgtname;
 
     warn "\nMODULE = $tgtname\n" if $verbose;
 
diff -BurpN busybox-1.17.1/modutils/insmod.c busybox/modutils/insmod.c
--- busybox-1.17.1/modutils/insmod.c	2010-07-06 02:25:54.000000000 +0000
+++ busybox/modutils/insmod.c	2010-07-25 18:53:17.000000000 +0000
@@ -9,6 +9,8 @@
 
 #include "libbb.h"
 #include "modutils.h"
+#include <sys/utsname.h>
+#include <fnmatch.h>
 
 /* 2.6 style insmod has no options and required filename
  * (not module name - .ko can't be omitted) */
@@ -35,8 +37,11 @@
 int insmod_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
 int insmod_main(int argc UNUSED_PARAM, char **argv)
 {
+	struct utsname uts;
+	struct stat st;
 	char *filename;
-	int rc;
+	FILE *fp = NULL;
+	int rc, pos;
 
 	/* Compat note:
 	 * 2.6 style insmod has no options and required filename
@@ -55,6 +60,28 @@ int insmod_main(int argc UNUSED_PARAM, c
 	if (!filename)
 		bb_show_usage();
 
+	/* Get a filedesc for the module.  Check we we have a complete path */
+	if (stat(filename, &st) < 0 || !S_ISREG(st.st_mode) ||
+		(fp = fopen(filename, "r")) == NULL) {
+                /* Hmm.  Could not open it.  First search under /lib/modules/`uname -r` */
+    		xchdir("/lib/modules");
+	        uname(&uts);
+    		xchdir(uts.release);
+
+		pos = strlen(filename) - 2;
+		if (get_linux_version_code() < KERNEL_VERSION(2,6,0)) {
+			if (pos < 0) pos = 0;
+			if (strncmp(&filename[pos], ".o", 2) !=0 )
+				filename = xasprintf("%s.o", filename);
+		} else {
+			if (--pos < 0) pos = 0;
+			if (strncmp(&filename[pos], ".ko", 3) !=0 )
+				filename = xasprintf("%s.ko", filename);
+		}
+	}
+	if (fp != NULL)
+		fclose(fp);
+
 	rc = bb_init_module(filename, parse_cmdline_module_options(argv, /*quote_spaces:*/ 0));
 	if (rc)
 		bb_error_msg("can't insert '%s': %s", filename, moderror(rc));
diff -BurpN busybox-1.17.1/modutils/modutils-24.c busybox/modutils/modutils-24.c
--- busybox-1.17.1/modutils/modutils-24.c	2010-07-06 02:25:54.000000000 +0000
+++ busybox/modutils/modutils-24.c	2010-07-25 18:51:49.000000000 +0000
@@ -3801,7 +3801,7 @@ int FAST_FUNC bb_init_module_24(const ch
 		/* Load module into memory and unzip if compressed */
 		image = xmalloc_open_zipped_read_close(m_filename, &image_size);
 		if (!image)
-			return EXIT_FAILURE;
+			return (-errno);
 	}
 
 	m_name = xstrdup(bb_basename(m_filename));
@@ -3832,8 +3832,10 @@ int FAST_FUNC bb_init_module_24(const ch
 				"\twhile this kernel is version %s",
 				flag_force_load ? "warning: " : "",
 				m_name, m_strversion, uts.release);
-			if (!flag_force_load)
+			if (!flag_force_load) {
+				exit_status = ESRCH;
 				goto out;
+			}
 		}
 	}
 #endif
