--- trunk/gcc/ipa-inline-analysis.c	2013/08/14 01:08:16	201718
+++ trunk/gcc/ipa-inline-analysis.c	2013/08/14 04:44:37	201719
@@ -337,7 +337,7 @@ add_clause (conditions conditions, struc
          and thus there is no point for looking for them.  */
       if (cc1->code == CHANGED || cc1->code == IS_NOT_CONSTANT)
 	continue;
-      for (c2 = c1 + 1; c2 <= NUM_CONDITIONS; c2++)
+      for (c2 = c1 + 1; c2 < NUM_CONDITIONS; c2++)
 	if (clause & (1 << c2))
 	  {
 	    condition *cc1 =
