diff -ruN ipkg-utils-1.7.orig/ipkg-build ipkg-utils-1.7/ipkg-build
--- ipkg-utils-1.7.orig/ipkg-build	2004-10-27 03:28:14.079066615 +1000
+++ ipkg-utils-1.7/ipkg-build	2004-10-27 03:29:12.045491010 +1000
@@ -243,7 +243,7 @@
 mkdir $tmp_dir
 
 echo $CONTROL > $tmp_dir/tarX
-( cd $pkg_dir && tar $ogargs -czf $tmp_dir/data.tar.gz . -X $tmp_dir/tarX )
+( cd $pkg_dir && tar $ogargs -X $tmp_dir/tarX -czf $tmp_dir/data.tar.gz . )
 ( cd $pkg_dir/$CONTROL && tar $ogargs -czf $tmp_dir/control.tar.gz . )
 rm $tmp_dir/tarX
 
diff -ruN ipkg-utils-1.7.orig/ipkg.py ipkg-utils-1.7/ipkg.py
--- ipkg-utils-1.7.orig/ipkg.py	2003-09-05 11:49:48.000000000 +0400
+++ ipkg-utils-1.7/ipkg.py	2009-08-16 18:27:48.000000000 +0400
@@ -93,9 +93,9 @@
             self.filename = os.path.basename(fn)
 	    ## sys.stderr.write("  extracting control.tar.gz from %s\n"% (fn,)) 
             if self.isdeb:
-                control = os.popen("ar p "+fn+" control.tar.gz | tar xfzO - '*control'","r")
+                control = os.popen("ar p "+fn+" control.tar.gz | tar --wildcards -xfzO - '*control'","r")
             else:
-                control = os.popen("tar xfzO "+fn+" '*control.tar.gz' | tar xfzO - '*control'","r")
+                control = os.popen("tar --wildcards -xzOf "+fn+" '*control.tar.gz' | tar --wildcards -xzOf - '*control'","r")
             line = control.readline()
             while 1:
                 if not line: break
@@ -122,7 +122,7 @@
             if self.isdeb:
                 data = os.popen("ar p "+fn+" data.tar.gz | tar tfz -","r")
             else:
-                data = os.popen("tar xfzO "+fn+" '*data.tar.gz' | tar tfz -","r")
+                data = os.popen("tar --wildcards -xzOf "+fn+" '*data.tar.gz' | tar -tzf -","r")
             while 1:
                 line = data.readline()
                 if not line: break
@@ -251,7 +251,7 @@
 	file.write(buf)
 
 	self._setup_scratch_area()
-	cmd = "cd %s ; tar cvfz %s/control.tar.gz control" % (self.meta_dir,
+	cmd = "cd %s ; tar -cvzf %s/control.tar.gz control" % (self.meta_dir,
 							      self.scratch_dir)
 
 	cmd_out, cmd_in, cmd_err = os.popen3(cmd)
@@ -266,7 +266,7 @@
 	bits = "control.tar.gz"
 
 	if self.file_list:
-		cmd = "cd %s ; tar cvfz %s/data.tar.gz" % (self.file_dir,
+		cmd = "cd %s ; tar -cvzf %s/data.tar.gz" % (self.file_dir,
 					   		   self.scratch_dir)
 
 		cmd_out, cmd_in, cmd_err = os.popen3(cmd)
@@ -281,7 +281,7 @@
 		bits = bits + " data.tar.gz"
 
 	file = "%s_%s_%s.ipk" % (self.package, self.version, self.architecture)
-	cmd = "cd %s ; tar cvfz %s/%s %s" % (self.scratch_dir,
+	cmd = "cd %s ; tar -cvzf %s/%s %s" % (self.scratch_dir,
 					     dirname,
 					     file,
 					     bits)
