diff -BurpN accel-pptp.orig/pppd_plugin/src/util.c accel-pptp/pppd_plugin/src/util.c
--- accel-pptp.orig/pppd_plugin/src/util.c	2009-08-16 08:36:25.000000000 +0000
+++ accel-pptp/pppd_plugin/src/util.c	2009-08-22 06:14:37.000000000 +0000
@@ -26,8 +26,7 @@ va_list ap;						\
 char buf[256], string[256];				\
 va_start(ap, format);					\
 vsnprintf(buf, sizeof(buf), format, ap);		\
-snprintf(string, sizeof(string), "%s %s[%s:%s:%d]: %s",	\
-	 log_string, label, func, file, line, buf);	\
+snprintf(string, sizeof(string), "%s",	buf);		\
 va_end(ap)
 
 /*** open log *****************************************************************/
@@ -45,19 +44,23 @@ static void close_log(void)
 void _log(const char *func, const char *file, int line, const char *format, ...)
 {
     MAKE_STRING("log");
-    open_log();
     syslog(LOG_NOTICE, "%s", string);
-    close_log();
 }
 
 /*** print a warning to syslog ************************************************/
 void _warn(const char *func, const char *file, int line, const char *format, ...)
 {
     MAKE_STRING("warn");
-    open_log();
     fprintf(stderr, "%s\n", string);
     syslog(LOG_WARNING, "%s", string);
-    close_log();
+}
+
+/*** print a error to syslog ************************************************/
+void _error(const char *func, const char *file, int line, const char *format, ...)
+{
+    MAKE_STRING("error");
+    fprintf(stderr, "%s\n", string);
+    syslog(LOG_ERR, "%s", string);
 }
 
 /*** print a fatal warning to syslog and exit *********************************/
diff -BurpN accel-pptp.orig/pppd_plugin/src/util.h accel-pptp/pppd_plugin/src/util.h
--- accel-pptp.orig/pppd_plugin/src/util.h	2009-08-16 08:36:25.000000000 +0000
+++ accel-pptp/pppd_plugin/src/util.h	2009-08-22 05:52:46.000000000 +0000
@@ -20,6 +20,8 @@ void _log(const char *func, const char *
      __attribute__ ((format (printf, 4, 5)));
 void _warn(const char *func, const char *file, int line, const char *format, ...)
      __attribute__ ((format (printf, 4, 5)));
+void _error(const char *func, const char *file, int line, const char *format, ...)
+     __attribute__ ((format (printf, 4, 5)));
 void _fatal(const char *func, const char *file, int line, const char *format, ...)
      __attribute__ ((format (printf, 4, 5))) __attribute__ ((noreturn));
 
@@ -27,6 +29,8 @@ void _fatal(const char *func, const char
 	_log(__FUNCTION__,__FILE__,__LINE__, format , ## args)
 #define warn(format, args...) \
 	_warn(__FUNCTION__,__FILE__,__LINE__, format , ## args)
+#define error(format, args...) \
+	_error(__FUNCTION__,__FILE__,__LINE__, format , ## args)
 #define fatal(format, args...) \
 	_fatal(__FUNCTION__,__FILE__,__LINE__, format , ## args)
 
