--- miniupnpd.orig/getifaddr.c	2008-10-15 10:16:28.000000000 +0000
+++ miniupnpd/getifaddr.c	2011-06-11 10:33:38.000000000 +0000
@@ -29,6 +29,8 @@ getifaddr(const char * ifname, char * bu
 	int ifrlen;
 	struct sockaddr_in * addr;
 	ifrlen = sizeof(ifr);
+	int ret = -1;
+
 	if(!ifname || ifname[0]=='\0')
 		return -1;
 	s = socket(PF_INET, SOCK_DGRAM, 0);
@@ -38,20 +40,28 @@ getifaddr(const char * ifname, char * bu
 		return -1;
 	}
 	strncpy(ifr.ifr_name, ifname, IFNAMSIZ);
+	if(ioctl(s, SIOCGIFFLAGS, &ifr) < 0)
+	{
+		syslog(LOG_DEBUG, "ioctl(s, SIOCGIFFLAGS, ...): %m");
+		goto err;
+	} else
+	if ((ifr.ifr_flags & IFF_UP) == 0)
+		goto err;
 	if(ioctl(s, SIOCGIFADDR, &ifr, &ifrlen) < 0)
 	{
 		syslog(LOG_ERR, "ioctl(s, SIOCGIFADDR, ...): %m");
-		close(s);
-		return -1;
+		goto err;
 	}
 	addr = (struct sockaddr_in *)&ifr.ifr_addr;
 	if(!inet_ntop(AF_INET, &addr->sin_addr, buf, len))
 	{
 		syslog(LOG_ERR, "inet_ntop(): %m");
-		close(s);
-		return -1;
+		goto err;
 	}
+	ret = 0;
+
+err:
 	close(s);
-	return 0;
+	return ret;
 }
 
