diff -urbBp -x'*.o' wpa_supplicant-0.6.8/wpa_supplicant/wpas_glue.c wpa_supplicant/wpa_supplicant/wpas_glue.c
--- wpa_supplicant-0.6.8/wpa_supplicant/wpas_glue.c	2009-02-15 21:00:00.000000000 +0300
+++ wpa_supplicant/wpa_supplicant/wpas_glue.c	2009-02-18 12:28:17.000000000 +0300
@@ -550,8 +550,10 @@ int wpa_supplicant_init_eapol(struct wpa
 	ctx->eapol_done_cb = wpa_supplicant_notify_eapol_done;
 	ctx->eapol_send = wpa_supplicant_eapol_send;
 	ctx->set_wep_key = wpa_eapol_set_wep_key;
+#ifndef CONFIG_NO_CONFIG_BLOBS
 	ctx->set_config_blob = wpa_supplicant_set_config_blob;
 	ctx->get_config_blob = wpa_supplicant_get_config_blob;
+#endif /* CONFIG_NO_CONFIG_BLOBS */
 	ctx->aborted_cached = wpa_supplicant_aborted_cached;
 #ifdef EAP_TLS_OPENSSL
 	ctx->opensc_engine_path = wpa_s->conf->opensc_engine_path;
