Port of pppd commit cab58617fd9d328029fffabc788020264b4fa91f
Subject: [PATCH] rp_pppoe: Copy acName and pppd_pppoe_service after option parsing

--- a/src/plugin.c	2009-05-05 23:44:56.000000000 +0400
+++ b/src/plugin.c	2010-03-25 22:10:58.000000000 +0300
@@ -109,12 +109,6 @@ PPPOEInitDevice(void)
 	fatal("Could not allocate memory for PPPoE session");
     }
     memset(conn, 0, sizeof(PPPoEConnection));
-    if (acName) {
-	SET_STRING(conn->acName, acName);
-    }
-    if (pppd_pppoe_service) {
-	SET_STRING(conn->serviceName, pppd_pppoe_service);
-    }
     SET_STRING(conn->ifName, devnam);
     conn->discoverySocket = -1;
     conn->sessionSocket = -1;
@@ -149,6 +143,12 @@ PPPOEConnectDevice(void)
 	return -1;
     }
 
+    if (acName) {
+	SET_STRING(conn->acName, acName);
+    }
+    if (pppd_pppoe_service) {
+	SET_STRING(conn->serviceName, pppd_pppoe_service);
+    }
     strlcpy(ppp_devnam, devnam, sizeof(ppp_devnam));
     if (existingSession) {
 	unsigned int mac[ETH_ALEN];
