prevent pppd prematory exit on pppoe plugin recoverable errors

diff -BurpN rp-pppoe-3.11.orig/src/plugin.c rp-pppoe-3.11/src/plugin.c
--- rp-pppoe-3.11.orig/src/plugin.c	2012-11-03 19:26:51.000000000 +0600
+++ rp-pppoe-3.11/src/plugin.c	2012-11-03 19:29:52.330719146 +0600
@@ -519,7 +519,9 @@ rp_fatal(char const *str)
 void
 sysErr(char const *str)
 {
-    rp_fatal(str);
+    char buf[1024];
+    sprintf(buf, "%.256s: %.256s", str, strerror(errno));
+    printErr(buf);
 }
 
 void pppoe_check_options(void)
