diff -urpN usb-modeswitch-1.2.5/usb_modeswitch.c usb-modeswitch/usb_modeswitch.c
--- usb-modeswitch-1.2.5/usb_modeswitch.c	2012-11-09 02:36:43.000000000 +0400
+++ usb-modeswitch/usb_modeswitch.c	2013-05-13 21:17:06.235973346 +0400
@@ -83,7 +83,7 @@ int devnum=-1, busnum=-1;
 int ret;
 
 char DetachStorageOnly=0, HuaweiMode=0, SierraMode=0, SonyMode=0, GCTMode=0, KobilMode=0;
-char SequansMode=0, MobileActionMode=0, CiscoMode=0, QisdaMode=0;
+char SequansMode=0, MobileActionMode=0, CiscoMode=0, QisdaMode=0, QuantaMode=0;
 char verbose=0, show_progress=1, ResetUSB=0, CheckSuccess=0, config_read=0;
 char NeedResponse=0, NoDriverLoading=0, InquireDevice=1, sysmode=0, mbim=0;
 
@@ -124,6 +124,7 @@ static struct option long_options[] = {
 	{"sierra-mode",			no_argument, 0, 'S'},
 	{"sony-mode",			no_argument, 0, 'O'},
 	{"qisda-mode",			no_argument, 0, 'B'},
+	{"quanta-mode",			no_argument, 0, 'E'},
 	{"kobil-mode",			no_argument, 0, 'T'},
 	{"gct-mode",			no_argument, 0, 'G'},
 	{"sequans-mode",		no_argument, 0, 'N'},
@@ -160,6 +161,7 @@ void readConfigFile(const char *configFi
 	ParseParamBool(configFilename, SierraMode);
 	ParseParamBool(configFilename, SonyMode);
 	ParseParamBool(configFilename, QisdaMode);
+	ParseParamBool(configFilename, QuantaMode);
 	ParseParamBool(configFilename, GCTMode);
 	ParseParamBool(configFilename, KobilMode);
 	ParseParamBool(configFilename, SequansMode);
@@ -218,6 +220,7 @@ void printConfig()
 	fprintf (output,"SierraMode=%i\n",			(int)SierraMode);
 	fprintf (output,"SonyMode=%i\n",			(int)SonyMode);
 	fprintf (output,"QisdaMode=%i\n",		(int)QisdaMode);
+	fprintf (output,"QuantaMode=%i\n",		(int)QuantaMode);
 	fprintf (output,"GCTMode=%i\n",			(int)GCTMode);
 	fprintf (output,"KobilMode=%i\n",		(int)KobilMode);
 	fprintf (output,"SequansMode=%i\n",		(int)SequansMode);
@@ -272,7 +275,7 @@ int readArguments(int argc, char **argv)
 
 	while (1)
 	{
-		c = getopt_long (argc, argv, "hejWQDndHSOBGTNALRItv:p:V:P:C:m:M:2:3:w:r:c:i:u:a:s:f:b:g:",
+		c = getopt_long (argc, argv, "hejWQDndHSOBEGTNALRItv:p:V:P:C:m:M:2:3:w:r:c:i:u:a:s:f:b:g:",
 						long_options, &option_index);
 
 		/* Detect the end of the options. */
@@ -299,6 +302,7 @@ int readArguments(int argc, char **argv)
 			case 'S': SierraMode = 1; break;
 			case 'O': SonyMode = 1; break;
 			case 'B': QisdaMode = 1; break;
+			case 'E': QuantaMode = 1; break;
 			case 'G': GCTMode = 1; break;
 			case 'T': KobilMode = 1; break;
 			case 'N': SequansMode = 1; break;
@@ -550,7 +554,7 @@ int main(int argc, char **argv)
 
 	/* Some scenarios are exclusive, so check for unwanted combinations */
  	specialMode = DetachStorageOnly + HuaweiMode + SierraMode + SonyMode + QisdaMode + KobilMode
-		+ SequansMode + MobileActionMode + CiscoMode;
+		+ SequansMode + MobileActionMode + CiscoMode + QuantaMode;
 	if ( specialMode > 1 ) {
 		SHOW_PROGRESS(output,"Invalid mode combination. Check your configuration. Aborting.\n\n");
 		exit(1);
@@ -596,6 +600,9 @@ int main(int argc, char **argv)
 		detachDriver();
 		switchKobilMode();
 	}
+	if(QuantaMode) {
+		switchQuantaMode();
+	}
 	if (SequansMode) {
 		switchSequansMode();
 	}
@@ -1072,6 +1079,11 @@ int switchQisdaMode () {
 	return 1;
 }
 
+void switchQuantaMode() {
+	SHOW_PROGRESS(output,"Sending Quanta control message ...\n");
+	usb_control_msg(devh, USB_TYPE_VENDOR | USB_RECIP_DEVICE | USB_ENDPOINT_IN, 0xff, 0, 0, buffer, 0, 1000);
+	SHOW_PROGRESS(output," OK, Quanta control message sent\n");
+}
 
 int switchSonyMode ()
 {
@@ -1925,6 +1937,7 @@ void printHelp()
 	" -T, --kobil-mode              apply a special procedure\n"
 	" -L, --cisco-mode              apply a special procedure\n"
 	" -B, --qisda-mode              apply a special procedure\n"
+	" -E, --quanta-mode             apply a special procedure\n"
 	" -R, --reset-usb               reset the device after all other actions\n"
 	" -Q, --quiet                   don't show progress or error messages\n"
 	" -W, --verbose                 print all settings and debug output\n"
diff -urpN usb-modeswitch-1.2.5/usb_modeswitch.h usb-modeswitch/usb_modeswitch.h
--- usb-modeswitch-1.2.5/usb_modeswitch.h	2012-11-09 02:36:57.000000000 +0400
+++ usb-modeswitch/usb_modeswitch.h	2013-05-13 21:16:29.046605496 +0400
@@ -36,6 +36,7 @@ void switchSierraMode();
 void switchGCTMode();
 int switchKobilMode();
 int switchQisdaMode();
+void switchQuantaMode();
 int switchSequansMode();
 int switchActionMode();
 int switchSonyMode();
