diff -BurpN -x'*.o' -x'*.orig' tools/upnp/Makefile gateway/upnp/Makefile
--- tools/upnp/Makefile	1970-01-01 00:00:00.000000000 +0000
+++ gateway/upnp/Makefile	2009-09-26 07:59:55.000000000 +0000
@@ -0,0 +1,34 @@
+#
+# Linux upnp Makefile
+#
+# Copyright 2004, Broadcom Corporation
+# All Rights Reserved.
+# 
+# THIS SOFTWARE IS OFFERED "AS IS", AND BROADCOM GRANTS NO WARRANTIES OF ANY
+# KIND, EXPRESS OR IMPLIED, BY STATUTE, COMMUNICATION OR OTHERWISE. BROADCOM
+# SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
+# FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
+#
+# $Id$
+#
+
+UPNP	:= igd/linux
+CFLAGS	+= -g -O2 $(EXTRACFLAGS)
+#DEBUG	:= 1
+export CFLAGS
+
+all: upnp
+
+install: all
+	install -d $(INSTALLDIR)/usr/sbin
+	install -m 755 $(UPNP)/upnp $(INSTALLDIR)/usr/sbin
+	$(STRIP) $(INSTALLDIR)/usr/sbin/upnp
+
+clean:
+	$(MAKE) -C $(UPNP) clean
+	rm -f $(UPNP)/upnp
+
+upnp: FORCE
+	$(MAKE) -C $(UPNP) DEBUG=$(DEBUG) SRCBASE=$(SRCBASE) TOP=$(TOP) TARGET_PREFIX=$(CROSS_COMPILE)
+
+FORCE:
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/igd.c gateway/upnp/igd/igd.c
--- tools/upnp/igd/igd.c	2004-11-19 13:34:33.000000000 +0000
+++ gateway/upnp/igd/igd.c	2009-09-25 19:07:17.000000000 +0000
@@ -18,6 +18,24 @@
 
 #include <signal.h>
 
+#if defined(linux)
+
+/* Allow some time for the page to reload before killing ourselves */
+static int
+kill_after(pid_t pid, int sig, unsigned int after)
+{
+        if (fork() == 0) {
+                sleep(after);
+                return kill(pid, sig);
+        }
+        return 0;
+}
+#define sys_restart() kill_after(1, SIGHUP, 3)
+#define sys_reboot() kill_after(1, SIGTERM, 3)
+
+#endif /* linux */
+
+
 extern DeviceTemplate LANDeviceTemplate;
 extern DeviceTemplate WANDeviceTemplate;
 
@@ -44,7 +62,9 @@ Error IGDErrors[] = {
 int IGDevice_Init(PDevice igdev, device_state_t state, va_list ap)
 {
     char *wan_ifname = &g_wandevs[0];
+#if INCLUDE_LANDEVICE
     char *lan_ifname = &g_landevs[0];
+#endif
     PDevice subdev;
 
     switch (state) {
@@ -93,7 +113,7 @@ int WANDevice_Init(PDevice pdev, device_
 	if (ifname) {
 	    pdata = (PWANDevicePrivateData) malloc(sizeof(WANDevicePrivateData));
 	    if (pdata) {
-		strcpy(pdata->ifname, nvram_safe_get("wan_ifname"));
+		strcpy(pdata->ifname, ifname);
 		pdev->opaque = (void *) pdata;
 	    }
 	}
@@ -111,7 +131,7 @@ int WANDevice_Init(PDevice pdev, device_
 
 int LANDevice_Init(PDevice pdev, device_state_t state, va_list ap)
 {
-    PWANDevicePrivateData pdata;
+    PLANDevicePrivateData pdata;
     char *ifname = NULL;
 
     switch (state) {
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/igd.h gateway/upnp/igd/igd.h
--- tools/upnp/igd/igd.h	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/igd.h	2009-09-25 19:07:17.000000000 +0000
@@ -68,9 +68,9 @@ typedef struct _WANCommonPrivateData {
 
 } WANCommonPrivateData, *PWANCommonPrivateData;
 
-extern void osl_igd_disable(char *ifname);
-extern void osl_igd_enable(char *ifname);
+extern void bump_generation();
 extern char *igd_pri_wan_var(char *prefix, int len, char *var);
+extern void req_nvram_commit(int force);
 
 #define SOAP_CONNECTIONNOTCONFIGURED	706 
 #define SOAP_DISCONNECTINPROGRESS	707 
@@ -81,21 +81,4 @@ extern char *igd_pri_wan_var(char *prefi
 #define SOAP_CONFLICTINMAPPINGENTRY	718 
 #define SOAP_ONLYPERMANENTLEASESSUPPORTED	725
 
-#if defined(linux)
-
-/* Allow some time for the page to reload before killing ourselves */
-static int
-kill_after(pid_t pid, int sig, unsigned int after)
-{
-	if (fork() == 0) {
-		sleep(after);
-		return kill(pid, sig);
-	}
-	return 0;
-}
-#define sys_restart() kill_after(1, SIGHUP, 3)
-#define sys_reboot() kill_after(1, SIGTERM, 3)
-
-#endif /* linux */
-
 #endif /* _igd_h_ */
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/igd_desc.c gateway/upnp/igd/igd_desc.c
--- tools/upnp/igd/igd_desc.c	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/igd_desc.c	2009-09-25 19:07:17.000000000 +0000
@@ -23,13 +23,11 @@
 
 #define INCLUDE_LAYER3
 #define INCLUDE_IPCONNECTION
-#define INCLUDE_PPPCONNECTION
-
-
+/* #define INCLUDE_PPPCONNECTION -- really broken */
 
 extern int WANDevice_Init(PDevice pdev, device_state_t state, va_list ap);
 extern int LANDevice_Init(PDevice pdev, device_state_t state, va_list ap);
-extern int IGDevice_Init(PDevice igdev, device_state_t state);
+extern int IGDevice_Init(PDevice igdev, device_state_t state, va_list ap);
 
 
 /* Global structure for storing the state table for this device */
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/ipt.c gateway/upnp/igd/ipt.c
--- tools/upnp/igd/ipt.c	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/ipt.c	2009-09-25 19:07:17.000000000 +0000
@@ -126,7 +126,6 @@ int AddPortMapping( UFILE *uclient, PSer
     bool found_match;
     netconf_nat_t e;
     mapping_t mapping;
-    char *wan_ip;
 
     // bypass port mapping when NAT is disabled
     if (nvram_invmatch("wan_nat_x", "1")) return 0;
@@ -137,15 +136,9 @@ int AddPortMapping( UFILE *uclient, PSer
 	    continue;
 	} 
 
-	if (nvram_invmatch("wan_ipaddr_t", ""))
-	{
-	     wan_ip = nvram_safe_get("wan_ipaddr_t");
-	}
-	else wan_ip = ac->params[0].value;
-    
 	parse_status = (int) parse_dnat(&e, 
 					ac->params[2].value, /* NewProtocol */
-					wan_ip, //ac->params[0].value, /* NewRemoteHost */
+					ac->params[0].value, /* NewRemoteHost */
 					ac->params[1].value, NULL, /* NewExternalPort */
 					ac->params[4].value, /* NewInternalClient */
 					ac->params[3].value, NULL /* NewInternalPort */
@@ -312,8 +305,6 @@ int GetSpecificPortMappingEntry(UFILE *u
     static char Enabled[2];
     static char Description[60];
 
-printf("Get Port: [%s]\n", ac->params[0].value);
-
     parse_status = (int) parse_dnat(&e, 
 			       ac->params[2].value,	/* NewProtocol */
 			       ac->params[0].value,	/* NewRemoteHost */
@@ -400,18 +391,10 @@ netconf_nat_t *parse_dnat(netconf_nat_t 
     entry->match.src.ports[1] = htons(0xffff);
 
     // parse the external ip address
-    // 2004/07/12 by Joey, change to src ip match
     if (strlen(RemoteHost)) 
     {
-	inet_aton("255.255.255.255", &entry->match.dst.netmask);
-	inet_aton(RemoteHost , &entry->match.dst.ipaddr);
-    }
-    else if (nvram_invmatch("wan_ipaddr_t", ""))
-    {
-    	// set the destination ip address
-    	// 2004/07/12, by Joey, set dst ip 
-	inet_aton("255.255.255.255", &entry->match.dst.netmask);
-	inet_aton(nvram_safe_get("wan_ipaddr_t"), &entry->match.dst.ipaddr);
+	inet_aton("255.255.255.255", &entry->match.src.netmask);
+	inet_aton(RemoteHost, &entry->match.src.ipaddr);
     }
 
     // parse the external ports
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/lanhostconfig.c gateway/upnp/igd/lanhostconfig.c
--- tools/upnp/igd/lanhostconfig.c	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/lanhostconfig.c	2009-09-25 19:07:17.000000000 +0000
@@ -52,7 +52,7 @@ int LANHostConfig_SetDomainName(UFILE *u
     uint success = TRUE; /* assume no error will occur */
 
     nvram_set("lan_domain", ac->params[0].value);
-    nvram_commit();
+    req_nvram_commit(0);
     // restart ?? 
     
     return success;
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/linux/Makefile gateway/upnp/igd/linux/Makefile
--- tools/upnp/igd/linux/Makefile	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/linux/Makefile	2009-09-26 08:01:24.000000000 +0000
@@ -32,7 +32,7 @@ COMPILER = $(TARGET_PREFIX)g++
 STRIP = $(TARGET_PREFIX)strip
 SIZE = $(TARGET_PREFIX)size
 
-VPATH.h = .:..:../../include:$(SRCBASE)/include:$(SRCBASE)/router/shared
+VPATH.h = .:..:../../include:$(SRCBASE)/include:$(TOP)/shared
 
 vpath %.c .:..:$(SRCBASE)/shared/netconf
 vpath %.h $(VPATH.h)
@@ -65,8 +65,8 @@ SOURCES.OBJ := $(SOURCES)
 SOURCES.OBJ := $(patsubst %.c,$(OBJDIR)/%.o,$(SOURCES.OBJ))
 
 TARGET = upnp
-LIBS = -L../../upnp/linux -L$(SRCBASE)/router/netconf \
-	-L$(SRCBASE)/router/nvram -L$(SRCBASE)/router/shared -lupnp -lnetconf -lnvram -lshared
+LIBS = -L../../upnp/linux -L$(TOP)/netconf \
+	-L$(TOP)/nvram -L$(TOP)/shared -lupnp -lnetconf -lnvram -lshared
 
 all : $(OBJDIR) $(TARGET)
 
@@ -89,20 +89,20 @@ ifneq ($(BUILD_LIBS),)
 
 $(TARGET) :: libnetconf.so
 
-$(TARGET) :: $(SRCBASE)/router/iptables/libiptables.a 
+$(TARGET) :: $(TOP)/iptables/libiptables.a 
 
-$(TARGET) :: $(SRCBASE)/router/nvram/libnvram.a 
+$(TARGET) :: $(TOP)/nvram/libnvram.a 
 
-$(SRCBASE)/router/iptables/libiptables.a : FORCE
-	$(MAKE) -C $(SRCBASE)/router/iptables PLATFORM=x86 CC=$(CC) LD=$(LD) SRCBASE=../.. TOP=..
-$(SRCBASE)/router/nvram/libnvram.a : FORCE
-	$(MAKE) -C $(SRCBASE)/router/nvram PLATFORM=x86 CC=$(CC) LD=$(LD) SRCBASE=../.. TOP=..
+$(TOP)/iptables/libiptables.a : FORCE
+	$(MAKE) -C $(TOP)/iptables PLATFORM=x86 CC=$(CC) LD=$(LD) SRCBASE=../.. TOP=..
+$(TOP)/nvram/libnvram.a : FORCE
+	$(MAKE) -C $(TOP)/nvram PLATFORM=x86 CC=$(CC) LD=$(LD) SRCBASE=../.. TOP=..
 
-$(SRCBASE)/router/netconf/libnetconf.so : $(SRCBASE)/router/iptables/libiptables.a FORCE
-	$(MAKE) -C $(SRCBASE)/router/netconf DEBUG=1 PLATFORM=x86 CC=$(CC) LD=$(LD) SRCBASE=../.. TOP=..
+$(TOP)/netconf/libnetconf.so : $(TOP)/iptables/libiptables.a FORCE
+	$(MAKE) -C $(TOP)/netconf DEBUG=1 PLATFORM=x86 CC=$(CC) LD=$(LD) SRCBASE=../.. TOP=..
 
-libnetconf.so : $(SRCBASE)/router/netconf/libnetconf.so
-	#	cp $(SRCBASE)/router/netconf/libnetconf.so /router/usr/lib/
+libnetconf.so : $(TOP)/netconf/libnetconf.so
+	#	cp $(TOP)/netconf/libnetconf.so /router/usr/lib/
 
 endif
 
@@ -122,9 +122,9 @@ clean: 
 	rm -f upnp
 	$(MAKE) -C ../../upnp/linux clean
 ifneq ($(BUILD_LIBS),)
-	$(MAKE) -C $(SRCBASE)/router/netconf clean
-	$(MAKE) -C $(SRCBASE)/router/iptables clean
-	$(MAKE) -C $(SRCBASE)/router/nvram clean
+	$(MAKE) -C $(TOP)/netconf clean
+	$(MAKE) -C $(TOP)/iptables clean
+	$(MAKE) -C $(TOP)/nvram clean
 endif
 
 fptest: fptest.o
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/linux/linux_igd.c gateway/upnp/igd/linux/linux_igd.c
--- tools/upnp/igd/linux/linux_igd.c	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/linux/linux_igd.c	2009-09-26 07:31:05.000000000 +0000
@@ -169,13 +169,13 @@ uint osl_max_bitrates(char *devname, ulo
 		speed = 0;
 	    }
 	} else {
-	    UPNP_ERROR(("ioctl(SIOCETHTOOL) failed in " __FUNCTION__));
+	    UPNP_ERROR(("ioctl(SIOCETHTOOL) failed in %s", __FUNCTION__));
 	}
 
 	/* close the control socket */
 	close(fd);
     } else {
-	UPNP_ERROR(("cannot open socket in " __FUNCTION__));
+	UPNP_ERROR(("cannot open socket in %s", __FUNCTION__));
     }
 
     *rx = *tx = speed;
@@ -378,32 +378,14 @@ bool osl_wan_isup(char *devname)
 {
     struct in_addr inaddr = {0};
     bool status = FALSE;
-	char tmp[100];
 
-    if (strcasecmp(nvram_safe_get(igd_pri_wan_var(tmp, sizeof(tmp), "proto")), "disabled") != 0) {
-#ifdef REMOVE
-	if (osl_ifaddr(nvram_safe_get(igd_pri_wan_var(tmp, sizeof(tmp), "ifname")), &inaddr)) 
+	if (osl_ifaddr(devname, &inaddr)) 
 	{
 	    if (inaddr.s_addr != 0) {
 		status = TRUE;
 	    }
 	}
-#else
-	char *wan_if;
 
-	if (nvram_match(tmp, "pppoe") || nvram_match(tmp, "pptp"))
-		wan_if = nvram_safe_get(igd_pri_wan_var(tmp, sizeof(tmp), "pppoe_ifname"));
-	else
-		wan_if = nvram_safe_get(igd_pri_wan_var(tmp, sizeof(tmp), "ifname"));
-
-	if (osl_ifaddr(wan_if, &inaddr)) 
-	{
-		if (inaddr.s_addr != 0) {
-			status = TRUE;
-	    	}		
-	}
-#endif 
-    }
     return status;
 }
 
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/linux/linux_main.c gateway/upnp/igd/linux/linux_main.c
--- tools/upnp/igd/linux/linux_main.c	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/linux/linux_main.c	2009-09-25 19:07:17.000000000 +0000
@@ -43,6 +43,56 @@ extern struct net_connection *net_connec
 extern int global_exit_now;
 extern struct iface *global_lans;
 
+#include <syslog.h>
+#include <bcmnvram.h>
+
+static timer_t t_nvram1;
+static volatile int nvcommit_state = 0;
+
+
+#ifdef DEBUG
+static void logmessage(char *fmt, ...)
+{
+    va_list args;
+    char buf[512];
+
+    va_start(args, fmt);
+    vsnprintf(buf, sizeof(buf), fmt, args);
+    va_end(args);
+
+    openlog("UPNP", 0, 0);
+    syslog(0, buf);
+    closelog();
+}
+#else
+ #define logmessage(...)
+#endif
+
+static inline void do_nvram_commit()
+{
+    logmessage("do nvram_commit");
+
+    nvcommit_state = 2;
+    nvram_commit();
+    nvcommit_state = 0;
+}
+
+void req_nvram_commit(int force)
+{
+    if (nvcommit_state != 2) {
+	if (force)
+	    do_nvram_commit();
+	else
+	    nvcommit_state = 1;
+    }
+}
+
+static void nvram_commit_helper(timer_t t, void *arg)
+{
+    if (nvcommit_state == 1)
+	do_nvram_commit();
+}
+
 static void
 reap(int sig)
 {
@@ -56,12 +106,12 @@ reap(int sig)
 int main(int argc, char *argv[])
 {
     extern char g_wandevs[];
-    extern char g_landevs[];
     extern DeviceTemplate IGDeviceTemplate;
     char **argp = &argv[1];
     char *wanif = NULL;
     char *lanif = NULL;
     int daemonize = 0;
+    struct itimerspec timer;
 
     while (argp < &argv[argc]) {
 	if (strcasecmp(*argp, "-L") == 0) {
@@ -96,11 +146,19 @@ int main(int argc, char *argv[])
 	   a SIGTERM after sending SIGUSR1 to the dhcp process (to
 	   renew a lease).  Ignore SIGTERM to avoid being killed when
 	   this happens.  */
-	//	signal(SIGTERM, SIG_IGN);
+	signal(SIGTERM, SIG_IGN);
 	signal(SIGUSR1, SIG_IGN);
 
+	memset(&timer, 0, sizeof(timer));
+        timer.it_interval.tv_sec = 60;
+        timer.it_value.tv_sec = 60;
+        t_nvram1 = enqueue_event(&timer, (event_callback_t) nvram_commit_helper, NULL);
+
 	fprintf(stderr, "calling upnp_main\n");
 	upnp_main(&IGDeviceTemplate, lanif);
+
+	timer_delete(t_nvram1);
+
     }
     
     return 0;
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/mapmgr.c gateway/upnp/igd/mapmgr.c
--- tools/upnp/igd/mapmgr.c	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/mapmgr.c	2009-09-26 08:04:39.000000000 +0000
@@ -95,7 +95,7 @@ static bool mapmgr_add_map(map_set_t *ps
     for (i = 0; i < NFDBITS; i++) {
 	if (!FD_ISSET(i, &ports.map) && !FD_ISSET(i, &ranges.map) ) {
 	    foundit = set_forward_port(i, (netconf_nat_t*)m);
-	    nvram_commit();
+	    req_nvram_commit(0);
 	    FD_SET(i, &pset->map);
 	    pset->count++;
 	    break;
@@ -118,7 +118,7 @@ static bool mapmgr_delete_map(map_set_t 
 		FD_CLR(i, &pset->map);
 		pset->count--;
 		foundit = del_forward_port(i);
-		nvram_commit();
+		req_nvram_commit(0);
 		break;
 	    }
 	}
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/wancommon.c gateway/upnp/igd/wancommon.c
--- tools/upnp/igd/wancommon.c	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/wancommon.c	2009-09-25 19:07:17.000000000 +0000
@@ -129,10 +129,6 @@ ServiceTemplate Template_WANCommonInterf
     "urn:upnp-org:serviceId:WANCommonIFC"
 };
 
-static const char *tzs[] = {
-    "GMT-12", "GMT-11", "GMT-10", "GMT-9", "GMT-8", "GMT-7", "GMT-6", "GMT-5", "GMT-4", "GMT-3", "GMT-2", "GMT-1", "GMT", 
-    "GMT1", "GMT2", "GMT3", "GMT4", "GMT5", "GMT6", "GMT7", "GMT8", "GMT9", "GMT10", "GMT11", "GMT12", "GMT-13"  };
-
 static int WANCommonInterfaceConfig_Init(PService psvc, service_state_t state)
 {
     PWANDevicePrivateData pdevdata;
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/wanipc.c gateway/upnp/igd/wanipc.c
--- tools/upnp/igd/wanipc.c	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/wanipc.c	2009-09-26 07:44:25.000000000 +0000
@@ -474,7 +474,7 @@ static int ForceTermination(UFILE *uclie
 {
     uint success = TRUE; /* assume no error will occur */
     PWANIPConnectionData pdata = (PWANIPConnectionData) psvc->opaque;
-    PWANDevicePrivateData pdevdata = (PWANDevicePrivateData) psvc->device->parent->opaque;
+    //PWANDevicePrivateData pdevdata = (PWANDevicePrivateData) psvc->device->parent->opaque;
     char wanproto[100];
 
     /* Our ConnectionType is always IP_Routed, so I don't need to check that here. */
@@ -490,7 +490,7 @@ static int ForceTermination(UFILE *uclie
 	/* Save the wan_proto into NVRAM to restore when the igd is enabled */
 	nvram_set("upnp_wan_proto", nvram_safe_get(wanproto));
 	nvram_set(wanproto, "disabled");
-	nvram_commit();
+	req_nvram_commit(1);
 	igd_restart(3);
     }
     
@@ -513,7 +513,7 @@ static int RequestConnection(UFILE *ucli
 	    nvram_set(wanproto, "dhcp");	/* pick dhcp as defualt */
 	else
 	    nvram_set(wanproto, str);
-	nvram_commit();
+	req_nvram_commit(1);
 	igd_restart(3);
 	pdata->connection_status = IP_CONNECTING;
 	pdata->connection_timeout = CONNECTION_TIMEOUT;
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/igd/wanppp.c gateway/upnp/igd/wanppp.c
--- tools/upnp/igd/wanppp.c	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/igd/wanppp.c	2009-09-25 19:07:17.000000000 +0000
@@ -78,7 +78,7 @@ int WANPPPConnection_ConfigureConnection
     } else {
 	nvram_set("wan_pppoe_username", ac->params[0].value);
 	nvram_set("wan_pppoe_passwd", ac->params[1].value);
-	nvram_commit();
+	req_nvram_commit(0);
     }
 
     return success;
@@ -91,7 +91,7 @@ int WANPPPConnection_SetIdleDisconnectTi
     uint success = TRUE; /* assume no error will occur */
 
     nvram_set("wan_pppoe_idletime", ac->params[0].value);
-    nvram_commit();
+    req_nvram_commit(0);
 
     return success;
 }
@@ -107,7 +107,7 @@ int WANPPPConnection_RequestConnection(U
 
     printf("PPP_CONNECTING %s\n", tmp);
 
-    nvram_commit();
+    req_nvram_commit(0);
     pppoe_up();
     
     return TRUE;
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/include/upnp.h gateway/upnp/include/upnp.h
--- tools/upnp/include/upnp.h	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/include/upnp.h	2009-09-26 17:40:30.000000000 +0000
@@ -24,7 +24,7 @@
 // #define  HTTP_PORT 80
 
 // number of seconds to wait before refershing device advertisements.
-#define SSDP_REFRESH   30
+#define SSDP_REFRESH   60
 #define UPNP_REFRESH   ((SSDP_REFRESH * 2)/3)
 #define NOTIFY_RECEIPT_TIMEOUT 30
 #define HTTP_REQUEST_TIMEOUT   15
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/include/upnp_dbg.h gateway/upnp/include/upnp_dbg.h
--- tools/upnp/include/upnp_dbg.h	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/include/upnp_dbg.h	2009-09-26 07:36:39.000000000 +0000
@@ -16,6 +16,8 @@
 #ifndef _upnp_dbg_
 #define _upnp_dbg_
 
+#if !defined(DEBUG)
+
 #define	UPNP_MAPPING(args)
 #define	UPNP_ERROR(args)
 #define	UPNP_TRACE(args) 
@@ -45,7 +47,39 @@
 #define UPNP_TRACE_ACTION_ON()	0
 #define UPNP_DUMP_ACTION_ON()	0
 
+#else	// defined(DEBUG)
+
+#define dbgprintf(args...)			fprintf(stderr, args)
+
+#define	UPNP_MAPPING(args...)			dbgprintf args
+#define	UPNP_ERROR(args...)			dbgprintf args
+#define	UPNP_TRACE(args...) 			dbgprintf args
+#define	UPNP_PRHDRS(i, s, h, p, n, l)
+#define	UPNP_PRPKT(m, b, n)
+#define	UPNP_INFORM(args...)			dbgprintf args
+#define UPNP_TRACE_ACTION(svc, ac)
+#define	UPNP_HTTP(args)
+#define	UPNP_SOCKET(args...)			dbgprintf args
+
+#define UPNP_MAPPING_ON()	0
+#define UPNP_RESPONSE_ON()	0
+#define UPNP_ERROR_ON()		0
+#define UPNP_PRHDRS_ON()	0
+#define UPNP_PRPKT_ON()		0
+#define UPNP_INFORM_ON()	0
+#define UPNP_PRINTRX_ON()	0
+#define UPNP_PRINTTX_ON()	0
+#define UPNP_HTTP_TRACE_ON()	0
+#define UPNP_HTTP_HDRS_ON()	0
+
+#define	UPNP_PREVENT(args)
+#define	UPNP_SUBSCRIBE(args)
 
+#define UPNP_ACTION(psvc, ac, args, nargs)
+#define UPNP_RESPONSE(ns, ac, args, nargs)
+#define UPNP_TRACE_ACTION_ON()	0
+#define UPNP_DUMP_ACTION_ON()	0
 
+#endif	// defined(DEBUG)
 
 #endif /* _upnp_dbg_ */
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/include/upnp_osl.h gateway/upnp/include/upnp_osl.h
--- tools/upnp/include/upnp_osl.h	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/include/upnp_osl.h	2009-09-25 19:07:17.000000000 +0000
@@ -23,6 +23,7 @@
 
 /* forward declaration - defined in upnp.h */
 struct _if_stats;
+struct iface;
 
 
 typedef enum { OSL_LINK_DOWN = 0, OSL_LINK_UP = 1 } osl_link_t;
@@ -40,5 +41,6 @@ extern void osl_igd_enable(char *devname
 extern bool osl_wan_isup(char *devname);
 extern bool osl_lan_isup(char *devname);
 extern bool osl_set_macaddr(char *devname, char spoofed[]);
+extern int osl_join_multicast(struct iface *pif, int fd, ulong ipaddr, ushort port);
 
 #endif	/* _upnp_osl_h_ */
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/upnp/device.c gateway/upnp/upnp/device.c
--- tools/upnp/upnp/device.c	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/upnp/device.c	2010-03-30 19:44:09.000000000 +0000
@@ -192,9 +192,17 @@ void device_xml(PDevice pdev, UFILE *up)
 
     // generate XML for any subdevices in this device.
     device_devicelist(pdev, up);
-	
+
     if (ISROOT(pdev)) {
-	uprintf(up, "<presentationURL>http://%s</presentationURL>\r\n", nvram_safe_get("lan_ipaddr"));
+	char buffer[32];
+	char *url = nvram_safe_get("lan_ipaddr");
+	int port = atoi(nvram_safe_get("http_lanport"));
+
+	if (*url && port && port != 80) {
+	    sprintf(buffer, "%s:%d", url, port);
+	    url = buffer;
+	}
+	uprintf(up, "<presentationURL>http://%s</presentationURL>\r\n", url);
     }
 
     uprintf(up, "</device>\r\n");
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/upnp/gena.c gateway/upnp/upnp/gena.c
--- tools/upnp/upnp/gena.c	2004-11-11 10:29:40.000000000 +0000
+++ gateway/upnp/upnp/gena.c	2009-09-25 19:07:17.000000000 +0000
@@ -424,7 +424,7 @@ static struct net_connection *make_gena_
 
     optval.l_onoff = 1;
     optval.l_linger = 0;
-    setsockopt(sr, SOL_SOCKET, SO_LINGER, &optval, sizeof (optval));
+    setsockopt(fd, SOL_SOCKET, SO_LINGER, &optval, sizeof (optval));
 
     // if we have not yet resolved the address of this host, do so now.
     // cache the result in the subscription.
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/upnp/linux/Makefile gateway/upnp/upnp/linux/Makefile
--- tools/upnp/upnp/linux/Makefile	2004-11-11 10:29:39.000000000 +0000
+++ gateway/upnp/upnp/linux/Makefile	2009-09-26 08:01:53.000000000 +0000
@@ -25,9 +25,9 @@ STRIP = $(TARGET_PREFIX)strip
 SIZE = $(TARGET_PREFIX)size
 RANLIB = $(TARGET_PREFIX)ranlib
 
-VPATH.h=.:..:../../include:$(SRCBASE)/include:$(SRCBASE)/router/shared
+VPATH.h=.:..:../../include:$(SRCBASE)/include:$(TOP)/shared
 
-vpath %.c .:..:$(SRCBASE)/router/shared:$(SRCBASE)/shared/netconf
+vpath %.c .:..:$(TOP)/shared:$(SRCBASE)/shared/netconf
 vpath %.h $(VPATH.h)
 
 ifeq ($(DEBUG),1)
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/upnp/service.c gateway/upnp/upnp/service.c
--- tools/upnp/upnp/service.c	2004-11-11 10:29:40.000000000 +0000
+++ gateway/upnp/upnp/service.c	2009-09-26 07:39:57.000000000 +0000
@@ -17,6 +17,9 @@
 extern char *getsvcval(PService psvc, int i);
 extern void soap_response(UFILE*, const char *, const char *, const char *, pvar_entry_t, int);
 
+const char *type2str(vartype_t type);
+
+
 PService init_service(PServiceTemplate svctmpl, PDevice pdev)
 {
     PService psvc;
@@ -87,12 +90,11 @@ void mark_changed(PService psvc, int var
  */
 void service_xml(PService psvc, UFILE *up)
 {
-    const char *type2str(vartype_t type);
     PFSVCXML func;
     PVarTemplate pv;
     PAction *ac;
     PParam pa;
-    char *tstr;
+    const char *tstr;
 
     uprintf(up, 
 	    "<?xml version=\"1.0\"?>\r\n"
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/upnp/ssdp.c gateway/upnp/upnp/ssdp.c
--- tools/upnp/upnp/ssdp.c	2004-11-11 10:29:40.000000000 +0000
+++ gateway/upnp/upnp/ssdp.c	2009-09-26 07:28:44.000000000 +0000
@@ -87,7 +87,7 @@ void ssdp_receive(caction_t flag, struct
     struct sockaddr_in srcaddr;
     int nbytes, addrlen;
     
-    UPNP_TRACE((__FUNCTION__ "\n"));
+    UPNP_TRACE(("%s\n", __FUNCTION__));
     if (flag == CONNECTION_RECV) {
 	addrlen = sizeof(srcaddr);
 	memset(&srcaddr, 0, addrlen);
@@ -113,7 +113,7 @@ static void process_msearch_all(char *st
 {
     PDevice pdev;
 
-    UPNP_TRACE((__FUNCTION__ "\n"));
+    UPNP_TRACE(("%s\n", __FUNCTION__));
     forall_devices(pdev) {
 	advertise_device(pdev, SSDP_REPLY, pif, srcaddr,  addrlen);
     }
@@ -136,7 +136,7 @@ static void process_msearch_rootdevice(c
     char tmpbuf[100];
     PDevice pdev;
     
-    UPNP_TRACE((__FUNCTION__ "\n"));
+    UPNP_TRACE(("%s\n", __FUNCTION__));
     if ((upkt = usopen(NULL, 0))) {
 	
 	for (pdev = root_devices; pdev; pdev = pdev->next) {
@@ -159,7 +159,7 @@ static void process_msearch_uuid(char *s
     UFILE *upkt;
     PDevice pdev;
 
-    UPNP_TRACE((__FUNCTION__ "\n"));
+    UPNP_TRACE(("%s\n", __FUNCTION__));
     if ((upkt = usopen(NULL, 0))) {
 	if ((pdev = find_dev_by_udn(st)) != NULL) {
 
@@ -182,7 +182,7 @@ static void process_msearch_devicetype(c
     PDevice pdev;
     char tmpbuf[100];
 
-    UPNP_TRACE((__FUNCTION__ "\n"));
+    UPNP_TRACE(("%s\n", __FUNCTION__));
     if ((upkt = usopen(NULL, 0))) {
 	forall_devices(pdev) {
 	    if (strcmp(pdev->template->type, st) == 0) {
@@ -211,7 +211,7 @@ static void process_msearch_service(char
     PService psvc;
     char tmpbuf[100];
     
-    UPNP_TRACE((__FUNCTION__ "\n"));
+    UPNP_TRACE(("%s\n", __FUNCTION__));
     p = strstr(st, ":service:");
     if (p) {
 	name = p+strlen(":service:");
@@ -258,7 +258,7 @@ static void process_msearch(char *msg, s
     char *man = NULL;  // the MAN: header
     long int mxval;
 
-    UPNP_TRACE((__FUNCTION__ "\n"));
+    UPNP_TRACE(("%s\n", __FUNCTION__));
     if ( (body = strstr(msg, "\r\n\r\n" )) != NULL )
 	body += 4;
     else if ( (body = strstr(msg, "\r\n" )) != NULL )
@@ -410,7 +410,7 @@ static void ssdp_packet(
     time_t now;
     char date[100];
 
-    UPNP_TRACE((__FUNCTION__ "\n"));
+    UPNP_TRACE(("%s\n", __FUNCTION__));
     switch (ptype) {
     case SSDP_REPLY:
 	uprintf(up, "HTTP/1.1 200 OK\r\n");
diff -BurpN -x'*.o' -x'*.orig' tools/upnp/upnp/upnp.c gateway/upnp/upnp/upnp.c
--- tools/upnp/upnp/upnp.c	2004-11-11 10:29:40.000000000 +0000
+++ gateway/upnp/upnp/upnp.c	2009-09-25 19:07:17.000000000 +0000
@@ -95,6 +95,9 @@ int upnp_main(PDeviceTemplate pdevtmpl, 
 
     signal(SIGINT, interrupt_handler);
     signal(SIGTERM, interrupt_handler);
+    
+    /* prevent socket errors to cause termination */
+    signal(SIGPIPE, SIG_IGN);
 
     memset(&timer, 0, sizeof(timer));
     timer.it_interval.tv_sec = 30;
