 
 Upgrade rp-pppoe plugin to 3.10
 
diff -ubBp ppp-2.4.5-pre/pppd/plugins/rp-pppoe/discovery.c ppp/pppd/plugins/rp-pppoe/discovery.c
--- ppp-2.4.5-pre/pppd/plugins/rp-pppoe/discovery.c	2008-11-26 13:02:29.000000000 +0300
+++ ppp/pppd/plugins/rp-pppoe/discovery.c	2009-02-09 17:20:11.000000000 +0300
@@ -22,6 +22,7 @@ static char const RCSID[] =
 #ifdef HAVE_SYS_TIME_H
 #include <sys/time.h>
 #endif
+#include <time.h>
 
 #ifdef HAVE_SYS_UIO_H
 #include <sys/uio.h>
diff -ubBp ppp-2.4.5-pre/pppd/plugins/rp-pppoe/Makefile.linux ppp/pppd/plugins/rp-pppoe/Makefile.linux
--- ppp-2.4.5-pre/pppd/plugins/rp-pppoe/Makefile.linux	2008-11-26 13:02:29.000000000 +0300
+++ ppp/pppd/plugins/rp-pppoe/Makefile.linux	2009-02-09 17:22:47.000000000 +0300
@@ -23,7 +23,7 @@ PPPDVERSION = $(shell awk -F '"' '/VERSI
 INSTALL	= install
 
 # Version is set ONLY IN THE MAKEFILE!  Don't delete this!
-RP_VERSION=3.8p
+RP_VERSION=3.10p
 
 COPTS=-O2 -g
 CFLAGS=$(COPTS) -I../../../include '-DRP_VERSION="$(RP_VERSION)"'
