Updates from Broadcom SDK 5.10.140.0

diff -urBp a/include/bcmdefs.h b/include/bcmdefs.h
--- a/include/bcmdefs.h	2010-01-29 17:37:48.000000000 +0300
+++ b/include/bcmdefs.h	2010-02-11 22:34:40.000000000 +0300
@@ -1,14 +1,15 @@
 /*
  * Misc system wide definitions
  *
- * Copyright (C) 2008, Broadcom Corporation
+ * Copyright (C) 2009, Broadcom Corporation
  * All Rights Reserved.
  * 
  * THIS SOFTWARE IS OFFERED "AS IS", AND BROADCOM GRANTS NO WARRANTIES OF ANY
  * KIND, EXPRESS OR IMPLIED, BY STATUTE, COMMUNICATION OR OTHERWISE. BROADCOM
  * SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
  * FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
- * $Id: bcmdefs.h,v 13.43.2.4.12.1 2008/11/03 23:30:41 Exp $
+ *
+ * $Id: bcmdefs.h,v 13.43.2.12 2009/03/12 21:32:38 Exp $
  */
 
 #ifndef	_bcmdefs_h_
@@ -158,12 +159,40 @@ typedef unsigned long dmaaddr_t;
 	} while (0)
 #endif /* BCMDMA64OSL */
 
+/* One physical DMA segment */
+typedef struct  {
+	dmaaddr_t addr;
+	uint32	  length;
+} hnddma_seg_t;
+
+#if defined(MACOSX)
+/* In MacOS, the OS API may return large number of segments. Setting this number lower
+ * will result in failure of dma map
+ */
+#define MAX_DMA_SEGS 8
+#else
+#define MAX_DMA_SEGS 4
+#endif
+
+
+typedef struct {
+	void *oshdmah; /* Opaque handle for OSL to store its information */
+	uint origsize; /* Size of the virtual packet */
+	uint nsegs;
+	hnddma_seg_t segs[MAX_DMA_SEGS];
+} hnddma_seg_map_t;
+
 /* packet headroom necessary to accomodate the largest header in the system, (i.e TXOFF).
  * By doing, we avoid the need  to allocate an extra buffer for the header when bridging to WL.
  * There is a compile time check in wlc.c which ensure that this value is at least as big
  * as TXOFF. This value is used in dma_rxfill (hnddma.c).
  */
-#define BCMEXTRAHDROOM 160
+
+#if defined(BCM_RPC_NOCOPY) || defined(BCM_RCP_TXNOCOPY)
+#define BCMEXTRAHDROOM 220
+#else
+#define BCMEXTRAHDROOM 172
+#endif
 
 /* Headroom required for dongle-to-host communication.  Packets allocated
  * locally in the dongle (e.g. for CDC ioctls or RNDIS messages) should
@@ -179,6 +208,10 @@ typedef unsigned long dmaaddr_t;
 #endif
 
 
+#if defined(BCMDBG_ASSERT) || defined(BCMASSERT_LOG)
+#define BCMASSERT_SUPPORT
+#endif /* BCMDBG_ASSERT || BCMASSERT_LOG */
+
 /* Brett's nifty macros for doing definition and get/set of bitfields
  * Usage example, e.g. a three-bit field (bits 4-6):
  *    #define <NAME>_M	BITFIELD_MASK(3)
diff -urBp a/include/bcmdevs.h b/include/bcmdevs.h
--- a/include/bcmdevs.h	2010-01-29 17:37:48.000000000 +0300
+++ b/include/bcmdevs.h	2010-02-12 23:50:51.000000000 +0300
@@ -1,7 +1,7 @@
 /*
  * Broadcom device-specific manifest constants.
  *
- * Copyright (C) 2008, Broadcom Corporation
+ * Copyright (C) 2009, Broadcom Corporation
  * All Rights Reserved.
  * 
  * THIS SOFTWARE IS OFFERED "AS IS", AND BROADCOM GRANTS NO WARRANTIES OF ANY
@@ -9,7 +9,7 @@
  * SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
  * FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
  *
- * $Id: bcmdevs.h,v 13.182.2.17.8.7 2009/05/20 02:39:20 Exp $
+ * $Id: bcmdevs.h,v 13.182.2.37 2009/11/06 02:52:26 Exp $
  */
 
 #ifndef	_BCMDEVS_H
@@ -42,6 +42,7 @@
 #define BCM_DNGL_BL_PID_4328    0xbd12
 #define BCM_DNGL_BL_PID_4322    0xbd13
 #define BCM_DNGL_BDC_PID        0xbdc
+#define BCM_DNGL_BL_PID_4319    0xbd16
 
 /* PCI Device ID's */
 #define	BCM4210_DEVICE_ID	0x1072		/* never used */
@@ -87,6 +88,11 @@
 #define BCM4319_D11N_ID		0x4337		/* 4319 802.11n dualband device */
 #define BCM4319_D11N2G_ID	0x4338		/* 4319 802.11n 2.4G device */
 #define BCM4319_D11N5G_ID	0x4339		/* 4319 802.11n 5G device */
+#define BCM43225_D11N2G_ID	0x4357		/* 43225 802.11n 2.4GHz device */
+#define BCM43421_D11N_ID	0xA99D		/* 43421 802.11n dualband device */
+#define	BCM4315_D11DUAL_ID	0x4334		/* 4315 802.11a/g id */
+#define	BCM4315_D11G_ID		0x4335		/* 4315 802.11g id */
+#define	BCM4315_D11A_ID		0x4336		/* 4315 802.11a id */
 
 
 #define	BCMGPRS_UART_ID		0x4333		/* Uart id used by 4306/gprs card */
@@ -145,8 +151,13 @@
 #define	BCM4322_CHIP_ID		0x4322		/* 4322 chipcommon chipid */
 #define	BCM43221_CHIP_ID	43221		/* 43221 chipcommon chipid (OTP chipid) */
 #define	BCM43231_CHIP_ID	43231		/* 43231 chipcommon chipid (OTP chipid) */
+#define	BCM43111_CHIP_ID	43111		/* 43111 chipcommon chipid (OTP chipid) */
+#define	BCM43112_CHIP_ID	43112		/* 43112 chipcommon chipid (OTP chipid) */
 #define	BCM43222_CHIP_ID	43222		/* 43222 chipcommon chipid */
 #define	BCM43224_CHIP_ID	43224		/* 43224 chipcommon chipid */
+#define	BCM43225_CHIP_ID	43225		/* 43225 chipcommon chipid (OTP chipid) */
+#define	BCM43421_CHIP_ID	43421		/* 43421 chipcommon chipid */
+#define	BCM4342_CHIP_ID		4342		/* 4342 chipcommon chipid (OTP, RBBU) */
 #define	BCM4325_CHIP_ID		0x4325		/* 4325 chip id */
 #define	BCM4328_CHIP_ID		0x4328		/* 4328 chip id */
 #define	BCM4402_CHIP_ID		0x4402		/* 4402 chipid */
@@ -155,12 +166,15 @@
 #define	BCM4712_CHIP_ID		0x4712		/* 4712 chipcommon chipid */
 #define	BCM4716_CHIP_ID		0x4716		/* 4716 chipcommon chipid */
 #define	BCM47162_CHIP_ID	47162		/* 47162 chipcommon chipid */
+#define	BCM4748_CHIP_ID		0x4748		/* 4716 chipcommon chipid (OTP, RBBU) */
 #define BCM4785_CHIP_ID		0x4785		/* 4785 chipcommon chipid */
 #define	BCM5350_CHIP_ID		0x5350		/* 5350 chipcommon chipid */
 #define	BCM5352_CHIP_ID		0x5352		/* 5352 chipcommon chipid */
 #define	BCM5354_CHIP_ID		0x5354		/* 5354 chipcommon chipid */
 #define BCM5365_CHIP_ID		0x5365		/* 5365 chipcommon chipid */
 #define	BCM5356_CHIP_ID		0x5356		/* 5356 chipcommon chipid */
+#define	BCM4319_CHIP_ID		0x4319		/* 4319 chip id */
+#define	BCM4329_CHIP_ID		0x4329		/* 4329 chipcommon chipid */
 
 
 /* Package ID's */
@@ -182,12 +196,14 @@
 #define HDLSIM5350_PKG_ID	1		/* HDL simulator package id for a 5350 */
 #define HDLSIM_PKG_ID		14		/* HDL simulator package id */
 #define HWSIM_PKG_ID		15		/* Hardware simulator package id */
+#define BCM43224_FAB_CSM	0x8		/* the chip is manufactured by CSM */
+#define BCM43224_FAB_SMIC	0xa		/* the chip is manufactured by SMIC */
 
 #define PCIXX21_FLASHMEDIA0_ID	0x8033		/* TI PCI xx21 Standard Host Controller */
 #define PCIXX21_SDIOH0_ID	0x8034		/* TI PCI xx21 Standard Host Controller */
 
 /* boardflags */
-#define	BFL_BTCOEXIST		0x00000001  /* Board implements Bluetooth coexistence */
+#define	BFL_BTC2WIRE		0x00000001  /* Board implements old 2wire Bluetooth coexistence */
 #define	BFL_PACTRL		0x00000002  /* Board has gpio 9 controlling the PA */
 #define	BFL_AIRLINEMODE		0x00000004  /* Board implements gpio 13 radio disable indication */
 #define	BFL_ENETROBO		0x00000010  /* Board has robo switch or core */
@@ -201,7 +217,7 @@
 #define BFL_FEM			0x00000800  /* Board supports the Front End Module */
 #define BFL_EXTLNA		0x00001000  /* Board has an external LNA in 2.4GHz band */
 #define BFL_HGPA		0x00002000  /* Board has a high gain PA */
-#define	BFL_BTCMOD		0x00004000  /* Board's BTCOEXIST is in the alternate gpios */
+#define	BFL_BTC2WIRE_ALTGPIO	0x00004000  /* Board's BTC 2wire is in the alternate gpios */
 #define	BFL_ALTIQ		0x00008000  /* Alternate I/Q settings */
 #define BFL_NOPA		0x00010000  /* Board has no PA */
 #define BFL_RSSIINV		0x00020000  /* Board's RSSI uses positive slope(not TSSI) */
@@ -213,7 +229,6 @@
 #define BFL_NOCBUCK		0x00800000  /* Power topology doesn't use CBUCK */
 #define BFL_CCKFAVOREVM		0x01000000  /* Favor CCK EVM over spectral mask */
 #define BFL_PALDO		0x02000000  /* Power topology uses PALDO */
-#define BFL_UCPWRCTL_MININDX	0x08000000  /* Enforce min power index to avoid FEM damage */
 #define BFL_EXTLNA_5GHz		0x10000000  /* Board has an external LNA in 5GHz band */
 
 /* boardflags2 */
@@ -230,9 +245,13 @@
 #define BFL2_GPLL_WAR	        0x00000400  /* Flag to implement alternative G-band PLL settings */
 #define BFL2_TRISTATE_LED	0x00000800  /* Tri-state the LED */
 #define BFL2_SINGLEANT_CCK	0x00001000  /* Tx CCK pkts on Ant 0 only */
-#define BFL2_2G_SPUR_WAR	0x00002000  /* Board has a WAR to reduce and avoid clock-harmonic */
+#define BFL2_2G_SPUR_WAR	0x00002000  /* Board has a WAR to reduce and avoid clock-harmonic
+					     * spurs in 2G band
+					     */
 #define BFL2_BPHY_ALL_TXCORES	0x00004000  /* Transmit bphy frames using all tx cores */
 #define BFL2_FCC_BANDEDGE_WAR	0x00008000  /* using 40Mhz LPF for 20Mhz bandedge channels */
+#define BFL2_GPLL_WAR2	        0x00010000  /* Flag to widen G-band PLL loop b/w */
+#define BFL2_INTERNDET_TXIQCAL  0x00040000  /* Use internal envelope detector for TX IQCAL */
 
 /* board specific GPIO assignment, gpio 0-3 are also customer-configurable led */
 #define	BOARD_GPIO_BTC3W_IN	0x850	/* bit 4 is RF_ACTIVE, bit 6 is STATUS, bit 11 is PRI */
@@ -245,6 +264,11 @@
 #define BOARD_GPIO_12		0x1000	/* gpio 12 */
 #define BOARD_GPIO_13		0x2000	/* gpio 13 */
 
+#define GPIO_BTC4W_OUT_4312  0x010  /* bit 4 is BT_IODISABLE */
+#define GPIO_BTC4W_OUT_43224  0x020  /* bit 5 is BT_IODISABLE */
+#define GPIO_BTC4W_OUT_43225  0x0e0  /* bit 5 BT_IODISABLE, bit 6 SW_BT, bit 7 SW_WL */
+#define GPIO_BTC4W_OUT_43421  0x020  /* bit 5 is BT_IODISABLE */
+
 #define	PCI_CFG_GPIO_SCS	0x10	/* PCI config space bit 4 for 4306c0 slow clock source */
 #define PCI_CFG_GPIO_HWRAD	0x20	/* PCI config space GPIO 13 for hw radio disable */
 #define PCI_CFG_GPIO_XTAL	0x40	/* PCI config space GPIO 14 for Xtal powerup */
@@ -434,6 +458,15 @@
 /* 4716 boards */
 #define BCM94716NR2_SSID	0x04cd
 
+/* 4319 boards */
+#define BCM94319DEVBU_SSID	0X04e5
+#define BCM94319USBNP4L_SSID	0X04e6
+#define BCM94319WLUSBN4L_SSID	0X04e7
+#define BCM94319SDG_SSID	0X04ea
+#define BCM94319LCUSBSDN4L_SSID	0X04eb
+#define BCM94319LCSDN4L_SSID	0X0507
+#define BCM94319LSUSBN4L_SSID	0X0508
+
 /* # of GPIO pins */
 #define GPIO_NUMPINS		16
 
diff -urBp a/include/bcmendian.h b/include/bcmendian.h
--- a/include/bcmendian.h	2010-01-29 17:37:48.000000000 +0300
+++ b/include/bcmendian.h	2010-02-11 22:34:40.000000000 +0300
@@ -1,7 +1,7 @@
 /*
  * local version of endian.h - byte order defines
  *
- * Copyright (C) 2008, Broadcom Corporation
+ * Copyright (C) 2009, Broadcom Corporation
  * All Rights Reserved.
  * 
  * THIS SOFTWARE IS OFFERED "AS IS", AND BROADCOM GRANTS NO WARRANTIES OF ANY
@@ -9,7 +9,7 @@
  * SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
  * FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
  *
- *  $Id: bcmendian.h,v 1.31 2005/09/03 22:24:22 Exp $
+ * $Id: bcmendian.h,v 1.31.296.2 2009/10/12 23:03:37 Exp $
 */
 
 #ifndef _BCMENDIAN_H_
@@ -72,6 +72,9 @@ bcmswap16_buf(uint16 *buf, uint len)
 #ifndef hton16
 #ifndef IL_BIGENDIAN
 #define HTON16(i) BCMSWAP16(i)
+#define NTOH16(i) BCMSWAP16(i)
+#define HTON32(i) BCMSWAP32(i)
+#define NTOH32(i) BCMSWAP32(i)
 #define	hton16(i) bcmswap16(i)
 #define	hton32(i) bcmswap32(i)
 #define	ntoh16(i) bcmswap16(i)
@@ -82,6 +85,9 @@ bcmswap16_buf(uint16 *buf, uint len)
 #define htol32(i) (i)
 #else
 #define HTON16(i) (i)
+#define NTOH16(i) (i)
+#define HTON32(i) (i)
+#define NTOH32(i) (i)
 #define	hton16(i) (i)
 #define	hton32(i) (i)
 #define	ntoh16(i) (i)
@@ -145,54 +151,58 @@ hton32_ua_store(uint32 val, uint8 *bytes
 	bytes[0] = val>>24;
 }
 
+#define _LTOH16_UA(cp)	((cp)[0] | ((cp)[1] << 8))
+#define _LTOH32_UA(cp)	((cp)[0] | ((cp)[1] << 8) | ((cp)[2] << 16) | ((cp)[3] << 24))
+#define _NTOH16_UA(cp)	(((cp)[0] << 8) | (cp)[1])
+#define _NTOH32_UA(cp)	(((cp)[0] << 24) | ((cp)[1] << 16) | ((cp)[2] << 8) | (cp)[3])
+
 /*
-* load 16-bit value from unaligned little endian byte array.
-*/
+ * Load 16-bit value from unaligned little-endian byte array.
+ */
 static INLINE uint16
-ltoh16_ua(void *bytes)
+ltoh16_ua(const void *bytes)
 {
-	return (((uint8*)bytes)[1]<<8)+((uint8 *)bytes)[0];
+	return _LTOH16_UA((const uint8 *)bytes);
 }
 
 /*
-* load 32-bit value from unaligned little endian byte array.
-*/
+ * Load 32-bit value from unaligned little-endian byte array.
+ */
 static INLINE uint32
-ltoh32_ua(void *bytes)
+ltoh32_ua(const void *bytes)
 {
-	return (((uint8*)bytes)[3]<<24)+(((uint8*)bytes)[2]<<16)+
-	       (((uint8*)bytes)[1]<<8)+((uint8*)bytes)[0];
+	return _LTOH32_UA((const uint8 *)bytes);
 }
 
 /*
-* load 16-bit value from unaligned big(network) endian byte array.
-*/
+ * Load 16-bit value from unaligned big-(network-)endian byte array.
+ */
 static INLINE uint16
-ntoh16_ua(void *bytes)
+ntoh16_ua(const void *bytes)
 {
-	return (((uint8*)bytes)[0]<<8)+((uint8*)bytes)[1];
+	return _NTOH16_UA((const uint8 *)bytes);
 }
 
 /*
-* load 32-bit value from unaligned big(network) endian byte array.
-*/
+ * Load 32-bit value from unaligned big-(network-)endian byte array.
+ */
 static INLINE uint32
-ntoh32_ua(void *bytes)
+ntoh32_ua(const void *bytes)
 {
-	return (((uint8*)bytes)[0]<<24)+(((uint8*)bytes)[1]<<16)+
-	       (((uint8*)bytes)[2]<<8)+((uint8*)bytes)[3];
+	return _NTOH32_UA((const uint8 *)bytes);
 }
 
-#define ltoh_ua(ptr) (\
-	sizeof(*(ptr)) == sizeof(uint8) ?  *(uint8 *)ptr : \
-	sizeof(*(ptr)) == sizeof(uint16) ? (((uint8 *)ptr)[1]<<8)+((uint8 *)ptr)[0] : \
-	(((uint8 *)ptr)[3]<<24)+(((uint8 *)ptr)[2]<<16)+(((uint8 *)ptr)[1]<<8)+((uint8 *)ptr)[0] \
-)
-
-#define ntoh_ua(ptr) (\
-	sizeof(*(ptr)) == sizeof(uint8) ?  *(uint8 *)ptr : \
-	sizeof(*(ptr)) == sizeof(uint16) ? (((uint8 *)ptr)[0]<<8)+((uint8 *)ptr)[1] : \
-	(((uint8 *)ptr)[0]<<24)+(((uint8 *)ptr)[1]<<16)+(((uint8 *)ptr)[2]<<8)+((uint8 *)ptr)[3] \
-)
+#define ltoh_ua(ptr) \
+	(sizeof(*(ptr)) == sizeof(uint8) ? *(const uint8 *)ptr : \
+	 sizeof(*(ptr)) == sizeof(uint16) ? _LTOH16_UA((const uint8 *)ptr) : \
+	 sizeof(*(ptr)) == sizeof(uint32) ? _LTOH32_UA((const uint8 *)ptr) : \
+	 0xfeedf00d)
+
+#define ntoh_ua(ptr) \
+	(sizeof(*(ptr)) == sizeof(uint8) ? *(const uint8 *)ptr : \
+	 sizeof(*(ptr)) == sizeof(uint16) ? _NTOH16_UA((const uint8 *)ptr) : \
+	 sizeof(*(ptr)) == sizeof(uint32) ? _NTOH32_UA((const uint8 *)ptr) : \
+	 0xfeedf00d)
+
 
 #endif /* _BCMENDIAN_H_ */
diff -urBp a/include/bcmwifi.h b/include/bcmwifi.h
--- a/include/bcmwifi.h	2010-01-29 17:37:48.000000000 +0300
+++ b/include/bcmwifi.h	2010-02-11 22:34:40.000000000 +0300
@@ -3,7 +3,7 @@
  * This header file housing the define and function prototype use by
  * both the wl driver, tools & Apps.
  *
- * Copyright (C) 2008, Broadcom Corporation
+ * Copyright (C) 2009, Broadcom Corporation
  * All Rights Reserved.
  * 
  * THIS SOFTWARE IS OFFERED "AS IS", AND BROADCOM GRANTS NO WARRANTIES OF ANY
@@ -11,7 +11,7 @@
  * SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
  * FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
  *
- * $Id: bcmwifi.h,v 1.19.2.1 2008/07/02 00:20:05 Exp $
+ * $Id: bcmwifi.h,v 1.19.2.3 2009/05/21 21:22:19 Exp $
  */
 
 #ifndef	_bcmwifi_h_
@@ -27,6 +27,7 @@ typedef uint16 chanspec_t;
 #define CH_EWA_VALID			0x04
 #define CH_20MHZ_APART			4
 #define CH_10MHZ_APART			2
+#define CH_5MHZ_APART			1	/* 2G band channels are 5 Mhz apart */
 #define CH_MAX_2G_CHANNEL		14	/* Max channel in 2G band */
 #define WLC_MAX_2G_CHANNEL		CH_MAX_2G_CHANNEL /* legacy define */
 #define	MAXCHANNEL		224	/* max # supported channels. The max channel no is 216,
@@ -94,6 +95,7 @@ typedef uint16 chanspec_t;
 #define CHSPEC_CTL_CHAN(chspec)  ((CHSPEC_SB_LOWER(chspec)) ? \
 				  (LOWER_20_SB((chspec & WL_CHANSPEC_CHAN_MASK))) : \
 				  (UPPER_20_SB((chspec & WL_CHANSPEC_CHAN_MASK))))
+#define CHSPEC2WLC_BAND(chspec) (CHSPEC_IS5G((chspec))? WLC_BAND_5G: WLC_BAND_2G)
 
 #define CHANSPEC_STR_LEN    8
 
diff -urBp a/include/cfe_osl.h b/include/cfe_osl.h
--- a/include/cfe_osl.h	2010-01-29 17:37:48.000000000 +0300
+++ b/include/cfe_osl.h	2010-02-11 22:34:40.000000000 +0300
@@ -9,7 +9,7 @@
  * or duplicated in any form, in whole or in part, without the prior
  * written permission of Broadcom Corporation.
  *
- * $Id: cfe_osl.h,v 1.49.2.1.12.3 2009/01/22 19:44:31 Exp $
+ * $Id: cfe_osl.h,v 1.49.2.6 2009/11/04 01:17:49 Exp $
  */
 
 #ifndef _cfe_osl_h_
@@ -31,7 +31,13 @@
 #include <bcmstdlib.h>
 
 /* assert and panic */
+#ifdef BCMDBG_ASSERT
+#define ASSERT(exp) \
+	do { if (!(exp)) osl_assert(#exp, __FILE__, __LINE__); } while (0)
+extern void osl_assert(char *exp, char *file, int line);
+#else /* BCMDBG_ASSERT */
 #define	ASSERT(exp)		do {} while (0)
+#endif /* BCMDBG_ASSERT */
 
 /* PCMCIA attribute space access macros */
 #define	OSL_PCMCIA_READ_ATTR(osh, offset, buf, size) \
@@ -127,6 +133,14 @@ extern void osl_detach(osl_t *osh);
 #define OSL_CACHED(a)		(a)
 #endif
 
+#ifdef __mips__
+#define OSL_PREF_RANGE_LD(va, sz) prefetch_range_PREF_LOAD_RETAINED(va, sz)
+#define OSL_PREF_RANGE_ST(va, sz) prefetch_range_PREF_STORE_RETAINED(va, sz)
+#else /* __mips__ */
+#define OSL_PREF_RANGE_LD(va, sz)
+#define OSL_PREF_RANGE_ST(va, sz)
+#endif /* __mips__ */
+
 /* host/bus architecture-specific address byte swap */
 #define BUS_SWAP32(v)		(v)
 
@@ -230,7 +244,7 @@ extern uchar *osl_pktpull(struct lbuf *l
 extern struct lbuf *osl_pktdup(struct lbuf *lb);
 extern int osl_error(int bcmerror);
 
-/* get system up time in miliseconds */
-#define OSL_SYSUPTIME()		(0)
+/* Global ASSERT type */
+extern uint32 g_assert_type;
 
 #endif	/* _cfe_osl_h_ */
diff -urBp a/include/linux_osl.h b/include/linux_osl.h
--- a/include/linux_osl.h	2009-10-10 20:49:25.000000000 +0400
+++ b/include/linux_osl.h	2010-02-12 23:53:03.000000000 +0300
@@ -9,7 +9,7 @@
  * SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
  * FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
  *
- * $Id: linux_osl.h,v 13.133.2.7.12.4 2009/03/01 16:54:44 Exp $
+ * $Id: linux_osl.h,v 13.133.2.18 2009/11/04 16:51:58 Exp $
  */
 
 #ifndef _linux_osl_h_
@@ -23,17 +23,26 @@
 extern osl_t *osl_attach(void *pdev, uint bustype, bool pkttag);
 extern void osl_detach(osl_t *osh);
 
+/* Global ASSERT type */
+extern uint32 g_assert_type;
+
 /* ASSERT */
+#if defined(BCMDBG_ASSERT) || defined(BCMASSERT_LOG)
+	#define ASSERT(exp) \
+	  do { if (!(exp)) osl_assert(#exp, __FILE__, __LINE__); } while (0)
+extern void osl_assert(char *exp, char *file, int line);
+#else
 	#ifdef __GNUC__
 		#define GCC_VERSION \
 			(__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__)
 		#if GCC_VERSION > 30100
 			#define ASSERT(exp)	do {} while (0)
 		#else
-			/* ASSERT could causes segmentation fault on GCC3.1, use empty instead */
+			/* ASSERT could cause segmentation fault on GCC3.1, use empty instead */
 			#define ASSERT(exp)
 		#endif /* GCC_VERSION > 30100 */
 	#endif /* __GNUC__ */
+#endif /* BCMDBG_ASSERT || BCMASSERT_LOG */
 
 /* microsecond delay */
 #define	OSL_DELAY(usec)		osl_delay(usec)
@@ -99,6 +108,9 @@ typedef struct {
 	extern void osl_mfree(osl_t *osh, void *addr, uint size);
 	extern uint osl_malloced(osl_t *osh);
 
+#define NATIVE_MALLOC(osh, size)		kmalloc(size, GFP_ATOMIC)
+#define NATIVE_MFREE(osh, addr, size)	kfree(addr)
+
 #define	MALLOC_FAILED(osh)	osl_malloc_failed((osh))
 extern uint osl_malloc_failed(osl_t *osh);
 
@@ -144,9 +156,6 @@ extern void osl_dma_unmap(osl_t *osh, ui
 	#define SELECT_BUS_READ(osh, mmap_op, bus_op) mmap_op
 #endif 
 
-/* get system up time in milliseconds */
-#define OSL_SYSUPTIME()		(0)
-
 #define OSL_ERROR(bcmerror)	osl_error(bcmerror)
 extern int osl_error(int bcmerror);
 
@@ -159,6 +168,7 @@ extern int osl_error(int bcmerror);
  */
 #ifndef BINOSL
 
+#define OSL_SYSUPTIME()		((uint32)jiffies * (1000 / HZ))
 #ifndef printf
 #define	printf(fmt, args...)	printk(fmt , ## args)
 #endif
@@ -289,6 +299,14 @@ extern void osl_writel(osl_t *osh, volat
 #define OSL_CACHED(va)		((void *)va)
 #endif /* mips */
 
+#ifdef __mips__
+#define OSL_PREF_RANGE_LD(va, sz) prefetch_range_PREF_LOAD_RETAINED(va, sz)
+#define OSL_PREF_RANGE_ST(va, sz) prefetch_range_PREF_STORE_RETAINED(va, sz)
+#else /* __mips__ */
+#define OSL_PREF_RANGE_LD(va, sz)
+#define OSL_PREF_RANGE_ST(va, sz)
+#endif /* __mips__ */
+
 /* get processor cycle count */
 #if defined(mips)
 #if defined DSLCPE_DELAY
@@ -583,12 +601,24 @@ extern void osl_writeb(uint8 v, volatile
 extern void osl_writew(uint16 v, volatile uint16 *r);
 extern void osl_writel(uint32 v, volatile uint32 *r);
 
+/* system up time in ms */
+#define OSL_SYSUPTIME()		osl_sysuptime()
+extern uint32 osl_sysuptime(void);
+
 /* uncached/cached virtual address */
 #define OSL_UNCACHED(va)	osl_uncached((va))
 extern void *osl_uncached(void *va);
 #define OSL_CACHED(va)		osl_cached((va))
 extern void *osl_cached(void *va);
 
+#ifdef __mips__
+#define OSL_PREF_RANGE_LD(va, sz) prefetch_range_PREF_LOAD_RETAINED(va, sz)
+#define OSL_PREF_RANGE_ST(va, sz) prefetch_range_PREF_STORE_RETAINED(va, sz)
+#else /* __mips__ */
+#define OSL_PREF_RANGE_LD(va, sz)
+#define OSL_PREF_RANGE_ST(va, sz)
+#endif /* __mips__ */
+
 /* get processor cycle count */
 #define OSL_GETCYCLES(x)	((x) = osl_getcycles())
 extern uint osl_getcycles(void);
@@ -661,7 +691,12 @@ extern uint osl_pktalloced(osl_t *osh);
 
 
 /* ASSERT */
+#ifdef BCMDBG_ASSERT
+	#include <assert.h>
+	#define ASSERT assert
+#else /* BCMDBG_ASSERT */
 	#define ASSERT(exp)	do {} while (0)
+#endif /* BCMDBG_ASSERT */
 
 /* MALLOC and MFREE */
 #define MALLOC(o, l) malloc(l)
diff -urBp a/include/min_osl.h b/include/min_osl.h
--- a/include/min_osl.h	2010-01-29 17:37:48.000000000 +0300
+++ b/include/min_osl.h	2010-02-11 22:34:40.000000000 +0300
@@ -9,7 +9,7 @@
  * SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
  * FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
  *
- * $Id: min_osl.h,v 13.22.2.1.6.2 2008/11/19 01:45:21 Exp $
+ * $Id: min_osl.h,v 13.22.2.6 2009/11/04 01:17:49 Exp $
  */
 
 #ifndef _min_osl_h_
@@ -122,6 +122,14 @@ extern void *malloc(uint size);
 #define	OSL_CACHED(va)		((void *)(va))
 #endif
 
+#ifdef __mips__
+#define OSL_PREF_RANGE_LD(va, sz) prefetch_range_PREF_LOAD_RETAINED(va, sz)
+#define OSL_PREF_RANGE_ST(va, sz) prefetch_range_PREF_STORE_RETAINED(va, sz)
+#else /* __mips__ */
+#define OSL_PREF_RANGE_LD(va, sz)
+#define OSL_PREF_RANGE_ST(va, sz)
+#endif /* __mips__ */
+
 /* host/bus architecture-specific address byte swap */
 #define BUS_SWAP32(v)		(v)
 
@@ -173,7 +181,7 @@ extern int osl_error(int);
 #define PKTFRMNATIVE(osh, lb)		((void *)NULL)
 #define PKTTONATIVE(osh, p)		((struct lbuf *)NULL)
 
-/* get system up time in miliseconds */
-#define OSL_SYSUPTIME()			(0)
+/* Global ASSERT type */
+extern uint32 g_assert_type;
 
 #endif	/* _min_osl_h_ */
diff -urBp a/include/osl.h b/include/osl.h
--- a/include/osl.h	2010-01-29 17:37:48.000000000 +0300
+++ b/include/osl.h	2010-02-11 22:34:40.000000000 +0300
@@ -9,7 +9,7 @@
  * SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
  * FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
  *
- * $Id: osl.h,v 13.38.2.1 2008/07/02 00:20:05 Exp $
+ * $Id: osl.h,v 13.38.2.5 2009/11/04 01:38:53 Exp $
  */
 
 #ifndef _osl_h_
@@ -30,6 +30,51 @@ typedef unsigned int (*osl_rreg_fn_t)(vo
 typedef void  (*osl_wreg_fn_t)(void *ctx, void *reg, unsigned int val, unsigned int size);
 #endif
 
+#ifdef __mips__
+#define PREF_LOAD		0
+#define PREF_STORE		1
+#define PREF_LOAD_STREAMED	4
+#define PREF_STORE_STREAMED	5
+#define PREF_LOAD_RETAINED	6
+#define PREF_STORE_RETAINED	7
+#define PREF_WBACK_INV		25
+#define PREF_PREPARE4STORE	30
+
+#define MAKE_PREFETCH_FN(hint) \
+static inline void prefetch_##hint(const void *addr) \
+{ \
+	__asm__ __volatile__( \
+	"       .set    mips4           \n" \
+	"       pref    %0, (%1)        \n" \
+	"       .set    mips0           \n" \
+	: \
+	: "i" (hint), "r" (addr)); \
+}
+
+#define MAKE_PREFETCH_RANGE_FN(hint) \
+static inline void prefetch_range_##hint(const void *addr, int len) \
+{ \
+	int size = len; \
+	while (size > 0) { \
+		prefetch_##hint(addr); \
+		size -= 32; \
+	} \
+}
+
+MAKE_PREFETCH_FN(PREF_LOAD)
+MAKE_PREFETCH_RANGE_FN(PREF_LOAD)
+MAKE_PREFETCH_FN(PREF_STORE)
+MAKE_PREFETCH_RANGE_FN(PREF_STORE)
+MAKE_PREFETCH_FN(PREF_LOAD_STREAMED)
+MAKE_PREFETCH_RANGE_FN(PREF_LOAD_STREAMED)
+MAKE_PREFETCH_FN(PREF_STORE_STREAMED)
+MAKE_PREFETCH_RANGE_FN(PREF_STORE_STREAMED)
+MAKE_PREFETCH_FN(PREF_LOAD_RETAINED)
+MAKE_PREFETCH_RANGE_FN(PREF_LOAD_RETAINED)
+MAKE_PREFETCH_FN(PREF_STORE_RETAINED)
+MAKE_PREFETCH_RANGE_FN(PREF_STORE_RETAINED)
+#endif /* __mips__ */
+
 #if defined(__ECOS)
 #include <ecos_osl.h>
 #elif  defined(DOS)
@@ -37,7 +82,11 @@ typedef void  (*osl_wreg_fn_t)(void *ctx
 #elif defined(PCBIOS)
 #include <pcbios_osl.h>
 #elif defined(linux)
+#ifdef USER_MODE
+#include <usermode_osl.h>
+#else
 #include <linux_osl.h>
+#endif
 #elif defined(NDIS)
 #include <ndis_osl.h>
 #elif defined(_CFE_)
@@ -57,4 +106,11 @@ typedef void  (*osl_wreg_fn_t)(void *ctx
 /* handy */
 #define	SET_REG(osh, r, mask, val)	W_REG((osh), (r), ((R_REG((osh), r) & ~(mask)) | (val)))
 
+#if !defined(OSL_SYSUPTIME)
+#define OSL_SYSUPTIME() (0)
+#define OSL_SYSUPTIME_SUPPORT FALSE
+#else
+#define OSL_SYSUPTIME_SUPPORT TRUE
+#endif /* OSL_SYSUPTIME */
+
 #endif	/* _osl_h_ */
