diff -BurpN accel-pptp.orig/pppd_plugin/src/pptp.c accel-pptp/pppd_plugin/src/pptp.c
--- accel-pptp.orig/pppd_plugin/src/pptp.c	2009-10-03 09:08:18.000000000 +0000
+++ accel-pptp/pppd_plugin/src/pptp.c	2010-08-20 17:43:33.000000000 +0000
@@ -67,6 +67,7 @@ char *pptp_client = NULL;
 char *pptp_phone = NULL;
 int pptp_sock=-1;
 int pptp_timeout=100000;
+int log_level = 0;
 struct in_addr localbind = { INADDR_NONE };
 struct rtentry rt;
 
@@ -93,6 +94,8 @@ static option_t Options[] =
       "PPTP socket" },
     { "pptp_phone", o_string, &pptp_phone,
       "PPTP Phone number" },
+    { "loglevel", o_int, &log_level,
+      "debugging level (0=low, 1=default, 2=high)"},
     { NULL }
 };
 
diff -BurpN accel-pptp.orig/pppd_plugin/src/util.c accel-pptp/pppd_plugin/src/util.c
--- accel-pptp.orig/pppd_plugin/src/util.c	2010-08-20 17:30:50.000000000 +0000
+++ accel-pptp/pppd_plugin/src/util.c	2010-08-20 17:42:08.000000000 +0000
@@ -26,8 +26,7 @@ va_list ap;						\
 char buf[256], string[256];				\
 va_start(ap, format);					\
 vsnprintf(buf, sizeof(buf), format, ap);		\
-snprintf(string, sizeof(string), "%s %s[%s:%s:%d]: %s",	\
-	 log_string, label, func, file, line, buf);	\
+snprintf(string, sizeof(string), "%s",	buf);		\
 va_end(ap)
 
 /*** open log *****************************************************************/
@@ -44,20 +43,19 @@ static void close_log(void)
 /*** print a message to syslog ************************************************/
 void _log(const char *func, const char *file, int line, const char *format, ...)
 {
-    MAKE_STRING("log");
-    open_log();
-    syslog(LOG_NOTICE, "%s", string);
-    close_log();
+    if (log_level > 0)
+    {
+	MAKE_STRING("log");
+	syslog(LOG_NOTICE, "%s", string);
+    }
 }
 
 /*** print a warning to syslog ************************************************/
 void _warn(const char *func, const char *file, int line, const char *format, ...)
 {
     MAKE_STRING("warn");
-    open_log();
     fprintf(stderr, "%s\n", string);
     syslog(LOG_WARNING, "%s", string);
-    close_log();
 }
 
 /*** print a fatal warning to syslog and exit *********************************/
