From 38d5811e793b478d1c64cc03ea1d3ace4f437701 Mon Sep 17 00:00:00 2001
From: Vladislav Grishenko <themiron@mail.ru>
Date: Mon, 1 Sep 2014 15:15:56 +0600
Subject: [PATCH 06/10] pptp: use predefined named framing/bearer values, cosmetics

---
 accel-pppd/ctrl/pptp/pptp.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/accel-pppd/ctrl/pptp/pptp.c b/accel-pppd/ctrl/pptp/pptp.c
index f428afc..19deb3d 100644
--- a/accel-pppd/ctrl/pptp/pptp.c
+++ b/accel-pppd/ctrl/pptp/pptp.c
@@ -192,8 +192,8 @@ static int send_pptp_start_ctrl_conn_rply(struct pptp_conn_t *conn, int res_code
 		.version = htons(PPTP_VERSION),
 		.result_code = res_code,
 		.error_code = err_code,
-		.framing_cap = htonl(3),
-		.bearer_cap = htonl(3),
+		.framing_cap = htonl(PPTP_FRAME_ANY),
+		.bearer_cap = htonl(PPTP_BEARER_ANY),
 		.max_channels = htons(1),
 		.firmware_rev = htons(PPTP_FIRMWARE_VERSION),
 	};
-- 
1.7.2.5

