--- accel-pptp.orig/pppd_plugin/src/pptp.c	2009-10-03 09:09:55.000000000 +0000
+++ accel-pptp/pppd_plugin/src/pptp.c	2011-07-21 15:12:16.017301000 +0000
@@ -280,6 +280,8 @@ static int open_callmgr(int call_id,stru
                 }
                 default: /* parent */
                     waitpid(pid, &status, 0);
+                    if (WIFEXITED(status))
+                        status = WEXITSTATUS(status);
                     if (status!= 0)
 		    {
 			close(fd);
