diff -ur nfs-utils-1.0.9/support/nfs/svc_socket.c nfs-utils/support/nfs/svc_socket.c
--- nfs-utils-1.0.9/support/nfs/svc_socket.c	2002-09-13 07:50:56.000000000 +0400
+++ nfs-utils/support/nfs/svc_socket.c	2005-02-12 23:46:00.936811456 +0300
@@ -39,9 +39,14 @@ svc_socket (u_long number,
 {
   struct sockaddr_in addr;
   socklen_t len = sizeof (struct sockaddr_in);
+#ifdef __UCLIBC_HAS_REENTRANT_RPC__
   char rpcdata [1024], servdata [1024];
   struct rpcent rpcbuf, *rpcp;
   struct servent servbuf, *servp = NULL;
+#else
+  struct rpcent *rpcp;
+  struct servent *servp;
+#endif
   int sock, ret;
   const char *proto = protocol == IPPROTO_TCP ? "tcp" : "udp";
 
@@ -63,13 +68,21 @@ svc_socket (u_long number,
   memset (&addr, 0, sizeof (addr));
   addr.sin_family = AF_INET;
 
+#ifdef __UCLIBC_HAS_REENTRANT_RPC__
   ret = getrpcbynumber_r (number, &rpcbuf, rpcdata, sizeof rpcdata,
 			  &rpcp);
+#else
+  ret = ((rpcp = getrpcbynumber(number)) == NULL);
+#endif
   if (ret == 0 && rpcp != NULL)
     {
       /* First try name.  */
+#ifdef __UCLIBC_HAS_REENTRANT_RPC__
       ret = getservbyname_r (rpcp->r_name, proto, &servbuf, servdata,
 			     sizeof servdata, &servp);
+#else
+      ret = ((servp = getservbyname(rpcp->r_name, proto)) == NULL);
+#endif
       if ((ret != 0 || servp == NULL) && rpcp->r_aliases)
 	{
 	  const char **a;
@@ -80,8 +93,12 @@ svc_socket (u_long number,
 	  /* Then we try aliases.  */
 	  for (a = (const char **) rpcp->r_aliases; *a != NULL; a++) 
 	    {
+#ifdef __UCLIBC_HAS_REENTRANT_RPC__
 	      ret = getservbyname_r (*a, proto, &servbuf, servdata,
 				     sizeof servdata, &servp);
+#else
+	      ret = ((servp = getservbyname(*a, proto)) == NULL);
+#endif
 	      if (ret == 0 && servp != NULL)
 		break;
 	    }
diff -ur nfs-utils-1.0.9/utils/statd/statd.c nfs-utils/utils/statd/statd.c
--- nfs-utils-1.0.9/utils/statd/statd.c	2004-12-06 04:25:27.000000000 +0300
+++ nfs-utils/utils/statd/statd.c	2005-02-12 23:44:04.711480392 +0300
@@ -27,6 +27,8 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 
+#include <sys/stat.h>
+
 /* Added to enable specification of state directory path at run-time
  * j_carlos_gomez@yahoo.com
  */
