diff -urBp igmpproxy-0.1_beta5/src/callout.c igmpproxy/src/callout.c
--- igmpproxy-0.1_beta5/src/callout.c	2009-05-13 23:43:18.000000000 +0400
+++ igmpproxy/src/callout.c	2009-09-15 11:23:19.000000000 +0400
@@ -76,21 +76,34 @@ void free_all_callouts() {
  */
 void age_callout_queue(int elapsed_time) {
     struct timeOutQueue *ptr;
+    struct timeOutQueue *_queue = NULL;
+    struct timeOutQueue *last = NULL;
     int i = 0;
 
-    for (ptr = queue; ptr; ptr = queue, i++) {
+    for (ptr = queue; ptr; ptr = ptr->next) {
         if (ptr->time > elapsed_time) {
             ptr->time -= elapsed_time;
-            return;
+            break;
         } else {
             elapsed_time -= ptr->time;
-            queue = queue->next;
-            my_log(LOG_DEBUG, 0, "About to call timeout %d (#%d)", ptr->id, i);
-
-            if (ptr->func)
-                ptr->func(ptr->data);
-            free(ptr);
-        }
+            if (_queue == NULL)
+                _queue = ptr;
+            last = ptr;
+         }
+    }
+
+    queue = ptr;
+    if (last) {
+        last->next = NULL;
+    }
+
+    /* process existing events */
+    for (ptr = _queue; ptr; ptr = _queue, i++) {
+        _queue = _queue->next;
+        my_log(LOG_DEBUG, 0, "About to call timeout %d (#%d)", ptr->id, i);
+        if (ptr->func)
+             ptr->func(ptr->data);
+        free(ptr);
     }
 }
 
