--- igmpproxy.orig/src/igmpproxy.h	2011-05-27 15:56:41.000000000 +0000
+++ igmpproxy/src/igmpproxy.h	2012-06-04 17:27:01.000000000 +0000
@@ -92,8 +92,8 @@ extern char     *send_buf;
 
 extern char     s1[];
 extern char     s2[];
-extern char		s3[];
-extern char		s4[];
+extern char     s3[];
+extern char     s4[];
 
 
 
--- igmpproxy.orig/src/request.c	2011-05-27 15:53:55.000000000 +0000
+++ igmpproxy/src/request.c	2012-06-04 17:27:01.000000000 +0000
@@ -100,7 +100,7 @@ void acceptGroupReport(uint32_t src, uin
 	my_log(LOG_INFO, 0, "The group address %s may not be requested from this interface. Ignoring.", inetFmt(group, s1));
     } else {
         // Log the state of the interface the report was recieved on.
-        my_log(LOG_INFO, 0, "Mebership report was recieved on %s. Ignoring.",
+        my_log(LOG_INFO, 0, "Membership report was recieved on %s. Ignoring.",
             sourceVif->state==IF_STATE_UPSTREAM?"the upstream interface":"a disabled interface");
     }
 
