--- igmpproxy/src/callout.c.orig	2006-11-22 00:54:14.000000000 +0300
+++ igmpproxy/src/callout.c	2006-11-24 20:58:40.633271656 +0300
@@ -76,22 +76,36 @@
  */
 void age_callout_queue(int elapsed_time) {
     struct timeOutQueue *ptr;
+    struct timeOutQueue *_queue = NULL;
+    struct timeOutQueue *last = NULL;
     int i = 0;
 
-    for (ptr = queue; ptr; ptr = queue, i++) {
+    for (ptr = queue; ptr; ptr = ptr->next) {
         if (ptr->time > elapsed_time) {
             ptr->time -= elapsed_time;
-            return;
+            break;
         } else {
             elapsed_time -= ptr->time;
-            queue = queue->next;
-            IF_DEBUG log(LOG_DEBUG, 0, "About to call timeout %d (#%d)", ptr->id, i);
-
-            if (ptr->func)
-                ptr->func(ptr->data);
-            free(ptr);
+            if (_queue == NULL)
+                _queue = ptr;
+            last = ptr;
         }
     }
+    
+    queue = ptr;
+    if (last) {
+        last->next = NULL;
+    }
+
+    /* process existing events */
+    for (ptr = _queue; ptr; ptr = _queue, i++) {
+        _queue = _queue->next;
+        IF_DEBUG log(LOG_DEBUG, 0, "About to call timeout %d (#%d)", ptr->id, i);
+
+        if (ptr->func)
+             ptr->func(ptr->data);
+        free(ptr);
+    }
 }
 
 /**
