From 708dd4c98662670f3104b71c2fc5eef82bb726a8
From: Alexey Fomenko
Date: Wed, 02 Mar 2011 03:01:24 +0000
Subject: add missing fclose's

Signed-off-by: Alexey Fomenko <ext-alexey.fomenko@nokia.com>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/networking/route.c b/networking/route.c
index b7d08dd..ac4fbb4 100644
--- a/networking/route.c
+++ b/networking/route.c
@@ -537,6 +537,7 @@ void FAST_FUNC bb_displayroutes(int noresolve, int netstatfmt)
 			printf("%-6d %-2d %7d %s\n", metric, ref, use, devname);
 		}
 	}
+	fclose(fp);
 }
 
 #if ENABLE_FEATURE_IPV6
@@ -625,6 +626,7 @@ static void INET6_displayroutes(void)
 			}
 		} while (1);
 	}
+	fclose(fp);
 }
 
 #endif
--
cgit v0.8.2.1
