From be168b119750beacc0d0212607c6fa3ee87f238c
From: Vladimir Dronnikov
Date: Mon, 05 Oct 2009 00:18:01 +0000
Subject: wget: fix for code 302

Signed-off-by: Vladimir Dronnikov <dronnikov@gmail.com>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/networking/wget.c b/networking/wget.c
index b8cd549..0f99e8d 100644
--- a/networking/wget.c
+++ b/networking/wget.c
@@ -370,8 +370,10 @@ static void parse_url(char *src_url, struct host_info *h)
 		h->path = sp;
 	}
 
+	// We used to set h->user to NULL here, but this interferes
+	// with handling of code 302 ("object was moved")
+
 	sp = strrchr(h->host, '@');
-	h->user = NULL;
 	if (sp != NULL) {
 		h->user = h->host;
 		*sp = '\0';
@@ -692,6 +694,7 @@ int wget_main(int argc UNUSED_PARAM, char **argv)
 
 	/* TODO: compat issue: should handle "wget URL1 URL2..." */
 
+	target.user = NULL;
 	parse_url(argv[optind], &target);
 
 	/* Use the proxy if necessary */
@@ -708,6 +709,7 @@ int wget_main(int argc UNUSED_PARAM, cha
 	if (use_proxy) {
 		proxy = getenv(target.is_ftp ? "ftp_proxy" : "http_proxy");
 		if (proxy && proxy[0]) {
+			server.user = NULL;
 			parse_url(proxy, &server);
 		} else {
 			use_proxy = 0;
--
cgit v0.8.2.1
