From 79950a6bbe080441fbffc450e1e75f103e968839
From: Denys Vlasenko
Date: Mon, 08 Mar 2010 21:03:24 +0000
Subject: wc: compat: make "wc -c ONE_FILE" not print leading spaces. +2 bytes

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/coreutils/wc.c b/coreutils/wc.c
index 08f3c2d..7116842 100644
--- a/coreutils/wc.c
+++ b/coreutils/wc.c
@@ -88,6 +88,8 @@ int wc_main(int argc UNUSED_PARAM, char **argv)
 	if (!argv[0]) {
 		*--argv = (char *) bb_msg_standard_input;
 		fname_fmt = "\n";
+	}
+	if (!argv[1]) { /* zero or one filename? */
 		if (!((print_type-1) & print_type)) /* exactly one option? */
 			start_fmt = "%"COUNT_FMT;
 	}
--
cgit v0.8.2.1
