From 46abfc0da14e5214fdf823edaf7643a9aa9eae65
From: Maciek Borzecki
Date: Tue, 16 Mar 2010 11:41:29 +0000
Subject: brctl: fix parsing of timespec to allow 0 (setfd, sethello..)

Signed-off-by: Maciek Borzecki <maciek.borzecki@gmail.com>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
diff --git a/networking/brctl.c b/networking/brctl.c
index 57074cd..77bb8f1 100644
--- a/networking/brctl.c
+++ b/networking/brctl.c
@@ -48,8 +48,9 @@ static ALWAYS_INLINE void strtotimeval(struct timeval *tv,
 {
 	double secs;
 #if BRCTL_USE_INTERNAL
-	secs = /*bb_*/strtod(time_str, NULL);
-	if (!secs)
+	char *endptr;
+	secs = /*bb_*/strtod(time_str, &endptr);
+	if (endptr == time_str)
 #else
 	if (sscanf(time_str, "%lf", &secs) != 1)
 #endif
--
cgit v0.8.2.1
