diff -urBp busybox-1.13.2/Makefile busybox/Makefile
--- busybox-1.13.2/Makefile	2008-12-31 06:32:54.000000000 +0300
+++ busybox/Makefile	2009-02-07 14:27:30.000000000 +0300
@@ -314,7 +314,7 @@
 
 # Use LINUXINCLUDE when you must reference the include/ directory.
 # Needed to be compatible with the O= option
-CFLAGS		:= $(CFLAGS)
+CFLAGS		:= $(CFLAGS) $(EXTRA_CFLAGS)
 # Added only to final link stage of busybox binary
 CFLAGS_busybox	:= $(CFLAGS_busybox)
 CPPFLAGS	:= $(CPPFLAGS)
diff -urBp busybox-1.13.2/include/applets.h busybox/include/applets.h
--- busybox-1.13.2/include/applets.h	2009-02-06 13:58:50.000000000 +0300
+++ busybox/include/applets.h	2009-02-07 20:33:24.000000000 +0300
 USE_BLKID(APPLET(blkid, _BB_DIR_SBIN, _BB_SUID_NEVER))
 USE_BRCTL(APPLET(brctl, _BB_DIR_USR_SBIN, _BB_SUID_NEVER))
 USE_BUNZIP2(APPLET(bunzip2, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
+USE_HTTPD(APPLET_ODDNAME(busybox_httpd, httpd, _BB_DIR_USR_SBIN, _BB_SUID_NEVER, httpd))
 USE_BUNZIP2(APPLET_ODDNAME(bzcat, bunzip2, _BB_DIR_USR_BIN, _BB_SUID_NEVER, bzcat))
 USE_BZIP2(APPLET(bzip2, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
 USE_CAL(APPLET(cal, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
@@ -178,7 +179,6 @@
 USE_HEXDUMP(APPLET_NOEXEC(hexdump, hexdump, _BB_DIR_USR_BIN, _BB_SUID_NEVER, hexdump))
 USE_HOSTID(APPLET_NOFORK(hostid, hostid, _BB_DIR_USR_BIN, _BB_SUID_NEVER, hostid))
 USE_HOSTNAME(APPLET(hostname, _BB_DIR_BIN, _BB_SUID_NEVER))
-USE_HTTPD(APPLET(httpd, _BB_DIR_USR_SBIN, _BB_SUID_NEVER))
 USE_HUSH(APPLET(hush, _BB_DIR_BIN, _BB_SUID_NEVER))
 USE_HWCLOCK(APPLET(hwclock, _BB_DIR_SBIN, _BB_SUID_NEVER))
 USE_ID(APPLET(id, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
diff -urBp busybox-1.13.2/include/libbb.h busybox/include/libbb.h
--- busybox-1.13.2/include/libbb.h	2008-12-31 06:06:45.000000000 +0300
+++ busybox/include/libbb.h	2009-02-07 20:45:25.000000000 +0300
@@ -1366,9 +1366,9 @@
 extern const char bb_busybox_exec_path[];
 /* util-linux manpage says /sbin:/bin:/usr/sbin:/usr/bin,
  * but I want to save a few bytes here */
-extern const char bb_PATH_root_path[]; /* "PATH=/sbin:/usr/sbin:/bin:/usr/bin" */
+extern const char bb_PATH_root_path[]; /* "PATH=/opt/sbin:/sbin:/usr/sbin:/opt/bin:/usr/local/bin:/bin:/usr/bin" */
 #define bb_default_root_path (bb_PATH_root_path + sizeof("PATH"))
-#define bb_default_path      (bb_PATH_root_path + sizeof("PATH=/sbin:/usr/sbin"))
+#define bb_default_path      (bb_PATH_root_path + sizeof("PATH=/opt/sbin:/sbin:/usr/sbin"))
 
 extern const int const_int_0;
 extern const int const_int_1;
diff -urBp busybox-1.13.2/libbb/messages.c busybox/libbb/messages.c
--- busybox-1.13.2/libbb/messages.c	2008-11-09 20:28:09.000000000 +0300
+++ busybox/libbb/messages.c	2009-02-07 20:45:15.000000000 +0300
@@ -48,7 +48,7 @@
 /* util-linux manpage says /sbin:/bin:/usr/sbin:/usr/bin,
  * but I want to save a few bytes here. Check libbb.h before changing! */
 const char bb_PATH_root_path[] ALIGN1 =
-	"PATH=/sbin:/usr/sbin:/bin:/usr/bin" BB_ADDITIONAL_PATH;
+	"PATH=/opt/sbin:/sbin:/usr/sbin:/opt/bin:/usr/local/bin:/bin:/usr/bin" BB_ADDITIONAL_PATH;
 
 
 const int const_int_1 = 1;
diff -urBp busybox-1.13.2/libbb/procps.c busybox/libbb/procps.c
--- busybox-1.13.2/libbb/procps.c	2008-11-09 20:28:09.000000000 +0300
+++ busybox/libbb/procps.c	2009-02-07 20:47:51.000000000 +0300
@@ -166,7 +166,7 @@
 procps_status_t* FAST_FUNC procps_scan(procps_status_t* sp, int flags)
 {
 	struct dirent *entry;
-	char buf[PROCPS_BUFSIZE];
+	char buf[PROCPS_BUFSIZE+1];
 	char filename[sizeof("/proc//cmdline") + sizeof(int)*3];
 	char *filename_tail;
 	long tasknice;
@@ -224,6 +224,7 @@ procps_status_t* FAST_FUNC procps_scan(p
 			n = read_to_buf(filename, buf);
 			if (n < 0)
 				break;
+			buf[n] = '\0';
 			cp = strrchr(buf, ')'); /* split into "PID (cmd" and "<rest>" */
 			/*if (!cp || cp[1] != ' ')
 				break;*/
