diff -BurpN busybox-1.14.2/Makefile busybox/Makefile
--- busybox-1.14.2/Makefile	2009-06-21 22:46:58.000000000 +0000
+++ busybox/Makefile	2009-07-22 14:53:46.000000000 +0000
@@ -315,7 +315,7 @@ AFLAGS_KERNEL	=
 
 # Use LINUXINCLUDE when you must reference the include/ directory.
 # Needed to be compatible with the O= option
-CFLAGS		:= $(CFLAGS)
+CFLAGS		:= $(CFLAGS) $(EXTRA_CFLAGS)
 # Added only to final link stage of busybox binary
 CFLAGS_busybox	:= $(CFLAGS_busybox)
 CPPFLAGS	:= $(CPPFLAGS)
diff -BurpN busybox-1.14.2/include/applets.h busybox/include/applets.h
--- busybox-1.14.2/include/applets.h	2009-06-21 22:40:29.000000000 +0000
+++ busybox/include/applets.h	2009-07-22 14:53:46.000000000 +0000
@@ -84,6 +84,7 @@ USE_BBCONFIG(APPLET(bbconfig, _BB_DIR_BI
 USE_BLKID(APPLET(blkid, _BB_DIR_SBIN, _BB_SUID_NEVER))
 USE_BRCTL(APPLET(brctl, _BB_DIR_USR_SBIN, _BB_SUID_NEVER))
 USE_BUNZIP2(APPLET(bunzip2, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
+USE_HTTPD(APPLET_ODDNAME(busybox_httpd, httpd, _BB_DIR_USR_SBIN, _BB_SUID_NEVER, httpd))
 USE_BUNZIP2(APPLET_ODDNAME(bzcat, bunzip2, _BB_DIR_USR_BIN, _BB_SUID_NEVER, bzcat))
 USE_BZIP2(APPLET(bzip2, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
 USE_CAL(APPLET(cal, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
@@ -182,7 +183,6 @@ USE_HEAD(APPLET(head, _BB_DIR_USR_BIN, _
 USE_HEXDUMP(APPLET_NOEXEC(hexdump, hexdump, _BB_DIR_USR_BIN, _BB_SUID_NEVER, hexdump))
 USE_HOSTID(APPLET_NOFORK(hostid, hostid, _BB_DIR_USR_BIN, _BB_SUID_NEVER, hostid))
 USE_HOSTNAME(APPLET(hostname, _BB_DIR_BIN, _BB_SUID_NEVER))
-USE_HTTPD(APPLET(httpd, _BB_DIR_USR_SBIN, _BB_SUID_NEVER))
 USE_HUSH(APPLET(hush, _BB_DIR_BIN, _BB_SUID_NEVER))
 USE_HWCLOCK(APPLET(hwclock, _BB_DIR_SBIN, _BB_SUID_NEVER))
 USE_ID(APPLET(id, _BB_DIR_USR_BIN, _BB_SUID_NEVER))
diff -BurpN busybox-1.14.2/include/libbb.h busybox/include/libbb.h
--- busybox-1.14.2/include/libbb.h	2009-06-21 22:46:43.000000000 +0000
+++ busybox/include/libbb.h	2009-07-22 14:53:46.000000000 +0000
@@ -1429,9 +1429,9 @@ extern const char bb_dev_null[];
 extern const char bb_busybox_exec_path[];
 /* util-linux manpage says /sbin:/bin:/usr/sbin:/usr/bin,
  * but I want to save a few bytes here */
-extern const char bb_PATH_root_path[]; /* "PATH=/sbin:/usr/sbin:/bin:/usr/bin" */
+extern const char bb_PATH_root_path[]; /* "PATH=/opt/sbin:/sbin:/usr/sbin:/opt/bin:/usr/local/bin:/bin:/usr/bin" */
 #define bb_default_root_path (bb_PATH_root_path + sizeof("PATH"))
-#define bb_default_path      (bb_PATH_root_path + sizeof("PATH=/sbin:/usr/sbin"))
+#define bb_default_path      (bb_PATH_root_path + sizeof("PATH=/opt/sbin:/sbin:/usr/sbin"))
 
 extern const int const_int_0;
 extern const int const_int_1;
diff -BurpN busybox-1.14.2/libbb/messages.c busybox/libbb/messages.c
--- busybox-1.14.2/libbb/messages.c	2009-06-21 22:32:00.000000000 +0000
+++ busybox/libbb/messages.c	2009-07-22 14:53:46.000000000 +0000
@@ -48,7 +48,7 @@ const char bb_default_login_shell[] ALIG
 /* util-linux manpage says /sbin:/bin:/usr/sbin:/usr/bin,
  * but I want to save a few bytes here. Check libbb.h before changing! */
 const char bb_PATH_root_path[] ALIGN1 =
-	"PATH=/sbin:/usr/sbin:/bin:/usr/bin" BB_ADDITIONAL_PATH;
+	"PATH=/opt/sbin:/sbin:/usr/sbin:/opt/bin:/usr/local/bin:/bin:/usr/bin" BB_ADDITIONAL_PATH;
 
 
 const int const_int_1 = 1;
