diff -BurpN busybox-1.14.2/libbb/procps.c busybox/libbb/procps.c
--- busybox-1.14.2/libbb/procps.c	2009-06-21 22:40:29.000000000 +0000
+++ busybox/libbb/procps.c	2009-07-22 14:53:46.000000000 +0000
@@ -166,7 +166,7 @@ void BUG_comm_size(void);
 procps_status_t* FAST_FUNC procps_scan(procps_status_t* sp, int flags)
 {
 	struct dirent *entry;
-	char buf[PROCPS_BUFSIZE];
+	char buf[PROCPS_BUFSIZE+1];
 	char filename[sizeof("/proc//cmdline") + sizeof(int)*3];
 	char *filename_tail;
 	long tasknice;
@@ -224,6 +224,7 @@ procps_status_t* FAST_FUNC procps_scan(p
 			n = read_to_buf(filename, buf);
 			if (n < 0)
 				break;
+			buf[n] = '\0';
 			cp = strrchr(buf, ')'); /* split into "PID (cmd" and "<rest>" */
 			/*if (!cp || cp[1] != ' ')
 				break;*/
