diff -x'.*' -x'*.[oa]' -urBp busybox-1.16/include/volume_id.h busybox/include/volume_id.h
--- busybox-1.16/include/volume_id.h	2010-03-20 05:58:07.000000000 +0300
+++ busybox/include/volume_id.h	2010-04-29 12:39:27.000000000 +0400
@@ -23,8 +23,9 @@ char *get_devname_from_uuid(const char *
 void display_uuid_cache(void);
 
 /* Returns:
- * 0: no UUID= or LABEL= prefix found
- * 1: UUID= or LABEL= prefix found. In this case,
- *    *fsname is replaced if device with such UUID or LABEL is found
+ * 0 : no UUID= or LABEL= prefix found
+ * 1 : UUID= or LABEL= prefix found. In this case,
+ *     *fsname is replaced if device with such UUID or LABEL is found
+ * -2: UUID= or LABEL= prefixes cannot be resolved
  */
 int resolve_mount_spec(char **fsname);
diff -x'.*' -x'*.[oa]' -urBp busybox-1.16/util-linux/volume_id/get_devname.c busybox/util-linux/volume_id/get_devname.c
--- busybox-1.16/util-linux/volume_id/get_devname.c	2010-03-20 05:58:07.000000000 +0300
+++ busybox/util-linux/volume_id/get_devname.c	2010-04-29 12:41:27.000000000 +0400
@@ -269,11 +269,12 @@ int resolve_mount_spec(char **fsname)
 		tmp = get_devname_from_uuid(*fsname + 5);
 	else if (strncmp(*fsname, "LABEL=", 6) == 0)
 		tmp = get_devname_from_label(*fsname + 6);
-
-	if (tmp == *fsname)
+	else
 		return 0; /* no UUID= or LABEL= prefix found */
 
-	if (tmp)
-		*fsname = tmp;
+	if (!tmp)
+		return -2; /* device defined by UUID= or LABEL= wasn't found */
+
+	*fsname = tmp;
 	return 1;
 }
