 Fix incorrect summary CPU percentage utilization on 2.4 kernels
 
diff -uBp busybox-1.13.2/procps/top.c busybox/procps/top.c
--- busybox-1.13.2/procps/top.c	2008-11-09 20:28:20.000000000 +0300
+++ busybox/procps/top.c	2009-02-23 19:23:08.000000000 +0300
@@ -200,7 +200,7 @@ static NOINLINE int read_cpu_jiffy(FILE
 			&p_jif->usr, &p_jif->nic, &p_jif->sys, &p_jif->idle,
 			&p_jif->iowait, &p_jif->irq, &p_jif->softirq,
 			&p_jif->steal);
-	if (ret > 4) {
+	if (ret >= 4) {
 		p_jif->total = p_jif->usr + p_jif->nic + p_jif->sys + p_jif->idle
 			+ p_jif->iowait + p_jif->irq + p_jif->softirq + p_jif->steal;
 		/* procps 2.x does not count iowait as busy time */
