diff -BurpN busybox-1.14.2/modutils/modprobe.c busybox/modutils/modprobe.c
--- busybox-1.14.2/modutils/modprobe.c	2009-06-21 22:46:43.000000000 +0000
+++ busybox/modutils/modprobe.c	2009-07-22 15:54:28.000000000 +0000
@@ -311,7 +311,7 @@ static void load_modules_dep(void)
 			G.num_unresolved_deps--;
 			llist_add_to(&m->deps, xstrdup(tokens[0]));
 			if (tokens[1])
-				string_to_llist(tokens[1], &m->deps, " ");
+				string_to_llist(tokens[1], &m->deps, " \t");
 		} else
 			DBG("skipping dep line");
 	}
