diff -urNBp busybox.orig/networking/udhcp/common.h busybox/networking/udhcp/common.h
--- busybox.orig/networking/udhcp/common.h	2011-05-18 04:00:00.000000000 +0400
+++ busybox/networking/udhcp/common.h	2011-05-20 13:54:28.000000000 +0400
@@ -251,6 +251,7 @@ struct option_set *udhcp_find_option(str
 /*** Logging ***/
 
 #if defined CONFIG_UDHCP_DEBUG && CONFIG_UDHCP_DEBUG >= 1
+# define IF_UDHCP_VERBOSE(...) __VA_ARGS__
 extern unsigned dhcp_verbose;
 # define log1(...) do { if (dhcp_verbose >= 1) bb_info_msg(__VA_ARGS__); } while (0)
 # if CONFIG_UDHCP_DEBUG >= 2
@@ -266,6 +267,7 @@ void udhcp_dump_packet(struct dhcp_packe
 #  define log3(...) ((void)0)
 # endif
 #else
+# define IF_UDHCP_VERBOSE(...)
 # define udhcp_dump_packet(...) ((void)0)
 # define log1(...) ((void)0)
 # define log2(...) ((void)0)
diff -urNBp busybox.orig/networking/udhcp/dhcpc.c busybox/networking/udhcp/dhcpc.c
--- busybox.orig/networking/udhcp/dhcpc.c	2011-05-18 04:00:00.000000000 +0400
+++ busybox/networking/udhcp/dhcpc.c	2011-05-20 13:55:56.000000000 +0400
@@ -1113,11 +1113,7 @@ int udhcpc_main(int argc UNUSED_PARAM, c
 
 	/* Parse command line */
 	/* O,x: list; -T,-t,-A take numeric param */
-	opt_complementary = "O::x::T+:t+:A+"
-#if defined CONFIG_UDHCP_DEBUG && CONFIG_UDHCP_DEBUG >= 1
-		":vv"
-#endif
-		;
+	opt_complementary = "O::x::T+:t+:A+" IF_UDHCP_VERBOSE(":vv") ;
 	IF_LONG_OPTS(applet_long_options = udhcpc_longopts;)
 	opt = getopt32(argv, "CV:H:h:F:i:np:qRr:s:T:t:SA:O:ox:fB"
 		USE_FOR_MMU("b")
@@ -1131,9 +1127,7 @@ int udhcpc_main(int argc UNUSED_PARAM, c
 		, &list_O
 		, &list_x
 		IF_FEATURE_UDHCP_PORT(, &str_P)
-#if defined CONFIG_UDHCP_DEBUG && CONFIG_UDHCP_DEBUG >= 1
-		, &dhcp_verbose
-#endif
+		IF_UDHCP_VERBOSE(, &dhcp_verbose)
 		);
 	if (opt & (OPT_h|OPT_H))
 		client_config.hostname = alloc_dhcp_option(DHCP_HOST_NAME, str_h, 0);
diff -urNBp busybox.orig/networking/udhcp/dhcpd.c busybox/networking/udhcp/dhcpd.c
--- busybox.orig/networking/udhcp/dhcpd.c	2010-10-13 22:19:07.000000000 +0400
+++ busybox/networking/udhcp/dhcpd.c	2011-05-20 13:56:19.000000000 +0400
@@ -305,9 +305,7 @@ int udhcpd_main(int argc UNUSED_PARAM, c
 #endif
 	opt = getopt32(argv, "fSv"
 		IF_FEATURE_UDHCP_PORT("P:", &str_P)
-#if defined CONFIG_UDHCP_DEBUG && CONFIG_UDHCP_DEBUG >= 1
-		, &dhcp_verbose
-#endif
+		IF_UDHCP_VERBOSE(, &dhcp_verbose)
 		);
 	if (!(opt & 1)) { /* no -f */
 		bb_daemonize_or_rexec(0, argv);
