Add workaround for IPV6_RECVPKTINFO in recv_from_to()

--- a/libbb/udp_io.c	2009-09-26 17:14:57.000000000 +0400
+++ a/libbb/udp_io.c	2009-11-18 21:18:59.000000000 +0300
@@ -156,8 +156,11 @@ recv_from_to(int fd, void *buf, size_t l
 		}
 # if ENABLE_FEATURE_IPV6 && defined(IPV6_PKTINFO)
 		if (cmsgptr->cmsg_level == IPPROTO_IPV6
-		 && cmsgptr->cmsg_type == IPV6_PKTINFO
-		) {
+		 && (cmsgptr->cmsg_type == IPV6_PKTINFO
+#if defined(IPV6_2292PKTINFO) && defined(IPV6_RECVPKTINFO)
+            		 || cmsgptr->cmsg_type == IPV6_2292PKTINFO
+#endif
+		)) {
 #  define pktinfo(cmsgptr) ( (struct in6_pktinfo*)(CMSG_DATA(cmsgptr)) )
 			to->sa_family = AF_INET6;
 			/*to6->sin6_addr = pktinfo(cmsgptr)->ipi6_addr; - may be unaligned */
