--- vsftpd.orig/parseconf.c	2008-01-08 13:23:50.000000000 +0300
+++ vsftpd/parseconf.c	2008-01-08 17:54:55.000000000 +0300
@@ -44,6 +44,7 @@
   { "anon_upload_enable", &tunable_anon_upload_enable },
   { "anon_mkdir_write_enable", &tunable_anon_mkdir_write_enable },
   { "anon_other_write_enable", &tunable_anon_other_write_enable },
+  { "anon_allow_writable_root", &tunable_anon_allow_writable_root },
   { "chown_uploads", &tunable_chown_uploads },
   { "connect_from_port_20", &tunable_connect_from_port_20 },
   { "xferlog_enable", &tunable_xferlog_enable },
--- vsftpd-2.1.0/tunables.c	2009-02-22 11:00:15.000000000 +0300
+++ vsftpd/tunables.c	2009-02-22 11:01:35.000000000 +0300
@@ -17,6 +17,7 @@
 int tunable_anon_upload_enable;
 int tunable_anon_mkdir_write_enable;
 int tunable_anon_other_write_enable;
+int tunable_anon_allow_writable_root;
 int tunable_chown_uploads;
 int tunable_connect_from_port_20;
 int tunable_xferlog_enable;
@@ -153,6 +154,7 @@
   tunable_anon_upload_enable = 0;
   tunable_anon_mkdir_write_enable = 0;
   tunable_anon_other_write_enable = 0;
+  tunable_anon_allow_writable_root = 0;
   tunable_chown_uploads = 0;
   tunable_connect_from_port_20 = 0;
   tunable_xferlog_enable = 0;
--- vsftpd.orig/tunables.h	2008-01-08 13:23:50.000000000 +0300
+++ vsftpd/tunables.h	2008-01-08 17:55:28.000000000 +0300
@@ -12,6 +12,7 @@
 extern int tunable_anon_upload_enable;        /* Enable STOR for anon users */
 extern int tunable_anon_mkdir_write_enable;   /* MKD for anon */
 extern int tunable_anon_other_write_enable;   /* APPE DELE RMD RNFR for anon */
+extern int tunable_anon_allow_writable_root;  /* allow writable / for anon */
 extern int tunable_chown_uploads;             /* chown() anon uploaded files */
 extern int tunable_connect_from_port_20;      /* PORT connects from port 20 */
 extern int tunable_xferlog_enable;            /* Log transfers to a file */
--- vsftpd.orig/secutil.c	2011-12-17 23:44:20.000000000 +0400
+++ vsftpd/secutil.c	2012-01-01 20:48:05.000000000 +0400
@@ -11,6 +11,7 @@
 #include "sysstr.h"
 #include "utility.h"
 #include "sysdeputil.h"
+#include "tunables.h"
 
 void
 vsf_secutil_change_credentials(const struct mystr* p_user_str,
@@ -134,7 +135,7 @@ vsf_secutil_change_credentials(const str
    */
   if (options & VSF_SECUTIL_OPTION_CHROOT)
   {
-    if (vsf_sysutil_write_access("/"))
+    if (!tunable_anon_allow_writable_root && vsf_sysutil_write_access("/"))
     {
       die("vsftpd: refusing to run with writable root inside chroot()");
     }
