--- vsftpd.orig/sysutil.c	2008-01-08 13:23:50.000000000 +0300
+++ vsftpd/sysutil.c	2008-01-08 22:30:53.000000000 +0300
@@ -2613,17 +2613,21 @@
 #ifdef LOG_FTP
   facility = LOG_FTP;
 #endif
-  openlog("vsftpd", LOG_NDELAY, facility);
+  openlog("vsftpd", LOG_NDELAY | LOG_PID, facility);
 }
 
 void
 vsf_sysutil_syslog(const char* p_text, int severe)
 {
+  char *text;
   int prio = LOG_INFO;
   if (severe)
   {
     prio = LOG_WARNING;
   }
+  /* skip date & pid */
+  if ((text = strstr(p_text, "[pid ")) && (text = strchr(text, ']')) && (text[1] == ' '))
+    p_text = text + 2;
   syslog(prio, "%s", p_text);
 }
 
