--- rp-l2tp.orig/dgram.c	2003-01-11 08:29:03.000000000 +0300
+++ rp-l2tp/dgram.c	2007-10-22 23:06:27.416957928 +0400
@@ -535,7 +535,8 @@
     unsigned char *len_ptr = NULL;
 
     DBG(l2tp_db(DBG_XMIT_RCV,
-	   "dgram_send_to_wire() -> %s\n",
+	   "dgram_send_to_wire(%s:%d) -> %s\n", 
+	   inet_ntoa(to->sin_addr), ntohs(to->sin_port),
 	   l2tp_debug_describe_dgram(dgram)));
     buf[0] = dgram->bits;
     buf[1] = dgram->version;
--- rp-l2tp.orig/main.c	2002-09-30 23:45:00.000000000 +0400
+++ rp-l2tp/main.c	2007-10-22 19:15:12.610550272 +0400
@@ -23,6 +23,7 @@
 #include <signal.h>
 #include <fcntl.h>
 #include <stdlib.h>
+#include <syslog.h>
 
 static void
 usage(int argc, char *argv[], int exitcode)
@@ -63,6 +64,7 @@
 	}
     }
 
+    openlog(argv[0], LOG_PID, LOG_DAEMON);
     l2tp_random_init();
     l2tp_tunnel_init(es);
     l2tp_peer_init();
--- rp-l2tp.orig/utils.c	2002-09-30 23:45:00.000000000 +0400
+++ rp-l2tp/utils.c	2007-10-22 19:50:21.137005528 +0400
@@ -27,6 +27,7 @@
 #include <dlfcn.h>
 #include <string.h>
 #include <errno.h>
+#include <syslog.h>
 
 #include "l2tp.h"
 
@@ -88,6 +89,8 @@
     va_end(ap);
     errmsg[MAX_ERRMSG_LEN-1] = 0;
     fprintf(stderr, "Error: %s\n", errmsg);
+    
+    vsyslog(LOG_ERR, fmt, ap);
 }
 
 /**********************************************************************
