--- rp-l2tp/tunnel.c.orig	2007-10-24 21:41:53.250600304 +0400
+++ rp-l2tp/tunnel.c	2007-10-24 21:43:26.853370520 +0400
@@ -665,8 +665,8 @@
 
     if (!tunnel) {
 	/* TODO: Send error message back? */
-	l2tp_set_errmsg("Invalid control message - unknown tunnel ID %d",
-		   (int) dgram->tid);
+	DBG(l2tp_db(DBG_TUNNEL, "Invalid control message - unknown tunnel ID %d",
+		   (int) dgram->tid));
 	return;
     }
 
