diff -ur rp-l2tp.orig/configure rp-l2tp/configure
--- rp-l2tp.orig/configure	2002-09-30 22:46:52.000000000 +0400
+++ rp-l2tp/configure	2005-12-30 20:05:39.349692032 +0300
@@ -1766,6 +1766,7 @@
 s%@infodir@%$infodir%g
 s%@mandir@%$mandir%g
 s%@CC@%$CC%g
+s%@AR@%$AR%g
 s%@INSTALL_PROGRAM@%$INSTALL_PROGRAM%g
 s%@INSTALL_SCRIPT@%$INSTALL_SCRIPT%g
 s%@INSTALL_DATA@%$INSTALL_DATA%g
diff -ur rp-l2tp.orig/handlers/Makefile.in rp-l2tp/handlers/Makefile.in
--- rp-l2tp.orig/handlers/Makefile.in	2002-09-30 23:45:00.000000000 +0400
+++ rp-l2tp/handlers/Makefile.in	2005-12-30 20:05:39.275703280 +0300
@@ -19,19 +19,19 @@
 all: $(HANDLERS) l2tp-control
 
 l2tp-control: cmd-control.o
-	gcc -o l2tp-control cmd-control.o
+	@CC@ -o l2tp-control cmd-control.o
 
 cmd-control.o: cmd-control.c
-	gcc -c -o $@ $^
+	@CC@ -c -o $@ $^
 
 %.o: %.c
-	gcc $(CFLAGS) -fPIC -c -o $@ $<
+	@CC@ $(CFLAGS) -fPIC -c -o $@ $<
 
 sync-pppd.so: pty.o sync-pppd.o
-	gcc -shared -o $@ $^
+	@CC@ -shared -o $@ $^
 
 cmd.so: cmd.o dstring.o
-	gcc -shared -o $@ $^
+	@CC@ -shared -o $@ $^
 
 clean:
 	rm -f *.so *.o *~
diff -ur rp-l2tp.orig/libevent/Makefile.in rp-l2tp/libevent/Makefile.in
--- rp-l2tp.orig/libevent/Makefile.in	2002-09-30 23:45:00.000000000 +0400
+++ rp-l2tp/libevent/Makefile.in	2005-12-30 20:05:39.275703280 +0300
@@ -18,20 +18,20 @@
 
 libevent.a: $(OBJS)
 	rm -f libevent.a
-	ar -cq libevent.a $(OBJS)
+	@AR@ -cq libevent.a $(OBJS)
 	@RANLIB@ libevent.a
 
 event.o: event.c $(HDRS)
-	gcc $(CFLAGS) -c -o event.o event.c
+	@CC@ $(CFLAGS) -c -o event.o event.c
 
 hash.o: hash.c $(HDRS)
-	gcc $(CFLAGS) -c -o hash.o hash.c
+	@CC@ $(CFLAGS) -c -o hash.o hash.c
 
 event_sig.o: event_sig.c $(HDRS)
-	gcc $(CFLAGS) -c -o event_sig.o event_sig.c
+	@CC@ $(CFLAGS) -c -o event_sig.o event_sig.c
 
 event_tcp.o: event_tcp.c $(HDRS)
-	gcc $(CFLAGS) -c -o event_tcp.o event_tcp.c
+	@CC@ $(CFLAGS) -c -o event_tcp.o event_tcp.c
 
 clean: FORCE
 	rm -f *.a *.o *~
diff -ur rp-l2tp.orig/Makefile.in rp-l2tp/Makefile.in
--- rp-l2tp.orig/Makefile.in	2004-07-01 19:17:11.000000000 +0400
+++ rp-l2tp/Makefile.in	2005-12-30 20:06:45.321662768 +0300
@@ -30,7 +30,7 @@
 EXTRA_LIBS=@LIBEVENT@/*.o -ldl
 
 SRCS=$(OBJS:.o=.c)
-CFLAGS=-g -I.. -I@LIBEVENT@ -Wall -Wstrict-prototypes -ansi -pedantic -D_GNU_SOURCE -DVERSION=\"$(VERSION)\" -DPREFIX=\"$(prefix)\"
+CFLAGS=@CFLAGS@ -I.. -I@LIBEVENT@ -Wall -Wstrict-prototypes -ansi -pedantic -D_GNU_SOURCE -DVERSION=\"$(VERSION)\" -DPREFIX=\"$(prefix)\"
 
 TARGETS=l2tpd libl2tp.a handlers
 
@@ -38,8 +38,8 @@
 
 libl2tp.a: $(OBJS)
 	rm -f $@
-	ar -rc $@ $^
-	ranlib $@ || true
+	@AR@ -rc $@ $^
+	@RANLIB@ $@ || true
 
 l2tpd: libl2tp.a libevent/libevent.a
 	$(MAKE) -C handlers
