--- src/shared/bcmrobo.c.orig	2007-10-16 00:17:58.457531336 +0400
+++ src/shared/bcmrobo.c	2007-10-16 00:22:10.053282968 +0400
@@ -74,7 +74,9 @@
 #define REG_VLAN_CTRL4	0x04	/* VLAN Control 4 register */
 #define REG_VLAN_CTRL5	0x05	/* VLAN Control 5 register */
 #define REG_VLAN_ACCESS	0x06	/* VLAN Table Access register */
+#define REG_VLAN_ACCESS_5365	0x08	/* 5365 VLAN Table Access register */
 #define REG_VLAN_WRITE	0x08	/* VLAN Write register */
+#define REG_VLAN_WRITE_5365	0x0A	/* 5365 VLAN Write register */
 #define REG_VLAN_READ	0x0C	/* VLAN Read register */
 #define REG_VLAN_PTAG0	0x10	/* VLAN Default Port Tag register - port 0 */
 #define REG_VLAN_PTAG1	0x12	/* VLAN Default Port Tag register - port 1 */
@@ -1003,7 +1005,21 @@ bcm_robo_config_vlan(robo_info_t *robo, uint8 *mac_addr)
 
 		val32 = (untag |			/* untag enable */
 		         member);			/* vlan members */
-		if (robo->devid == DEVID5325) {
+		if (robo->sih->chip == BCM5365_CHIP_ID) 
+		{
+			/* VLAN Write Register (Page 0x34, Address 0x0A) */
+			val32 = ((1 << 14) |	/* valid write */
+				 (untag << 1) |	/* untag enable */
+				 member);		/* vlan members */
+			robo->ops->write_reg(robo, PAGE_VLAN, REG_VLAN_WRITE_5365, &val32,
+			                     sizeof(val32));
+			/* VLAN Table Access Register (Page 0x34, Address 0x08) */
+			val16 = ((1 << 13) | 	/* start command */
+				 (1 << 12) |	/* write state */
+				 vid);		/* vlan id */
+			robo->ops->write_reg(robo, PAGE_VLAN, REG_VLAN_ACCESS_5365, &val16,
+			                     sizeof(val16));
+		} else if (robo->devid == DEVID5325) {
 			if (robo->corerev < 3) {
 				val32 |= ((1 << 20) |		/* valid write */
 				          ((vid >> 4) << 12));	/* vlan id bit[11:4] */
@@ -1150,6 +1165,13 @@ bcm_robo_enable_switch(robo_info_t *robo)
 		robo->ops->write_reg(robo, PAGE_CTRL, REG_CTRL_IMP, &val8, sizeof(val8));
 	}
 
+	/* Enable WAN port (#0) on the asus wl-500g deluxe boxes */
+	if (robo->sih->chip == BCM5365_CHIP_ID) 
+	{
+		val8 = 0;
+		robo->ops->write_reg(robo, PAGE_CTRL, REG_CTRL_PORT0, &val8, sizeof(val8));
+	}
+
 	/* Disable management interface access */
 	if (robo->ops->disable_mgmtif)
 		robo->ops->disable_mgmtif(robo);
