Create sysfs 'device' link

diff --git src/et.orig/sys/et_linux.c src/et/sys/et_linux.c
--- src/et.orig/sys/et_linux.c
+++ src/et/sys/et_linux.c
@@ -208,6 +208,9 @@ et_probe(struct pci_dev *pdev, const str
 		osl_detach(osh);
 		return -ENOMEM;
 	}
+
+	SET_MODULE_OWNER(dev);
+	SET_NETDEV_DEV(dev, &pdev->dev);
 #else
 	if (!(dev = (struct net_device *) MALLOC(osh, sizeof(struct net_device)))) {
 		ET_ERROR(("%s: out of memory, malloced %d bytes\n", __func__,
