diff -urBp igmpproxy.orig/src/igmpproxy.c igmpproxy/src/igmpproxy.c
--- igmpproxy.orig/src/igmpproxy.c	2009-09-15 19:25:13.000000000 +0400
+++ igmpproxy/src/igmpproxy.c	2009-09-15 19:56:19.000000000 +0400
@@ -126,6 +126,20 @@ int main( int ArgCn, char *ArgVc[] ) {
             break;
         }
 
+	if ( !Log2Stderr ) {
+
+	    // Only daemon goes past this line...
+	    if (fork()) exit(0);
+
+	    // Detach daemon from terminal
+	    if ( close( 0 ) < 0 || close( 1 ) < 0 || close( 2 ) < 0
+		 || open( "/dev/null", 0 ) != 0 || dup2( 0, 1 ) < 0 || dup2( 0, 2 ) < 0
+		 || setpgrp() < 0
+	       ) {
+		my_log( LOG_ERR, errno, "failed to detach daemon" );
+	    }
+	}
+
         // Go to the main loop.
         igmpProxyRun();
     
