diff -urBp a/src/igmpproxy.h igmpproxy/src/igmpproxy.h
--- a/src/igmpproxy.h	2011-05-27 19:33:24.000000000 +0400
+++ igmpproxy/src/igmpproxy.h	2011-05-27 19:56:41.000000000 +0400
@@ -142,15 +142,15 @@ struct SubnetList {
 
 struct IfDesc {
     char                Name[IF_NAMESIZE];
-    struct in_addr      InAdr;          /* == 0 for non IP interfaces */            
+    struct in_addr      InAdr;          /* == 0 for non IP interfaces */
     short               Flags;
     short               state;
+    short               index;
+    uint8_t             robustness;
+    uint8_t             threshold;   /* ttl limit */
+    unsigned int        ratelimit;
     struct SubnetList*  allowednets;
     struct SubnetList*  allowedgroups;
-    unsigned int        robustness;
-    unsigned char       threshold;   /* ttl limit */
-    unsigned int        ratelimit; 
-    unsigned int        index;
 };
 
 // Keeps common configuration settings 
@@ -165,7 +165,7 @@ struct Config {
     unsigned int        lastMemberQueryInterval;
     unsigned int        lastMemberQueryCount;
     // Set if upstream leave messages should be sent instantly..
-    unsigned short      fastUpstreamLeave;
+    unsigned int        fastUpstreamLeave;
 };
 
 // Defines the Index of the upstream VIF...
@@ -176,15 +176,15 @@ extern int upStreamVif;
 void buildIfVc( void );
 struct IfDesc *getIfByName( const char *IfName );
 struct IfDesc *getIfByIx( unsigned Ix );
-struct IfDesc *getIfByAddress( uint32_t Ix );
+struct IfDesc *getIfByAddress( uint32_t ipaddr );
 int isAdressValidForIf(struct IfDesc* intrface, uint32_t ipaddr);
 
 /* mroute-api.c
  */
 struct MRouteDesc {
+    vifi_t          InVif;
+    uint8_t         TtlVc[ MAX_MC_VIFS ];
     struct in_addr  OriginAdr, McAdr;
-    short           InVif;
-    uint8_t           TtlVc[ MAX_MC_VIFS ];
 };
 
 // IGMP socket as interface for the mrouted API
diff -urBp a/src/request.c igmpproxy/src/request.c
--- a/src/request.c	2011-05-27 19:33:24.000000000 +0400
+++ igmpproxy/src/request.c	2011-05-27 19:53:55.000000000 +0400
@@ -194,7 +194,7 @@ void sendGroupSpecificMemberQuery(void *
 void sendGeneralMembershipQuery() {
     struct  Config  *conf = getCommonConfig();
     struct  IfDesc  *Dp;
-    int             Ix;
+    unsigned        Ix;
 
     // Loop through all downstream vifs...
     for ( Ix = 0; (Dp = getIfByIx(Ix)); Ix++ ) {
