diff -BurpN libusb.orig/libusb/Makefile.am libusb/libusb/Makefile.am
--- libusb.orig/libusb/Makefile.am	2008-06-17 23:13:02.000000000 +0000
+++ libusb/libusb/Makefile.am	2011-01-13 15:28:36.005997000 +0000
@@ -4,6 +4,6 @@ lib_LTLIBRARIES = libusb.la
 libusb_la_SOURCES = core.c usbi.h
 libusb_la_CFLAGS = -fvisibility=hidden $(AM_CFLAGS) $(LIBUSB_1_0_CFLAGS)
 libusb_la_LIBADD = $(LIBUSB_1_0_LIBS)
-libusb_la_LDFLAGS = -version-info $(LT_MAJOR):$(LT_REVISION):$(LT_AGE) \
+libusb_la_LDFLAGS = -avoid-version \
 	-release 0.1
 
diff -BurpN libusb.orig/libusb/Makefile.in libusb/libusb/Makefile.in
--- libusb.orig/libusb/Makefile.in	2009-07-09 21:15:51.000000000 +0000
+++ libusb/libusb/Makefile.in	2011-01-13 15:28:48.075997000 +0000
@@ -204,7 +204,7 @@ lib_LTLIBRARIES = libusb.la
 libusb_la_SOURCES = core.c usbi.h
 libusb_la_CFLAGS = -fvisibility=hidden $(AM_CFLAGS) $(LIBUSB_1_0_CFLAGS)
 libusb_la_LIBADD = $(LIBUSB_1_0_LIBS)
-libusb_la_LDFLAGS = -version-info $(LT_MAJOR):$(LT_REVISION):$(LT_AGE) \
+libusb_la_LDFLAGS = -avoid-version \
 	-release 0.1
 
 all: all-am
