diff -uBp router/rc/common_ex.c gateway/rc/common_ex.c
--- router/rc/common_ex.c	2009-02-20 10:18:33.000000000 +0300
+++ gateway/rc/common_ex.c	2009-09-21 16:01:01.000000000 +0400
@@ -709,25 +709,6 @@ char *pppstatus(char *buf)
 }
 
 
-/* 
- * Kills process whose PID is stored in plaintext in pidfile
- * @param	pidfile	PID file, signal
- * @return	0 on success and errno on failure
- */
-int
-kill_pidfile_s(char *pidfile, int sig)
-{
-	FILE *fp = fopen(pidfile, "r");
-	char buf[256];
-
-	if (fp && fgets(buf, sizeof(buf), fp)) {
-		pid_t pid = strtoul(buf, NULL, 0);
-		fclose(fp);
-		return kill(pid, sig);
-  	} else
-		return errno;
-}
-
 #ifdef RT2400_SUPPORT
 void write_rt2400_conf(void)
 {
diff -uBp router/rc/services_ex.c gateway/rc/services_ex.c
--- router/rc/services_ex.c	2009-05-23 21:50:18.000000000 +0400
+++ gateway/rc/services_ex.c	2009-09-21 16:08:30.000000000 +0400
@@ -1989,11 +2002,7 @@ int service_handle(void)
 			nvram_match("wan0_proto", "bigpond"))
 		{		
 			snprintf(tmp, sizeof(tmp), "/var/run/udhcpc%d.pid", 0);
-			if ((str = file2str(tmp))) {
-				pid = atoi(str);
-				free(str);			
-				kill(pid, SIGUSR2);
-			}
+			kill_pidfile_s(tmp, SIGUSR2);
 		}
 		else 
 		{			
@@ -2010,11 +2019,7 @@ int service_handle(void)
 			nvram_match("wan0_proto", "bigpond"))
 		{
 			snprintf(tmp, sizeof(tmp), "/var/run/udhcpc%d.pid", 0);
-			if ((str = file2str(tmp))) {
-				pid = atoi(str);
-				free(str);
-				kill(pid, SIGUSR1);
-			}
+			kill_pidfile_s(tmp, SIGUSR1);
 		}
 		else 
 		{
diff -uBp router/rc/watchdog.c gateway/rc/watchdog.c
--- router/rc/watchdog.c	2009-03-02 21:49:13.000000000 +0300
+++ gateway/rc/watchdog.c	2009-09-21 16:19:50.000000000 +0400
@@ -594,69 +594,36 @@ int rcamd_processcheck()
 
 int notice_rcamd(int flag)
 {
-	 int rcamdpid=-1;
-	 //printf("Send signal : %d %d\n", rcamdpid, flag);
-	 if (rcamdpid==-1)
-	 {
-			FILE *fp;
+	int ret = -1;
 
-			if ((fp=fopen("/var/run/rcamd.pid","r"))!=NULL)
-			{
-				fscanf(fp,"%d", &rcamdpid);
-				fclose(fp);
-			}			
-	 }
-	 if (rcamdpid!=-1) 
-	 {
-		 if (flag)
-			kill(rcamdpid, SIGUSR1);
-		 else
-			kill(rcamdpid, SIGUSR2);	
-
-		 return 1;
-	 }
-	 return 0;
+	if (flag)
+		ret = kill_pidfile_s("/var/run/rcamd.pid", SIGUSR1);
+	else
+		ret = kill_pidfile_s("/var/run/rcamd.pid", SIGUSR2);
+	
+	return (ret == 0);
 }
 
 int refresh_rcamd(void)
 {
-	FILE *fp;
-	int rcamdpid=-1;
-
-	if ((fp=fopen("/var/run/rcamd.pid","r"))!=NULL)
+	if (kill_pidfile_s("/var/run/rcamd.pid", SIGUSR1) == 0)
 	{
-		fscanf(fp,"%d", &rcamdpid);
-		fclose(fp);
 		unlink("/var/run/rcamd.pid");
-		kill(rcamdpid, SIGUSR1);
 	}			
 	else 
 	{	
 		eval("killall", "rcamd");
 	}
 
-	if ((fp=fopen("/var/run/rcamdmain.pid","r"))!=NULL)
-	{
-		fscanf(fp,"%d", &rcamdpid);
-		fclose(fp);
-		kill(rcamdpid, SIGUSR1);
-	}		
+	kill_pidfile_s("/var/run/rcamdmain.pid", SIGUSR1);
 	return 0;
 }
 
 int refresh_wave(void)
 {
-	FILE *fp;
-	int wavepid=-1;
-
 	eval("killall", "waveserver");
 
-	if ((fp=fopen("/var/run/waveservermain.pid","r"))!=NULL)
-	{
-		fscanf(fp,"%d", &wavepid);
-		fclose(fp);
-		kill(wavepid, SIGUSR1);
-	}			
+	kill_pidfile_s("/var/run/waveservermain.pid", SIGUSR1);
 	return 0;
 }
 
