diff -BurpN gateway.orig/rc/services_ex.c gateway/rc/services_ex.c
--- gateway.orig/rc/services_ex.c	2009-01-05 23:26:40.000000000 +0500
+++ gateway/rc/services_ex.c	2009-02-22 18:20:11.000000000 +0500
@@ -854,7 +854,8 @@ start_usb(void)
 	eval("insmod", "usbcore");
 	eval("insmod", "usb-ohci");
 	eval("insmod", "usb-uhci");
-	eval("insmod", "ehci-hcd");
+	if (nvram_invmatch("usb20_disable_x", "1"))
+		eval("insmod", "ehci-hcd");
 
 	/* mount usbfs */
 	mount("usbfs", "/proc/bus/usb", "usbfs", MS_MGC_VAL, NULL);
