diff -urBp rc.old/services_ex.c rc/services_ex.c
--- router/rc/services_ex.c	2009-05-13 22:24:54.000000000 +0400
+++ gateway/rc/services_ex.c	2009-05-13 22:42:26.000000000 +0400
@@ -1861,6 +1861,11 @@ hotplug_usb(void)
 		}
 		else if(strncmp(interface, "1/", 2)==0)
 			return 0;
+
+		/* ignore hubs */
+		if (strncmp(interface, "9/", 2)==0)
+			return 0;
+
 		i=0;
 		isweb = WEB_NONE;
 		while(PWCLIST[i]!=NULL)
@@ -1886,7 +1891,7 @@ hotplug_usb(void)
 	else return 0;
 
 usbhandler:	
-	if (strstr(action, "add"))
+	if (strncasecmp(action, "add", 3))
 	{		
 		nvram_set("usb_device", "1");
 		if (isweb==WEB_NONE)
