diff -BurpN '-x*.o' router/rc/ppp.c gateway/rc/ppp.c
--- router/rc/ppp.c	2008-03-19 22:07:24.000000000 +0500
+++ gateway/rc/ppp.c	2009-02-01 01:18:24.000000000 +0500
@@ -92,7 +92,7 @@
 
 	wan_up(wan_ifname);
 
-	logmessage(nvram_safe_get("wan_proto_t"), "connect to ISP");
+	logmessage(nvram_safe_get("wan_proto_t"), "connected to ISP");
 	wanmessage("");
 
 	dprintf("done\n");
diff -uBp router/rc/network.c gateway/rc/network.c
--- router/rc/network.c	2009-02-17 21:15:16.000000000 +0300
+++ gateway/rc/network.c	2009-02-21 23:09:16.000000000 +0300
@@ -789,6 +789,8 @@ start_wan(void)
 			if (demand) 
 			{
 				int timeout = 5;
+				char *ping_argv[] = { "ping", "-c1", "", NULL};
+
 				/* Wait for pppx to be created */
 				while (ifconfig(wan_ifname, IFUP, NULL, NULL) && timeout--)
 					sleep(1);
@@ -817,6 +819,10 @@ start_wan(void)
 				*/
 
 				preset_wan_routes(wan_ifname);
+
+				/* Stimulate link up */
+				ping_argv[2] = nvram_safe_get(strcat_r(prefix, "gateway", tmp));
+				_eval(ping_argv, NULL, 0, &pid);
 			}
 #ifdef ASUS_EXT
 			nvram_set("wan_ifname_t", wan_ifname);
