diff -Bbup services.c.orig services.c
--- gateway/rc/services.c.orig	2008-03-19 20:07:24.000000000 +0300
+++ gateway/rc/services.c	2009-02-19 17:01:53.000000000 +0300
@@ -169,6 +169,26 @@ stop_telnetd(void)
 	return ret;
 }
 
+int
+start_dropbear(void)
+{
+	if (nvram_match("ssh_enable", "0"))
+		return 0;
+
+	int ret = eval("dropbearstart");
+
+	dprintf("done\n");
+	return ret;
+}
+
+int
+stop_dropbear(void)
+{
+	int ret = eval("killall", "dropbear");
+
+	dprintf("done\n");
+	return ret;
+}
 
 int
 start_snmpd(void)
@@ -343,6 +363,7 @@ int
 start_services(void)
 {
 	start_telnetd();
+	start_dropbear();
 	start_httpd();
 	start_dns();
 	start_dhcpd();
@@ -376,6 +397,7 @@ stop_services(void)
 	stop_dhcpd();
 	stop_dns();
 	stop_httpd();
+	stop_dropbear();
 	stop_telnetd();
 
 	dprintf("done\n");
diff -Bbup firewall_ex.c.orig firewall_ex.c
--- gateway/rc/firewall_ex.c.orig	2009-01-28 22:21:29.000000000 +0300
+++ gateway/rc/firewall_ex.c	2009-02-19 16:51:11.000000000 +0300
@@ -782,6 +782,8 @@ filter_setting(char *wan_if, char *wan_i
 		}
 
 		// Firewall between WAN and Local
+		if (nvram_match("ssh_enable", "1"))
+			fprintf(fp, "-A INPUT -p tcp --syn --dport %s -j %s\n", nvram_safe_get("ssh_port"), logaccept);
 		if (nvram_invmatch("http_lanport", "80"))
 			fprintf(fp, "-A INPUT -p tcp -m tcp -d %s --dport 80 -j %s\n", nvram_safe_get("lan_ipaddr"), logaccept);
 		if (nvram_match("misc_http_x", "1"))
