diff -BurpN gateway.orig/rc/watchdog.c gateway/rc/watchdog.c
--- gateway.orig/rc/watchdog.c	2009-03-02 23:50:10.000000000 +0500
+++ gateway/rc/watchdog.c	2009-03-02 23:49:13.000000000 +0500
@@ -384,8 +384,8 @@ int ntp_timesync(void)
 		      		sync_interval = (atoi(nvram_safe_get("ntp_interval_x")) ? : 2) * 360;
 			  	logmessage("ntp client", "Synchronizing time with %s...", nvram_safe_get("ntp_servers"));
 
-				stop_upnp();
-				start_upnp();
+				//stop_upnp();
+				//start_upnp();
 		   	}	
 		  	else sync_interval=1;
 	
