diff -BurpN '-x*.o' router/rc/firewall_ex.c gateway/rc/firewall_ex.c
--- router/rc/firewall_ex.c	2008-12-28 13:56:49.000000000 +0500
+++ gateway/rc/firewall_ex.c	2009-02-24 23:01:31.000000000 +0500
@@ -659,9 +659,9 @@ void nat_setting(char *wan_if, char *wan
 			fprintf(fp, "-A POSTROUTING -o %s ! -s %s -j MASQUERADE\n", 
 	   			nvram_get("wan0_ifname"), nvram_get("wanx_ipaddr"));
 
-		// masquerade lan to lan
+		// snat lan to lan
 		ip2class(lan_ip, nvram_safe_get("lan_netmask"), lan_class);
-		fprintf(fp, "-A POSTROUTING -o %s -s %s -d %s -j MASQUERADE\n", lan_if, lan_class, lan_class);
+		fprintf(fp, "-A POSTROUTING -o %s -s %s -d %s -j SNAT --to-source %s\n", lan_if, lan_class, lan_class, lan_ip);
 	}
 
 	/* Trigger port setting */
