--- igmpproxy-0.1_beta4/src/igmpproxy.c    2009-06-10 10:30:34.000000000 +0200
+++ igmpproxy/src/igmpproxy.c 2009-06-10 11:07:07.000000000 +0200
@@ -71,8 +71,10 @@
 */   
 int main( int ArgCn, char *ArgVc[] ) {
 
+    int c;
+
     // Parse the commandline options and setup basic settings..
-    for (int c; (c = getopt(ArgCn, ArgVc, "vdh")) != -1;) {
+    for (c; (c = getopt(ArgCn, ArgVc, "vdh")) != -1;) {
         switch (c) {
         case 'd':
             Log2Stderr = true;
