--- igmpproxy/src/igmp.c.orig	2007-10-20 17:52:40.670596600 +0400
+++ igmpproxy/src/igmp.c	2007-10-20 18:06:25.299233992 +0400
@@ -117,6 +117,13 @@
 
     //IF_DEBUG log(LOG_DEBUG, 0, "Got a IGMP request to process...");
 
+    /* filter local multicast 239.255.255.250 */
+    if (dst == htonl(0xEFFFFFFA))
+    {
+        log(LOG_NOTICE, 0, "The IGMP message was local multicast. Ignoring.");
+        return;
+    }
+
     /* 
      * this is most likely a message from the kernel indicating that
      * a new src grp pair message has arrived and so, it would be 
