--- a/CHANGES	2010/03/05 12:15:33	1.97
+++ b/CHANGES	2010/03/10 07:57:36	1.98
@@ -1,4 +1,8 @@
-$Id: CHANGES,v 1.97 2010/03/05 12:15:33 psavola Exp $
+$Id: CHANGES,v 1.98 2010/03/10 07:57:36 psavola Exp $
+
+10/03/2010	Fix radvd skipping multiple interfaces when
+                UnicastOnly is on or AdvSendAdvert is off.
+		This got broken in radvd 1.3. Patch from Cedric BAIL.
 
 05/03/2010	Release 1.6.
 
--- a/radvd.c	2010/01/28 13:34:26	1.41
+++ b/radvd.c	2010/03/10 07:57:36	1.42
@@ -1,5 +1,5 @@
 /*
- *   $Id: radvd.c,v 1.41 2010/01/28 13:34:26 psavola Exp $
+ *   $Id: radvd.c,v 1.42 2010/03/10 07:57:36 psavola Exp $
  *
  *   Authors:
  *    Pedro Roque		<roque@di.fc.ul.pt>
@@ -382,12 +382,12 @@
 	for(iface=IfaceList; iface; iface=iface->next)
 	{
 		if( iface->UnicastOnly )
-			break;
+			continue;
 
 		init_timer(&iface->tm, timer_handler, (void *) iface);
 
 		if (!iface->AdvSendAdvert)
-			break;
+			continue;
 
 		/* send an initial advertisement */
 		if (send_ra_forall(sock, iface, NULL) == 0) {
